// ng-devui/esm2020/data-table/tmpl/data-table-cell-tmpl.component.mjs
import { ChangeDetectionStrategy, Component, ContentChild, Input } from '@angular/core';
import { DataTableCellEditTmplComponent } from './data-table-cell-edit-tmpl.component';
import { DataTableCellViewTmplComponent } from './data-table-cell-view-tmpl.component';
import * as i0 from "@angular/core";
export class DataTableCellTmplComponent {
}
DataTableCellTmplComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DataTableCellTmplComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
DataTableCellTmplComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: DataTableCellTmplComponent, selector: "d-cell-template", inputs: { type: "type" }, queries: [{ propertyName: "view", first: true, predicate: DataTableCellViewTmplComponent, descendants: true }, { propertyName: "edit", first: true, predicate: DataTableCellEditTmplComponent, descendants: true }], ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DataTableCellTmplComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'd-cell-template',
                    template: '',
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], propDecorators: { type: [{
                type: Input
            }], view: [{
                type: ContentChild,
                args: [DataTableCellViewTmplComponent]
            }], edit: [{
                type: ContentChild,
                args: [DataTableCellEditTmplComponent]
            }] } });
