// ng-devui/esm2020/data-table/tmpl/data-table-column-tmpl.component.mjs
import { ChangeDetectionStrategy, Component, ContentChild, EventEmitter, Input, Output, TemplateRef } from '@angular/core';
import { DataTableCellEditTmplComponent } from './data-table-cell-edit-tmpl.component';
import { DataTableCellViewTmplComponent } from './data-table-cell-view-tmpl.component';
import { DataTableHeadCellTmplComponent } from './data-table-head-cell-tmpl.component';
import * as i0 from "@angular/core";
export class DataTableColumnTmplComponent {
    constructor() {
        this.nestedColumnIndent = 16;
        this.order = Number.MAX_VALUE;
        this.nestedColumn = false;
        this.filterChange = new EventEmitter();
        this.filterToggle = new EventEmitter();
        this.filterMultiple = true;
        this.orderChange = new EventEmitter();
        this.widthChange = new EventEmitter();
        // 鼠标是否移动到当前表头单元格
        this.selected = false;
        // @deprecated
        this.fieldType = 'text';
    }
    ngOnInit() { }
    ngOnChanges(changes) {
        if (changes['order']) {
            this.orderChange.emit(changes['order'].currentValue);
        }
        if (changes['width']) {
            this.widthChange.emit(changes['width'].currentValue);
        }
    }
    ngOnDestroy() {
        this.orderChange.unsubscribe();
    }
    // column.extraOptions?.dateFormat
    emitFilterData(filterData) {
        this.filterChange.emit(filterData);
    }
    emitFilterToggle(data) {
        this.filterToggle.emit(data);
    }
}
DataTableColumnTmplComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DataTableColumnTmplComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
DataTableColumnTmplComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: DataTableColumnTmplComponent, selector: "d-column", inputs: { advancedHeader: "advancedHeader", maxWidth: "maxWidth", minWidth: "minWidth", field: "field", header: "header", sortable: "sortable", editable: "editable", filterable: "filterable", closeFilterWhenScroll: "closeFilterWhenScroll", cellClass: "cellClass", nestedColumnIndent: "nestedColumnIndent", width: "width", fixedLeft: "fixedLeft", fixedRight: "fixedRight", extraOptions: "extraOptions", order: "order", nestedColumn: "nestedColumn", filterList: "filterList", filterMultiple: "filterMultiple", filterIconActive: "filterIconActive", filterBoxWidth: "filterBoxWidth", filterBoxHeight: "filterBoxHeight", beforeFilter: "beforeFilter", customFilterTemplate: "customFilterTemplate", extraFilterTemplate: "extraFilterTemplate", searchFn: "searchFn", fieldType: "fieldType" }, outputs: { filterChange: "filterChange", filterToggle: "filterToggle" }, queries: [{ propertyName: "cellCmp", first: true, predicate: DataTableCellViewTmplComponent, descendants: true }, { propertyName: "cellEditCmp", first: true, predicate: DataTableCellEditTmplComponent, descendants: true }, { propertyName: "headCellTmpl", first: true, predicate: DataTableHeadCellTmplComponent, descendants: true }], usesOnChanges: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DataTableColumnTmplComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'd-column',
                    template: '',
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { advancedHeader: [{
                type: Input
            }], maxWidth: [{
                type: Input
            }], minWidth: [{
                type: Input
            }], field: [{
                type: Input
            }], header: [{
                type: Input
            }], sortable: [{
                type: Input
            }], editable: [{
                type: Input
            }], filterable: [{
                type: Input
            }], closeFilterWhenScroll: [{
                type: Input
            }], cellClass: [{
                type: Input
            }], nestedColumnIndent: [{
                type: Input
            }], width: [{
                type: Input
            }], fixedLeft: [{
                type: Input
            }], fixedRight: [{
                type: Input
            }], extraOptions: [{
                type: Input
            }], order: [{
                type: Input
            }], nestedColumn: [{
                type: Input
            }], filterList: [{
                type: Input
            }], filterChange: [{
                type: Output
            }], filterToggle: [{
                type: Output
            }], filterMultiple: [{
                type: Input
            }], filterIconActive: [{
                type: Input
            }], filterBoxWidth: [{
                type: Input
            }], filterBoxHeight: [{
                type: Input
            }], beforeFilter: [{
                type: Input
            }], cellCmp: [{
                type: ContentChild,
                args: [DataTableCellViewTmplComponent]
            }], cellEditCmp: [{
                type: ContentChild,
                args: [DataTableCellEditTmplComponent]
            }], headCellTmpl: [{
                type: ContentChild,
                args: [DataTableHeadCellTmplComponent]
            }], customFilterTemplate: [{
                type: Input
            }], extraFilterTemplate: [{
                type: Input
            }], searchFn: [{
                type: Input
            }], fieldType: [{
                type: Input
            }] } });
