// ng-devui/esm2020/data-table/tmpl/data-table-cell-view-tmpl.component.mjs
import { ChangeDetectionStrategy, Component, ContentChild, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
export class DataTableCellViewTmplComponent {
}
DataTableCellViewTmplComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DataTableCellViewTmplComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
DataTableCellViewTmplComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: DataTableCellViewTmplComponent, selector: "d-cell", queries: [{ propertyName: "template", first: true, predicate: TemplateRef, descendants: true }], ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DataTableCellViewTmplComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'd-cell',
                    template: '',
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], propDecorators: { template: [{
                type: ContentChild,
                args: [TemplateRef]
            }] } });
