// ng-devui/esm2020/dragdrop/drag-drop.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { BatchDraggableDirective } from './directives/batch-draggable.directive';
import { DragPreviewCloneDomRefComponent } from './directives/drag-preivew-clone-from-domRef.component';
import { DragPreviewComponent } from './directives/drag-preview.component';
import { DragPreviewDirective } from './directives/drag-preview.directive';
import { DragSyncDirective } from './directives/drag-sync.directive';
import { DragDropSyncBoxDirective } from './directives/dragdrop-sync-box.directive';
import { DraggableDirective } from './directives/draggable.directive';
import { DropScrollEnhancedSideDirective } from './directives/drop-scroll-enhance-side.directive';
import { DropScrollEnhancedDirective } from './directives/drop-scroll-enhance.directive';
import { DropSortSyncDirective } from './directives/drop-sort-sync.directive';
import { DroppableDirective } from './directives/droppable.directive';
import { SortableDirective } from './directives/sortable.directive';
import { DragDropService } from './services/drag-drop.service';
import * as i0 from "@angular/core";
export class DragDropModule {
}
DragDropModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DragDropModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DragDropModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DragDropModule, declarations: [DraggableDirective,
        DroppableDirective,
        SortableDirective,
        DropScrollEnhancedDirective,
        DropScrollEnhancedSideDirective,
        BatchDraggableDirective,
        DragSyncDirective,
        DropSortSyncDirective,
        DragDropSyncBoxDirective,
        DragPreviewDirective,
        DragPreviewComponent,
        DragPreviewCloneDomRefComponent], imports: [CommonModule], exports: [DraggableDirective,
        DroppableDirective,
        SortableDirective,
        DropScrollEnhancedDirective,
        DropScrollEnhancedSideDirective,
        BatchDraggableDirective,
        DragSyncDirective,
        DropSortSyncDirective,
        DragDropSyncBoxDirective,
        DragPreviewDirective,
        DragPreviewComponent,
        DragPreviewCloneDomRefComponent] });
DragDropModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DragDropModule, providers: [
        DragDropService
    ], imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DragDropModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [
                        DraggableDirective,
                        DroppableDirective,
                        SortableDirective,
                        DropScrollEnhancedDirective,
                        DropScrollEnhancedSideDirective,
                        BatchDraggableDirective,
                        DragSyncDirective,
                        DropSortSyncDirective,
                        DragDropSyncBoxDirective,
                        DragPreviewDirective,
                        DragPreviewComponent,
                        DragPreviewCloneDomRefComponent,
                    ],
                    exports: [
                        DraggableDirective,
                        DroppableDirective,
                        SortableDirective,
                        DropScrollEnhancedDirective,
                        DropScrollEnhancedSideDirective,
                        BatchDraggableDirective,
                        DragSyncDirective,
                        DropSortSyncDirective,
                        DragDropSyncBoxDirective,
                        DragPreviewDirective,
                        DragPreviewComponent,
                        DragPreviewCloneDomRefComponent,
                    ],
                    providers: [
                        DragDropService
                    ],
                }]
        }] });
