// ng-devui/esm2020/tags-input/tags.input.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { ToggleMenuModule } from 'ng-devui/toggle-menu';
import { TagsInputComponent } from './tags.input.component';
import * as i0 from "@angular/core";
export class TagsInputModule {
}
TagsInputModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TagsInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TagsInputModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TagsInputModule, declarations: [TagsInputComponent], imports: [CommonModule, FormsModule, ToggleMenuModule], exports: [TagsInputComponent] });
TagsInputModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TagsInputModule, imports: [[CommonModule, FormsModule, ToggleMenuModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TagsInputModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, FormsModule, ToggleMenuModule],
                    exports: [TagsInputComponent],
                    declarations: [TagsInputComponent],
                }]
        }] });
