// ng-devui/esm2020/portal/portal.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { PortalComponent } from './portal.component';
import * as i0 from "@angular/core";
export class PortalModule {
}
PortalModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PortalModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PortalModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PortalModule, declarations: [PortalComponent], imports: [CommonModule], exports: [PortalComponent] });
PortalModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PortalModule, imports: [[
            CommonModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PortalModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                    ],
                    declarations: [
                        PortalComponent,
                    ],
                    exports: [
                        PortalComponent,
                    ],
                }]
        }] });
