// ng-devui/esm2020/window-ref/window-ref.service.mjs
import { Injectable } from '@angular/core';
import { DocumentRef } from './document-ref.service';
import * as i0 from "@angular/core";
import * as i1 from "./document-ref.service";
export class WindowRef {
    constructor(documentRef) {
        this.documentRef = documentRef;
    }
    get window() {
        return this.document.defaultView;
    }
    get document() {
        return this.documentRef.document;
    }
    get pageXOffset() {
        return this.window.pageXOffset;
    }
    get pageYOffset() {
        return this.window.pageYOffset;
    }
    get innerHeight() {
        return this.window.innerHeight;
    }
    get innerWidth() {
        return this.window.innerWidth;
    }
    getComputedStyle(element) {
        return this.window.getComputedStyle(element);
    }
    getBoundingClientRect(elementRef) {
        return elementRef.nativeElement && elementRef.nativeElement.getBoundingClientRect();
    }
}
WindowRef.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: WindowRef, deps: [{ token: i1.DocumentRef }], target: i0.ɵɵFactoryTarget.Injectable });
WindowRef.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: WindowRef });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: WindowRef, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.DocumentRef }]; } });
