// ng-devui/esm2020/datepicker-pro/lib/footer-panel.component.mjs
import { Component, Input, TemplateRef } from '@angular/core';
import { I18nService } from 'ng-devui/i18n';
import { DatepickerProService } from '../datepicker-pro.service';
import * as i0 from "@angular/core";
import * as i1 from "../datepicker-pro.service";
import * as i2 from "ng-devui/i18n";
import * as i3 from "ng-devui/button";
import * as i4 from "@angular/common";
export class FooterPanelComponent {
    constructor(pickerSrv, i18n) {
        this.pickerSrv = pickerSrv;
        this.i18n = i18n;
        this.setI18nText();
    }
    get isRange() {
        return this.pickerSrv.isRange;
    }
    get confirmDisable() {
        if (this.isRange) {
            return (this.pickerSrv.currentActiveInput === 'start' && !this.pickerSrv.curRangeDate[0]) ||
                (this.pickerSrv.currentActiveInput === 'end' && !this.pickerSrv.curRangeDate[1]);
        }
        else {
            return false;
        }
    }
    setI18nText() {
        this.i18nText = this.i18n.getI18nText().common;
        this.i18nSubscription = this.i18n.langChange().subscribe((data) => {
            this.i18nText = data.common;
        });
    }
    ensureDate() {
        if (this.pickerSrv.isRange) {
            if (this.pickerSrv.currentActiveInput === 'start') {
                this.pickerSrv.currentActiveInput = 'end';
                this.pickerSrv.activeInputChange.next('end');
            }
            else if (!this.pickerSrv.curRangeDate[0]) {
                this.pickerSrv.currentActiveInput = 'start';
                this.pickerSrv.activeInputChange.next('start');
            }
            else {
                this.close(true);
            }
        }
        else {
            this.close(true);
        }
    }
    close(isConfirm = false) {
        this.pickerSrv.closeDropdownEvent.next(isConfirm);
    }
    ngOnDestroy() {
        this.i18nSubscription.unsubscribe();
    }
}
FooterPanelComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: FooterPanelComponent, deps: [{ token: i1.DatepickerProService }, { token: i2.I18nService }], target: i0.ɵɵFactoryTarget.Component });
FooterPanelComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: FooterPanelComponent, selector: "d-datepicker-footer-panel", inputs: { footerTemplate: "footerTemplate" }, ngImport: i0, template: "<div class=\"devui-datepicekr-pro-footer\">\n  <ng-template [ngTemplateOutlet]=\"footerTemplate || default\"> </ng-template>\n</div>\n\n<ng-template #default>\n  <d-button bsStyle=\"primary\" [disabled]=\"confirmDisable\" (click)=\"ensureDate()\"> {{ i18nText.btnOk }} </d-button>\n  <d-button class=\"devui-cancel-button\" bsStyle=\"common\" *ngIf=\"isRange\" (click)=\"close()\"> {{ i18nText.btnCancel }} </d-button>\n</ng-template>\n", styles: [".devui-datepicekr-pro-footer{padding:8px;text-align:center;border-top:1px solid #f2f2f3;border-top:1px solid var(--devui-dividing-line, #f2f2f3)}.devui-datepicekr-pro-footer .devui-cancel-button{margin-left:4px}\n"], components: [{ type: i3.ButtonComponent, selector: "d-button", inputs: ["id", "type", "bsStyle", "shape", "bsSize", "bsPosition", "bordered", "icon", "disabled", "showLoading", "width", "autofocus", "loadingTemplateRef"], outputs: ["btnClick"] }], directives: [{ type: i4.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet"] }, { type: i4.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: FooterPanelComponent, decorators: [{
            type: Component,
            args: [{ selector: 'd-datepicker-footer-panel', preserveWhitespaces: false, template: "<div class=\"devui-datepicekr-pro-footer\">\n  <ng-template [ngTemplateOutlet]=\"footerTemplate || default\"> </ng-template>\n</div>\n\n<ng-template #default>\n  <d-button bsStyle=\"primary\" [disabled]=\"confirmDisable\" (click)=\"ensureDate()\"> {{ i18nText.btnOk }} </d-button>\n  <d-button class=\"devui-cancel-button\" bsStyle=\"common\" *ngIf=\"isRange\" (click)=\"close()\"> {{ i18nText.btnCancel }} </d-button>\n</ng-template>\n", styles: [".devui-datepicekr-pro-footer{padding:8px;text-align:center;border-top:1px solid #f2f2f3;border-top:1px solid var(--devui-dividing-line, #f2f2f3)}.devui-datepicekr-pro-footer .devui-cancel-button{margin-left:4px}\n"] }]
        }], ctorParameters: function () { return [{ type: i1.DatepickerProService }, { type: i2.I18nService }]; }, propDecorators: { footerTemplate: [{
                type: Input
            }] } });
