// ng-devui/esm2020/tabs/tab.component.mjs
import { Component, ContentChild, Input, TemplateRef, ViewChild } from '@angular/core';
import { TabContentDirective } from './tab-content.directive';
import { TabTitleDirective } from './tab-title.directive';
import * as i0 from "@angular/core";
export class TabComponent {
    constructor() {
        this.disabled = false;
    }
    /**
     * @deprecated
     * 用id替代
     */
    set tabId(value) {
        this.id = value;
    }
}
TabComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TabComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TabComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: TabComponent, selector: "d-tab", inputs: { tabId: "tabId", id: "id", title: "title", disabled: "disabled" }, queries: [{ propertyName: "contentTpl", first: true, predicate: TabContentDirective, descendants: true }, { propertyName: "titleTpl", first: true, predicate: TabTitleDirective, descendants: true }], viewQueries: [{ propertyName: "innerContent", first: true, predicate: ["innerContent"], descendants: true, static: true }], ngImport: i0, template: `<ng-template #innerContent>
    <ng-content></ng-content>
  </ng-template>
  `, isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TabComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'd-tab',
                    template: `<ng-template #innerContent>
    <ng-content></ng-content>
  </ng-template>
  `,
                    preserveWhitespaces: false,
                }]
        }], propDecorators: { tabId: [{
                type: Input
            }], id: [{
                type: Input
            }], title: [{
                type: Input
            }], disabled: [{
                type: Input
            }], contentTpl: [{
                type: ContentChild,
                args: [TabContentDirective]
            }], titleTpl: [{
                type: ContentChild,
                args: [TabTitleDirective]
            }], innerContent: [{
                type: ViewChild,
                args: ['innerContent', { static: true }]
            }] } });
