// ng-devui/esm2020/tabs/tabs.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { DropDownModule } from 'ng-devui/dropdown';
import { TabContentDirective } from './tab-content.directive';
import { TabTitleDirective } from './tab-title.directive';
import { TabComponent } from './tab.component';
import { TabsComponent } from './tabs.component';
import * as i0 from "@angular/core";
export class TabsModule {
}
TabsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TabsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TabsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TabsModule, declarations: [TabsComponent, TabComponent, TabTitleDirective, TabContentDirective], imports: [CommonModule, DropDownModule], exports: [TabsComponent, TabComponent, TabTitleDirective, TabContentDirective] });
TabsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TabsModule, providers: [], imports: [[CommonModule, DropDownModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TabsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, DropDownModule],
                    exports: [TabsComponent, TabComponent, TabTitleDirective, TabContentDirective],
                    declarations: [TabsComponent, TabComponent, TabTitleDirective, TabContentDirective],
                    providers: [],
                }]
        }] });
