// ng-devui/esm2020/panel/panel.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { PanelBodyComponent } from './panel-body.component';
import { PanelFooterComponent } from './panel-footer.component';
import { PanelHeaderComponent } from './panel-header.component';
import { PanelComponent } from './panel.component';
import * as i0 from "@angular/core";
export class PanelModule {
}
PanelModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PanelModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PanelModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PanelModule, declarations: [PanelComponent,
        PanelHeaderComponent,
        PanelBodyComponent,
        PanelFooterComponent], imports: [CommonModule], exports: [PanelComponent,
        PanelHeaderComponent,
        PanelBodyComponent,
        PanelFooterComponent] });
PanelModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PanelModule, providers: [], imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PanelModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    exports: [
                        PanelComponent,
                        PanelHeaderComponent,
                        PanelBodyComponent,
                        PanelFooterComponent
                    ],
                    declarations: [
                        PanelComponent,
                        PanelHeaderComponent,
                        PanelBodyComponent,
                        PanelFooterComponent,
                    ],
                    providers: []
                }]
        }] });
