// ng-devui/esm2020/panel/panel-body.component.mjs
import { Component, HostBinding } from '@angular/core';
import * as i0 from "@angular/core";
export class PanelBodyComponent {
    constructor() {
        this.default = true;
    }
}
PanelBodyComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PanelBodyComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
PanelBodyComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: PanelBodyComponent, selector: "d-panel-body", host: { properties: { "class.d-panel-body": "this.default" } }, ngImport: i0, template: `<ng-content></ng-content>`, isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PanelBodyComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'd-panel-body',
                    template: `<ng-content></ng-content>`,
                    preserveWhitespaces: false,
                }]
        }], propDecorators: { default: [{
                type: HostBinding,
                args: ['class.d-panel-body']
            }] } });
