// ng-devui/esm2020/drawer/drawer.moudule.mjs
import { ScrollingModule } from '@angular/cdk/scrolling';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { OverlayContainerModule } from 'ng-devui/overlay-container';
import { PortalModule } from 'ng-devui/portal';
import { DrawerComponent, DrawerContentDirective } from './drawer.component';
import { DrawerService } from './drawer.service';
import * as i0 from "@angular/core";
export class DrawerModule {
}
DrawerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DrawerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DrawerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DrawerModule, declarations: [DrawerComponent,
        DrawerContentDirective], imports: [CommonModule,
        ScrollingModule,
        PortalModule,
        OverlayContainerModule], exports: [DrawerComponent] });
DrawerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DrawerModule, providers: [DrawerService], imports: [[
            CommonModule,
            ScrollingModule,
            PortalModule,
            OverlayContainerModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DrawerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        ScrollingModule,
                        PortalModule,
                        OverlayContainerModule
                    ],
                    exports: [DrawerComponent],
                    declarations: [
                        DrawerComponent,
                        DrawerContentDirective
                    ],
                    providers: [DrawerService],
                }]
        }] });
