// ng-devui/esm2020/drawer/drawer.service.mjs
import { ComponentFactoryResolver, Injectable } from '@angular/core';
import { OverlayContainerRef } from 'ng-devui/overlay-container';
import { assign, isUndefined } from 'lodash-es';
import { DrawerComponent } from './drawer.component';
import * as i0 from "@angular/core";
import * as i1 from "ng-devui/overlay-container";
export class DrawerService {
    constructor(overlayContainerRef, componentFactoryResolver) {
        this.overlayContainerRef = overlayContainerRef;
        this.componentFactoryResolver = componentFactoryResolver;
    }
    open({ drawerContentComponent, injector, componentFactoryResolver, id, zIndex, width, fullScreen, // @deprecated
    data, isCover, clickDoms, onClose, afterOpened, backdropCloseable, escKeyCloseable, beforeHidden, destroyOnHide = true, position = 'right', bodyScrollable = true, showAnimation = true, contentTemplate }) {
        const componentFactoryResolver_ = componentFactoryResolver || this.componentFactoryResolver;
        const drawerRef = this.overlayContainerRef.createComponent(componentFactoryResolver_.resolveComponentFactory(DrawerComponent), injector);
        assign(drawerRef.instance, {
            id,
            width,
            zIndex,
            isCover,
            clickDoms,
            fullScreen,
            beforeHidden,
            afterOpened,
            escKeyCloseable,
            position,
            backdropCloseable: isUndefined(backdropCloseable) ? true : backdropCloseable,
            bodyScrollable,
            showAnimation,
            contentTemplate
        });
        let drawerContentRef;
        if (drawerContentComponent) {
            drawerContentRef = drawerRef.instance.drawerContentHost.viewContainerRef.createComponent(componentFactoryResolver_.resolveComponentFactory(drawerContentComponent), 0, injector);
            assign(drawerContentRef.instance, data);
        }
        drawerRef.instance.onHidden = () => {
            if (onClose) {
                onClose();
            }
            if (destroyOnHide) {
                setTimeout(() => {
                    drawerRef.hostView.destroy();
                });
            }
        };
        drawerRef.instance.destroy = () => {
            if (!destroyOnHide && drawerRef.instance.animateState === 'void') {
                drawerRef.hostView.destroy();
            }
        };
        drawerRef.instance.show();
        return {
            drawerInstance: drawerRef.instance,
            drawerContentInstance: drawerContentRef ? drawerContentRef.instance : null
        };
    }
}
DrawerService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DrawerService, deps: [{ token: i1.OverlayContainerRef }, { token: i0.ComponentFactoryResolver }], target: i0.ɵɵFactoryTarget.Injectable });
DrawerService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DrawerService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DrawerService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.OverlayContainerRef }, { type: i0.ComponentFactoryResolver }]; } });
