// ng-devui/esm2020/toast/toast.service.mjs
import { ComponentFactoryResolver, Injectable } from '@angular/core';
import { OverlayContainerRef } from 'ng-devui/overlay-container';
import { assign } from 'lodash-es';
import { ToastComponent } from './toast.component';
import * as i0 from "@angular/core";
import * as i1 from "ng-devui/overlay-container";
export class ToastService {
    constructor(overlayContainerRef, componentFactoryResolver) {
        this.overlayContainerRef = overlayContainerRef;
        this.componentFactoryResolver = componentFactoryResolver;
    }
    open({ value, life = 5000, lifeMode = 'global', sticky = false, style, styleClass, injector, 
    /**
     * @deprecated
     */
    component, componentFactoryResolver, } = {}) {
        const finalComponentFactoryResolver = componentFactoryResolver || this.componentFactoryResolver;
        const toastRef = this.overlayContainerRef.createComponent(finalComponentFactoryResolver.resolveComponentFactory(ToastComponent), injector);
        assign(toastRef.instance, {
            lifeMode: lifeMode,
            sticky: sticky,
            style: style,
            styleClass: styleClass,
            value: value,
            life: life,
        });
        toastRef.instance.close = (index) => {
            if (index !== undefined && index > -1) {
                toastRef.instance.removeIndexThrottle(index);
            }
            else if (index !== undefined) {
                toastRef.instance.removeMsgThrottle(index);
            }
            else {
                setTimeout(() => {
                    toastRef.instance.removeAll();
                    setTimeout(() => {
                        if (toastRef) {
                            toastRef.destroy();
                        }
                    }, 300);
                });
            }
        };
        toastRef.instance.onHidden = () => {
            if (toastRef?.hostView) {
                toastRef.hostView.destroy();
            }
        };
        toastRef.instance.show();
        return {
            toastInstance: toastRef.instance,
        };
    }
}
ToastService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: ToastService, deps: [{ token: i1.OverlayContainerRef }, { token: i0.ComponentFactoryResolver }], target: i0.ɵɵFactoryTarget.Injectable });
ToastService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: ToastService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: ToastService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.OverlayContainerRef }, { type: i0.ComponentFactoryResolver }]; } });
