// ng-devui/esm2020/tree/auto-focus.directive.mjs
import { Directive, ElementRef } from '@angular/core';
import * as i0 from "@angular/core";
export class AutofocusDirective {
    constructor(elementRef) {
        this.elementRef = elementRef;
    }
    ngOnInit() {
        setTimeout(() => {
            this.elementRef.nativeElement.focus();
            this.elementRef.nativeElement.select();
        });
    }
}
AutofocusDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AutofocusDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
AutofocusDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: AutofocusDirective, selector: "[dTreeAutoFocus]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AutofocusDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[dTreeAutoFocus]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; } });
