// ng-devui/esm2020/utils/popper/popper.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { PopperComponent } from './popper.component';
import * as i0 from "@angular/core";
export class PopperModule {
}
PopperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PopperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PopperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PopperModule, declarations: [PopperComponent], imports: [CommonModule], exports: [PopperComponent] });
PopperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PopperModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PopperModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [PopperComponent],
                    exports: [PopperComponent]
                }]
        }] });
