// ng-devui/esm2020/gantt/gantt-bar-parent/gantt-bar-parent.component.mjs
import { ChangeDetectorRef, Component, ElementRef, Input, ViewChild } from '@angular/core';
import { GanttService } from '../gantt.service';
import * as i0 from "@angular/core";
import * as i1 from "../gantt.service";
import * as i2 from "@angular/common";
import * as i3 from "ng-devui/tooltip";
export class GanttBarParentComponent {
    constructor(cdr, ganttService) {
        this.cdr = cdr;
        this.ganttService = ganttService;
        this.progressRate = 0;
        this.tipHovered = false;
        this.left = 0;
        this.width = 0;
        this.max = 100;
        this.min = 0;
    }
    ngOnInit() {
        if (this.progressRate === null) {
            this.setValue(this.ensureValueInRange(null));
        }
        this.duration = this.ganttService.getDuration(this.startDate, this.endDate) + 'd';
        this.ganttScaleStatusHandler = this.ganttService.ganttScaleConfigChange.subscribe((config) => {
            if (config.startDate) {
                this.left = this.ganttService.getDatePostionOffset(this.startDate);
            }
            if (config.unit) {
                this.left = this.ganttService.getDatePostionOffset(this.startDate);
                this.width = this.ganttService.getDurationWidth(this.startDate, this.endDate);
            }
        });
    }
    ngOnChanges(changes) {
        if (changes['progressRate'] && this.progressRate > 0) {
            this.updateTrackAndHandle();
        }
        if (changes['startDate']) {
            this.left = this.ganttService.getDatePostionOffset(this.startDate);
            this.width = this.ganttService.getDurationWidth(this.startDate, this.endDate);
        }
        if (changes['endDate']) {
            this.width = this.ganttService.getDurationWidth(this.startDate, this.endDate);
        }
    }
    ngAfterViewInit() {
        if (this.progressRate && this.progressRate > 0) {
            this.updateTrackAndHandle();
        }
    }
    setValue(value) {
        if (this.progressRate !== value) {
            this.progressRate = value;
            this.updateTrackAndHandle();
        }
    }
    ensureValueInRange(value) {
        let safeValue;
        if (!this.valueMustBeValid(value)) {
            safeValue = this.min;
        }
        else {
            safeValue = this.clamp(this.min, value, this.max);
        }
        return safeValue;
    }
    valueMustBeValid(value) {
        return !isNaN(typeof value !== 'number' ? parseFloat(value) : value);
    }
    clamp(min, n, max) {
        return Math.max(min, Math.min(n, max));
    }
    updateTrackAndHandle() {
        const value = this.progressRate;
        const offset = this.valueToOffset(value);
        this.updateStyle(offset / 100);
        this.cdr.markForCheck();
    }
    valueToOffset(value) {
        return ((value - this.min) / (this.max - this.min)) * 100;
    }
    updateStyle(percentage) {
        this.percentage = Math.min(1, Math.max(0, percentage));
        if (this.ganttBarTrack && this.ganttBarTrack.nativeElement) {
            this.ganttBarTrack.nativeElement.style.width = `${this.percentage * 100}%`;
        }
        if (this.ganttBarProgress && this.ganttBarProgress.nativeElement) {
            this.ganttBarProgress.nativeElement.style.left = `${this.percentage * 100}%`;
        }
    }
    ngOnDestroy() {
        if (this.ganttScaleStatusHandler) {
            this.ganttScaleStatusHandler.unsubscribe();
            this.ganttScaleStatusHandler = null;
        }
    }
}
GanttBarParentComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: GanttBarParentComponent, deps: [{ token: i0.ChangeDetectorRef }, { token: i1.GanttService }], target: i0.ɵɵFactoryTarget.Component });
GanttBarParentComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: GanttBarParentComponent, selector: "d-gantt-bar-parent", inputs: { startDate: "startDate", endDate: "endDate", progressRate: "progressRate", data: "data", id: "id", tip: "tip" }, viewQueries: [{ propertyName: "ganttBarProgress", first: true, predicate: ["ganttBarProgress"], descendants: true }, { propertyName: "ganttBarTrack", first: true, predicate: ["ganttBarTrack"], descendants: true }, { propertyName: "ganttBarRail", first: true, predicate: ["ganttBarRail"], descendants: true }], usesOnChanges: true, ngImport: i0, template: "<div\n  *ngIf=\"startDate && endDate\"\n  #ganttBar\n  class=\"devui-gantt-bar-parent\"\n  [style.left]=\"left + 'px'\"\n  [style.width]=\"width + 'px'\"\n  dTooltip\n  [content]=\"tip\"\n>\n  <div class=\"devui-gantt-bar-rail\" #ganttBarRail></div>\n  <div class=\"devui-gantt-bar-track\" [ngClass]=\"{ head: percentage > 0, tail: percentage === 1 }\" #ganttBarTrack></div>\n</div>\n", styles: [".devui-gantt-bar-parent{width:100%;box-sizing:border-box;height:24px;z-index:3;position:relative}.devui-gantt-bar-parent .devui-gantt-bar-rail{position:absolute;background:#eaecf0;height:12px;width:100%;margin-top:4px}.devui-gantt-bar-parent .devui-gantt-bar-rail:before{width:0;height:0;border:3px transparent solid;border-top-color:#eaecf0;border-left-color:#eaecf0;position:absolute;left:0;bottom:-6px;content:\"\"}.devui-gantt-bar-parent .devui-gantt-bar-rail:after{width:0;height:0;border:3px transparent solid;border-top-color:#eaecf0;border-right-color:#eaecf0;position:absolute;right:0;bottom:-6px;content:\"\"}.devui-gantt-bar-parent .devui-gantt-bar-track{position:absolute;background-color:#cacfd8;height:12px;margin-top:4px;width:0}.devui-gantt-bar-parent .devui-gantt-bar-track.head:before{width:0;height:0;border:3px transparent solid;border-top-color:#cacfd8;border-left-color:#cacfd8;position:absolute;left:0;bottom:-6px;content:\"\"}.devui-gantt-bar-parent .devui-gantt-bar-track.tail:after{width:0;height:0;border:3px transparent solid;border-top-color:#cacfd8;border-right-color:#cacfd8;position:absolute;right:0;bottom:-6px;content:\"\"}\n"], directives: [{ type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i3.TooltipDirective, selector: "[dTooltip]", inputs: ["content", "position", "showAnimation", "showAnimate", "mouseEnterDelay", "mouseLeaveDelay"], exportAs: ["dTooltip"] }, { type: i2.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: GanttBarParentComponent, decorators: [{
            type: Component,
            args: [{ selector: 'd-gantt-bar-parent', template: "<div\n  *ngIf=\"startDate && endDate\"\n  #ganttBar\n  class=\"devui-gantt-bar-parent\"\n  [style.left]=\"left + 'px'\"\n  [style.width]=\"width + 'px'\"\n  dTooltip\n  [content]=\"tip\"\n>\n  <div class=\"devui-gantt-bar-rail\" #ganttBarRail></div>\n  <div class=\"devui-gantt-bar-track\" [ngClass]=\"{ head: percentage > 0, tail: percentage === 1 }\" #ganttBarTrack></div>\n</div>\n", styles: [".devui-gantt-bar-parent{width:100%;box-sizing:border-box;height:24px;z-index:3;position:relative}.devui-gantt-bar-parent .devui-gantt-bar-rail{position:absolute;background:#eaecf0;height:12px;width:100%;margin-top:4px}.devui-gantt-bar-parent .devui-gantt-bar-rail:before{width:0;height:0;border:3px transparent solid;border-top-color:#eaecf0;border-left-color:#eaecf0;position:absolute;left:0;bottom:-6px;content:\"\"}.devui-gantt-bar-parent .devui-gantt-bar-rail:after{width:0;height:0;border:3px transparent solid;border-top-color:#eaecf0;border-right-color:#eaecf0;position:absolute;right:0;bottom:-6px;content:\"\"}.devui-gantt-bar-parent .devui-gantt-bar-track{position:absolute;background-color:#cacfd8;height:12px;margin-top:4px;width:0}.devui-gantt-bar-parent .devui-gantt-bar-track.head:before{width:0;height:0;border:3px transparent solid;border-top-color:#cacfd8;border-left-color:#cacfd8;position:absolute;left:0;bottom:-6px;content:\"\"}.devui-gantt-bar-parent .devui-gantt-bar-track.tail:after{width:0;height:0;border:3px transparent solid;border-top-color:#cacfd8;border-right-color:#cacfd8;position:absolute;right:0;bottom:-6px;content:\"\"}\n"] }]
        }], ctorParameters: function () { return [{ type: i0.ChangeDetectorRef }, { type: i1.GanttService }]; }, propDecorators: { ganttBarProgress: [{
                type: ViewChild,
                args: ['ganttBarProgress']
            }], ganttBarTrack: [{
                type: ViewChild,
                args: ['ganttBarTrack']
            }], ganttBarRail: [{
                type: ViewChild,
                args: ['ganttBarRail']
            }], startDate: [{
                type: Input
            }], endDate: [{
                type: Input
            }], progressRate: [{
                type: Input
            }], data: [{
                type: Input
            }], id: [{
                type: Input
            }], tip: [{
                type: Input
            }] } });
