// ng-devui/esm2020/datepicker-pro/datepicker-pro.component.mjs
import { Component, ContentChild, ElementRef, EventEmitter, forwardRef, Input, Output, TemplateRef, ViewChild } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { I18nService } from 'ng-devui/i18n';
import { DefaultDateConverter } from 'ng-devui/utils';
import { fromEvent, Subject } from 'rxjs';
import { debounceTime, takeUntil } from 'rxjs/operators';
import { DatepickerProService } from './datepicker-pro.service';
import * as i0 from "@angular/core";
import * as i1 from "ng-devui/i18n";
import * as i2 from "./datepicker-pro.service";
import * as i3 from "ng-devui/dropdown";
import * as i4 from "./datepicker-panel.component";
import * as i5 from "@angular/common";
import * as i6 from "@angular/forms";
export class DatepickerProComponent {
    constructor(i18n, pickerSrv) {
        this.i18n = i18n;
        this.pickerSrv = pickerSrv;
        this.mode = 'date';
        this.showTime = false;
        this.disabled = false;
        this.autoOpen = false;
        this.showAnimation = true;
        this.appendToBody = true;
        this.allowClear = true;
        this.dropdownToggle = new EventEmitter();
        this.confirmEvent = new EventEmitter();
        this.dateValue = '';
        this.unsubscribe$ = new Subject();
        this.isOpen = false;
        this.onChange = (_) => null;
        this.onTouched = () => null;
        this.i18nText = this.i18n.getI18nText().datePickerPro;
        this.datepickerConvert = new DefaultDateConverter();
    }
    set calenderRange(value) {
        this.pickerSrv.calendarRange = value || [1970, 2099];
    }
    set minDate(value) {
        if (!value) {
            return;
        }
        this.pickerSrv.minDate = value;
    }
    set maxDate(value) {
        if (!value) {
            return;
        }
        this.pickerSrv.maxDate = value;
    }
    set markedRangeDateList(value) {
        this.pickerSrv.markedRangeDateList = value;
    }
    ;
    set markedDateList(value) {
        this.pickerSrv.markedDateList = value;
    }
    set markDateInfoTemplate(tmp) {
        this.pickerSrv.markDateInfoTemplate = tmp;
    }
    ;
    get dateConfig() {
        return {
            dateConverter: this.datepickerConvert,
            min: this.pickerSrv.minDate || new Date(this.pickerSrv.calendarRange[0] + '/01/01'),
            max: this.pickerSrv.maxDate || new Date(this.pickerSrv.calendarRange[1] + '/12/31'),
            format: {
                date: this.format || 'y/MM/dd',
                time: this.format || 'y/MM/dd HH:mm:ss',
                month: this.format || 'y-MM',
                year: 'y'
            }
        };
    }
    get curFormat() {
        if (this.mode === 'year') {
            return this.dateConfig.format.year;
        }
        else if (this.mode === 'month') {
            return this.dateConfig.format.month;
        }
        else {
            return this.showTime ? this.dateConfig.format.time : this.dateConfig.format.date;
        }
    }
    ngOnInit() {
        this.initSrvStatus();
        this.setI18nText();
        setTimeout(() => {
            this.isOpen = this.autoOpen;
        });
    }
    ngAfterViewInit() {
        this.initObservable();
    }
    initSrvStatus() {
        this.pickerSrv.showTime = this.showTime;
        this.pickerSrv.isRange = false;
    }
    initObservable() {
        this.pickerSrv.selectedDateChange.pipe(takeUntil(this.unsubscribe$)).subscribe(change => {
            this.dateValue = this.formatDateToString(change.value);
            this.pickerSrv.curDate = change.value;
            this.onChange(change.value);
        });
        this.pickerSrv.selectedTimeChange.pipe(takeUntil(this.unsubscribe$)).subscribe(time => {
            if (this.dateValue) {
                const curTime = this.datepickerConvert.parse(this.dateValue, this.curFormat).setHours(time.hour, time.min, time.seconds);
                const curDate = new Date(curTime);
                this.pickerSrv.curDate = curDate;
                this.dateValue = this.formatDateToString(curDate);
                this.onChange(curDate);
            }
            else {
                this.writeValue(new Date(new Date().setHours(time.hour, time.min, time.seconds)));
                this.onChange(this.pickerSrv.curDate);
            }
        });
        this.pickerSrv.closeDropdownEvent.pipe(takeUntil(this.unsubscribe$)).subscribe(isConfirm => {
            this.isOpen = false;
            this.dropdownToggle.emit(false);
            if (isConfirm) {
                this.confirmEvent.emit(this.pickerSrv.curDate);
            }
        });
        if (!this.hostTemplate) {
            fromEvent(this.datepickerInput.nativeElement, 'input').pipe(takeUntil(this.unsubscribe$), debounceTime(300)).subscribe((event) => {
                if (!this.dateValue) {
                    return;
                }
                const inputDate = this.datepickerConvert.parse(this.dateValue, this.curFormat);
                if (inputDate instanceof Date && inputDate.getTime() === this.pickerSrv.curDate.getTime()) {
                    return;
                }
                if (this.validateDate(this.dateValue)) {
                    this.pickerSrv.curDate = inputDate;
                    this.pickerSrv.updateDateValue.next({
                        type: 'single',
                        value: inputDate
                    });
                    if (this.showTime) {
                        this.pickerSrv.updateTimeChange.next({
                            hour: inputDate.getHours(),
                            min: inputDate.getMinutes(),
                            seconds: inputDate.getSeconds()
                        });
                    }
                }
            });
            fromEvent(this.datepickerInput.nativeElement, 'blur').pipe(takeUntil(this.unsubscribe$)).subscribe(() => {
                if (!this.validateDate(this.dateValue)) {
                    this.dateValue = this.pickerSrv.curDate ?
                        this.datepickerConvert.format(this.pickerSrv.curDate, this.curFormat, this.i18nLocale) :
                        '';
                }
            });
        }
    }
    setI18nText() {
        this.i18nLocale = this.i18n.getI18nText().locale;
        this.i18n.langChange().pipe(takeUntil(this.unsubscribe$)).subscribe((data) => {
            this.i18nLocale = data.locale;
            this.i18nText = data.datePickerPro;
        });
    }
    validateDate(value) {
        const valueDate = this.datepickerConvert.parse(value, this.curFormat);
        const valueFormat = valueDate && !isNaN(valueDate.getTime()) &&
            this.datepickerConvert.format(valueDate, this.curFormat, this.i18nLocale);
        if (!valueDate || value !== valueFormat ||
            (value === valueFormat && !this.pickerSrv.dateInRange(valueDate))) {
            return false;
        }
        else {
            return true;
        }
    }
    formatDateToString(date) {
        return this.datepickerConvert.format(date, this.curFormat);
    }
    clear(event, isHandle) {
        if (event) {
            event.stopPropagation();
        }
        if (this.disabled && isHandle) {
            return;
        }
        this.pickerSrv.updateDateValue.next({
            type: 'single',
            value: null
        });
        this.pickerSrv.updateTimeChange.next({
            hour: null,
            min: null,
            seconds: null
        });
        this.dateValue = null;
        this.pickerSrv.curDate = null;
        if (event) {
            this.onChange(this.pickerSrv.curDate);
        }
    }
    onToggle(isOpen) {
        if (isOpen !== this.isOpen || isOpen) {
            this.dropdownToggle.emit(isOpen);
        }
        this.isOpen = isOpen;
        this.pickerSrv.toggleEvent.next(isOpen);
    }
    openDropdown(event) {
        if (this.isOpen) {
            event.stopPropagation();
        }
        this.isOpen = true;
        setTimeout(() => {
            if (this.datepickerInput?.nativeElement) {
                this.datepickerInput.nativeElement.focus();
            }
        });
    }
    writeValue(value) {
        if (!value || !this.pickerSrv.dateInRange(new Date(value))) {
            this.clear();
            return;
        }
        this.dateValue = this.formatDateToString(value);
        this.pickerSrv.curDate = value;
        this.pickerSrv.updateDateValue.next({
            type: 'single',
            value
        });
        if (this.showTime) {
            this.pickerSrv.updateTimeChange.next({
                hour: value.getHours(),
                min: value.getMinutes(),
                seconds: value.getSeconds()
            });
        }
    }
    registerOnChange(fn) {
        this.onChange = fn;
    }
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    ngOnDestroy() {
        this.unsubscribe$.next();
        this.unsubscribe$.complete();
    }
}
DatepickerProComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DatepickerProComponent, deps: [{ token: i1.I18nService }, { token: i2.DatepickerProService }], target: i0.ɵɵFactoryTarget.Component });
DatepickerProComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: DatepickerProComponent, selector: "d-datepicker-pro", inputs: { mode: "mode", showTime: "showTime", disabled: "disabled", autoOpen: "autoOpen", format: "format", cssClass: "cssClass", showAnimation: "showAnimation", appendToBody: "appendToBody", width: "width", placeholder: "placeholder", allowClear: "allowClear", calenderRange: "calenderRange", minDate: "minDate", maxDate: "maxDate", markedRangeDateList: "markedRangeDateList", markedDateList: "markedDateList" }, outputs: { dropdownToggle: "dropdownToggle", confirmEvent: "confirmEvent" }, providers: [
        DatepickerProService,
        {
            provide: NG_VALUE_ACCESSOR,
            useExisting: forwardRef(() => DatepickerProComponent),
            multi: true
        }
    ], queries: [{ propertyName: "customTemplate", first: true, predicate: ["customTemplate"], descendants: true }, { propertyName: "footerTemplate", first: true, predicate: ["footerTemplate"], descendants: true }, { propertyName: "hostTemplate", first: true, predicate: ["hostTemplate"], descendants: true }, { propertyName: "markDateInfoTemplate", first: true, predicate: ["markDateInfoTemplate"], descendants: true }], viewQueries: [{ propertyName: "datepickerInput", first: true, predicate: ["dateInput"], descendants: true }], ngImport: i0, template: "<div\n  class=\"devui-datepicker-pro-wrapper {{ cssClass }}\"\n  *ngIf=\"appendToBody; else normal\"\n  dDropDown\n  appendToBody\n  [disabled]=\"disabled\"\n  [isOpen]=\"isOpen\"\n  [trigger]=\"'manually'\"\n  (toggleEvent)=\"onToggle($event)\"\n  (click)=\"openDropdown($event)\"\n  [closeScope]=\"'blank'\"\n  [showAnimation]=\"showAnimation\"\n  [ngStyle]=\"{ width: width || '250px' }\"\n>\n  <div dDropDownToggle>\n    <ng-template [ngTemplateOutlet]=\"hostTemplate || default\" [ngTemplateOutletContext]=\"{ value: dateValue }\"></ng-template>\n  </div>\n\n  <div class=\"devui-datepicker-pro-dropdown-panel\" [ngStyle]=\"{ paddingBottom: !showTime && !footerTemplate ? 0 : '4px' }\" dDropDownMenu>\n    <d-datepicker-panel\n      [isRangeType]=\"false\"\n      [showTime]=\"showTime\"\n      [customTemplate]=\"customTemplate\"\n      [footerTemplate]=\"footerTemplate\"\n      [mode]=\"mode\"\n    >\n    </d-datepicker-panel>\n  </div>\n</div>\n\n<ng-template #normal>\n  <div\n    class=\"devui-datepicker-pro-wrapper {{ cssClass }}\"\n    dDropDown\n    [disabled]=\"disabled\"\n    [isOpen]=\"isOpen\"\n    [trigger]=\"'manually'\"\n    (toggleEvent)=\"onToggle($event)\"\n    (click)=\"openDropdown($event)\"\n    [closeScope]=\"'blank'\"\n    [showAnimation]=\"showAnimation\"\n    [ngStyle]=\"{ width: width || '250px' }\"\n  >\n    <div dDropDownToggle>\n      <ng-template [ngTemplateOutlet]=\"hostTemplate || default\" [ngTemplateOutletContext]=\"{ value: dateValue }\"></ng-template>\n    </div>\n\n    <div class=\"devui-datepicker-pro-dropdown-panel\" [ngStyle]=\"{ paddingBottom: !showTime && !footerTemplate ? 0 : '4px' }\" dDropDownMenu>\n      <d-datepicker-panel\n        [isRangeType]=\"false\"\n        [showTime]=\"showTime\"\n        [customTemplate]=\"customTemplate\"\n        [footerTemplate]=\"footerTemplate\"\n        [mode]=\"mode\"\n      >\n      </d-datepicker-panel>\n    </div>\n  </div>\n</ng-template>\n\n<ng-template #default>\n  <div\n    class=\"devui-single-picker devui-input-group devui-dropdown-origin\"\n    [ngClass]=\"{\n      'devui-disabled': disabled,\n      'devui-has-value': dateValue\n    }\"\n  >\n    <i class=\"devui-single-picker-icon\">\n      <svg width=\"16px\" height=\"16px\" viewBox=\"0 0 16 16\" version=\"1.1\">\n        <g stroke=\"none\" stroke-width=\"1\" fill=\"none\" fill-rule=\"evenodd\">\n          <path\n            d=\"M5,0 L5,1 L11,1 L11,0 L12,0 L12,1 L14,1 C15.1045695,1 16,1.8954305 16,3 L16,14 C16,15.1045695 15.1045695,16 14,16 L2,16 C0.8954305,16 0,15.1045695 0,14 L0,3 C0,1.8954305 0.8954305,1 2,1 L4,1 L4,0 L5,0 Z M15,5 L1,5 L1,14 C1,14.5522847 1.44771525,15 2,15 L14,15 C14.5522847,15 15,14.5522847 15,14 L15,5 Z M6,7 C6.55228475,7 7,7.44771525 7,8 L7,10 C7,10.5522847 6.55228475,11 6,11 L4,11 C3.44771525,11 3,10.5522847 3,10 L3,8 C3,7.44771525 3.44771525,7 4,7 L6,7 Z M6,8 L4,8 L4,10 L6,10 L6,8 Z M4,2 L2,2 C1.44771525,2 1,2.44771525 1,3 L1,4 L15,4 L15,3 C15,2.44771525 14.5522847,2 14,2 L12,2 L12,3 L11,3 L11,2 L5,2 L5,3 L4,3 L4,2 Z\"\n            fill=\"#BABBC0\"\n          ></path>\n        </g>\n      </svg>\n    </i>\n    <input\n      #dateInput\n      class=\"devui-input devui-form-control\"\n      [disabled]=\"disabled\"\n      [placeholder]=\"placeholder || i18nText?.placeholder\"\n      [(ngModel)]=\"dateValue\"\n      autocomplete=\"off\"\n    />\n    <div *ngIf=\"allowClear\" class=\"devui-input-group-addon close-icon-wrapper\" (click)=\"clear($event, true)\">\n      <div class=\"close-icon\">\n        <svg\n          width=\"16px\"\n          height=\"16px\"\n          viewBox=\"0 0 16 16\"\n          version=\"1.1\"\n          xmlns=\"http://www.w3.org/2000/svg\"\n          xmlns:xlink=\"http://www.w3.org/1999/xlink\"\n        >\n          <g stroke=\"none\" stroke-width=\"1\" fill=\"none\" fill-rule=\"evenodd\">\n            <path\n              d=\"M8,1 C11.8659932,1 15,4.13400675 15,8 C15,11.8659932 11.8659932,15 8,15 C4.13400675,15 1,11.8659932 1,8 C1,4.13400675 4.13400675,1 8,1 Z M5.87867966,5.17157288 C5.68341751,4.97631073 5.36683502,4.97631073 5.17157288,5.17157288 C4.99800652,5.34513923 4.97872137,5.61456363 5.11371742,5.80943177 L5.17157288,5.87867966 L7.29218611,7.99929289 L5.17157288,10.1213203 C4.97631073,10.3165825 4.97631073,10.633165 5.17157288,10.8284271 C5.34513923,11.0019935 5.61456363,11.0212786 5.80943177,10.8862826 L5.87867966,10.8284271 L7.99929289,8.70639967 L10.1213203,10.8284271 C10.3165825,11.0236893 10.633165,11.0236893 10.8284271,10.8284271 C11.0019935,10.6548608 11.0212786,10.3854364 10.8862826,10.1905682 L10.8284271,10.1213203 L8.70710678,8 L10.8284271,5.87867966 C11.0236893,5.68341751 11.0236893,5.36683502 10.8284271,5.17157288 C10.6548608,4.99800652 10.3854364,4.97872137 10.1905682,5.11371742 L10.1213203,5.17157288 L8,7.29289322 L5.87867966,5.17157288 Z\"\n              fill-rule=\"nonzero\"\n            ></path>\n          </g>\n        </svg>\n      </div>\n    </div>\n  </div>\n</ng-template>\n", styles: [".devui-datepicker-pro-wrapper{display:inline-block;background-color:#fff;background-color:var(--devui-base-bg, #ffffff)}.devui-datepicker-pro-wrapper .devui-single-picker{padding-left:8px;position:relative;box-sizing:border-box;width:100%;min-height:24px}.devui-datepicker-pro-wrapper .devui-single-picker .devui-input{padding:4px 8px;width:calc(100% - 16px)}.devui-datepicker-pro-wrapper .devui-single-picker-icon{vertical-align:bottom}.devui-datepicker-pro-wrapper .devui-single-picker .close-icon-wrapper{padding:0 8px;vertical-align:baseline;visibility:hidden}.devui-datepicker-pro-wrapper .devui-single-picker .close-icon-wrapper .close-icon{width:16px;height:16px;display:inline-block;line-height:12px}.devui-datepicker-pro-wrapper .devui-single-picker .close-icon-wrapper .close-icon svg{vertical-align:middle;position:relative;top:-.1em}.devui-datepicker-pro-wrapper .devui-single-picker .close-icon-wrapper .close-icon svg path{fill:#d7d8da;fill:var(--devui-shape-icon-fill, #d7d8da)}.devui-datepicker-pro-wrapper .devui-single-picker .close-icon-wrapper .close-icon:hover svg path{fill:#babbc0;fill:var(--devui-shape-icon-fill-active, #babbc0)}.devui-datepicker-pro-wrapper .devui-single-picker.devui-has-value:hover:not(.devui-disabled) .close-icon-wrapper{visibility:visible}.devui-datepicker-pro-wrapper .devui-single-picker:not(.devui-disabled) .close-icon-wrapper{cursor:pointer}.devui-datepicker-pro-wrapper .devui-input-group>.devui-input{display:inline-block}.devui-datepicker-pro-wrapper .devui-input-group>.devui-input::-ms-clear{display:none}\n"], components: [{ type: i3.DropDownAppendToBodyComponent, selector: "[dDropDown][appendToBody]", inputs: ["alignOrigin", "appendToBodyDirections"] }, { type: i4.DatepickerPanelComponent, selector: "d-datepicker-panel", inputs: ["isRangeType", "showTime", "showCustom", "customTemplate", "footerTemplate", "mode"] }], directives: [{ type: i5.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i3.DropDownDirective, selector: "[dDropDown]", inputs: ["isOpen", "disabled", "showAnimation", "trigger", "closeScope", "closeOnMouseLeaveMenu"], outputs: ["toggleEvent"], exportAs: ["d-dropdown"] }, { type: i5.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }, { type: i3.DropDownToggleDirective, selector: "[dDropDownToggle]", inputs: ["toggleOnFocus", "autoFocus"], exportAs: ["d-dropdown-toggle"] }, { type: i5.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet"] }, { type: i3.DropDownMenuDirective, selector: "[dDropDownMenu]", exportAs: ["d-dropdown-menu"] }, { type: i5.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { type: i6.DefaultValueAccessor, selector: "input:not([type=checkbox])[formControlName],textarea[formControlName],input:not([type=checkbox])[formControl],textarea[formControl],input:not([type=checkbox])[ngModel],textarea[ngModel],[ngDefaultControl]" }, { type: i6.NgControlStatus, selector: "[formControlName],[ngModel],[formControl]" }, { type: i6.NgModel, selector: "[ngModel]:not([formControlName]):not([formControl])", inputs: ["name", "disabled", "ngModel", "ngModelOptions"], outputs: ["ngModelChange"], exportAs: ["ngModel"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DatepickerProComponent, decorators: [{
            type: Component,
            args: [{ selector: 'd-datepicker-pro', providers: [
                        DatepickerProService,
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(() => DatepickerProComponent),
                            multi: true
                        }
                    ], preserveWhitespaces: false, template: "<div\n  class=\"devui-datepicker-pro-wrapper {{ cssClass }}\"\n  *ngIf=\"appendToBody; else normal\"\n  dDropDown\n  appendToBody\n  [disabled]=\"disabled\"\n  [isOpen]=\"isOpen\"\n  [trigger]=\"'manually'\"\n  (toggleEvent)=\"onToggle($event)\"\n  (click)=\"openDropdown($event)\"\n  [closeScope]=\"'blank'\"\n  [showAnimation]=\"showAnimation\"\n  [ngStyle]=\"{ width: width || '250px' }\"\n>\n  <div dDropDownToggle>\n    <ng-template [ngTemplateOutlet]=\"hostTemplate || default\" [ngTemplateOutletContext]=\"{ value: dateValue }\"></ng-template>\n  </div>\n\n  <div class=\"devui-datepicker-pro-dropdown-panel\" [ngStyle]=\"{ paddingBottom: !showTime && !footerTemplate ? 0 : '4px' }\" dDropDownMenu>\n    <d-datepicker-panel\n      [isRangeType]=\"false\"\n      [showTime]=\"showTime\"\n      [customTemplate]=\"customTemplate\"\n      [footerTemplate]=\"footerTemplate\"\n      [mode]=\"mode\"\n    >\n    </d-datepicker-panel>\n  </div>\n</div>\n\n<ng-template #normal>\n  <div\n    class=\"devui-datepicker-pro-wrapper {{ cssClass }}\"\n    dDropDown\n    [disabled]=\"disabled\"\n    [isOpen]=\"isOpen\"\n    [trigger]=\"'manually'\"\n    (toggleEvent)=\"onToggle($event)\"\n    (click)=\"openDropdown($event)\"\n    [closeScope]=\"'blank'\"\n    [showAnimation]=\"showAnimation\"\n    [ngStyle]=\"{ width: width || '250px' }\"\n  >\n    <div dDropDownToggle>\n      <ng-template [ngTemplateOutlet]=\"hostTemplate || default\" [ngTemplateOutletContext]=\"{ value: dateValue }\"></ng-template>\n    </div>\n\n    <div class=\"devui-datepicker-pro-dropdown-panel\" [ngStyle]=\"{ paddingBottom: !showTime && !footerTemplate ? 0 : '4px' }\" dDropDownMenu>\n      <d-datepicker-panel\n        [isRangeType]=\"false\"\n        [showTime]=\"showTime\"\n        [customTemplate]=\"customTemplate\"\n        [footerTemplate]=\"footerTemplate\"\n        [mode]=\"mode\"\n      >\n      </d-datepicker-panel>\n    </div>\n  </div>\n</ng-template>\n\n<ng-template #default>\n  <div\n    class=\"devui-single-picker devui-input-group devui-dropdown-origin\"\n    [ngClass]=\"{\n      'devui-disabled': disabled,\n      'devui-has-value': dateValue\n    }\"\n  >\n    <i class=\"devui-single-picker-icon\">\n      <svg width=\"16px\" height=\"16px\" viewBox=\"0 0 16 16\" version=\"1.1\">\n        <g stroke=\"none\" stroke-width=\"1\" fill=\"none\" fill-rule=\"evenodd\">\n          <path\n            d=\"M5,0 L5,1 L11,1 L11,0 L12,0 L12,1 L14,1 C15.1045695,1 16,1.8954305 16,3 L16,14 C16,15.1045695 15.1045695,16 14,16 L2,16 C0.8954305,16 0,15.1045695 0,14 L0,3 C0,1.8954305 0.8954305,1 2,1 L4,1 L4,0 L5,0 Z M15,5 L1,5 L1,14 C1,14.5522847 1.44771525,15 2,15 L14,15 C14.5522847,15 15,14.5522847 15,14 L15,5 Z M6,7 C6.55228475,7 7,7.44771525 7,8 L7,10 C7,10.5522847 6.55228475,11 6,11 L4,11 C3.44771525,11 3,10.5522847 3,10 L3,8 C3,7.44771525 3.44771525,7 4,7 L6,7 Z M6,8 L4,8 L4,10 L6,10 L6,8 Z M4,2 L2,2 C1.44771525,2 1,2.44771525 1,3 L1,4 L15,4 L15,3 C15,2.44771525 14.5522847,2 14,2 L12,2 L12,3 L11,3 L11,2 L5,2 L5,3 L4,3 L4,2 Z\"\n            fill=\"#BABBC0\"\n          ></path>\n        </g>\n      </svg>\n    </i>\n    <input\n      #dateInput\n      class=\"devui-input devui-form-control\"\n      [disabled]=\"disabled\"\n      [placeholder]=\"placeholder || i18nText?.placeholder\"\n      [(ngModel)]=\"dateValue\"\n      autocomplete=\"off\"\n    />\n    <div *ngIf=\"allowClear\" class=\"devui-input-group-addon close-icon-wrapper\" (click)=\"clear($event, true)\">\n      <div class=\"close-icon\">\n        <svg\n          width=\"16px\"\n          height=\"16px\"\n          viewBox=\"0 0 16 16\"\n          version=\"1.1\"\n          xmlns=\"http://www.w3.org/2000/svg\"\n          xmlns:xlink=\"http://www.w3.org/1999/xlink\"\n        >\n          <g stroke=\"none\" stroke-width=\"1\" fill=\"none\" fill-rule=\"evenodd\">\n            <path\n              d=\"M8,1 C11.8659932,1 15,4.13400675 15,8 C15,11.8659932 11.8659932,15 8,15 C4.13400675,15 1,11.8659932 1,8 C1,4.13400675 4.13400675,1 8,1 Z M5.87867966,5.17157288 C5.68341751,4.97631073 5.36683502,4.97631073 5.17157288,5.17157288 C4.99800652,5.34513923 4.97872137,5.61456363 5.11371742,5.80943177 L5.17157288,5.87867966 L7.29218611,7.99929289 L5.17157288,10.1213203 C4.97631073,10.3165825 4.97631073,10.633165 5.17157288,10.8284271 C5.34513923,11.0019935 5.61456363,11.0212786 5.80943177,10.8862826 L5.87867966,10.8284271 L7.99929289,8.70639967 L10.1213203,10.8284271 C10.3165825,11.0236893 10.633165,11.0236893 10.8284271,10.8284271 C11.0019935,10.6548608 11.0212786,10.3854364 10.8862826,10.1905682 L10.8284271,10.1213203 L8.70710678,8 L10.8284271,5.87867966 C11.0236893,5.68341751 11.0236893,5.36683502 10.8284271,5.17157288 C10.6548608,4.99800652 10.3854364,4.97872137 10.1905682,5.11371742 L10.1213203,5.17157288 L8,7.29289322 L5.87867966,5.17157288 Z\"\n              fill-rule=\"nonzero\"\n            ></path>\n          </g>\n        </svg>\n      </div>\n    </div>\n  </div>\n</ng-template>\n", styles: [".devui-datepicker-pro-wrapper{display:inline-block;background-color:#fff;background-color:var(--devui-base-bg, #ffffff)}.devui-datepicker-pro-wrapper .devui-single-picker{padding-left:8px;position:relative;box-sizing:border-box;width:100%;min-height:24px}.devui-datepicker-pro-wrapper .devui-single-picker .devui-input{padding:4px 8px;width:calc(100% - 16px)}.devui-datepicker-pro-wrapper .devui-single-picker-icon{vertical-align:bottom}.devui-datepicker-pro-wrapper .devui-single-picker .close-icon-wrapper{padding:0 8px;vertical-align:baseline;visibility:hidden}.devui-datepicker-pro-wrapper .devui-single-picker .close-icon-wrapper .close-icon{width:16px;height:16px;display:inline-block;line-height:12px}.devui-datepicker-pro-wrapper .devui-single-picker .close-icon-wrapper .close-icon svg{vertical-align:middle;position:relative;top:-.1em}.devui-datepicker-pro-wrapper .devui-single-picker .close-icon-wrapper .close-icon svg path{fill:#d7d8da;fill:var(--devui-shape-icon-fill, #d7d8da)}.devui-datepicker-pro-wrapper .devui-single-picker .close-icon-wrapper .close-icon:hover svg path{fill:#babbc0;fill:var(--devui-shape-icon-fill-active, #babbc0)}.devui-datepicker-pro-wrapper .devui-single-picker.devui-has-value:hover:not(.devui-disabled) .close-icon-wrapper{visibility:visible}.devui-datepicker-pro-wrapper .devui-single-picker:not(.devui-disabled) .close-icon-wrapper{cursor:pointer}.devui-datepicker-pro-wrapper .devui-input-group>.devui-input{display:inline-block}.devui-datepicker-pro-wrapper .devui-input-group>.devui-input::-ms-clear{display:none}\n"] }]
        }], ctorParameters: function () { return [{ type: i1.I18nService }, { type: i2.DatepickerProService }]; }, propDecorators: { mode: [{
                type: Input
            }], showTime: [{
                type: Input
            }], disabled: [{
                type: Input
            }], autoOpen: [{
                type: Input
            }], format: [{
                type: Input
            }], cssClass: [{
                type: Input
            }], showAnimation: [{
                type: Input
            }], appendToBody: [{
                type: Input
            }], width: [{
                type: Input
            }], placeholder: [{
                type: Input
            }], allowClear: [{
                type: Input
            }], dropdownToggle: [{
                type: Output
            }], confirmEvent: [{
                type: Output
            }], calenderRange: [{
                type: Input
            }], minDate: [{
                type: Input
            }], maxDate: [{
                type: Input
            }], markedRangeDateList: [{
                type: Input
            }], markedDateList: [{
                type: Input
            }], customTemplate: [{
                type: ContentChild,
                args: ['customTemplate']
            }], footerTemplate: [{
                type: ContentChild,
                args: ['footerTemplate']
            }], hostTemplate: [{
                type: ContentChild,
                args: ['hostTemplate']
            }], markDateInfoTemplate: [{
                type: ContentChild,
                args: ['markDateInfoTemplate']
            }], datepickerInput: [{
                type: ViewChild,
                args: ['dateInput']
            }] } });
