// ng-devui/esm2020/dragdrop/directives/drag-preview.directive.mjs
import { ComponentFactoryResolver, Directive, Input, TemplateRef } from '@angular/core';
import { OverlayContainerRef } from 'ng-devui/overlay-container';
import { DragDropService } from '../services/drag-drop.service';
import { DragPreviewComponent } from './drag-preview.component';
import * as i0 from "@angular/core";
import * as i1 from "ng-devui/overlay-container";
import * as i2 from "../services/drag-drop.service";
export class DragPreviewDirective {
    constructor(componentFactoryResolver, overlayContainerRef, dragDropService) {
        this.componentFactoryResolver = componentFactoryResolver;
        this.overlayContainerRef = overlayContainerRef;
        this.dragDropService = dragDropService;
        this.dragPreviewOptions = {
            skipBatchPreview: false
        };
    }
    createPreview() {
        const finalComponentFactoryResolver = this.componentFactoryResolver;
        const previewRef = this.overlayContainerRef.createComponent(finalComponentFactoryResolver.resolveComponentFactory(DragPreviewComponent));
        this.previewRef = previewRef;
        this.updateData();
        return this.previewRef;
    }
    updateData() {
        Object.assign(this.previewRef.instance, {
            templateRef: this.dragPreviewTemplate,
            data: this.dragPreviewData,
            draggedEl: this.dragDropService.draggedEl,
            dragData: this.dragDropService.dragData,
            batchDragData: this.dragDropService.batchDragData && this.dragDropService.getBatchDragData(),
            dragSyncDOMElements: this.dragDropService.dragSyncGroupDirectives && this.getDragSyncDOMElements()
        });
        this.previewRef.instance.updateTemplate();
    }
    destroyPreview() {
        if (this.previewRef) {
            this.previewRef.hostView.destroy();
        }
    }
    getPreviewElement() {
        return this.previewRef && this.previewRef.instance.element;
    }
    getDragSyncDOMElements() {
        return this.dragDropService.dragSyncGroupDirectives.map(dir => dir.el.nativeElement);
    }
}
DragPreviewDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DragPreviewDirective, deps: [{ token: i0.ComponentFactoryResolver }, { token: i1.OverlayContainerRef }, { token: i2.DragDropService }], target: i0.ɵɵFactoryTarget.Directive });
DragPreviewDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: DragPreviewDirective, selector: "[dDraggable][dDragPreview]", inputs: { dragPreviewTemplate: ["dDragPreview", "dragPreviewTemplate"], dragPreviewData: "dragPreviewData", dragPreviewOptions: "dragPreviewOptions" }, exportAs: ["dDragPreview"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DragPreviewDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[dDraggable][dDragPreview]',
                    exportAs: 'dDragPreview'
                }]
        }], ctorParameters: function () { return [{ type: i0.ComponentFactoryResolver }, { type: i1.OverlayContainerRef }, { type: i2.DragDropService }]; }, propDecorators: { dragPreviewTemplate: [{
                type: Input,
                args: ['dDragPreview']
            }], dragPreviewData: [{
                type: Input
            }], dragPreviewOptions: [{
                type: Input
            }] } });
