// ng-devui/esm2020/dragdrop/directives/dragdrop-sync-box.directive.mjs
import { Directive } from '@angular/core';
import { Subscription } from 'rxjs';
import { DescendantRoot } from '../services/drag-drop-desc-reg.service';
import { DragSyncDescendantRegisterService, DropSortSyncDescendantRegisterService } from '../services/drag-drop-descendant-sync.service';
import { DragDropSyncService } from '../services/drag-drop-sync.service';
import * as i0 from "@angular/core";
import * as i1 from "../services/drag-drop-sync.service";
import * as i2 from "../services/drag-drop-descendant-sync.service";
export class DragDropSyncBoxDirective {
    constructor(dragDropSyncService, dragSyncDrs, dropSortSyncDrs) {
        this.dragDropSyncService = dragDropSyncService;
        this.dragSyncDrs = dragSyncDrs;
        this.dropSortSyncDrs = dropSortSyncDrs;
        this.sub = new Subscription();
    }
    ngOnInit() {
        this.dragSyncList = new DescendantRoot(this.dragSyncDrs);
        this.dropSyncList = new DescendantRoot(this.dropSortSyncDrs);
    }
    ngAfterViewInit() {
        this.dragSyncList.on();
        this.dropSyncList.on();
        this.dragDropSyncService.updateDragSyncList(this.dragSyncList);
        this.dragDropSyncService.updateDropSyncList(this.dropSyncList);
        this.sub.add(this.dragSyncList.changes.subscribe(list => this.dragDropSyncService.updateDragSyncList(list))).add(this.dropSyncList.changes.subscribe(list => this.dragDropSyncService.updateDropSyncList(list)));
    }
    ngOnDestroy() {
        if (this.sub) {
            this.sub.unsubscribe();
        }
        this.dragSyncList.off();
        this.dropSyncList.off();
    }
}
DragDropSyncBoxDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DragDropSyncBoxDirective, deps: [{ token: i1.DragDropSyncService }, { token: i2.DragSyncDescendantRegisterService }, { token: i2.DropSortSyncDescendantRegisterService }], target: i0.ɵɵFactoryTarget.Directive });
DragDropSyncBoxDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: DragDropSyncBoxDirective, selector: "[dDragDropSyncBox]", providers: [
        DragDropSyncService,
        DragSyncDescendantRegisterService,
        DropSortSyncDescendantRegisterService,
    ], exportAs: ["dDragDropSyncBox"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DragDropSyncBoxDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[dDragDropSyncBox]',
                    exportAs: 'dDragDropSyncBox',
                    providers: [
                        DragDropSyncService,
                        DragSyncDescendantRegisterService,
                        DropSortSyncDescendantRegisterService,
                    ]
                }]
        }], ctorParameters: function () { return [{ type: i1.DragDropSyncService }, { type: i2.DragSyncDescendantRegisterService }, { type: i2.DropSortSyncDescendantRegisterService }]; } });
