// ng-devui/esm2020/dragdrop/directives/drag-sync.directive.mjs
import { Directive, ElementRef, Input, Optional, Self } from '@angular/core';
import { Subscription } from 'rxjs';
import { DescendantChildren } from '../services/drag-drop-desc-reg.service';
import { DragSyncDescendantRegisterService } from '../services/drag-drop-descendant-sync.service';
import { DragDropSyncService } from '../services/drag-drop-sync.service';
import { DragDropService } from '../services/drag-drop.service';
import { DraggableDirective } from './draggable.directive';
import * as i0 from "@angular/core";
import * as i1 from "./draggable.directive";
import * as i2 from "../services/drag-drop-sync.service";
import * as i3 from "../services/drag-drop.service";
import * as i4 from "../services/drag-drop-descendant-sync.service";
export class DragSyncDirective extends DescendantChildren {
    constructor(el, draggable, dragDropSyncService, dragDropService, dragSyncDrs) {
        super(dragSyncDrs);
        this.el = el;
        this.draggable = draggable;
        this.dragDropSyncService = dragDropSyncService;
        this.dragDropService = dragDropService;
        this.dragSyncDrs = dragSyncDrs;
        this.dragSyncGroup = '';
        this.subscription = new Subscription();
        this.subDragElEvent = (bool) => {
            this.syncGroupDirectives.forEach(dir => this.renderDisplay(dir.el.nativeElement, bool));
        };
        this.descendantItem = this;
    }
    ngOnInit() {
        if (this.draggable) {
            this.subscription.add(this.draggable.dragElShowHideEvent.subscribe(this.subDragElEvent));
            this.subscription.add(this.draggable.beforeDragStartEvent.subscribe(() => {
                this.syncGroupDirectives = this.dragDropSyncService.getDragSyncByGroup(this.dragSyncGroup).filter(directive => directive !== this);
                this.dragDropService.dragSyncGroupDirectives = this.syncGroupDirectives;
            }));
            this.subscription.add(this.draggable.dropEndEvent.subscribe(() => {
                this.dragDropService.dragSyncGroupDirectives = undefined;
                this.syncGroupDirectives = undefined;
            }));
        }
        super.ngOnInit();
    }
    ngOnDestroy() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
        super.ngOnDestroy();
    }
    renderDisplay(nativeEl, bool) {
        nativeEl.style.display = bool ? '' : 'none';
    }
}
DragSyncDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DragSyncDirective, deps: [{ token: i0.ElementRef }, { token: i1.DraggableDirective, optional: true, self: true }, { token: i2.DragDropSyncService }, { token: i3.DragDropService }, { token: i4.DragSyncDescendantRegisterService }], target: i0.ɵɵFactoryTarget.Directive });
DragSyncDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: DragSyncDirective, selector: "[dDragSync]", inputs: { dragSyncGroup: ["dDragSync", "dragSyncGroup"] }, exportAs: ["dDragSync"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DragSyncDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[dDragSync]',
                    exportAs: 'dDragSync'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i1.DraggableDirective, decorators: [{
                    type: Optional
                }, {
                    type: Self
                }] }, { type: i2.DragDropSyncService }, { type: i3.DragDropService }, { type: i4.DragSyncDescendantRegisterService }]; }, propDecorators: { dragSyncGroup: [{
                type: Input,
                args: ['dDragSync']
            }] } });
