// ng-devui/esm2020/dragdrop/directives/drag-preview.component.mjs
import { ChangeDetectorRef, Component, ElementRef } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export class DragPreviewComponent {
    constructor(el, cdr) {
        this.el = el;
        this.cdr = cdr;
        this.element = el.nativeElement;
    }
    updateTemplate() {
        this.cdr.detectChanges();
    }
}
DragPreviewComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DragPreviewComponent, deps: [{ token: i0.ElementRef }, { token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Component });
DragPreviewComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: DragPreviewComponent, selector: "d-drag-preview", ngImport: i0, template: "<ng-template\n  [ngTemplateOutlet]=\"templateRef\"\n  [ngTemplateOutletContext]=\"{\n    data: data,\n    draggedEl: draggedEl,\n    dragData: dragData,\n    batchDragData: batchDragData,\n    dragSyncDOMElements: dragSyncDOMElements\n  }\"\n>\n</ng-template>\n", directives: [{ type: i1.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DragPreviewComponent, decorators: [{
            type: Component,
            args: [{ selector: 'd-drag-preview', preserveWhitespaces: false, template: "<ng-template\n  [ngTemplateOutlet]=\"templateRef\"\n  [ngTemplateOutletContext]=\"{\n    data: data,\n    draggedEl: draggedEl,\n    dragData: dragData,\n    batchDragData: batchDragData,\n    dragSyncDOMElements: dragSyncDOMElements\n  }\"\n>\n</ng-template>\n" }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.ChangeDetectorRef }]; } });
