// ng-devui/esm2020/dragdrop/directives/drag-preivew-clone-from-domRef.component.mjs
import { ChangeDetectorRef, Component, ElementRef, Input } from '@angular/core';
import { DragDropService } from '../services/drag-drop.service';
import * as i0 from "@angular/core";
import * as i1 from "../services/drag-drop.service";
export class DragPreviewCloneDomRefComponent {
    constructor(el, cdr, dragDropService) {
        this.el = el;
        this.cdr = cdr;
        this.dragDropService = dragDropService;
        this.copyStyle = true;
    }
    ngAfterViewInit() {
        if (!this.cloneNode) {
            this.createView();
        }
    }
    ngOnChanges(changes) {
        if (changes['domRef']) {
            if (this.cloneNode) {
                this.destroyView();
                this.createView();
            }
            else {
                this.createView();
            }
        }
    }
    ngOnDestroy() {
        if (this.cloneNode) {
            this.destroyView();
        }
    }
    createView() {
        if (this.domRef) {
            this.cloneNode = this.domRef.cloneNode(true);
            if (this.copyStyle) {
                this.dragDropService.copyStyle(this.domRef, this.cloneNode);
            }
            this.el.nativeElement.appendChild(this.cloneNode);
        }
    }
    destroyView() {
        if (this.cloneNode) {
            if (this.el.nativeElement.contains(this.cloneNode)) {
                this.el.nativeElement.removeChild(this.cloneNode);
            }
            this.cloneNode = undefined;
        }
    }
    updateTemplate() {
        this.cdr.detectChanges();
    }
}
DragPreviewCloneDomRefComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DragPreviewCloneDomRefComponent, deps: [{ token: i0.ElementRef }, { token: i0.ChangeDetectorRef }, { token: i1.DragDropService }], target: i0.ɵɵFactoryTarget.Component });
DragPreviewCloneDomRefComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: DragPreviewCloneDomRefComponent, selector: "d-drag-preview-clone-dom-ref", inputs: { domRef: "domRef", copyStyle: "copyStyle" }, usesOnChanges: true, ngImport: i0, template: '', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DragPreviewCloneDomRefComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'd-drag-preview-clone-dom-ref',
                    template: '',
                    preserveWhitespaces: false,
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.ChangeDetectorRef }, { type: i1.DragDropService }]; }, propDecorators: { domRef: [{
                type: Input
            }], copyStyle: [{
                type: Input
            }] } });
