// ng-devui/esm2020/dragdrop/directives/drop-scroll-enhance-side.directive.mjs
import { Directive, Input } from '@angular/core';
import { DropScrollEnhancedDirective } from './drop-scroll-enhance.directive';
import * as i0 from "@angular/core";
export class DropScrollEnhancedSideDirective extends DropScrollEnhancedDirective {
    constructor() {
        super(...arguments);
        this.sideDirection = 'v';
    }
    ngOnInit() {
        this.direction = this.sideDirection === 'v' ? 'h' : 'v';
    }
}
DropScrollEnhancedSideDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DropScrollEnhancedSideDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
DropScrollEnhancedSideDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: DropScrollEnhancedSideDirective, selector: "[dDropScrollEnhancedSide]", inputs: { sideDirection: ["direction", "sideDirection"] }, exportAs: ["dDropScrollEnhancedSide"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DropScrollEnhancedSideDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[dDropScrollEnhancedSide]',
                    exportAs: 'dDropScrollEnhancedSide'
                }]
        }], propDecorators: { sideDirection: [{
                type: Input,
                args: ['direction']
            }] } });
