// ng-devui/esm2020/dragdrop/directives/sortable.directive.mjs
import { Directive, ElementRef, HostBinding, Input } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Makes an element draggable by adding the draggable html attribute
 */
export class SortableDirective {
    constructor(el) {
        this.el = el;
        this.dSortDirection = 'v';
        this.dSortableZMode = false;
        this.dSortable = true;
    }
}
SortableDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: SortableDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
SortableDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: SortableDirective, selector: "[dSortable]", inputs: { dSortDirection: ["dSortable", "dSortDirection"], dSortableZMode: "dSortableZMode" }, host: { properties: { "attr.dsortable": "this.dSortDirection", "attr.d-sortable-zmode": "this.dSortableZMode", "attr.d-sortable": "this.dSortable" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: SortableDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[dSortable]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; }, propDecorators: { dSortDirection: [{
                type: HostBinding,
                args: ['attr.dsortable']
            }, {
                type: Input,
                args: ['dSortable']
            }], dSortableZMode: [{
                type: HostBinding,
                args: ['attr.d-sortable-zmode']
            }, {
                type: Input
            }], dSortable: [{
                type: HostBinding,
                args: ['attr.d-sortable']
            }] } });
