// ng-devui/esm2020/layout/class.directive.mjs
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { DBreakpoints } from './layout.types';
import { DScreenMediaQueryService } from './screen-media-query.service';
import * as i0 from "@angular/core";
import * as i1 from "./screen-media-query.service";
export class DClassDirective {
    constructor(elementRef, renderer, screenQueryService) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.screenQueryService = screenQueryService;
        this.destroy$ = new Subject();
        this.executedClassList = [];
    }
    ngOnInit() {
        this.screenQueryService.getPoint()
            .pipe(takeUntil(this.destroy$))
            .subscribe(({ currentPoint }) => {
            this.updateClass(currentPoint);
        });
    }
    /* TODO: 参数需要优化，不仅可以设置断点，还可以设置通用 */
    updateClass(currentPoint) {
        let finalClassList = [];
        if (!Array.isArray(this.dClass)) {
            for (const point of DBreakpoints) {
                if (this.dClass[point]) {
                    finalClassList = [...finalClassList, ...this.dClass[point]];
                }
                if (currentPoint === point) {
                    break;
                }
            }
        }
        else if (this.dClass) {
            finalClassList = [...this.dClass];
        }
        this.executedClassList.forEach(className => {
            this.renderer.removeClass(this.elementRef.nativeElement, className);
        });
        finalClassList.forEach(className => {
            this.renderer.addClass(this.elementRef.nativeElement, className);
        });
        this.executedClassList = finalClassList;
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
}
DClassDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DClassDirective, deps: [{ token: i0.ElementRef }, { token: i0.Renderer2 }, { token: i1.DScreenMediaQueryService }], target: i0.ɵɵFactoryTarget.Directive });
DClassDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: DClassDirective, selector: "[dClass]", inputs: { dClass: "dClass" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DClassDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: `[dClass]`,
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.Renderer2 }, { type: i1.DScreenMediaQueryService }]; }, propDecorators: { dClass: [{
                type: Input
            }] } });
