// ng-devui/esm2020/layout/flex.directive.mjs
import { Directive, ElementRef, HostBinding, Input, Renderer2 } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { updateClassList } from './layout-utils';
import { DBreakpoints } from './layout.types';
import { DScreenMediaQueryService } from './screen-media-query.service';
import * as i0 from "@angular/core";
import * as i1 from "./screen-media-query.service";
export class DFlexDirective {
    constructor(elementRef, renderer, screenQueryService) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.screenQueryService = screenQueryService;
        this.destroy$ = new Subject();
    }
    get flexRow() {
        return this.dFlexContainer === 'row';
    }
    get flexColumn() {
        return this.dFlexContainer === 'column';
    }
    get flex() {
        return !!this.dFlexContainer;
    }
    parseFlex(flex) {
        if (typeof flex === 'number') {
            return `${flex}`;
        }
        else if (typeof flex === 'string') {
            if (/^\d+(\.\d+)?(px|em|rem|%)$/.test(flex)) {
                return `0 0 ${flex}`;
            }
        }
        return flex;
    }
    updateFlex(currentPoint) {
        let finalFlex = '';
        if (this.dFlex) {
            if (typeof this.dFlex === 'object') {
                for (const point of DBreakpoints) {
                    if (this.dFlex[point]) {
                        finalFlex = this.dFlex[point];
                    }
                    if (currentPoint === point) {
                        break;
                    }
                }
            }
            else {
                finalFlex = this.dFlex;
            }
        }
        this.renderer.setStyle(this.elementRef.nativeElement, 'flex', this.parseFlex(finalFlex));
    }
    ngOnInit() {
        this.screenQueryService.getPoint()
            .pipe(takeUntil(this.destroy$))
            .subscribe(({ currentPoint }) => {
            this.updateFlex(currentPoint);
        });
        updateClassList(this, this.elementRef, this.renderer);
    }
    ngOnChanges() {
        updateClassList(this, this.elementRef, this.renderer);
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
}
DFlexDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DFlexDirective, deps: [{ token: i0.ElementRef }, { token: i0.Renderer2 }, { token: i1.DScreenMediaQueryService }], target: i0.ɵɵFactoryTarget.Directive });
DFlexDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: DFlexDirective, selector: "[dFlex], d-row, d-col", inputs: { dFlex: "dFlex", dFlexContainer: "dFlexContainer", dOrder: "dOrder", dAlign: "dAlign", dAlignSelf: "dAlignSelf", dJustify: "dJustify", dFlexWrap: "dFlexWrap" }, host: { properties: { "class.dl-flex-row": "this.flexRow", "class.dl-flex-column": "this.flexColumn", "class.dl-d-flex": "this.flex" } }, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DFlexDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: `[dFlex], d-row, d-col`,
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.Renderer2 }, { type: i1.DScreenMediaQueryService }]; }, propDecorators: { flexRow: [{
                type: HostBinding,
                args: ['class.dl-flex-row']
            }], flexColumn: [{
                type: HostBinding,
                args: ['class.dl-flex-column']
            }], flex: [{
                type: HostBinding,
                args: ['class.dl-d-flex']
            }], dFlex: [{
                type: Input
            }], dFlexContainer: [{
                type: Input
            }], dOrder: [{
                type: Input
            }], dAlign: [{
                type: Input
            }], dAlignSelf: [{
                type: Input
            }], dJustify: [{
                type: Input
            }], dFlexWrap: [{
                type: Input
            }] } });
