// ng-devui/esm2020/layout/layout.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { DClassDirective } from './class.directive';
import { DColComponent } from './col.component';
import { AsideComponent } from './elements/aside.component';
import { ContentComponent } from './elements/content.component';
import { FooterComponent } from './elements/footer.component';
import { HeaderComponent } from './elements/header.component';
import { DFlexDirective } from './flex.directive';
import { DGutterDirective } from './gutter.directive';
import { LayoutComponent } from './layout.component';
import { DRowComponent } from './row.component';
import { DSpaceDirective } from './space.directive';
import { DStyleDirective } from './style.directive';
import * as i0 from "@angular/core";
export class LayoutModule {
}
LayoutModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: LayoutModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
LayoutModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: LayoutModule, declarations: [LayoutComponent,
        HeaderComponent,
        ContentComponent,
        FooterComponent,
        AsideComponent,
        DClassDirective,
        DStyleDirective,
        DFlexDirective,
        DRowComponent,
        DColComponent,
        DSpaceDirective,
        DGutterDirective], imports: [CommonModule], exports: [LayoutComponent,
        HeaderComponent,
        ContentComponent,
        FooterComponent,
        AsideComponent,
        DClassDirective,
        DStyleDirective,
        DFlexDirective,
        DRowComponent,
        DColComponent,
        DSpaceDirective,
        DGutterDirective] });
LayoutModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: LayoutModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: LayoutModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        LayoutComponent,
                        HeaderComponent,
                        ContentComponent,
                        FooterComponent,
                        AsideComponent,
                        DClassDirective,
                        DStyleDirective,
                        DFlexDirective,
                        DRowComponent,
                        DColComponent,
                        DSpaceDirective,
                        DGutterDirective,
                    ],
                    exports: [
                        LayoutComponent,
                        HeaderComponent,
                        ContentComponent,
                        FooterComponent,
                        AsideComponent,
                        DClassDirective,
                        DStyleDirective,
                        DFlexDirective,
                        DRowComponent,
                        DColComponent,
                        DSpaceDirective,
                        DGutterDirective,
                    ],
                    imports: [CommonModule],
                }]
        }] });
