// ng-devui/esm2020/layout/screen-media-query.service.mjs
import { Injectable } from '@angular/core';
import { fromEvent, ReplaySubject, Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { DBreakpoints, DBreakpointsMap } from './layout.types';
import * as i0 from "@angular/core";
export class DScreenMediaQueryService {
    constructor() {
        this.pointChangeSub = new ReplaySubject(1);
        this.destroy$ = new Subject();
    }
    // 可以传入一个基准point，返回数据结构{ currentPoint, 变大or变小or没变，比基准point大or小or一样 }
    getPoint() {
        if (!this.currentPoint) {
            this.currentPoint = this.getCurrentPoint();
            this.pointChangeSub.next({
                currentPoint: this.currentPoint,
                change: 0,
                compare: this.comparePoints(this.currentPoint),
            });
            fromEvent(window, 'resize')
                .pipe(takeUntil(this.destroy$))
                .subscribe(() => {
                const tempPoint = this.getCurrentPoint();
                if (this.currentPoint !== tempPoint) {
                    const change = this.comparePoints(tempPoint, this.currentPoint);
                    this.currentPoint = tempPoint;
                    this.pointChangeSub.next({
                        currentPoint: this.currentPoint,
                        change: change,
                        compare: this.comparePoints(tempPoint),
                    });
                }
            });
        }
        return this.pointChangeSub;
    }
    // 无p2，则全量对比
    comparePoints(p1, p2) {
        let index1;
        let index2;
        for (let i = 0; i < DBreakpoints.length; i++) {
            if (p1 === DBreakpoints[i]) {
                index1 = i;
            }
            if (p2 === DBreakpoints[i]) {
                index2 = i;
            }
        }
        if (!p2) {
            const res = {};
            DBreakpoints.forEach((point, index) => {
                res[point] = index1 - index;
            });
            return res;
        }
        return index1 - index2;
    }
    getCurrentPoint() {
        const currentScreenWidth = window.innerWidth;
        for (let i = 0; i < DBreakpoints.length; i++) {
            if (DBreakpointsMap[DBreakpoints[i]] >= currentScreenWidth || i === DBreakpoints.length - 1) {
                return DBreakpoints[i];
            }
        }
        return;
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
}
DScreenMediaQueryService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DScreenMediaQueryService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
DScreenMediaQueryService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DScreenMediaQueryService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DScreenMediaQueryService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
