// ng-devui/esm2020/layout/space.directive.mjs
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { DBreakpoints } from './layout.types';
import { DScreenMediaQueryService } from './screen-media-query.service';
import * as i0 from "@angular/core";
import * as i1 from "./screen-media-query.service";
export class DSpaceDirective {
    constructor(elementRef, renderer, screenQueryService) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.screenQueryService = screenQueryService;
        this.destroy$ = new Subject();
        this.executedSpace = [null, null];
    }
    ngOnInit() {
        this.screenQueryService.getPoint()
            .pipe(takeUntil(this.destroy$))
            .subscribe(({ currentPoint }) => {
            this.updateSpace(currentPoint);
        });
    }
    getCurrentSpace(currentPoint) {
        let finalSpace = [null, null];
        if (Array.isArray(this.dSpace) || typeof this.dSpace === 'number') {
            finalSpace = this._transSpaceToArray(this.dSpace);
        }
        else {
            for (const point of DBreakpoints) {
                if (this.dSpace[point] !== undefined) {
                    finalSpace = this._transSpaceToArray(this.dSpace[point]);
                }
                if (currentPoint === point) {
                    break;
                }
            }
        }
        return finalSpace;
    }
    _transSpaceToArray(space) {
        let finalSpace = space;
        if (typeof space === 'number') {
            finalSpace = [space, null];
        }
        if (this.dSpaceDirection === 'horizontal') {
            finalSpace = finalSpace.reverse();
        }
        return finalSpace;
    }
    updateSpace(currentPoint) {
        const finalSpace = this.getCurrentSpace(currentPoint);
        const items = this.elementRef.nativeElement.children;
        for (let i = 0; i < items.length - 1; i++) {
            if (finalSpace[0] !== null) {
                this.renderer.setStyle(items[i], 'margin-bottom', finalSpace[0] + 'px');
            }
            else if (this.executedSpace[0] !== null) {
                this.renderer.removeStyle(items[i], 'margin-bottom');
            }
            if (finalSpace[1] !== null) {
                this.renderer.setStyle(items[i], 'margin-right', finalSpace[1] + 'px');
            }
            else if (this.executedSpace[1] !== null) {
                this.renderer.removeStyle(items[i], 'margin-right');
            }
        }
        this.executedSpace = finalSpace;
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
}
DSpaceDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DSpaceDirective, deps: [{ token: i0.ElementRef }, { token: i0.Renderer2 }, { token: i1.DScreenMediaQueryService }], target: i0.ɵɵFactoryTarget.Directive });
DSpaceDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: DSpaceDirective, selector: "[dSpace]", inputs: { dSpace: "dSpace", dSpaceDirection: "dSpaceDirection" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DSpaceDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: `[dSpace]`,
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.Renderer2 }, { type: i1.DScreenMediaQueryService }]; }, propDecorators: { dSpace: [{
                type: Input
            }], dSpaceDirection: [{
                type: Input
            }] } });
