// ng-devui/esm2020/layout/style.directive.mjs
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { DBreakpoints } from './layout.types';
import { DScreenMediaQueryService } from './screen-media-query.service';
import * as i0 from "@angular/core";
import * as i1 from "./screen-media-query.service";
export class DStyleDirective {
    constructor(elementRef, renderer, screenQueryService) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.screenQueryService = screenQueryService;
        this.destroy$ = new Subject();
        this.styleObject = {};
    }
    ngOnInit() {
        this.screenQueryService.getPoint()
            .pipe(takeUntil(this.destroy$))
            .subscribe(({ currentPoint }) => {
            this.updateStyle(currentPoint);
        });
    }
    updateStyle(currentPoint) {
        let finalStyleObject = {};
        let isResponse = false;
        if (this.dStyle) {
            for (const point of DBreakpoints) {
                if (this.dStyle[point]) {
                    finalStyleObject = { ...finalStyleObject, ...this.dStyle[point] };
                    isResponse = true;
                }
                if (currentPoint === point) {
                    break;
                }
            }
            if (!isResponse) {
                finalStyleObject = { ...this.dStyle };
            }
        }
        Object.keys(this.styleObject).forEach(key => {
            this.renderer.removeStyle(this.elementRef.nativeElement, key);
        });
        Object.keys(finalStyleObject).forEach(key => {
            this.renderer.setStyle(this.elementRef.nativeElement, key, finalStyleObject[key]);
        });
        this.styleObject = finalStyleObject;
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
}
DStyleDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DStyleDirective, deps: [{ token: i0.ElementRef }, { token: i0.Renderer2 }, { token: i1.DScreenMediaQueryService }], target: i0.ɵɵFactoryTarget.Directive });
DStyleDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: DStyleDirective, selector: "[dStyle]", inputs: { dStyle: "dStyle" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DStyleDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: `[dStyle]`,
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.Renderer2 }, { type: i1.DScreenMediaQueryService }]; }, propDecorators: { dStyle: [{
                type: Input
            }] } });
