// ng-devui/esm2020/accordion/accordion.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { AccordionItemHreflinkComponent } from './accordion-item-hreflink.component';
import { AccordionItemRouterlinkComponent } from './accordion-item-routerlink.component';
import { AccordionItemComponent } from './accordion-item.component';
import { AccordionListComponent } from './accordion-list.component';
import { AccordionMenuComponent } from './accordion-menu.component';
import { AccordionComponent } from './accordion.component';
import * as i0 from "@angular/core";
export class AccordionModule {
}
AccordionModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AccordionModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AccordionModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AccordionModule, declarations: [AccordionComponent,
        AccordionListComponent,
        AccordionMenuComponent,
        AccordionItemComponent,
        AccordionItemHreflinkComponent,
        AccordionItemRouterlinkComponent], imports: [CommonModule,
        RouterModule], exports: [AccordionComponent,
        AccordionListComponent,
        AccordionMenuComponent,
        AccordionItemComponent,
        AccordionItemHreflinkComponent,
        AccordionItemRouterlinkComponent] });
AccordionModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AccordionModule, imports: [[
            CommonModule,
            RouterModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AccordionModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        RouterModule,
                    ],
                    declarations: [
                        AccordionComponent,
                        AccordionListComponent,
                        AccordionMenuComponent,
                        AccordionItemComponent,
                        AccordionItemHreflinkComponent,
                        AccordionItemRouterlinkComponent,
                    ],
                    exports: [
                        AccordionComponent,
                        AccordionListComponent,
                        AccordionMenuComponent,
                        AccordionItemComponent,
                        AccordionItemHreflinkComponent,
                        AccordionItemRouterlinkComponent,
                    ],
                }]
        }] });
