// ng-devui/esm2020/accordion/accordion-menu.component.mjs
import { Component, forwardRef, HostBinding, Inject, ViewEncapsulation } from '@angular/core';
import { AccordionBaseComponent } from './accordion-base-component.class';
import { ACCORDION_MENU } from './accordion-menu-token';
import { ACCORDION } from './accordion-token';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export class AccordionMenuComponent extends AccordionBaseComponent {
    constructor(accordion) {
        super(accordion);
        this.accordion = accordion;
        this.defaultClasses = true;
    }
    get menuItemTemplate() {
        return this.accordion.menuItemTemplate;
    }
    get open() {
        return (this.keyOpen === undefined && this.accordion.autoOpenActiveMenu)
            ? this.childActived
            : this.keyOpen;
    }
    get keyOpen() {
        return this.item && this.item[this.accordion.openKey];
    }
    get children() {
        return this.item && this.item[this.accordion.childrenKey];
    }
    get childActived() {
        return this.routerLinkActived || this.hasActiveChildren;
    }
    get routerLinkActived() {
        return this.accordionListFromView && this.accordionListFromView.routerLinkActived;
    }
    get hasActiveChildren() {
        return this.accordionListFromView && this.accordionListFromView.hasActiveChildren;
    }
    toggle(event) {
        this.accordion.menuToggleFn({
            item: this.item,
            open: !this.open,
            parent: this.parent,
            event: event
        });
    }
}
AccordionMenuComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AccordionMenuComponent, deps: [{ token: ACCORDION }], target: i0.ɵɵFactoryTarget.Component });
AccordionMenuComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: AccordionMenuComponent, selector: "d-accordion-menu", host: { properties: { "class.devui-accordion-menu-item": "this.defaultClasses", "class.open": "this.open", "class.devui-router-active": "this.routerLinkActived", "class.devui-has-active-item": "this.hasActiveChildren" } }, providers: [{
            provide: ACCORDION_MENU,
            useExisting: forwardRef(() => AccordionMenuComponent)
        }], usesInheritance: true, ngImport: i0, template: "<div\n  class=\"devui-accordion-item-title devui-over-flow-ellipsis\"\n  [ngClass]=\"{\n    open: open,\n    active: childActived,\n    disabled: disabled\n  }\"\n  title=\"{{ title }}\"\n  (click)=\"!disabled && toggle($event)\"\n>\n  <div\n    class=\"devui-accordion-splitter\"\n    [ngClass]=\"{\n      'devui-parent-list': deepth === 0\n    }\"\n    [ngStyle]=\"{ left: deepth * 20 + 10 + 'px' }\"\n  ></div>\n  <ng-container *ngIf=\"!menuItemTemplate\">\n    {{ title }}\n  </ng-container>\n  <ng-template\n    *ngIf=\"menuItemTemplate\"\n    [ngTemplateOutlet]=\"menuItemTemplate\"\n    [ngTemplateOutletContext]=\"{\n      parent: parent,\n      item: item,\n      deepth: deepth\n    }\"\n  ></ng-template>\n  <span class=\"devui-accordion-open-icon\">\n    <svg\n      width=\"1em\"\n      height=\"1em\"\n      viewBox=\"0 0 16 16\"\n      version=\"1.1\"\n      xmlns=\"http://www.w3.org/2000/svg\"\n      xmlns:xlink=\"http://www.w3.org/1999/xlink\"\n    >\n      <g stroke=\"none\" stroke-width=\"1\" fill=\"none\" fill-rule=\"evenodd\">\n        <path\n          d=\"M12.1464466,6.85355339 L8.35355339,10.6464466 C8.15829124,10.8417088 7.84170876,10.8417088 7.64644661,10.6464466 L3.85355339,6.85355339 C3.65829124,6.65829124 3.65829124,6.34170876 3.85355339,6.14644661 C3.94732158,6.05267842 4.07449854,6 4.20710678,6 L11.7928932,6 C12.0690356,6 12.2928932,6.22385763 12.2928932,6.5 C12.2928932,6.63260824 12.2402148,6.7597852 12.1464466,6.85355339 Z\"\n          fill-rule=\"nonzero\"\n        ></path>\n      </g>\n    </svg>\n  </span>\n</div>\n", directives: [{ type: i1.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { type: i1.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }, { type: i1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i1.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet"] }], encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AccordionMenuComponent, decorators: [{
            type: Component,
            args: [{ selector: 'd-accordion-menu', encapsulation: ViewEncapsulation.None, preserveWhitespaces: false, providers: [{
                            provide: ACCORDION_MENU,
                            useExisting: forwardRef(() => AccordionMenuComponent)
                        }], template: "<div\n  class=\"devui-accordion-item-title devui-over-flow-ellipsis\"\n  [ngClass]=\"{\n    open: open,\n    active: childActived,\n    disabled: disabled\n  }\"\n  title=\"{{ title }}\"\n  (click)=\"!disabled && toggle($event)\"\n>\n  <div\n    class=\"devui-accordion-splitter\"\n    [ngClass]=\"{\n      'devui-parent-list': deepth === 0\n    }\"\n    [ngStyle]=\"{ left: deepth * 20 + 10 + 'px' }\"\n  ></div>\n  <ng-container *ngIf=\"!menuItemTemplate\">\n    {{ title }}\n  </ng-container>\n  <ng-template\n    *ngIf=\"menuItemTemplate\"\n    [ngTemplateOutlet]=\"menuItemTemplate\"\n    [ngTemplateOutletContext]=\"{\n      parent: parent,\n      item: item,\n      deepth: deepth\n    }\"\n  ></ng-template>\n  <span class=\"devui-accordion-open-icon\">\n    <svg\n      width=\"1em\"\n      height=\"1em\"\n      viewBox=\"0 0 16 16\"\n      version=\"1.1\"\n      xmlns=\"http://www.w3.org/2000/svg\"\n      xmlns:xlink=\"http://www.w3.org/1999/xlink\"\n    >\n      <g stroke=\"none\" stroke-width=\"1\" fill=\"none\" fill-rule=\"evenodd\">\n        <path\n          d=\"M12.1464466,6.85355339 L8.35355339,10.6464466 C8.15829124,10.8417088 7.84170876,10.8417088 7.64644661,10.6464466 L3.85355339,6.85355339 C3.65829124,6.65829124 3.65829124,6.34170876 3.85355339,6.14644661 C3.94732158,6.05267842 4.07449854,6 4.20710678,6 L11.7928932,6 C12.0690356,6 12.2928932,6.22385763 12.2928932,6.5 C12.2928932,6.63260824 12.2402148,6.7597852 12.1464466,6.85355339 Z\"\n          fill-rule=\"nonzero\"\n        ></path>\n      </g>\n    </svg>\n  </span>\n</div>\n" }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [ACCORDION]
                }] }]; }, propDecorators: { defaultClasses: [{
                type: HostBinding,
                args: ['class.devui-accordion-menu-item']
            }], open: [{
                type: HostBinding,
                args: ['class.open']
            }], routerLinkActived: [{
                type: HostBinding,
                args: ['class.devui-router-active']
            }], hasActiveChildren: [{
                type: HostBinding,
                args: ['class.devui-has-active-item']
            }] } });
