// ng-devui/esm2020/accordion/accordion-list.component.mjs
import { Component, forwardRef, Host, HostBinding, Inject, Input, Optional, QueryList, SkipSelf, ViewChildren, ViewEncapsulation } from '@angular/core';
import { expandCollapse, expandCollapseForDomDestroy } from 'ng-devui/utils';
import { AccordionItemRouterlinkComponent } from './accordion-item-routerlink.component';
import { ACCORDION_LIST } from './accordion-list-token';
import { ACCORDION_MENU } from './accordion-menu-token';
import { ACCORDION } from './accordion-token';
import * as i0 from "@angular/core";
import * as i1 from "./accordion-menu.component";
import * as i2 from "./accordion-item-routerlink.component";
import * as i3 from "./accordion-item-hreflink.component";
import * as i4 from "./accordion-item.component";
import * as i5 from "@angular/common";
export class AccordionListComponent {
    constructor(parentComponent, accordion) {
        this.parentComponent = parentComponent;
        this.accordion = accordion;
        this.deepth = 0;
        this.menuToggleItemFn = (item, event) => {
            this.accordion.menuToggleFn({
                item: item,
                open: !item[this.accordion.openKey],
                parent: this.parent.parent,
                event: event
            });
        };
        this.itemClickItemFn = (item, event) => {
            this.accordion.itemClickFn({
                item: item,
                parent: this.parent,
                event: event
            });
        };
    }
    get animateState() {
        return this.accordion.showAnimation;
    }
    get loading() {
        return this.parent && this.parent[this.accordion.loadingKey];
    }
    get noContent() {
        return this.data === undefined || this.data === null || this.data.length === 0;
    }
    get linkTypeKey() {
        return this.accordion.linkTypeKey;
    }
    get childrenKey() {
        return this.accordion.childrenKey;
    }
    get activeKey() {
        return this.accordion.activeKey;
    }
    get itemTemplate() {
        return this.accordion.itemTemplate;
    }
    get menuItemTemplate() {
        return this.accordion.menuItemTemplate;
    }
    get innerListTemplate() {
        return this.accordion.innerListTemplate;
    }
    get loadingTemplate() {
        return this.accordion.loadingTemplate;
    }
    get noContentTemplate() {
        return this.accordion.noContentTemplate;
    }
    get linkType() {
        return this.accordion.linkType;
    }
    get i18nCommonText() {
        return this.accordion.i18nCommonText;
    }
    get showNoContent() {
        return this.accordion.showNoContent;
    }
    get linkDefaultTarget() {
        return this.accordion.linkDefaultTarget;
    }
    get routerLinkActived() {
        return (!!this.accordionItemRouterlinkQueryList
            && this.accordionItemRouterlinkQueryList.some(airlc => this.isLinkRouterActive(airlc))) || (!!this.accordionMenuQueryList
            && this.accordionMenuQueryList.some(amc => this.isMenuRouterActive(amc)));
    }
    get hasActiveChildren() {
        return (!!this.accordionMenuQueryList
            && this.accordionMenuQueryList.some(amc => this.isMenuDataActive(amc)))
            || (!!this.data && !!this.data.length
                && this.data.some(item => this.isItemData(item) && this.isItemDataActive(item)));
    }
    ngOnInit() {
        if (this.parentComponent) {
            setTimeout(() => { this.parentComponent.accordionListFromView = this; });
        }
    }
    ngOnDestroy() {
        if (this.parentComponent) {
            this.parentComponent.accordionListFromView = undefined;
        }
    }
    isLinkRouterActive(airlc) {
        return airlc.routerLinkActived;
    }
    isMenuRouterActive(amc) {
        return amc.routerLinkActived;
    }
    isMenuDataActive(amc) {
        return amc.hasActiveChildren;
    }
    isItemDataActive(item) {
        return !!item[this.activeKey];
    }
    isItemData(item) {
        return item[this.childrenKey] === undefined;
    }
}
AccordionListComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AccordionListComponent, deps: [{ token: ACCORDION_MENU, host: true, optional: true, skipSelf: true }, { token: ACCORDION }], target: i0.ɵɵFactoryTarget.Component });
AccordionListComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: AccordionListComponent, selector: "d-accordion-list", inputs: { data: "data", deepth: "deepth", parent: "parent" }, host: { properties: { "class.devui-accordion-show-animate": "this.animateState" } }, providers: [{
            provide: ACCORDION_LIST,
            useExisting: forwardRef(() => AccordionListComponent)
        }], viewQueries: [{ propertyName: "accordionMenuQueryList", predicate: ACCORDION_MENU, descendants: true }, { propertyName: "accordionItemRouterlinkQueryList", predicate: AccordionItemRouterlinkComponent, descendants: true }], ngImport: i0, template: "<ul class=\"devui-accordion-list\" *ngIf=\"!innerListTemplate || deepth === 0\">\n  <li class=\"devui-accordion-item\" *ngFor=\"let item of data\">\n    <!--\u83DC\u5355\u7C7B\u578B-->\n    <d-accordion-menu *ngIf=\"item[childrenKey] !== undefined\" [item]=\"item\" [deepth]=\"deepth\" [parent]=\"parent\"></d-accordion-menu>\n    <!--\u975E\u83DC\u5355\u7C7B\u578B-->\n    <d-accordion-list\n      *ngIf=\"item[childrenKey] !== undefined\"\n      class=\"devui-accordion-submenu\"\n      [deepth]=\"deepth + 1\"\n      [data]=\"item[childrenKey]\"\n      [parent]=\"item\"\n      [@collapse]=\"item.open ? 'expanded' : 'collapsed'\"\n      [@.disabled]=\"!animateState\"\n    >\n    </d-accordion-list>\n    <ng-container *ngIf=\"item[childrenKey] === undefined\">\n      <!--\u8DEF\u7531\u94FE\u63A5-->\n      <d-accordion-item-routerlink\n        *ngIf=\"linkType === 'routerLink'\"\n        [item]=\"item\"\n        [deepth]=\"deepth\"\n        [parent]=\"parent\"\n      ></d-accordion-item-routerlink>\n      <!--\u666E\u901A\u94FE\u63A5-->\n      <d-accordion-item-hreflink\n        *ngIf=\"linkType === 'hrefLink'\"\n        [item]=\"item\"\n        [deepth]=\"deepth\"\n        [parent]=\"parent\"\n      ></d-accordion-item-hreflink>\n      <!--\u52A8\u6001\u94FE\u63A5-->\n      <ng-container *ngIf=\"linkType === 'dependOnLinkTypeKey'\">\n        <d-accordion-item-routerlink\n          *ngIf=\"item[linkTypeKey] === 'routerLink'\"\n          [item]=\"item\"\n          [deepth]=\"deepth\"\n          [parent]=\"parent\"\n        ></d-accordion-item-routerlink>\n        <d-accordion-item-hreflink\n          *ngIf=\"item[linkTypeKey] === 'hrefLink'\"\n          [item]=\"item\"\n          [deepth]=\"deepth\"\n          [parent]=\"parent\"\n        ></d-accordion-item-hreflink>\n        <d-accordion-item\n          *ngIf=\"item[linkTypeKey] !== 'hrefLink' && item[linkTypeKey] !== 'routerLink'\"\n          [item]=\"item\"\n          [deepth]=\"deepth\"\n          [parent]=\"parent\"\n        ></d-accordion-item>\n      </ng-container>\n      <!--\u666E\u901A\u7C7B\u578B-->\n      <d-accordion-item *ngIf=\"!linkType || linkType === ''\" [item]=\"item\" [deepth]=\"deepth\" [parent]=\"parent\"></d-accordion-item>\n    </ng-container>\n  </li>\n</ul>\n<!--\u5217\u8868\u6A21\u677F-->\n<div *ngIf=\"innerListTemplate && deepth !== 0\">\n  <ng-template\n    [ngTemplateOutlet]=\"innerListTemplate\"\n    [ngTemplateOutletContext]=\"{\n      item: parent,\n      deepth: deepth,\n      itemClickFn: itemClickItemFn,\n      menuToggleFn: menuToggleItemFn\n    }\"\n  ></ng-template>\n</div>\n<!--\u65E0\u6570\u636E/\u52A0\u8F7D\u4E2D\u6A21\u677F-->\n<ul *ngIf=\"!innerListTemplate && (loading || (noContent && showNoContent))\">\n  <!--\u52A0\u8F7D\u4E2D-->\n  <li *ngIf=\"loading && !loadingTemplate\" class=\"devui-accordion-item\">\n    <div class=\"devui-accordion-item-title devui-over-flow-ellipsis\" [style.textIndent]=\"deepth * 20 + 'px'\">\n      {{ i18nCommonText?.loading }}\n    </div>\n  </li>\n  <!--\u81EA\u5B9A\u4E49\u52A0\u8F7D\u4E2D-->\n  <ng-template\n    *ngIf=\"loading && loadingTemplate\"\n    [ngTemplateOutlet]=\"loadingTemplate\"\n    [ngTemplateOutletContext]=\"{ item: parent, deepth: deepth }\"\n  ></ng-template>\n  <!--\u65E0\u6570\u636E-->\n  <li *ngIf=\"showNoContent && !loading && noContent && !noContentTemplate\" class=\"devui-accordion-item\">\n    <div class=\"devui-accordion-item-title devui-over-flow-ellipsis disabled\" [style.textIndent]=\"deepth * 20 + 'px'\">\n      {{ i18nCommonText?.noData }}\n    </div>\n  </li>\n  <!--\u81EA\u5B9A\u4E49\u65E0\u6570\u636E-->\n  <ng-template\n    *ngIf=\"showNoContent && !loading && noContent && noContentTemplate\"\n    [ngTemplateOutlet]=\"noContentTemplate\"\n    [ngTemplateOutletContext]=\"{ item: parent, deepth: deepth }\"\n  ></ng-template>\n</ul>\n", components: [{ type: i1.AccordionMenuComponent, selector: "d-accordion-menu" }, { type: AccordionListComponent, selector: "d-accordion-list", inputs: ["data", "deepth", "parent"] }, { type: i2.AccordionItemRouterlinkComponent, selector: "d-accordion-item-routerlink" }, { type: i3.AccordionItemHreflinkComponent, selector: "d-accordion-item-hreflink" }, { type: i4.AccordionItemComponent, selector: "d-accordion-item" }], directives: [{ type: i5.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i5.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i5.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet"] }], animations: [expandCollapse, expandCollapseForDomDestroy], encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AccordionListComponent, decorators: [{
            type: Component,
            args: [{ selector: 'd-accordion-list', encapsulation: ViewEncapsulation.None, animations: [expandCollapse, expandCollapseForDomDestroy], preserveWhitespaces: false, providers: [{
                            provide: ACCORDION_LIST,
                            useExisting: forwardRef(() => AccordionListComponent)
                        }], template: "<ul class=\"devui-accordion-list\" *ngIf=\"!innerListTemplate || deepth === 0\">\n  <li class=\"devui-accordion-item\" *ngFor=\"let item of data\">\n    <!--\u83DC\u5355\u7C7B\u578B-->\n    <d-accordion-menu *ngIf=\"item[childrenKey] !== undefined\" [item]=\"item\" [deepth]=\"deepth\" [parent]=\"parent\"></d-accordion-menu>\n    <!--\u975E\u83DC\u5355\u7C7B\u578B-->\n    <d-accordion-list\n      *ngIf=\"item[childrenKey] !== undefined\"\n      class=\"devui-accordion-submenu\"\n      [deepth]=\"deepth + 1\"\n      [data]=\"item[childrenKey]\"\n      [parent]=\"item\"\n      [@collapse]=\"item.open ? 'expanded' : 'collapsed'\"\n      [@.disabled]=\"!animateState\"\n    >\n    </d-accordion-list>\n    <ng-container *ngIf=\"item[childrenKey] === undefined\">\n      <!--\u8DEF\u7531\u94FE\u63A5-->\n      <d-accordion-item-routerlink\n        *ngIf=\"linkType === 'routerLink'\"\n        [item]=\"item\"\n        [deepth]=\"deepth\"\n        [parent]=\"parent\"\n      ></d-accordion-item-routerlink>\n      <!--\u666E\u901A\u94FE\u63A5-->\n      <d-accordion-item-hreflink\n        *ngIf=\"linkType === 'hrefLink'\"\n        [item]=\"item\"\n        [deepth]=\"deepth\"\n        [parent]=\"parent\"\n      ></d-accordion-item-hreflink>\n      <!--\u52A8\u6001\u94FE\u63A5-->\n      <ng-container *ngIf=\"linkType === 'dependOnLinkTypeKey'\">\n        <d-accordion-item-routerlink\n          *ngIf=\"item[linkTypeKey] === 'routerLink'\"\n          [item]=\"item\"\n          [deepth]=\"deepth\"\n          [parent]=\"parent\"\n        ></d-accordion-item-routerlink>\n        <d-accordion-item-hreflink\n          *ngIf=\"item[linkTypeKey] === 'hrefLink'\"\n          [item]=\"item\"\n          [deepth]=\"deepth\"\n          [parent]=\"parent\"\n        ></d-accordion-item-hreflink>\n        <d-accordion-item\n          *ngIf=\"item[linkTypeKey] !== 'hrefLink' && item[linkTypeKey] !== 'routerLink'\"\n          [item]=\"item\"\n          [deepth]=\"deepth\"\n          [parent]=\"parent\"\n        ></d-accordion-item>\n      </ng-container>\n      <!--\u666E\u901A\u7C7B\u578B-->\n      <d-accordion-item *ngIf=\"!linkType || linkType === ''\" [item]=\"item\" [deepth]=\"deepth\" [parent]=\"parent\"></d-accordion-item>\n    </ng-container>\n  </li>\n</ul>\n<!--\u5217\u8868\u6A21\u677F-->\n<div *ngIf=\"innerListTemplate && deepth !== 0\">\n  <ng-template\n    [ngTemplateOutlet]=\"innerListTemplate\"\n    [ngTemplateOutletContext]=\"{\n      item: parent,\n      deepth: deepth,\n      itemClickFn: itemClickItemFn,\n      menuToggleFn: menuToggleItemFn\n    }\"\n  ></ng-template>\n</div>\n<!--\u65E0\u6570\u636E/\u52A0\u8F7D\u4E2D\u6A21\u677F-->\n<ul *ngIf=\"!innerListTemplate && (loading || (noContent && showNoContent))\">\n  <!--\u52A0\u8F7D\u4E2D-->\n  <li *ngIf=\"loading && !loadingTemplate\" class=\"devui-accordion-item\">\n    <div class=\"devui-accordion-item-title devui-over-flow-ellipsis\" [style.textIndent]=\"deepth * 20 + 'px'\">\n      {{ i18nCommonText?.loading }}\n    </div>\n  </li>\n  <!--\u81EA\u5B9A\u4E49\u52A0\u8F7D\u4E2D-->\n  <ng-template\n    *ngIf=\"loading && loadingTemplate\"\n    [ngTemplateOutlet]=\"loadingTemplate\"\n    [ngTemplateOutletContext]=\"{ item: parent, deepth: deepth }\"\n  ></ng-template>\n  <!--\u65E0\u6570\u636E-->\n  <li *ngIf=\"showNoContent && !loading && noContent && !noContentTemplate\" class=\"devui-accordion-item\">\n    <div class=\"devui-accordion-item-title devui-over-flow-ellipsis disabled\" [style.textIndent]=\"deepth * 20 + 'px'\">\n      {{ i18nCommonText?.noData }}\n    </div>\n  </li>\n  <!--\u81EA\u5B9A\u4E49\u65E0\u6570\u636E-->\n  <ng-template\n    *ngIf=\"showNoContent && !loading && noContent && noContentTemplate\"\n    [ngTemplateOutlet]=\"noContentTemplate\"\n    [ngTemplateOutletContext]=\"{ item: parent, deepth: deepth }\"\n  ></ng-template>\n</ul>\n" }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Host
                }, {
                    type: SkipSelf
                }, {
                    type: Inject,
                    args: [ACCORDION_MENU]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [ACCORDION]
                }] }]; }, propDecorators: { animateState: [{
                type: HostBinding,
                args: ['class.devui-accordion-show-animate']
            }], data: [{
                type: Input
            }], deepth: [{
                type: Input
            }], parent: [{
                type: Input
            }], accordionMenuQueryList: [{
                type: ViewChildren,
                args: [ACCORDION_MENU]
            }], accordionItemRouterlinkQueryList: [{
                type: ViewChildren,
                args: [AccordionItemRouterlinkComponent]
            }] } });
