// ng-devui/esm2020/accordion/accordion-base-component.class.mjs
import { Directive, HostBinding, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class AccordionBaseComponent {
    constructor(accordion) {
        this.accordion = accordion;
        this.deepth = 0;
    }
    get disabled() {
        return this.item && this.item[this.accordion.disabledKey];
    }
    get title() {
        return this.item && this.item[this.accordion.titleKey];
    }
    get textIndent() {
        return this.deepth * 20 + 'px';
    }
}
AccordionBaseComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AccordionBaseComponent, deps: "invalid", target: i0.ɵɵFactoryTarget.Directive });
AccordionBaseComponent.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: AccordionBaseComponent, inputs: { item: "item", deepth: "deepth", parent: "parent" }, host: { properties: { "class.disabled": "this.disabled", "attr.title": "this.title", "style.textIndent": "this.textIndent" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AccordionBaseComponent, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: undefined }]; }, propDecorators: { item: [{
                type: Input
            }], deepth: [{
                type: Input
            }], parent: [{
                type: Input
            }], disabled: [{
                type: HostBinding,
                args: ['class.disabled']
            }], title: [{
                type: HostBinding,
                args: ['attr.title']
            }], textIndent: [{
                type: HostBinding,
                args: ['style.textIndent']
            }] } });
