// ng-devui/esm2020/accordion/accordion-item-routerlink.component.mjs
import { Component, HostBinding, HostListener, Inject, ViewChild, ViewEncapsulation } from '@angular/core';
import { Router, RouterLinkActive } from '@angular/router';
import { AccordionBaseLinkComponent } from './accordion-base-link-component.class';
import { ACCORDION } from './accordion-token';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@angular/common";
export class AccordionItemRouterlinkComponent extends AccordionBaseLinkComponent {
    constructor(accordion, router) {
        super(accordion);
        this.accordion = accordion;
        this.router = router;
    }
    get routerLinkActived() {
        return !!(this.routerLinkActiveDirective && this.routerLinkActiveDirective.isActive);
    }
    set urlTree(urlTree) {
        if (urlTree) {
            this.queryParams = urlTree.queryParams;
            this.fragment = urlTree.fragment;
        }
        else {
            this.queryParams = undefined;
            this.fragment = undefined;
        }
    }
    onClick(event) {
        if (!this.disabled) {
            this.accordion.linkItemClickFn({
                item: this.item,
                parent: this.parent,
                event: event
            });
        }
    }
    ngOnChanges(changes) {
        if (changes['item']) {
            if (this.link) {
                this.urlTree = this.router.parseUrl(this.link);
                const handelPath = (url) => {
                    let path = url;
                    const index = url.indexOf('?');
                    if (index > -1) {
                        path = url.slice(0, index);
                    }
                    return path;
                };
                this.path = handelPath(this.link);
            }
            else {
                this.urlTree = undefined;
                this.path = undefined;
            }
        }
    }
}
AccordionItemRouterlinkComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AccordionItemRouterlinkComponent, deps: [{ token: ACCORDION }, { token: i1.Router }], target: i0.ɵɵFactoryTarget.Component });
AccordionItemRouterlinkComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: AccordionItemRouterlinkComponent, selector: "d-accordion-item-routerlink", host: { listeners: { "click": "onClick($event)" }, properties: { "class.devui-router-active": "this.routerLinkActived" } }, viewQueries: [{ propertyName: "routerLinkActiveDirective", first: true, predicate: RouterLinkActive, descendants: true }], usesInheritance: true, usesOnChanges: true, ngImport: i0, template: "<ng-container *ngIf=\"!disabled\">\n  <a\n    class=\"devui-over-flow-ellipsis\"\n    [routerLink]=\"path\"\n    [queryParams]=\"queryParams\"\n    [fragment]=\"fragment\"\n    [routerLinkActive]=\"'devui-router-active'\"\n    target=\"{{ target }}\"\n    rel=\"noopener\"\n    title=\"{{ title }}\"\n  >\n    <div\n      class=\"devui-accordion-splitter\"\n      [ngClass]=\"{ 'devui-parent-list': deepth === 0 }\"\n      [ngStyle]=\"{ left: deepth * 20 + 10 + 'px' }\"\n    ></div>\n    <ng-container *ngIf=\"!itemTemplate\">\n      {{ title }}\n    </ng-container>\n    <ng-template\n      *ngIf=\"itemTemplate\"\n      [ngTemplateOutlet]=\"itemTemplate\"\n      [ngTemplateOutletContext]=\"{\n        parent: parent,\n        item: item,\n        deepth: deepth\n      }\"\n    ></ng-template>\n  </a>\n</ng-container>\n<ng-container *ngIf=\"disabled\">\n  <a class=\"devui-over-flow-ellipsis\" title=\"{{ title }}\">\n    <div\n      class=\"devui-accordion-splitter\"\n      [ngClass]=\"{ 'devui-parent-list': deepth === 0 }\"\n      [ngStyle]=\"{ left: deepth * 20 + 10 + 'px' }\"\n    ></div>\n    <ng-container *ngIf=\"!itemTemplate\">\n      {{ title }}\n    </ng-container>\n    <ng-template\n      *ngIf=\"itemTemplate\"\n      [ngTemplateOutlet]=\"itemTemplate\"\n      [ngTemplateOutletContext]=\"{\n        parent: parent,\n        item: item,\n        deepth: deepth\n      }\"\n    ></ng-template>\n  </a>\n</ng-container>\n", directives: [{ type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i1.RouterLinkWithHref, selector: "a[routerLink],area[routerLink]", inputs: ["target", "queryParams", "fragment", "queryParamsHandling", "preserveFragment", "skipLocationChange", "replaceUrl", "state", "relativeTo", "routerLink"] }, { type: i1.RouterLinkActive, selector: "[routerLinkActive]", inputs: ["routerLinkActiveOptions", "routerLinkActive"], outputs: ["isActiveChange"], exportAs: ["routerLinkActive"] }, { type: i2.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { type: i2.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }, { type: i2.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet"] }], encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AccordionItemRouterlinkComponent, decorators: [{
            type: Component,
            args: [{ selector: 'd-accordion-item-routerlink', encapsulation: ViewEncapsulation.None, preserveWhitespaces: false, template: "<ng-container *ngIf=\"!disabled\">\n  <a\n    class=\"devui-over-flow-ellipsis\"\n    [routerLink]=\"path\"\n    [queryParams]=\"queryParams\"\n    [fragment]=\"fragment\"\n    [routerLinkActive]=\"'devui-router-active'\"\n    target=\"{{ target }}\"\n    rel=\"noopener\"\n    title=\"{{ title }}\"\n  >\n    <div\n      class=\"devui-accordion-splitter\"\n      [ngClass]=\"{ 'devui-parent-list': deepth === 0 }\"\n      [ngStyle]=\"{ left: deepth * 20 + 10 + 'px' }\"\n    ></div>\n    <ng-container *ngIf=\"!itemTemplate\">\n      {{ title }}\n    </ng-container>\n    <ng-template\n      *ngIf=\"itemTemplate\"\n      [ngTemplateOutlet]=\"itemTemplate\"\n      [ngTemplateOutletContext]=\"{\n        parent: parent,\n        item: item,\n        deepth: deepth\n      }\"\n    ></ng-template>\n  </a>\n</ng-container>\n<ng-container *ngIf=\"disabled\">\n  <a class=\"devui-over-flow-ellipsis\" title=\"{{ title }}\">\n    <div\n      class=\"devui-accordion-splitter\"\n      [ngClass]=\"{ 'devui-parent-list': deepth === 0 }\"\n      [ngStyle]=\"{ left: deepth * 20 + 10 + 'px' }\"\n    ></div>\n    <ng-container *ngIf=\"!itemTemplate\">\n      {{ title }}\n    </ng-container>\n    <ng-template\n      *ngIf=\"itemTemplate\"\n      [ngTemplateOutlet]=\"itemTemplate\"\n      [ngTemplateOutletContext]=\"{\n        parent: parent,\n        item: item,\n        deepth: deepth\n      }\"\n    ></ng-template>\n  </a>\n</ng-container>\n" }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [ACCORDION]
                }] }, { type: i1.Router }]; }, propDecorators: { routerLinkActiveDirective: [{
                type: ViewChild,
                args: [RouterLinkActive]
            }], routerLinkActived: [{
                type: HostBinding,
                args: ['class.devui-router-active']
            }], onClick: [{
                type: HostListener,
                args: ['click', ['$event']]
            }] } });
