// ng-devui/esm2020/accordion/accordion-base-item-component.class.mjs
import { Directive, HostBinding } from '@angular/core';
import { AccordionBaseComponent } from './accordion-base-component.class';
import * as i0 from "@angular/core";
export class AccordionBaseItemComponent extends AccordionBaseComponent {
    constructor(accordion) {
        super(accordion);
        this.accordion = accordion;
        this.defaultClasses = true;
    }
    get itemTemplate() {
        return this.accordion.itemTemplate;
    }
    get active() {
        return this.item && this.item[this.accordion.activeKey];
    }
}
AccordionBaseItemComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AccordionBaseItemComponent, deps: "invalid", target: i0.ɵɵFactoryTarget.Directive });
AccordionBaseItemComponent.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: AccordionBaseItemComponent, host: { properties: { "class.active": "this.active", "class.devui-accordion-item-title": "this.defaultClasses", "class.devui-over-flow-ellipsis": "this.defaultClasses" } }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AccordionBaseItemComponent, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: undefined }]; }, propDecorators: { active: [{
                type: HostBinding,
                args: ['class.active']
            }], defaultClasses: [{
                type: HostBinding,
                args: ['class.devui-accordion-item-title']
            }, {
                type: HostBinding,
                args: ['class.devui-over-flow-ellipsis']
            }] } });
