// ng-devui/esm2020/accordion/accordion-base-link-component.class.mjs
import { Directive } from '@angular/core';
import { AccordionBaseItemComponent } from './accordion-base-item-component.class';
import * as i0 from "@angular/core";
export class AccordionBaseLinkComponent extends AccordionBaseItemComponent {
    constructor(accordion) {
        super(accordion);
        this.accordion = accordion;
    }
    get link() {
        return this.item && this.item[this.accordion.linkKey];
    }
    get target() {
        return this.item && this.item[this.accordion.linkTargetKey] || this.accordion.linkDefaultTarget;
    }
    get linkType() {
        return this.item && this.item[this.accordion.linkTypeKey] || '';
    }
}
AccordionBaseLinkComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AccordionBaseLinkComponent, deps: "invalid", target: i0.ɵɵFactoryTarget.Directive });
AccordionBaseLinkComponent.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: AccordionBaseLinkComponent, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AccordionBaseLinkComponent, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: undefined }]; } });
