// ng-devui/esm2020/accordion/accordion-item.component.mjs
import { Component, HostListener, Inject, ViewEncapsulation } from '@angular/core';
import { AccordionBaseItemComponent } from './accordion-base-item-component.class';
import { ACCORDION } from './accordion-token';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export class AccordionItemComponent extends AccordionBaseItemComponent {
    constructor(accordion) {
        super(accordion);
        this.accordion = accordion;
    }
    onClick(event) {
        if (!this.disabled) {
            this.accordion.itemClickFn({
                item: this.item,
                parent: this.parent,
                event: event
            });
        }
    }
}
AccordionItemComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AccordionItemComponent, deps: [{ token: ACCORDION }], target: i0.ɵɵFactoryTarget.Component });
AccordionItemComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: AccordionItemComponent, selector: "d-accordion-item", host: { listeners: { "click": "onClick($event)" } }, usesInheritance: true, ngImport: i0, template: "<div class=\"devui-accordion-splitter\" [ngClass]=\"{ 'devui-parent-list': deepth === 0 }\" [ngStyle]=\"{ left: deepth * 20 + 10 + 'px' }\"></div>\n<ng-container *ngIf=\"!itemTemplate\">\n  {{ title }}\n</ng-container>\n<ng-template\n  *ngIf=\"itemTemplate\"\n  [ngTemplateOutlet]=\"itemTemplate\"\n  [ngTemplateOutletContext]=\"{\n    parent: parent,\n    item: item,\n    deepth: deepth\n  }\"\n></ng-template>\n", directives: [{ type: i1.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { type: i1.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }, { type: i1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i1.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet"] }], encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AccordionItemComponent, decorators: [{
            type: Component,
            args: [{ selector: 'd-accordion-item', encapsulation: ViewEncapsulation.None, preserveWhitespaces: false, template: "<div class=\"devui-accordion-splitter\" [ngClass]=\"{ 'devui-parent-list': deepth === 0 }\" [ngStyle]=\"{ left: deepth * 20 + 10 + 'px' }\"></div>\n<ng-container *ngIf=\"!itemTemplate\">\n  {{ title }}\n</ng-container>\n<ng-template\n  *ngIf=\"itemTemplate\"\n  [ngTemplateOutlet]=\"itemTemplate\"\n  [ngTemplateOutletContext]=\"{\n    parent: parent,\n    item: item,\n    deepth: deepth\n  }\"\n></ng-template>\n" }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [ACCORDION]
                }] }]; }, propDecorators: { onClick: [{
                type: HostListener,
                args: ['click', ['$event']]
            }] } });
