// ng-devui/esm2020/card/card.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { CardActionsDirective, CardAvatarDirective, CardComponent, CardContentDirective, CardExtendComponent, CardHeaderComponent, CardMetaDirective, CardSubtitleDirective, CardTitleDirective } from './card.component';
import * as i0 from "@angular/core";
export class CardModule {
}
CardModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
CardModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardModule, declarations: [CardComponent, CardHeaderComponent, CardAvatarDirective, CardMetaDirective,
        CardActionsDirective, CardSubtitleDirective, CardTitleDirective, CardContentDirective, CardExtendComponent], imports: [CommonModule], exports: [CardComponent, CardHeaderComponent, CardAvatarDirective, CardMetaDirective,
        CardActionsDirective, CardSubtitleDirective, CardTitleDirective, CardContentDirective, CardExtendComponent] });
CardModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardModule, imports: [[
            CommonModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [CardComponent, CardHeaderComponent, CardAvatarDirective, CardMetaDirective,
                        CardActionsDirective, CardSubtitleDirective, CardTitleDirective, CardContentDirective, CardExtendComponent],
                    imports: [
                        CommonModule
                    ],
                    exports: [CardComponent, CardHeaderComponent, CardAvatarDirective, CardMetaDirective,
                        CardActionsDirective, CardSubtitleDirective, CardTitleDirective, CardContentDirective, CardExtendComponent]
                }]
        }] });
