// ng-devui/esm2020/datepicker/datepicker.module.mjs
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { DatePickerConfigService } from './date-picker.config.service';
import { ButtonModule } from 'ng-devui/button';
import { DateRangePickerComponent } from './date-range-picker.component';
import { DateRangePickerDirective } from './date-range-picker.directive';
import { DatePickerAppendToBodyComponent } from './datepicker-cdk-overlay.component';
import { DatepickerComponent } from './datepicker.component';
import { DatepickerDirective } from './datepicker.directive';
import { SingleDateRangePickerComponent } from './single-date-range-picker.component';
import { TwoDatePickerEndDirective } from './two-datepicker/two-datepicker-end.directive';
import { TwoDatepickerSingleComponent } from './two-datepicker/two-datepicker-single.component';
import { TwoDatePickerStartDirective } from './two-datepicker/two-datepicker-start.directive';
import { TwoDatePickerComponent } from './two-datepicker/two-datepicker.component';
import * as i0 from "@angular/core";
export class DatepickerModule {
}
DatepickerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DatepickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DatepickerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DatepickerModule, declarations: [DatepickerComponent,
        DatepickerDirective,
        DatePickerAppendToBodyComponent,
        DateRangePickerComponent,
        DateRangePickerDirective,
        SingleDateRangePickerComponent,
        TwoDatePickerComponent,
        TwoDatepickerSingleComponent,
        TwoDatePickerStartDirective,
        TwoDatePickerEndDirective], imports: [CommonModule,
        FormsModule,
        OverlayModule,
        ButtonModule], exports: [DatepickerComponent,
        DatepickerDirective,
        DatePickerAppendToBodyComponent,
        DateRangePickerComponent,
        DateRangePickerDirective,
        SingleDateRangePickerComponent,
        TwoDatePickerComponent,
        TwoDatepickerSingleComponent,
        TwoDatePickerStartDirective,
        TwoDatePickerEndDirective] });
DatepickerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DatepickerModule, providers: [DatePickerConfigService], imports: [[
            CommonModule,
            FormsModule,
            OverlayModule,
            ButtonModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DatepickerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        OverlayModule,
                        ButtonModule
                    ],
                    exports: [
                        DatepickerComponent,
                        DatepickerDirective,
                        DatePickerAppendToBodyComponent,
                        DateRangePickerComponent,
                        DateRangePickerDirective,
                        SingleDateRangePickerComponent,
                        TwoDatePickerComponent,
                        TwoDatepickerSingleComponent,
                        TwoDatePickerStartDirective,
                        TwoDatePickerEndDirective
                    ],
                    declarations: [
                        DatepickerComponent,
                        DatepickerDirective,
                        DatePickerAppendToBodyComponent,
                        DateRangePickerComponent,
                        DateRangePickerDirective,
                        SingleDateRangePickerComponent,
                        TwoDatePickerComponent,
                        TwoDatepickerSingleComponent,
                        TwoDatePickerStartDirective,
                        TwoDatePickerEndDirective
                    ],
                    providers: [DatePickerConfigService]
                }]
        }] });
