// ng-devui/esm2020/textarea/textarea.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TextareaDirective } from './textarea.directive';
import * as i0 from "@angular/core";
export class TextareaModule {
}
TextareaModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextareaModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TextareaModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextareaModule, declarations: [TextareaDirective], imports: [CommonModule], exports: [TextareaDirective] });
TextareaModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextareaModule, imports: [[
            CommonModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextareaModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                    ],
                    exports: [TextareaDirective],
                    declarations: [TextareaDirective],
                }]
        }] });
