// ng-devui/esm2020/textarea/textarea.directive.mjs
import { Directive, HostBinding, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class TextareaDirective {
    constructor() {
        this.resize = 'none';
    }
}
TextareaDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextareaDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TextareaDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: TextareaDirective, selector: "[dTextarea]", inputs: { resize: "resize", error: "error" }, host: { properties: { "style.resize": "this.resize", "class.error": "this.error" } }, exportAs: ["dTextarea"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextareaDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[dTextarea]',
                    exportAs: 'dTextarea',
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { resize: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['style.resize']
            }], error: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['class.error']
            }] } });
