// ng-devui/esm2020/popover/popover.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { OverlayContainerModule } from 'ng-devui/overlay-container';
import { PositioningModule } from 'ng-devui/position';
import { PopoverComponent } from './popover.component';
import { PopoverDirective } from './popover.directive';
import * as i0 from "@angular/core";
export class PopoverModule {
}
PopoverModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PopoverModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PopoverModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PopoverModule, declarations: [PopoverComponent, PopoverDirective], imports: [CommonModule,
        PositioningModule,
        OverlayContainerModule], exports: [PopoverComponent, PopoverDirective] });
PopoverModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PopoverModule, imports: [[
            CommonModule,
            PositioningModule,
            OverlayContainerModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PopoverModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PositioningModule,
                        OverlayContainerModule
                    ],
                    exports: [PopoverComponent, PopoverDirective],
                    declarations: [PopoverComponent, PopoverDirective],
                }]
        }] });
