// ng-devui/esm2020/overlay-container/overlay-container.module.mjs
import { NgModule } from '@angular/core';
import { WindowRefModule } from 'ng-devui/window-ref';
import { OverlayContainerRef } from './overlay-container-ref';
import * as i0 from "@angular/core";
export class OverlayContainerModule {
}
OverlayContainerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: OverlayContainerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
OverlayContainerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: OverlayContainerModule, imports: [WindowRefModule] });
OverlayContainerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: OverlayContainerModule, providers: [
        OverlayContainerRef,
    ], imports: [[WindowRefModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: OverlayContainerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [WindowRefModule],
                    exports: [],
                    declarations: [],
                    providers: [
                        OverlayContainerRef,
                    ],
                }]
        }] });
