// ng-devui/fesm2020/ng-devui-textarea.mjs
import * as i0 from '@angular/core';
import { Directive, Input, HostBinding, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

class TextareaDirective {
    constructor() {
        this.resize = 'none';
    }
}
TextareaDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextareaDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TextareaDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: TextareaDirective, selector: "[dTextarea]", inputs: { resize: "resize", error: "error" }, host: { properties: { "style.resize": "this.resize", "class.error": "this.error" } }, exportAs: ["dTextarea"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextareaDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[dTextarea]',
                    exportAs: 'dTextarea',
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { resize: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['style.resize']
            }], error: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['class.error']
            }] } });

class TextareaModule {
}
TextareaModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextareaModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TextareaModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextareaModule, declarations: [TextareaDirective], imports: [CommonModule], exports: [TextareaDirective] });
TextareaModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextareaModule, imports: [[
            CommonModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextareaModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                    ],
                    exports: [TextareaDirective],
                    declarations: [TextareaDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TextareaDirective, TextareaModule };
