// ng-devui/fesm2020/ng-devui-status.mjs
import * as i0 from '@angular/core';
import { Component, Input, NgModule } from '@angular/core';
import * as i1 from '@angular/common';
import { CommonModule } from '@angular/common';

class StatusComponent {
    constructor() {
        this.classMap = {};
    }
    get type() {
        return this._type;
    }
    set type(value) {
        this._type = value;
        this.setClassMap();
    }
    setClassMap() {
        this.classMap = 'devui-status-bg-' + this.type;
    }
}
StatusComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: StatusComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
StatusComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: StatusComponent, selector: "d-status", inputs: { type: "type" }, ngImport: i0, template: "<span class=\"devui-status\" [ngClass]=\"classMap\">\n  <ng-content></ng-content>\n</span>\n", styles: [":host{display:block}.devui-status{line-height:20px;height:20px;display:flex;align-items:center}.devui-status:before{display:inline-block;content:\"\";width:10px;height:10px;margin-right:5px;border-radius:100%}.devui-status.devui-status-bg-success:before{background:#50d4ab;background:var(--devui-success, #50d4ab)}.devui-status.devui-status-bg-error:before{background:#f66f6a;background:var(--devui-danger, #f66f6a)}.devui-status.devui-status-bg-warning:before{background:#fac20a;background:var(--devui-warning, #fac20a)}.devui-status.devui-status-bg-waiting:before{background:#beccfa;background:var(--devui-waiting, #beccfa)}.devui-status.devui-status-bg-running:before{background:#5e7ce0;background:var(--devui-info, #5e7ce0)}.devui-status.devui-status-bg-invalid:before{background:#f2f2f3;background:var(--devui-dividing-line, #f2f2f3)}.devui-status.devui-status-bg-initial:before{background:#e9edfa;background:var(--devui-initial, #e9edfa)}\n"], directives: [{ type: i1.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: StatusComponent, decorators: [{
            type: Component,
            args: [{ selector: 'd-status', preserveWhitespaces: false, template: "<span class=\"devui-status\" [ngClass]=\"classMap\">\n  <ng-content></ng-content>\n</span>\n", styles: [":host{display:block}.devui-status{line-height:20px;height:20px;display:flex;align-items:center}.devui-status:before{display:inline-block;content:\"\";width:10px;height:10px;margin-right:5px;border-radius:100%}.devui-status.devui-status-bg-success:before{background:#50d4ab;background:var(--devui-success, #50d4ab)}.devui-status.devui-status-bg-error:before{background:#f66f6a;background:var(--devui-danger, #f66f6a)}.devui-status.devui-status-bg-warning:before{background:#fac20a;background:var(--devui-warning, #fac20a)}.devui-status.devui-status-bg-waiting:before{background:#beccfa;background:var(--devui-waiting, #beccfa)}.devui-status.devui-status-bg-running:before{background:#5e7ce0;background:var(--devui-info, #5e7ce0)}.devui-status.devui-status-bg-invalid:before{background:#f2f2f3;background:var(--devui-dividing-line, #f2f2f3)}.devui-status.devui-status-bg-initial:before{background:#e9edfa;background:var(--devui-initial, #e9edfa)}\n"] }]
        }], propDecorators: { type: [{
                type: Input
            }] } });

class StatusModule {
}
StatusModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: StatusModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
StatusModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: StatusModule, declarations: [StatusComponent], imports: [CommonModule], exports: [StatusComponent] });
StatusModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: StatusModule, providers: [], imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: StatusModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    exports: [StatusComponent],
                    declarations: [StatusComponent],
                    providers: []
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { StatusComponent, StatusModule };
