// ng-devui/fesm2020/ng-devui-text-input.mjs
import * as i0 from '@angular/core';
import { Directive, Input, HostBinding, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

class TextDirective {
    constructor() {
        this.size = '';
    }
    get large() {
        return this.size === 'lg';
    }
    get small() {
        return this.size === 'sm';
    }
}
TextDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TextDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: TextDirective, selector: "[dTextInput]", inputs: { error: "error", size: "size" }, host: { properties: { "class.error": "this.error", "class.devui-textinput-lg": "this.large", "class.devui-textinput-sm": "this.small" } }, exportAs: ["dTextInput"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[dTextInput]',
                    exportAs: 'dTextInput',
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { error: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['class.error']
            }], size: [{
                type: Input
            }], large: [{
                type: HostBinding,
                args: ['class.devui-textinput-lg']
            }], small: [{
                type: HostBinding,
                args: ['class.devui-textinput-sm']
            }] } });

class TextInputModule {
}
TextInputModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TextInputModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextInputModule, declarations: [TextDirective], imports: [CommonModule], exports: [TextDirective] });
TextInputModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextInputModule, imports: [[
            CommonModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextInputModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                    ],
                    exports: [TextDirective],
                    declarations: [TextDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TextDirective, TextInputModule };
