// ng-devui/esm2020/steps-guide/steps-guide.service.mjs
import { Injectable } from '@angular/core';
import { BehaviorSubject, ReplaySubject } from 'rxjs';
import * as i0 from "@angular/core";
export class StepsGuideService {
    constructor() {
        this.showGuideSource = new ReplaySubject(1);
        this.currentIndexSource = new BehaviorSubject(0);
        this.showGuideObs = this.showGuideSource.asObservable();
        this.currentIndex = this.currentIndexSource.asObservable();
        // 引导步骤
        this.steps = [];
        this.currentStep = 0;
    }
    // 切换向导的显示和销毁
    showGuide(visible) {
        this.showGuideSource.next(visible);
    }
    setCurrentIndex(index) {
        this.currentStep = index;
        this.currentIndexSource.next(index);
    }
    getCurrentStep() {
        return this.currentStep;
    }
    setSteps(steps) {
        this.steps = steps;
    }
    getSteps() {
        return this.steps;
    }
}
StepsGuideService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: StepsGuideService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
StepsGuideService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: StepsGuideService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: StepsGuideService, decorators: [{
            type: Injectable
        }] });
