// ng-devui/esm2020/steps-guide/steps-guide.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { OverlayContainerRef } from 'ng-devui/overlay-container';
import { PositioningModule } from 'ng-devui/position';
import { SafePipeModule } from 'ng-devui/utils';
import { StepsGuideComponent } from './steps-guide.component';
import { StepsGuideDirective } from './steps-guide.directive';
import { StepsGuideService } from './steps-guide.service';
import * as i0 from "@angular/core";
export class StepsGuideModule {
}
StepsGuideModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: StepsGuideModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
StepsGuideModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: StepsGuideModule, declarations: [StepsGuideComponent, StepsGuideDirective], imports: [CommonModule,
        SafePipeModule,
        PositioningModule], exports: [StepsGuideDirective] });
StepsGuideModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: StepsGuideModule, providers: [
        OverlayContainerRef,
        StepsGuideService
    ], imports: [[
            CommonModule,
            SafePipeModule,
            PositioningModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: StepsGuideModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        SafePipeModule,
                        PositioningModule
                    ],
                    declarations: [StepsGuideComponent, StepsGuideDirective],
                    exports: [StepsGuideDirective],
                    providers: [
                        OverlayContainerRef,
                        StepsGuideService
                    ]
                }]
        }] });
