// ng-devui/esm2020/time-axis/time-axis.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { SafePipeModule } from 'ng-devui/utils';
import { TimeAxisItemComponent } from './time-axis-item/time-axis-item.component';
import { TimeAxisComponent } from './time-axis.component';
import * as i0 from "@angular/core";
export class TimeAxisModule {
}
TimeAxisModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TimeAxisModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TimeAxisModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TimeAxisModule, declarations: [TimeAxisComponent,
        TimeAxisItemComponent], imports: [CommonModule,
        FormsModule,
        SafePipeModule], exports: [TimeAxisComponent,
        TimeAxisItemComponent] });
TimeAxisModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TimeAxisModule, providers: [], imports: [[
            CommonModule,
            FormsModule,
            SafePipeModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TimeAxisModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        SafePipeModule
                    ],
                    exports: [
                        TimeAxisComponent,
                        TimeAxisItemComponent
                    ],
                    declarations: [
                        TimeAxisComponent,
                        TimeAxisItemComponent
                    ],
                    providers: []
                }]
        }] });
