// ng-devui/esm2020/splitter/splitter.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { PopoverModule } from 'ng-devui/popover';
import { ResizeDirective } from './resize.directive';
import { SplitterBarComponent } from './splitter-bar.component';
import { SplitterPaneComponent } from './splitter-pane.component';
import { SplitterComponent } from './splitter.component';
import * as i0 from "@angular/core";
export class SplitterModule {
}
SplitterModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: SplitterModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SplitterModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: SplitterModule, declarations: [SplitterComponent,
        SplitterPaneComponent,
        SplitterBarComponent,
        ResizeDirective], imports: [CommonModule,
        FormsModule,
        PopoverModule], exports: [SplitterComponent,
        SplitterPaneComponent,
        SplitterBarComponent,
        ResizeDirective] });
SplitterModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: SplitterModule, providers: [], imports: [[
            CommonModule,
            FormsModule,
            PopoverModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: SplitterModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        PopoverModule
                    ],
                    exports: [
                        SplitterComponent,
                        SplitterPaneComponent,
                        SplitterBarComponent,
                        ResizeDirective
                    ],
                    declarations: [
                        SplitterComponent,
                        SplitterPaneComponent,
                        SplitterBarComponent,
                        ResizeDirective
                    ],
                    providers: []
                }]
        }] });
