// ng-devui/esm2020/theme/media-query.mjs
import * as enquire from 'enquire.js';
import { ReplaySubject } from 'rxjs';
export class PrefersColorSchemeMediaQuery {
    constructor() {
        this.prefersColorSchemeSubject = new ReplaySubject(1);
        this.prefersColorSchemeChange = this.prefersColorSchemeSubject.asObservable();
        this.handleColorSchemeChange = (value) => {
            this.prefersColorSchemeSubject.next(value);
        };
    }
    register() {
        PrefersColorSchemeMediaQuery.enquire
            .register.bind(enquire)(PrefersColorSchemeMediaQuery.Query.light, {
            match: () => {
                this.handleColorSchemeChange('light');
            }
        })
            .register(PrefersColorSchemeMediaQuery.Query.dark, {
            match: () => {
                this.handleColorSchemeChange('dark');
            }
        });
        this.prefersColorSchemeSubject.next(this.getInitValue());
    }
    unregister() {
        PrefersColorSchemeMediaQuery.enquire
            .unregister(PrefersColorSchemeMediaQuery.Query.light)
            .unregister(PrefersColorSchemeMediaQuery.Query.dark);
        this.prefersColorSchemeSubject.complete();
    }
    getInitValue() {
        return window.matchMedia(PrefersColorSchemeMediaQuery.Query.light).matches && 'light'
            || window.matchMedia(PrefersColorSchemeMediaQuery.Query.dark).matches && 'dark'
            || 'no-preference';
    }
}
PrefersColorSchemeMediaQuery.enquire = enquire; // prevent code optimization excluding enquire out
/* eslint-disable-next-line @typescript-eslint/no-namespace */
(function (PrefersColorSchemeMediaQuery) {
    let Query;
    (function (Query) {
        Query["light"] = "screen and (prefers-color-scheme: light)";
        Query["dark"] = "screen and (prefers-color-scheme: dark)";
        Query["noPreferences"] = "screen and (prefers-color-scheme: light)";
    })(Query = PrefersColorSchemeMediaQuery.Query || (PrefersColorSchemeMediaQuery.Query = {}));
})(PrefersColorSchemeMediaQuery || (PrefersColorSchemeMediaQuery = {}));
