// ng-devui/esm2020/text-input/text-input.directive.mjs
import { Directive, HostBinding, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class TextDirective {
    constructor() {
        this.size = '';
    }
    get large() {
        return this.size === 'lg';
    }
    get small() {
        return this.size === 'sm';
    }
}
TextDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TextDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: TextDirective, selector: "[dTextInput]", inputs: { error: "error", size: "size" }, host: { properties: { "class.error": "this.error", "class.devui-textinput-lg": "this.large", "class.devui-textinput-sm": "this.small" } }, exportAs: ["dTextInput"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[dTextInput]',
                    exportAs: 'dTextInput',
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { error: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['class.error']
            }], size: [{
                type: Input
            }], large: [{
                type: HostBinding,
                args: ['class.devui-textinput-lg']
            }], small: [{
                type: HostBinding,
                args: ['class.devui-textinput-sm']
            }] } });
