// ng-devui/esm2020/text-input/text-input.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TextDirective } from './text-input.directive';
import * as i0 from "@angular/core";
export class TextInputModule {
}
TextInputModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TextInputModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextInputModule, declarations: [TextDirective], imports: [CommonModule], exports: [TextDirective] });
TextInputModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextInputModule, imports: [[
            CommonModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TextInputModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                    ],
                    exports: [TextDirective],
                    declarations: [TextDirective],
                }]
        }] });
