// ng-devui/esm2020/multi-auto-complete/multi-auto-complete.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { AutoCompleteModule } from 'ng-devui/auto-complete';
import { TagsModule } from 'ng-devui/tags';
import { MultiAutoCompleteComponent } from './multi-auto-complete.component';
import * as i0 from "@angular/core";
export class MultiAutoCompleteModule {
}
MultiAutoCompleteModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: MultiAutoCompleteModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MultiAutoCompleteModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: MultiAutoCompleteModule, declarations: [MultiAutoCompleteComponent], imports: [CommonModule, AutoCompleteModule, FormsModule, TagsModule], exports: [MultiAutoCompleteComponent] });
MultiAutoCompleteModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: MultiAutoCompleteModule, providers: [], imports: [[CommonModule, AutoCompleteModule, FormsModule, TagsModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: MultiAutoCompleteModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, AutoCompleteModule, FormsModule, TagsModule],
                    exports: [MultiAutoCompleteComponent],
                    declarations: [MultiAutoCompleteComponent],
                    providers: [],
                }]
        }] });
