// ng-devui/esm2020/transfer/transfer.module.mjs
import { ScrollingModule } from '@angular/cdk/scrolling';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CheckBoxModule } from 'ng-devui/checkbox';
import { DragDropModule } from 'ng-devui/dragdrop';
import { PopoverModule } from 'ng-devui/popover';
import { SearchModule } from 'ng-devui/search';
import { TransferComponent } from './transfer.component';
import * as i0 from "@angular/core";
export class TransferModule {
}
TransferModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TransferModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TransferModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TransferModule, declarations: [TransferComponent], imports: [CommonModule, FormsModule, ScrollingModule, SearchModule, CheckBoxModule, DragDropModule, PopoverModule], exports: [TransferComponent] });
TransferModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TransferModule, providers: [], imports: [[CommonModule, FormsModule, ScrollingModule, SearchModule, CheckBoxModule, DragDropModule, PopoverModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TransferModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, FormsModule, ScrollingModule, SearchModule, CheckBoxModule, DragDropModule, PopoverModule],
                    exports: [TransferComponent],
                    declarations: [TransferComponent],
                    providers: []
                }]
        }] });
