// ng-devui/esm2020/upload/single-upload-view.component.mjs
import { Component, EventEmitter, Input, Output, TemplateRef, ViewChild } from '@angular/core';
import { IUploadOptions, UploadStatus } from './file-uploader.types';
import { UploadComponent } from './upload.class';
import { UploadedFilesComponent } from './uploaded-files.component';
import * as i0 from "@angular/core";
import * as i1 from "./uploaded-files.component";
import * as i2 from "@angular/common";
export class SingleUploadViewComponent extends UploadComponent {
    constructor() {
        super(...arguments);
        this.uploadedFiles = [];
        this.deleteUploadedFileEvent = new EventEmitter();
        this.UploadStatus = UploadStatus;
        this.fileUploaders = [];
        // 解决templateContext 传递method.bind(this)引发模板中内嵌组件initialize问题
        this.deleteFileProxy = file => {
            this.deleteFile(file);
        };
    }
    addFile(file) {
        this.fileUploaders = [];
        let uploadOptions = this.uploadOptions;
        if (this.dynamicUploadOptionsFn) {
            uploadOptions = this.dynamicUploadOptionsFn(file, this.uploadOptions);
        }
        super.addFile(file, uploadOptions);
    }
    deleteFile(file) {
        super.deleteFile(file);
        this.deleteUploadedFileEvent.emit(file);
    }
    deletePreUploadFile(file) {
        super.deleteFile(file);
    }
    _onDeleteUploadedFile(filePath) {
        this.deleteUploadedFileEvent.emit(filePath);
    }
}
SingleUploadViewComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: SingleUploadViewComponent, deps: null, target: i0.ɵɵFactoryTarget.Component });
SingleUploadViewComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: SingleUploadViewComponent, selector: "d-single-upload-view", inputs: { uploadOptions: "uploadOptions", preloadFilesRef: "preloadFilesRef", uploadedFiles: "uploadedFiles", uploadedFilesRef: "uploadedFilesRef", filePath: "filePath", dynamicUploadOptionsFn: "dynamicUploadOptionsFn" }, outputs: { deleteUploadedFileEvent: "deleteUploadedFileEvent" }, viewQueries: [{ propertyName: "uploadedFilesComponent", first: true, predicate: ["dUploadedFiles"], descendants: true, static: true }], exportAs: ["dSingleUploadView"], usesInheritance: true, ngImport: i0, template: "<d-uploaded-files\n  #dUploadedFiles\n  [uploadedFiles]=\"uploadedFiles\"\n  [filePath]=\"filePath\"\n  [uploadedFilesRef]=\"uploadedFilesRef\"\n  (deleteUploadedFileEvent)=\"_onDeleteUploadedFile($event)\"\n>\n</d-uploaded-files>\n<ng-template\n  [ngTemplateOutlet]=\"preloadFilesRef\"\n  [ngTemplateOutletContext]=\"{ $implicit: this, fileUploaders: fileUploaders, UploadStatus: UploadStatus, deleteFile: deleteFileProxy }\"\n>\n</ng-template>\n", components: [{ type: i1.UploadedFilesComponent, selector: "d-uploaded-files", inputs: ["uploadedFiles", "uploadedFilesRef", "filePath"], outputs: ["deleteUploadedFileEvent"], exportAs: ["dUploadFiles"] }], directives: [{ type: i2.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: SingleUploadViewComponent, decorators: [{
            type: Component,
            args: [{ selector: 'd-single-upload-view', exportAs: 'dSingleUploadView', preserveWhitespaces: false, template: "<d-uploaded-files\n  #dUploadedFiles\n  [uploadedFiles]=\"uploadedFiles\"\n  [filePath]=\"filePath\"\n  [uploadedFilesRef]=\"uploadedFilesRef\"\n  (deleteUploadedFileEvent)=\"_onDeleteUploadedFile($event)\"\n>\n</d-uploaded-files>\n<ng-template\n  [ngTemplateOutlet]=\"preloadFilesRef\"\n  [ngTemplateOutletContext]=\"{ $implicit: this, fileUploaders: fileUploaders, UploadStatus: UploadStatus, deleteFile: deleteFileProxy }\"\n>\n</ng-template>\n" }]
        }], propDecorators: { uploadOptions: [{
                type: Input
            }], preloadFilesRef: [{
                type: Input
            }], uploadedFiles: [{
                type: Input
            }], uploadedFilesRef: [{
                type: Input
            }], filePath: [{
                type: Input
            }], dynamicUploadOptionsFn: [{
                type: Input
            }], deleteUploadedFileEvent: [{
                type: Output
            }], uploadedFilesComponent: [{
                type: ViewChild,
                args: ['dUploadedFiles', { static: true }]
            }] } });
