// ng-devui/esm2020/upload/upload.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ButtonModule } from 'ng-devui/button';
import { ProgressModule } from 'ng-devui/progress';
import { ToastModule } from 'ng-devui/toast';
import { FileDropDirective } from './file-drop.directive';
import { MultipleUploadViewComponent } from './multiple-upload-view.component';
import { MultipleUploadComponent } from './multiple-upload.component';
import { SelectFiles } from './select-files.utils';
import { SingleUploadViewComponent } from './single-upload-view.component';
import { SingleUploadComponent } from './single-upload.component';
import { UploadDirective } from './upload.directive';
import { UploadedFilesComponent } from './uploaded-files.component';
import * as i0 from "@angular/core";
export class UploadModule {
}
UploadModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: UploadModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
UploadModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: UploadModule, declarations: [MultipleUploadViewComponent,
        SingleUploadViewComponent,
        SingleUploadComponent,
        MultipleUploadComponent,
        UploadedFilesComponent,
        FileDropDirective,
        UploadDirective], imports: [CommonModule,
        ButtonModule,
        ToastModule,
        ProgressModule], exports: [MultipleUploadViewComponent,
        SingleUploadViewComponent,
        SingleUploadComponent,
        MultipleUploadComponent,
        UploadedFilesComponent,
        FileDropDirective,
        UploadDirective] });
UploadModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: UploadModule, providers: [SelectFiles], imports: [[
            CommonModule,
            ButtonModule,
            ToastModule,
            ProgressModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: UploadModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        ButtonModule,
                        ToastModule,
                        ProgressModule
                    ],
                    exports: [
                        MultipleUploadViewComponent,
                        SingleUploadViewComponent,
                        SingleUploadComponent,
                        MultipleUploadComponent,
                        UploadedFilesComponent,
                        FileDropDirective,
                        UploadDirective
                    ],
                    declarations: [
                        MultipleUploadViewComponent,
                        SingleUploadViewComponent,
                        SingleUploadComponent,
                        MultipleUploadComponent,
                        UploadedFilesComponent,
                        FileDropDirective,
                        UploadDirective
                    ],
                    providers: [SelectFiles],
                }]
        }] });
