// ng-devui/esm2020/upload/uploaded-files.component.mjs
import { Component, EventEmitter, Input, Output, TemplateRef, } from '@angular/core';
import { I18nService } from 'ng-devui/i18n';
import * as i0 from "@angular/core";
import * as i1 from "ng-devui/i18n";
import * as i2 from "@angular/common";
export class UploadedFilesComponent {
    constructor(i18n) {
        this.i18n = i18n;
        this.uploadedFiles = [];
        this.deleteUploadedFileEvent = new EventEmitter();
        // 解决templateContext 传递method.bind(this)引发模板中内嵌组件initialize问题
        this.deleteFileProxy = filePath => {
            this.deleteFile(filePath);
        };
    }
    ngOnInit() {
        this.i18nText = this.i18n.getI18nText().upload;
        this.i18nSubscription = this.i18n.langChange().subscribe((data) => {
            this.i18nText = data.upload;
        });
    }
    cleanUploadedFiles() {
        this.uploadedFiles = [];
    }
    addAndOverwriteFile(file) {
        this.cleanUploadedFiles();
        this.uploadedFiles.push(file);
    }
    addFile(file) {
        this.uploadedFiles.push(file);
    }
    deleteFile(filePath) {
        this.uploadedFiles = this.uploadedFiles.filter((file) => {
            return filePath !== file[this.filePath];
        });
        this.deleteUploadedFileEvent.emit(filePath);
    }
    ngOnDestroy() {
        if (this.i18nSubscription) {
            this.i18nSubscription.unsubscribe();
        }
    }
}
UploadedFilesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: UploadedFilesComponent, deps: [{ token: i1.I18nService }], target: i0.ɵɵFactoryTarget.Component });
UploadedFilesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: UploadedFilesComponent, selector: "d-uploaded-files", inputs: { uploadedFiles: "uploadedFiles", uploadedFilesRef: "uploadedFilesRef", filePath: "filePath" }, outputs: { deleteUploadedFileEvent: "deleteUploadedFileEvent" }, exportAs: ["dUploadFiles"], ngImport: i0, template: "<ng-template\n  [ngTemplateOutlet]=\"uploadedFilesRef\"\n  [ngTemplateOutletContext]=\"{ $implicit: this, uploadedFiles: uploadedFiles, filePath: filePath, deleteFile: deleteFileProxy }\"\n>\n</ng-template>\n", directives: [{ type: i2.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: UploadedFilesComponent, decorators: [{
            type: Component,
            args: [{ selector: 'd-uploaded-files', exportAs: 'dUploadFiles', preserveWhitespaces: false, template: "<ng-template\n  [ngTemplateOutlet]=\"uploadedFilesRef\"\n  [ngTemplateOutletContext]=\"{ $implicit: this, uploadedFiles: uploadedFiles, filePath: filePath, deleteFile: deleteFileProxy }\"\n>\n</ng-template>\n" }]
        }], ctorParameters: function () { return [{ type: i1.I18nService }]; }, propDecorators: { uploadedFiles: [{
                type: Input
            }], uploadedFilesRef: [{
                type: Input
            }], filePath: [{
                type: Input
            }], deleteUploadedFileEvent: [{
                type: Output
            }] } });
