// ng-devui/esm2020/cascader/cascader.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CheckBoxModule } from 'ng-devui/checkbox';
import { DropDownModule, } from 'ng-devui/dropdown';
import { SelectModule } from 'ng-devui/select';
import { TagsModule } from 'ng-devui/tags';
import { CascaderLiComponent } from './cascader-li.component';
import { CascaderComponent } from './cascader.component';
import * as i0 from "@angular/core";
export class CascaderModule {
}
CascaderModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CascaderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
CascaderModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CascaderModule, declarations: [CascaderComponent,
        CascaderLiComponent], imports: [CommonModule,
        SelectModule,
        FormsModule,
        DropDownModule,
        TagsModule,
        CheckBoxModule], exports: [CascaderComponent] });
CascaderModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CascaderModule, imports: [[
            CommonModule,
            SelectModule,
            FormsModule,
            DropDownModule,
            TagsModule,
            CheckBoxModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CascaderModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        CascaderComponent,
                        CascaderLiComponent
                    ],
                    imports: [
                        CommonModule,
                        SelectModule,
                        FormsModule,
                        DropDownModule,
                        TagsModule,
                        CheckBoxModule
                    ],
                    exports: [CascaderComponent],
                }]
        }] });
