// ng-devui/esm2020/common/common.module.mjs
import { ClipboardModule } from '@angular/cdk/clipboard';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AutoFocusDirective } from './auto-focus.directive';
import { ClipboardDirective } from './clipboard.directive';
import { DatePipe } from './date-pipe';
import { SimulateATagDirective } from './helper-utils';
import { IframeEventPropagateDirective } from './iframe-event-propagate.directive';
import * as i0 from "@angular/core";
export class DCommonModule {
}
DCommonModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DCommonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DCommonModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DCommonModule, declarations: [AutoFocusDirective,
        DatePipe,
        SimulateATagDirective,
        IframeEventPropagateDirective,
        ClipboardDirective], imports: [CommonModule,
        ClipboardModule], exports: [AutoFocusDirective,
        DatePipe,
        SimulateATagDirective,
        IframeEventPropagateDirective,
        ClipboardDirective] });
DCommonModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DCommonModule, providers: [], imports: [[
            CommonModule,
            ClipboardModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DCommonModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        ClipboardModule
                    ],
                    exports: [
                        AutoFocusDirective,
                        DatePipe,
                        SimulateATagDirective,
                        IframeEventPropagateDirective,
                        ClipboardDirective
                    ],
                    declarations: [
                        AutoFocusDirective,
                        DatePipe,
                        SimulateATagDirective,
                        IframeEventPropagateDirective,
                        ClipboardDirective
                    ],
                    providers: [],
                }]
        }] });
