// ng-devui/esm2020/common/iframe-event-propagate.directive.mjs
import { DOCUMENT } from '@angular/common';
import { Directive, ElementRef, Inject, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class IframeEventPropagateDirective {
    constructor(el, doc) {
        this.doc = doc;
        this.event = 'click';
        this.AddIframeContentDocumentClickListener = () => {
            const iframe = this.element.querySelector('iframe');
            if (iframe !== null) {
                if (iframe.contentDocument !== null) {
                    iframe.contentDocument.addEventListener(this.event, this.dispatchClickEvent);
                }
                else {
                    const loadHandler = () => {
                        iframe.contentDocument.addEventListener(this.event, this.dispatchClickEvent);
                        iframe.removeEventListener('load', loadHandler);
                    };
                    iframe.addEventListener('load', loadHandler);
                }
                this.element.removeEventListener('DOMSubtreeModified', this.AddIframeContentDocumentClickListener);
            }
        };
        this.dispatchClickEvent = ($event) => {
            const event = this.document.createEvent('MouseEvents');
            event.initEvent(this.event, true, true);
            event['originEvent'] = $event;
            this.element.dispatchEvent(event);
        };
        this.element = el.nativeElement;
        this.document = this.doc;
    }
    ngAfterViewInit() {
        this.element.addEventListener('DOMSubtreeModified', this.AddIframeContentDocumentClickListener);
        if (this.element.querySelector('iframe') !== null) {
            this.AddIframeContentDocumentClickListener();
        }
    }
}
IframeEventPropagateDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: IframeEventPropagateDirective, deps: [{ token: i0.ElementRef }, { token: DOCUMENT }], target: i0.ɵɵFactoryTarget.Directive });
IframeEventPropagateDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: IframeEventPropagateDirective, selector: "[dIframeEventPropagate]", inputs: { event: "event" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: IframeEventPropagateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[dIframeEventPropagate]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }]; }, propDecorators: { event: [{
                type: Input
            }] } });
