// ng-devui/esm2020/image-preview/image-preview.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { InputNumberModule } from 'ng-devui/input-number';
import { ModalModule } from 'ng-devui/modal';
import { TooltipModule } from 'ng-devui/tooltip';
import { SafePipeModule } from 'ng-devui/utils';
import { DImagePreviewComponent } from './image-preview.component';
import { ImagePreviewDirective } from './image-preview.directive';
import * as i0 from "@angular/core";
export class ImagePreviewModule {
}
ImagePreviewModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: ImagePreviewModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ImagePreviewModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: ImagePreviewModule, declarations: [DImagePreviewComponent, ImagePreviewDirective], imports: [CommonModule,
        ModalModule,
        SafePipeModule,
        TooltipModule,
        FormsModule,
        InputNumberModule], exports: [ImagePreviewDirective] });
ImagePreviewModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: ImagePreviewModule, imports: [[
            CommonModule,
            ModalModule,
            SafePipeModule,
            TooltipModule,
            FormsModule,
            InputNumberModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: ImagePreviewModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [DImagePreviewComponent, ImagePreviewDirective],
                    imports: [
                        CommonModule,
                        ModalModule,
                        SafePipeModule,
                        TooltipModule,
                        FormsModule,
                        InputNumberModule
                    ],
                    exports: [
                        ImagePreviewDirective
                    ],
                }]
        }] });
