// ng-devui/esm2020/image-preview/image-preview.directive.mjs
import { Directive, ElementRef, HostBinding, HostListener, Input } from '@angular/core';
import { ModalService } from 'ng-devui/modal';
import { Subject } from 'rxjs';
import { DImagePreviewComponent } from './image-preview.component';
import * as i0 from "@angular/core";
import * as i1 from "ng-devui/modal";
export class ImagePreviewDirective {
    constructor(elementRef, modalService) {
        this.elementRef = elementRef;
        this.modalService = modalService;
        this.disableDefault = false;
    }
    get defaultClasses() {
        return !this.disableDefault;
    }
    onClick($event) {
        if (this.disableDefault) {
            return;
        }
        const target = $event.target;
        if (target && target.nodeName.toLowerCase() === 'img') {
            this.imagePreView(target);
        }
    }
    ngOnInit() {
        if (this.customSub) {
            this.customSub.subscribe((target) => {
                this.imagePreView(target);
            });
        }
    }
    ngOnDestroy() {
        if (this.customSub) {
            this.customSub.unsubscribe();
        }
    }
    imagePreView(imageHTMLElement) {
        const modalRef = this.modalService.open({
            id: 'devui-image-preview-modal',
            component: DImagePreviewComponent,
            zIndex: this.zIndex,
            backDropZIndex: this.backDropZIndex,
            showAnimation: false,
            data: {
                targetImage: imageHTMLElement,
                images: Array.from(this.elementRef.nativeElement.querySelectorAll('img')),
                onClose: () => {
                    modalRef.modalInstance.hide();
                },
            },
        });
    }
}
ImagePreviewDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: ImagePreviewDirective, deps: [{ token: i0.ElementRef }, { token: i1.ModalService }], target: i0.ɵɵFactoryTarget.Directive });
ImagePreviewDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: ImagePreviewDirective, selector: "[dImagePreview]", inputs: { customSub: "customSub", disableDefault: "disableDefault", zIndex: "zIndex", backDropZIndex: "backDropZIndex" }, host: { listeners: { "click": "onClick($event)" }, properties: { "class.devui-image-preview-container": "this.defaultClasses" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: ImagePreviewDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[dImagePreview]',
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i1.ModalService }]; }, propDecorators: { customSub: [{
                type: Input
            }], disableDefault: [{
                type: Input
            }], zIndex: [{
                type: Input
            }], backDropZIndex: [{
                type: Input
            }], defaultClasses: [{
                type: HostBinding,
                args: ['class.devui-image-preview-container']
            }], onClick: [{
                type: HostListener,
                args: ['click', ['$event']]
            }] } });
