// ng-devui/esm2020/breadcrumb/breadcrumb.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { DropDownModule } from 'ng-devui/dropdown';
import { SearchModule } from 'ng-devui/search';
import { BreadCrumbItemComponent } from './breadcrumb-item/breadcrumb-item.component';
import { BreadCrumbComponent } from './breadcrumb.component';
import { BreadCrumbService } from './breadcrumb.service';
import * as i0 from "@angular/core";
export class BreadcrumbModule {
}
BreadcrumbModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: BreadcrumbModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
BreadcrumbModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: BreadcrumbModule, declarations: [BreadCrumbComponent, BreadCrumbItemComponent], imports: [CommonModule, RouterModule, DropDownModule, SearchModule], exports: [BreadCrumbComponent, BreadCrumbItemComponent] });
BreadcrumbModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: BreadcrumbModule, providers: [BreadCrumbService], imports: [[CommonModule, RouterModule, DropDownModule, SearchModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: BreadcrumbModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, RouterModule, DropDownModule, SearchModule],
                    exports: [BreadCrumbComponent, BreadCrumbItemComponent],
                    declarations: [BreadCrumbComponent, BreadCrumbItemComponent],
                    providers: [BreadCrumbService],
                }]
        }] });
