// ng-devui/esm2020/status/status.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { StatusComponent } from './status.component';
import * as i0 from "@angular/core";
export class StatusModule {
}
StatusModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: StatusModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
StatusModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: StatusModule, declarations: [StatusComponent], imports: [CommonModule], exports: [StatusComponent] });
StatusModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: StatusModule, providers: [], imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: StatusModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    exports: [StatusComponent],
                    declarations: [StatusComponent],
                    providers: []
                }]
        }] });
