// ng-devui/esm2020/utils/lazy-load/lazy-load.directive.mjs
import { Directive, ElementRef, EventEmitter, Input, Output } from '@angular/core';
import { fromEvent } from 'rxjs';
import { debounceTime, distinctUntilChanged } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class LazyLoadDirective {
    constructor(el) {
        this.el = el;
        // 启用懒加载，默认不启用
        this.enableLazyLoad = false;
        // 懒加载模式，默认列表模式
        this.contentMode = 'list';
        // 加载更多
        this.loadMore = new EventEmitter();
        // 触发懒加载的距离
        this.loadFactor = 5;
    }
    ngOnChanges(changes) {
        const element = this.target ? this.target : this.el.nativeElement;
        if (changes && changes['enableLazyLoad']) {
            if (changes.enableLazyLoad.currentValue) {
                const scrollEvent = fromEvent(element, 'scroll');
                let scrollEventFormat = scrollEvent;
                if (this.contentMode === 'list') {
                    scrollEventFormat = scrollEvent.pipe(debounceTime(300), distinctUntilChanged());
                }
                this.scrollSubscription = scrollEventFormat.subscribe(event => this.scrollList(event));
            }
            else if (this.scrollSubscription) {
                this.scrollSubscription.unsubscribe();
            }
            else {
                return;
            }
        }
    }
    ngAfterViewInit() {
        if (this.contentMode === 'img') {
            setTimeout(() => {
                const target = this.target ? this.target : this.el.nativeElement;
                const mockEvent = { target };
                this.scrollList(mockEvent);
            });
        }
    }
    ngOnDestroy() {
        if (this.scrollSubscription) {
            this.scrollSubscription.unsubscribe();
        }
    }
    scrollList(event) {
        const targetEl = event.target.scrollingElement ? event.target.scrollingElement : event.target;
        const clientHeight = targetEl.clientHeight;
        const scrollTop = targetEl.scrollTop;
        if (this.contentMode === 'img') {
            const rect = this.el.nativeElement.getBoundingClientRect();
            if (rect.top >= 0 && (clientHeight >= rect.top + this.loadFactor)) {
                if (this.imgLoadSrc) {
                    this.el.nativeElement.src = this.imgLoadSrc;
                }
                this.loadMore.emit(event);
                this.scrollSubscription.unsubscribe();
            }
        }
        else {
            const scrollHeight = targetEl.scrollHeight;
            if (scrollTop !== 0 && (scrollTop + clientHeight + this.loadFactor >= scrollHeight)) {
                this.loadMore.emit(event);
            }
        }
    }
}
LazyLoadDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: LazyLoadDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
LazyLoadDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: LazyLoadDirective, selector: "[dLazyLoad]", inputs: { enableLazyLoad: "enableLazyLoad", contentMode: "contentMode", target: "target", imgLoadSrc: "imgLoadSrc" }, outputs: { loadMore: "loadMore" }, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: LazyLoadDirective, decorators: [{
            type: Directive,
            args: [{ selector: '[dLazyLoad]' }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; }, propDecorators: { enableLazyLoad: [{
                type: Input
            }], contentMode: [{
                type: Input
            }], target: [{
                type: Input
            }], imgLoadSrc: [{
                type: Input
            }], loadMore: [{
                type: Output
            }] } });
