// ng-devui/esm2020/nav-sprite/nav-sprite.module.mjs
import { DragDropModule } from '@angular/cdk/drag-drop';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ButtonModule } from 'ng-devui/button';
import { TooltipModule } from 'ng-devui/tooltip';
import { NavSpriteComponent } from './nav-sprite.component';
import * as i0 from "@angular/core";
export class NavSpriteModule {
}
NavSpriteModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: NavSpriteModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NavSpriteModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: NavSpriteModule, declarations: [NavSpriteComponent], imports: [CommonModule, DragDropModule, ButtonModule, TooltipModule], exports: [NavSpriteComponent] });
NavSpriteModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: NavSpriteModule, imports: [[CommonModule, DragDropModule, ButtonModule, TooltipModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: NavSpriteModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, DragDropModule, ButtonModule, TooltipModule],
                    declarations: [NavSpriteComponent],
                    exports: [NavSpriteComponent],
                }]
        }] });
