// ng-devui/esm2020/layout/elements/aside.component.mjs
import { Component, HostBinding } from '@angular/core';
import * as i0 from "@angular/core";
export class AsideComponent {
    constructor() {
        this.default = true;
    }
}
AsideComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AsideComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
AsideComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: AsideComponent, selector: "d-aside", host: { properties: { "class.d-aside": "this.default" } }, exportAs: ["dAside"], ngImport: i0, template: '<ng-content></ng-content>', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AsideComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'd-aside',
                    exportAs: 'dAside',
                    template: '<ng-content></ng-content>',
                }]
        }], propDecorators: { default: [{
                type: HostBinding,
                args: ['class.d-aside']
            }] } });
