// ng-devui/esm2020/form/validator-directive/d-validate-sync.directive.mjs
import { Directive, Input, Self } from '@angular/core';
import { NgControl } from '@angular/forms';
import { DValidateSyncService } from '../services/d-validate-sync.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/forms";
import * as i2 from "../services/d-validate-sync.service";
export class DValidateSyncDirective {
    constructor(_cd, syncService) {
        this._cd = _cd;
        this.syncService = syncService;
        this._added = false;
    }
    ngOnChanges() {
        if (this.key && this._cd.control && !this._added) {
            this.syncService.addControl(this.key, this._cd.control);
            this._added = true;
        }
    }
    ngOnDestroy() {
        if (this.key && this._cd.control) {
            this.syncService.removeControl(this.key, this._cd.control);
        }
    }
}
DValidateSyncDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DValidateSyncDirective, deps: [{ token: i1.NgControl, self: true }, { token: i2.DValidateSyncService }], target: i0.ɵɵFactoryTarget.Directive });
DValidateSyncDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: DValidateSyncDirective, selector: "[dValidateSyncKey][formControlName],[dValidateSyncKey][ngModel],[dValidateSyncKey][formControl],\n    [dValidateSyncKey][formGroupName],[dValidateSyncKey][formArrayName],[dValidateSyncKey][ngModelGroup],\n    [dValidateSyncKey][formGroup],[dValidateSyncKey]form:not([ngNoForm]),[dValidateSyncKey][ngForm]\n  ", inputs: { key: ["dValidateSyncKey", "key"] }, exportAs: ["dValidateSync"], usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DValidateSyncDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: `[dValidateSyncKey][formControlName],[dValidateSyncKey][ngModel],[dValidateSyncKey][formControl],
    [dValidateSyncKey][formGroupName],[dValidateSyncKey][formArrayName],[dValidateSyncKey][ngModelGroup],
    [dValidateSyncKey][formGroup],[dValidateSyncKey]form:not([ngNoForm]),[dValidateSyncKey][ngForm]
  `,
                    exportAs: 'dValidateSync'
                }]
        }], ctorParameters: function () { return [{ type: i1.NgControl, decorators: [{
                    type: Self
                }] }, { type: i2.DValidateSyncService }]; }, propDecorators: { key: [{
                type: Input,
                args: ['dValidateSyncKey']
            }] } });
