// ng-devui/esm2020/loading/loading-backdrop.component.mjs
import { Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export class LoadingBackdropComponent {
    constructor() {
        this.backdrop = true;
    }
    ngOnInit() {
        if (this.target) {
            this.targetName = this.target.nodeName;
        }
    }
    ngOnChanges(changes) {
        if (changes['target']) {
            if (this.target) {
                this.targetName = this.target.nodeName;
            }
        }
    }
}
LoadingBackdropComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: LoadingBackdropComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
LoadingBackdropComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: LoadingBackdropComponent, selector: "d-loading-backdrop", inputs: { backdrop: "backdrop", target: "target", zIndex: "zIndex" }, usesOnChanges: true, ngImport: i0, template: `<div class="devui-loading-backdrop" [ngStyle]="{ 'z-index': zIndex }"
                  [ngClass]="{ 'devui-loading-full': targetName === 'BODY' }"
                  *ngIf="backdrop">
             </div>`, isInline: true, styles: [".devui-loading-backdrop{position:absolute;z-index:1079;z-index:calc(var(--devui-z-index-full-page-overlay, 1080) - 1);background-color:#d7d8da;background-color:var(--devui-line, #d7d8da);opacity:.3;left:0;top:0;bottom:0;right:0}.devui-loading-backdrop.devui-loading-full{position:fixed;width:100vw;height:100vh}\n"], directives: [{ type: i1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i1.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }, { type: i1.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: LoadingBackdropComponent, decorators: [{
            type: Component,
            args: [{ selector: 'd-loading-backdrop', template: `<div class="devui-loading-backdrop" [ngStyle]="{ 'z-index': zIndex }"
                  [ngClass]="{ 'devui-loading-full': targetName === 'BODY' }"
                  *ngIf="backdrop">
             </div>`, preserveWhitespaces: false, styles: [".devui-loading-backdrop{position:absolute;z-index:1079;z-index:calc(var(--devui-z-index-full-page-overlay, 1080) - 1);background-color:#d7d8da;background-color:var(--devui-line, #d7d8da);opacity:.3;left:0;top:0;bottom:0;right:0}.devui-loading-backdrop.devui-loading-full{position:fixed;width:100vw;height:100vh}\n"] }]
        }], propDecorators: { backdrop: [{
                type: Input
            }], target: [{
                type: Input
            }], zIndex: [{
                type: Input
            }] } });
