// ng-devui/esm2020/loading/loading.service.mjs
import { DOCUMENT } from '@angular/common';
import { ComponentFactoryResolver, Inject, Injectable, RendererFactory2 } from '@angular/core';
import { OverlayContainerRef } from 'ng-devui/overlay-container';
import { LoadingBackdropComponent } from './loading-backdrop.component';
import { LoadingComponent } from './loading.component';
import * as i0 from "@angular/core";
import * as i1 from "ng-devui/overlay-container";
export class LoadingService {
    constructor(overlayContainerRef, componentFactoryResolver, rendererFactory, doc) {
        this.overlayContainerRef = overlayContainerRef;
        this.componentFactoryResolver = componentFactoryResolver;
        this.rendererFactory = rendererFactory;
        this.doc = doc;
        this.renderer = this.rendererFactory.createRenderer(null, null);
        this.document = this.doc;
    }
    // loading 服务内的函数，外部就可以传入ILoadingOptions类型的参数调用它
    open({ target = this.document.body, backdrop = true, message, loadingTemplateRef, positionType = 'relative', view, injector, zIndex, } = {}) {
        const finalComponentFactoryResolver = this.componentFactoryResolver;
        let positionTypeOld = '';
        positionTypeOld = '' || target['style']['position'];
        let backdropRef;
        if (backdrop) {
            backdropRef = this.overlayContainerRef.createComponent(finalComponentFactoryResolver.resolveComponentFactory(LoadingBackdropComponent), injector);
            Object.assign(backdropRef.instance, {
                backdrop: backdrop,
                zIndex: zIndex,
                target: target ? target : this.document.body,
            });
            const viewRef = backdropRef.hostView;
            viewRef.rootNodes.forEach((node) => target.appendChild(node));
        }
        const loadingRef = this.overlayContainerRef.createComponent(finalComponentFactoryResolver.resolveComponentFactory(LoadingComponent), injector);
        Object.assign(loadingRef.instance, {
            message: message,
            zIndex: zIndex,
            loadingTemplateRef: loadingTemplateRef,
            top: view ? view.top : '50%',
            left: view ? view.left : '50%',
            isCustomPosition: !!view,
            target: target ? target : this.document.body,
        });
        this.renderer.setStyle(target, 'position', positionType);
        const viewRef1 = loadingRef.hostView;
        viewRef1.rootNodes.forEach((node) => target.appendChild(node));
        loadingRef.instance.close = () => {
            if (loadingRef) {
                loadingRef.destroy();
            }
            if (backdropRef) {
                backdropRef.destroy();
                backdropRef = null;
            }
            this.renderer.setStyle(target, 'position', positionTypeOld);
        };
        //  返回一个对象内含2个实例loadingInstance和backdropInstance
        return {
            loadingInstance: loadingRef.instance,
            backdropInstance: backdropRef ? backdropRef.instance : null,
        };
    }
}
LoadingService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: LoadingService, deps: [{ token: i1.OverlayContainerRef }, { token: i0.ComponentFactoryResolver }, { token: i0.RendererFactory2 }, { token: DOCUMENT }], target: i0.ɵɵFactoryTarget.Injectable });
LoadingService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: LoadingService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: LoadingService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: i1.OverlayContainerRef }, { type: i0.ComponentFactoryResolver }, { type: i0.RendererFactory2 }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }]; } });
