// ng-devui/esm2020/loading/loading.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { LoadingBackdropComponent } from './loading-backdrop.component';
import { LoadingComponent } from './loading.component';
import { LoadingDirective } from './loading.directive';
import { LoadingService } from './loading.service';
import * as i0 from "@angular/core";
export class LoadingModule {
}
LoadingModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: LoadingModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
LoadingModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: LoadingModule, declarations: [LoadingDirective,
        LoadingBackdropComponent,
        LoadingComponent], imports: [CommonModule], exports: [LoadingDirective,
        LoadingBackdropComponent,
        LoadingComponent] });
LoadingModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: LoadingModule, providers: [LoadingService], imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: LoadingModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    exports: [
                        LoadingDirective,
                        LoadingBackdropComponent,
                        LoadingComponent,
                    ],
                    declarations: [
                        LoadingDirective,
                        LoadingBackdropComponent,
                        LoadingComponent
                    ],
                    providers: [LoadingService],
                }]
        }] });
