// ng-devui/esm2020/fullscreen/fullscreen.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ButtonModule } from 'ng-devui/button';
import { FullscreenComponent } from './fullscreen.component';
import * as i0 from "@angular/core";
export class FullscreenModule {
}
FullscreenModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: FullscreenModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FullscreenModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: FullscreenModule, declarations: [FullscreenComponent], imports: [CommonModule,
        ButtonModule], exports: [FullscreenComponent] });
FullscreenModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: FullscreenModule, imports: [[
            CommonModule,
            ButtonModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: FullscreenModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        ButtonModule
                    ],
                    exports: [
                        FullscreenComponent
                    ],
                    declarations: [
                        FullscreenComponent
                    ]
                }]
        }] });
