// ng-devui/esm2020/form/form.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { PopoverModule } from 'ng-devui/popover';
import { FormControlComponent } from './form-control.component';
import { FormItemComponent } from './form-item.component';
import { FormLabelComponent } from './form-label.component';
import { DFormResetDirective, DFormSubmitDirective, FormOperationComponent } from './form-operation.component';
import { FormDirective } from './form.directive';
import { DValidateSyncDirective } from './validator-directive/d-validate-sync.directive';
import { DFormControlRuleDirective, DFormGroupRuleDirective } from './validator-directive/form-control-rules.directive';
import * as i0 from "@angular/core";
export class FormModule {
}
FormModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: FormModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FormModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: FormModule, declarations: [FormItemComponent,
        FormLabelComponent,
        FormControlComponent,
        FormOperationComponent,
        FormDirective,
        DFormGroupRuleDirective,
        DFormControlRuleDirective,
        DFormSubmitDirective,
        DFormResetDirective,
        DValidateSyncDirective], imports: [CommonModule, ReactiveFormsModule, FormsModule, PopoverModule], exports: [FormItemComponent,
        FormLabelComponent,
        FormControlComponent,
        FormOperationComponent,
        FormDirective,
        DFormGroupRuleDirective,
        DFormControlRuleDirective,
        DFormSubmitDirective,
        DFormResetDirective,
        DValidateSyncDirective] });
FormModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: FormModule, providers: [], imports: [[CommonModule, ReactiveFormsModule, FormsModule, PopoverModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: FormModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, ReactiveFormsModule, FormsModule, PopoverModule],
                    exports: [
                        FormItemComponent,
                        FormLabelComponent,
                        FormControlComponent,
                        FormOperationComponent,
                        FormDirective,
                        DFormGroupRuleDirective,
                        DFormControlRuleDirective,
                        DFormSubmitDirective,
                        DFormResetDirective,
                        DValidateSyncDirective
                    ],
                    declarations: [
                        FormItemComponent,
                        FormLabelComponent,
                        FormControlComponent,
                        FormOperationComponent,
                        FormDirective,
                        DFormGroupRuleDirective,
                        DFormControlRuleDirective,
                        DFormSubmitDirective,
                        DFormResetDirective,
                        DValidateSyncDirective
                    ],
                    providers: [],
                }]
        }] });
