// ng-devui/esm2020/dropdown/dropdown.moudule.mjs
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { WindowRefModule } from 'ng-devui/window-ref';
import { DropDownMenuDirective } from './dropdown-menu.directive';
import { DropDownToggleDirective } from './dropdown-toggle.directive';
import { DropDownAppendToBodyComponent } from './dropdown.component';
import { DropDownDirective } from './dropdown.directive';
import * as i0 from "@angular/core";
export class DropDownModule {
}
DropDownModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DropDownModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DropDownModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DropDownModule, declarations: [DropDownDirective,
        DropDownMenuDirective,
        DropDownToggleDirective,
        DropDownAppendToBodyComponent], imports: [CommonModule,
        OverlayModule,
        WindowRefModule], exports: [DropDownDirective,
        DropDownMenuDirective,
        DropDownToggleDirective,
        DropDownAppendToBodyComponent] });
DropDownModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DropDownModule, imports: [[
            CommonModule,
            OverlayModule,
            WindowRefModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DropDownModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        OverlayModule,
                        WindowRefModule
                    ],
                    exports: [
                        DropDownDirective,
                        DropDownMenuDirective,
                        DropDownToggleDirective,
                        DropDownAppendToBodyComponent
                    ],
                    declarations: [
                        DropDownDirective,
                        DropDownMenuDirective,
                        DropDownToggleDirective,
                        DropDownAppendToBodyComponent
                    ],
                }]
        }] });
