// ng-devui/esm2020/dropdown/dropdown.service.mjs
import { DOCUMENT } from '@angular/common';
import { Inject, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class DropDownService {
    constructor(doc) {
        this.doc = doc;
        this.documentClickTimeOut = null;
        this.closeDropdownBind = this.closeDropdown.bind(this);
        this.document = this.doc;
    }
    open(dropdownScope) {
        if (!this.openScope) {
            // 延时绑定document事件，防止事件冒泡导致立即触发
            this.documentClickTimeOut = setTimeout(() => {
                this.document.addEventListener('click', this.closeDropdownBind);
            });
        }
        this.openScope = dropdownScope;
    }
    close(dropdownScope) {
        if (this.openScope !== dropdownScope) {
            return;
        }
        this.openScope = null;
        clearTimeout(this.documentClickTimeOut);
        this.document.removeEventListener('click', this.closeDropdownBind);
    }
    closeDropdown(event) {
        if (!this.openScope) {
            return;
        }
        const menuEl = this.openScope.menuEl?.nativeElement;
        if (event && this.openScope.menuEl &&
            ((/input|textarea/i.test(event.target.tagName) && menuEl.contains(event.target))
                || this.openScope.closeScope === 'none'
                || (menuEl.contains(event.target) && this.openScope.closeScope === 'blank')
                || (this.openScope.dropdownChildren.some(children => children.toggleEl.nativeElement.contains(event.target))))) {
            return;
        }
        this.openScope.isOpen = false;
    }
}
DropDownService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DropDownService, deps: [{ token: DOCUMENT }], target: i0.ɵɵFactoryTarget.Injectable });
DropDownService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DropDownService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DropDownService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }]; } });
