// ng-devui/esm2020/dropdown/dropdown-toggle.directive.mjs
import { Directive, ElementRef, Host, HostBinding, HostListener, Input } from '@angular/core';
import { DropDownDirective } from './dropdown.directive';
import * as i0 from "@angular/core";
import * as i1 from "./dropdown.directive";
export class DropDownToggleDirective {
    constructor(dropdown, el) {
        this.dropdown = dropdown;
        this.el = el;
        this.addClass = true;
        this.toggleOnFocus = false;
        this.autoFocus = false;
        this.isMouseEvent = false;
    }
    get tabIndex() {
        return this.disabled ? null : 0;
    }
    get attrDisabled() {
        return this.disabled ? 'disabled' : null;
    }
    get disabled() {
        return this.dropdown && this.dropdown.disabled;
    }
    ngOnInit() {
        this.dropdown.dropDownToggle = this;
    }
    ngAfterViewInit() {
        if (this.autoFocus) {
            setTimeout(() => {
                this.el.nativeElement.focus();
            }, 0);
        }
    }
    toggleDropdown(event) {
        if (!this.disabled && this.dropdown.trigger !== 'manually') {
            this.dropdown.toggle();
        }
        return false;
    }
    // mousedown mouseup解决focus与click冲突问题
    setMouseEventTrue(event) {
        this.isMouseEvent = true;
    }
    setMouseEventFalse(event) {
        this.isMouseEvent = false;
    }
    toggleOnFocusFn(event) {
        if (this.toggleOnFocus && !this.disabled && !this.dropdown.isOpen && !this.isMouseEvent) {
            this.dropdown.toggle();
        }
    }
    toggle(event) {
        if (this.disabled || this.dropdown.trigger === 'manually' || event.defaultPrevented) {
            return;
        }
        this.dropdown.toggle();
    }
}
DropDownToggleDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DropDownToggleDirective, deps: [{ token: i1.DropDownDirective, host: true }, { token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
DropDownToggleDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: DropDownToggleDirective, selector: "[dDropDownToggle]", inputs: { toggleOnFocus: "toggleOnFocus", autoFocus: "autoFocus" }, host: { listeners: { "click": "toggleDropdown($event)", "mousedown": "setMouseEventTrue($event)", "mouseup": "setMouseEventFalse($event)", "focus": "toggleOnFocusFn($event)", "keydown.enter": "toggle($event)" }, properties: { "attr.tabIndex": "this.tabIndex", "attr.disabled": "this.attrDisabled", "class.devui-dropdown-toggle": "this.addClass" } }, exportAs: ["d-dropdown-toggle"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DropDownToggleDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[dDropDownToggle]',
                    exportAs: 'd-dropdown-toggle',
                }]
        }], ctorParameters: function () { return [{ type: i1.DropDownDirective, decorators: [{
                    type: Host
                }] }, { type: i0.ElementRef }]; }, propDecorators: { tabIndex: [{
                type: HostBinding,
                args: ['attr.tabIndex']
            }], attrDisabled: [{
                type: HostBinding,
                args: ['attr.disabled']
            }], addClass: [{
                type: HostBinding,
                args: ['class.devui-dropdown-toggle']
            }], toggleOnFocus: [{
                type: Input
            }], autoFocus: [{
                type: Input
            }], toggleDropdown: [{
                type: HostListener,
                args: ['click', ['$event']]
            }], setMouseEventTrue: [{
                type: HostListener,
                args: ['mousedown', ['$event']]
            }], setMouseEventFalse: [{
                type: HostListener,
                args: ['mouseup', ['$event']]
            }], toggleOnFocusFn: [{
                type: HostListener,
                args: ['focus', ['$event']]
            }], toggle: [{
                type: HostListener,
                args: ['keydown.enter', ['$event']]
            }] } });
