// ng-devui/esm2020/data-table/table/body/td/td.component.mjs
import { Component, ElementRef, EventEmitter, HostBinding, Input, Output } from '@angular/core';
import { fromEvent } from 'rxjs';
import { tap } from 'rxjs/operators';
import { EditableTip } from '../../../data-table.model';
import { TableTdService } from './td.service';
import * as i0 from "@angular/core";
import * as i1 from "./td.service";
import * as i2 from "@angular/common";
import * as i3 from "ng-devui/utils";
export class TableTdComponent {
    constructor(elementRef, tdService) {
        this.elementRef = elementRef;
        this.tdService = tdService;
        this.editableTip = EditableTip.hover;
        this.nestedColumn = false;
        this.nestedColumnIndent = 16;
        this.toggleChildTableEvent = new EventEmitter();
        this.editStatusEvent = new EventEmitter();
        /**
         * 编辑状态调整 @deprecated
         */
        this.editingChange = new EventEmitter();
        this.currentEditing = false;
    }
    get isEditable() {
        return this.editable && !this.editing && this.editableTip !== 'btn';
    }
    ngOnInit() { }
    ngOnChanges(changes) {
        if (changes['fixedLeft']) {
            if (this.fixedLeft) {
                this.stickyLeftClass = true;
                this.stickyLeftStyle = this.fixedLeft;
            }
            else {
                this.stickyLeftClass = false;
                this.stickyLeftStyle = null;
            }
        }
        if (changes['fixedRight']) {
            if (this.fixedRight) {
                this.stickyRightClass = true;
                this.stickyRightStyle = this.fixedRight;
            }
            else {
                this.stickyRightClass = false;
                this.stickyRightStyle = null;
            }
        }
        if (changes['editing'] && changes['editing'].currentValue && !this.currentEditing) {
            this.bindEditClickEvent();
        }
    }
    startEditing(event) {
        event.stopPropagation();
        event.preventDefault();
        let beforePromise = Promise.resolve(true);
        if (this.beforeEditStart) {
            const result = this.beforeEditStart(this.rowItem, this.field);
            if (typeof result !== 'undefined') {
                if (result.then) {
                    beforePromise = result;
                }
                else if (result.subscribe) {
                    beforePromise = result.toPromise();
                }
                else {
                    beforePromise = Promise.resolve(result);
                }
            }
        }
        beforePromise.then((canStart) => {
            if (canStart) {
                this.editing = true;
                this.currentEditing = true;
                this.editingChange.emit(true);
                this.editStatusEvent.emit(true);
                this.tdService.tableCellClickEvent.emit(event);
                this.bindEditClickEvent();
            }
        });
    }
    bindEditClickEvent() {
        this.documentClickSubscription = fromEvent(document, 'click').pipe(tap((e) => {
            e.stopPropagation();
        })).subscribe((clickEvent) => {
            if (!this.elementRef.nativeElement.contains(clickEvent.target) && !this.clickInTd) {
                this.finishCellEdit();
            }
            this.clickInTd = false;
        });
        this.tdMousedownSubscription = fromEvent(this.elementRef.nativeElement, 'mousedown').subscribe(event => {
            this.clickInTd = true;
        });
        this.tdMouseupSubscription = fromEvent(this.elementRef.nativeElement, 'mouseup').subscribe(event => {
            this.clickInTd = false;
        });
        this.tdClickSubscription = this.tdService.tableCellClickEvent.subscribe((clickEvent) => {
            if (!this.elementRef.nativeElement.contains(clickEvent.target)) {
                this.finishCellEdit();
            }
        });
    }
    removeEditClickEvent() {
        if (this.documentClickSubscription) {
            this.documentClickSubscription.unsubscribe();
            this.documentClickSubscription = null;
        }
        if (this.tdClickSubscription) {
            this.tdClickSubscription.unsubscribe();
            this.tdClickSubscription = null;
        }
    }
    finishCellEdit() {
        let beforePromise = Promise.resolve(true);
        if (this.beforeEditEnd) {
            const result = this.beforeEditEnd(this.rowItem, this.field);
            if (typeof result !== 'undefined') {
                if (result.then) {
                    beforePromise = result;
                }
                else if (result.subscribe) {
                    beforePromise = result.toPromise();
                }
                else {
                    beforePromise = Promise.resolve(result);
                }
            }
        }
        beforePromise.then((canEnd) => {
            if (canEnd) {
                this.editing = false;
                this.currentEditing = false;
                this.editingChange.emit(false);
                this.editStatusEvent.emit(false);
                this.removeEditClickEvent();
            }
        });
    }
    toggleChildTable(rowItem) {
        rowItem.$isChildTableOpen = !rowItem.$isChildTableOpen;
        this.toggleChildTableEvent.emit(rowItem.$isChildTableOpen);
    }
    ngOnDestroy() {
        if (this.documentClickSubscription) {
            this.documentClickSubscription.unsubscribe();
            this.documentClickSubscription = null;
        }
        if (this.tdClickSubscription) {
            this.tdClickSubscription.unsubscribe();
            this.tdClickSubscription = null;
        }
        if (this.tdMousedownSubscription) {
            this.tdMousedownSubscription.unsubscribe();
            this.tdMousedownSubscription = null;
        }
        if (this.tdMouseupSubscription) {
            this.tdMouseupSubscription.unsubscribe();
            this.tdMouseupSubscription = null;
        }
    }
}
TableTdComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TableTdComponent, deps: [{ token: i0.ElementRef }, { token: i1.TableTdService }], target: i0.ɵɵFactoryTarget.Component });
TableTdComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: TableTdComponent, selector: "[dTableCell]", inputs: { editable: "editable", editableTip: "editableTip", nestedColumn: "nestedColumn", nestedLayer: "nestedLayer", field: "field", rowItem: "rowItem", fixedLeft: "fixedLeft", fixedRight: "fixedRight", iconFoldTable: "iconFoldTable", iconUnFoldTable: "iconUnFoldTable", nestedColumnIndent: "nestedColumnIndent", beforeEditStart: "beforeEditStart", beforeEditEnd: "beforeEditEnd", editing: "editing" }, outputs: { toggleChildTableEvent: "toggleChildTableEvent", editStatusEvent: "editStatusEvent", editingChange: "editingChange" }, host: { properties: { "class.devui-sticky-left-cell": "this.stickyLeftClass", "class.devui-sticky-right-cell": "this.stickyRightClass", "style.left": "this.stickyLeftStyle", "style.right": "this.stickyRightStyle", "class.editable-cell": "this.isEditable" } }, usesOnChanges: true, ngImport: i0, template: "<ng-container *ngIf=\"editable && !editing\">\n  <div *ngIf=\"editableTip === 'btn'\" class=\"cell-container\">\n    <div class=\"cell-container-inner not-break modify-holder\">\n      <ng-template\n        [ngTemplateOutlet]=\"nestedTmp\"\n        [ngTemplateOutletContext]=\"{ nestedColumn: nestedColumn, rowItem: rowItem, nestedLayer: nestedLayer }\"\n      ></ng-template>\n      <ng-container *ngTemplateOutlet=\"content\"></ng-container>\n      <span class=\"cell-modify\" (click)=\"startEditing($event)\">\n        <svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 16 16\" width=\"16px\" height=\"16px\" version=\"1.1\">\n          <g fill=\"none\" fill-rule=\"evenodd\" stroke=\"none\" stroke-width=\"1\">\n            <path\n              d=\"M 9.48528 3.41421 L 10.8995 2 L 13.7279 4.82843 L 12.3137 6.24264 L 9.48528 3.41421 Z M 5.24264 13.3137 L 2.41421 13.3137 L 2.41421 10.4853 L 8.77817 4.12132 L 11.6066 6.94975 L 5.24264 13.3137 Z\"\n            />\n          </g>\n        </svg>\n      </span>\n    </div>\n  </div>\n\n  <div *ngIf=\"editableTip !== 'btn'\" class=\"cell-editable not-break\" (click)=\"startEditing($event)\">\n    <ng-template\n      [ngTemplateOutlet]=\"nestedTmp\"\n      [ngTemplateOutletContext]=\"{ nestedColumn: nestedColumn, rowItem: rowItem, nestedLayer: nestedLayer }\"\n    ></ng-template>\n    <ng-container *ngTemplateOutlet=\"content\"></ng-container>\n  </div>\n</ng-container>\n\n<ng-container *ngIf=\"!editable || editing\">\n  <ng-template\n    [ngTemplateOutlet]=\"nestedTmp\"\n    [ngTemplateOutletContext]=\"{ nestedColumn: nestedColumn, rowItem: rowItem, nestedLayer: nestedLayer }\"\n  ></ng-template>\n  <ng-container *ngTemplateOutlet=\"content\"></ng-container>\n</ng-container>\n\n<ng-template #content><ng-content></ng-content></ng-template>\n\n<ng-template #nestedTmp let-nestedColumn=\"nestedColumn\" let-rowItem=\"rowItem\" let-nestedLayer=\"nestedLayer\">\n  <ng-container *ngIf=\"nestedColumn\">\n    <span *ngIf=\"nestedLayer > 0\" class=\"cell-placeholder\" [ngStyle]=\"{ width: nestedLayer * nestedColumnIndent + 'px' }\"></span>\n    <span [style.visibility]=\"rowItem.children ? 'visible' : 'hidden'\" class=\"childtable-toggler\" (click)=\"toggleChildTable(rowItem)\">\n      <ng-container *ngIf=\"rowItem['$isChildTableOpen']\">\n        <svg\n          *ngIf=\"!iconFoldTable\"\n          width=\"16px\"\n          height=\"16px\"\n          viewBox=\"0 0 16 16\"\n          version=\"1.1\"\n          xmlns=\"http://www.w3.org/2000/svg\"\n          xmlns:xlink=\"http://www.w3.org/1999/xlink\"\n          class=\"svg-icon svg-icon-close\"\n        >\n          <g stroke=\"none\" stroke-width=\"1\" fill=\"none\" fill-rule=\"evenodd\">\n            <rect x=\"0.5\" y=\"0.5\" width=\"15\" height=\"15\" rx=\"2\"></rect>\n            <rect x=\"4\" y=\"7\" width=\"8\" height=\"2\"></rect>\n          </g>\n        </svg>\n        <span *ngIf=\"iconFoldTable\" [innerHTML]=\"iconFoldTable | safe: 'html'\"></span>\n      </ng-container>\n      <ng-container *ngIf=\"!rowItem['$isChildTableOpen']\">\n        <svg\n          *ngIf=\"!iconUnFoldTable\"\n          width=\"16px\"\n          height=\"16px\"\n          viewBox=\"0 0 16 16\"\n          version=\"1.1\"\n          xmlns=\"http://www.w3.org/2000/svg\"\n          xmlns:xlink=\"http://www.w3.org/1999/xlink\"\n          class=\"svg-icon svg-icon-open-folder\"\n        >\n          <g stroke=\"none\" stroke-width=\"1\" fill=\"none\" fill-rule=\"evenodd\">\n            <rect x=\"0.5\" y=\"0.5\" width=\"15\" height=\"15\" rx=\"2\"></rect>\n            <path\n              d=\"M8.75,4 L8.75,7.25 L12,7.25 L12,8.75 L8.749,8.75 L8.75,12 L7.25,12 L7.249,8.75 L4,8.75 L4,7.25 L7.25,7.25 L7.25,4 L8.75,4 Z\"\n            ></path>\n          </g>\n        </svg>\n        <span *ngIf=\"iconUnFoldTable\" [innerHTML]=\"iconUnFoldTable | safe: 'html'\"></span>\n      </ng-container>\n    </span>\n  </ng-container>\n</ng-template>\n", styles: ["svg.svg-icon-arrow>g>polygon{fill:#71757f;fill:var(--devui-icon-text, #71757f)}:host ::ng-deep d-button button.d-btn.d-btn-icon{padding:0!important}:host.editable-cell{position:relative}:host.editable-cell:hover:after{content:\"\";display:block;width:calc(100% - 32px);position:absolute;height:calc(100% - 14px);border:1px solid #d7d8da;border:1px solid var(--devui-line, #d7d8da);left:12px;top:8px;border-radius:2px;border-radius:var(--devui-border-radius, 2px);pointer-events:none}.cell-modify{position:absolute;right:0;cursor:pointer}.cell-modify>svg{vertical-align:middle}.cell-modify:hover>svg>g>g{fill:#71757f;fill:var(--devui-icon-text, #71757f)}.cell-container{position:relative;height:24px}.cell-container-inner{position:absolute;max-width:100%;width:100%;min-height:24px}.cell-container-inner.modify-holder{padding-right:20px}.cell-container-inner.modify-holder .cell-modify{display:none;cursor:pointer;position:absolute;right:0}.cell-container-inner.modify-holder .cell-modify>svg{vertical-align:middle}.cell-container-inner.modify-holder .cell-modify>svg>g{fill:#71757f;fill:var(--devui-icon-text, #71757f)}.cell-container-inner.modify-holder .cell-modify:hover>svg>g{fill:#252b3a;fill:var(--devui-icon-fill-active-hover, #252b3a)}.cell-container-inner.modify-holder:hover .cell-modify{display:inline}.not-break{white-space:nowrap;text-overflow:ellipsis;overflow:hidden;word-spacing:normal;word-wrap:normal}.cell-editable{cursor:pointer;margin-left:-5px;padding-left:5px}.cell-editable:hover{background:#f2f2f3;background:var(--devui-list-item-hover-bg, #f2f2f3)}.cell-placeholder{display:inline-block;height:16px}\n"], directives: [{ type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i2.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet"] }, { type: i2.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }], pipes: { "safe": i3.SafePipe } });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TableTdComponent, decorators: [{
            type: Component,
            args: [{ selector: '[dTableCell]', template: "<ng-container *ngIf=\"editable && !editing\">\n  <div *ngIf=\"editableTip === 'btn'\" class=\"cell-container\">\n    <div class=\"cell-container-inner not-break modify-holder\">\n      <ng-template\n        [ngTemplateOutlet]=\"nestedTmp\"\n        [ngTemplateOutletContext]=\"{ nestedColumn: nestedColumn, rowItem: rowItem, nestedLayer: nestedLayer }\"\n      ></ng-template>\n      <ng-container *ngTemplateOutlet=\"content\"></ng-container>\n      <span class=\"cell-modify\" (click)=\"startEditing($event)\">\n        <svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 16 16\" width=\"16px\" height=\"16px\" version=\"1.1\">\n          <g fill=\"none\" fill-rule=\"evenodd\" stroke=\"none\" stroke-width=\"1\">\n            <path\n              d=\"M 9.48528 3.41421 L 10.8995 2 L 13.7279 4.82843 L 12.3137 6.24264 L 9.48528 3.41421 Z M 5.24264 13.3137 L 2.41421 13.3137 L 2.41421 10.4853 L 8.77817 4.12132 L 11.6066 6.94975 L 5.24264 13.3137 Z\"\n            />\n          </g>\n        </svg>\n      </span>\n    </div>\n  </div>\n\n  <div *ngIf=\"editableTip !== 'btn'\" class=\"cell-editable not-break\" (click)=\"startEditing($event)\">\n    <ng-template\n      [ngTemplateOutlet]=\"nestedTmp\"\n      [ngTemplateOutletContext]=\"{ nestedColumn: nestedColumn, rowItem: rowItem, nestedLayer: nestedLayer }\"\n    ></ng-template>\n    <ng-container *ngTemplateOutlet=\"content\"></ng-container>\n  </div>\n</ng-container>\n\n<ng-container *ngIf=\"!editable || editing\">\n  <ng-template\n    [ngTemplateOutlet]=\"nestedTmp\"\n    [ngTemplateOutletContext]=\"{ nestedColumn: nestedColumn, rowItem: rowItem, nestedLayer: nestedLayer }\"\n  ></ng-template>\n  <ng-container *ngTemplateOutlet=\"content\"></ng-container>\n</ng-container>\n\n<ng-template #content><ng-content></ng-content></ng-template>\n\n<ng-template #nestedTmp let-nestedColumn=\"nestedColumn\" let-rowItem=\"rowItem\" let-nestedLayer=\"nestedLayer\">\n  <ng-container *ngIf=\"nestedColumn\">\n    <span *ngIf=\"nestedLayer > 0\" class=\"cell-placeholder\" [ngStyle]=\"{ width: nestedLayer * nestedColumnIndent + 'px' }\"></span>\n    <span [style.visibility]=\"rowItem.children ? 'visible' : 'hidden'\" class=\"childtable-toggler\" (click)=\"toggleChildTable(rowItem)\">\n      <ng-container *ngIf=\"rowItem['$isChildTableOpen']\">\n        <svg\n          *ngIf=\"!iconFoldTable\"\n          width=\"16px\"\n          height=\"16px\"\n          viewBox=\"0 0 16 16\"\n          version=\"1.1\"\n          xmlns=\"http://www.w3.org/2000/svg\"\n          xmlns:xlink=\"http://www.w3.org/1999/xlink\"\n          class=\"svg-icon svg-icon-close\"\n        >\n          <g stroke=\"none\" stroke-width=\"1\" fill=\"none\" fill-rule=\"evenodd\">\n            <rect x=\"0.5\" y=\"0.5\" width=\"15\" height=\"15\" rx=\"2\"></rect>\n            <rect x=\"4\" y=\"7\" width=\"8\" height=\"2\"></rect>\n          </g>\n        </svg>\n        <span *ngIf=\"iconFoldTable\" [innerHTML]=\"iconFoldTable | safe: 'html'\"></span>\n      </ng-container>\n      <ng-container *ngIf=\"!rowItem['$isChildTableOpen']\">\n        <svg\n          *ngIf=\"!iconUnFoldTable\"\n          width=\"16px\"\n          height=\"16px\"\n          viewBox=\"0 0 16 16\"\n          version=\"1.1\"\n          xmlns=\"http://www.w3.org/2000/svg\"\n          xmlns:xlink=\"http://www.w3.org/1999/xlink\"\n          class=\"svg-icon svg-icon-open-folder\"\n        >\n          <g stroke=\"none\" stroke-width=\"1\" fill=\"none\" fill-rule=\"evenodd\">\n            <rect x=\"0.5\" y=\"0.5\" width=\"15\" height=\"15\" rx=\"2\"></rect>\n            <path\n              d=\"M8.75,4 L8.75,7.25 L12,7.25 L12,8.75 L8.749,8.75 L8.75,12 L7.25,12 L7.249,8.75 L4,8.75 L4,7.25 L7.25,7.25 L7.25,4 L8.75,4 Z\"\n            ></path>\n          </g>\n        </svg>\n        <span *ngIf=\"iconUnFoldTable\" [innerHTML]=\"iconUnFoldTable | safe: 'html'\"></span>\n      </ng-container>\n    </span>\n  </ng-container>\n</ng-template>\n", styles: ["svg.svg-icon-arrow>g>polygon{fill:#71757f;fill:var(--devui-icon-text, #71757f)}:host ::ng-deep d-button button.d-btn.d-btn-icon{padding:0!important}:host.editable-cell{position:relative}:host.editable-cell:hover:after{content:\"\";display:block;width:calc(100% - 32px);position:absolute;height:calc(100% - 14px);border:1px solid #d7d8da;border:1px solid var(--devui-line, #d7d8da);left:12px;top:8px;border-radius:2px;border-radius:var(--devui-border-radius, 2px);pointer-events:none}.cell-modify{position:absolute;right:0;cursor:pointer}.cell-modify>svg{vertical-align:middle}.cell-modify:hover>svg>g>g{fill:#71757f;fill:var(--devui-icon-text, #71757f)}.cell-container{position:relative;height:24px}.cell-container-inner{position:absolute;max-width:100%;width:100%;min-height:24px}.cell-container-inner.modify-holder{padding-right:20px}.cell-container-inner.modify-holder .cell-modify{display:none;cursor:pointer;position:absolute;right:0}.cell-container-inner.modify-holder .cell-modify>svg{vertical-align:middle}.cell-container-inner.modify-holder .cell-modify>svg>g{fill:#71757f;fill:var(--devui-icon-text, #71757f)}.cell-container-inner.modify-holder .cell-modify:hover>svg>g{fill:#252b3a;fill:var(--devui-icon-fill-active-hover, #252b3a)}.cell-container-inner.modify-holder:hover .cell-modify{display:inline}.not-break{white-space:nowrap;text-overflow:ellipsis;overflow:hidden;word-spacing:normal;word-wrap:normal}.cell-editable{cursor:pointer;margin-left:-5px;padding-left:5px}.cell-editable:hover{background:#f2f2f3;background:var(--devui-list-item-hover-bg, #f2f2f3)}.cell-placeholder{display:inline-block;height:16px}\n"] }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i1.TableTdService }]; }, propDecorators: { stickyLeftClass: [{
                type: HostBinding,
                args: ['class.devui-sticky-left-cell']
            }], stickyRightClass: [{
                type: HostBinding,
                args: ['class.devui-sticky-right-cell']
            }], stickyLeftStyle: [{
                type: HostBinding,
                args: ['style.left']
            }], stickyRightStyle: [{
                type: HostBinding,
                args: ['style.right']
            }], editable: [{
                type: Input
            }], editableTip: [{
                type: Input
            }], nestedColumn: [{
                type: Input
            }], nestedLayer: [{
                type: Input
            }], field: [{
                type: Input
            }], rowItem: [{
                type: Input
            }], fixedLeft: [{
                type: Input
            }], fixedRight: [{
                type: Input
            }], iconFoldTable: [{
                type: Input
            }], iconUnFoldTable: [{
                type: Input
            }], nestedColumnIndent: [{
                type: Input
            }], beforeEditStart: [{
                type: Input
            }], beforeEditEnd: [{
                type: Input
            }], toggleChildTableEvent: [{
                type: Output
            }], editing: [{
                type: Input
            }], editStatusEvent: [{
                type: Output
            }], editingChange: [{
                type: Output
            }], isEditable: [{
                type: HostBinding,
                args: ['class.editable-cell']
            }] } });
