// ng-devui/esm2020/modal/modal.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ButtonModule } from 'ng-devui/button';
import { IconModule } from 'ng-devui/icon';
import { OverlayContainerModule } from 'ng-devui/overlay-container';
import { DocumentRef } from 'ng-devui/window-ref';
import { DialogService } from './dialog.service';
import { ModalContainerComponent } from './modal-container.component';
import { ModalFooterComponent } from './modal-footer.component';
import { ModalHeaderComponent } from './modal-header.component';
import { ModalComponent } from './modal.component';
import { ModalContainerDirective, ModalContentDirective } from './modal.directive';
import { ModalService } from './modal.service';
import { MovableDirective } from './movable.directive';
import * as i0 from "@angular/core";
export class ModalModule {
}
ModalModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: ModalModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ModalModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: ModalModule, declarations: [ModalComponent,
        ModalContainerComponent,
        ModalContainerDirective,
        ModalContentDirective,
        ModalHeaderComponent,
        ModalFooterComponent,
        MovableDirective], imports: [CommonModule,
        OverlayContainerModule,
        ButtonModule,
        IconModule], exports: [ModalComponent,
        ModalContainerComponent,
        ModalHeaderComponent,
        ModalFooterComponent,
        MovableDirective] });
ModalModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: ModalModule, providers: [
        ModalService,
        DialogService,
        DocumentRef,
    ], imports: [[
            CommonModule,
            OverlayContainerModule,
            ButtonModule,
            IconModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: ModalModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        OverlayContainerModule,
                        ButtonModule,
                        IconModule
                    ],
                    declarations: [
                        ModalComponent,
                        ModalContainerComponent,
                        ModalContainerDirective,
                        ModalContentDirective,
                        ModalHeaderComponent,
                        ModalFooterComponent,
                        MovableDirective
                    ],
                    exports: [
                        ModalComponent,
                        ModalContainerComponent,
                        ModalHeaderComponent,
                        ModalFooterComponent,
                        MovableDirective
                    ],
                    providers: [
                        ModalService,
                        DialogService,
                        DocumentRef,
                    ],
                }]
        }] });
