// ng-devui/esm2020/button/button.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { DCommonModule } from 'ng-devui/common';
import { LoadingModule } from 'ng-devui/loading';
import { ButtonGroupComponent } from './button-group.component';
import { ButtonComponent } from './button.component';
import * as i0 from "@angular/core";
export class ButtonModule {
}
ButtonModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: ButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ButtonModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: ButtonModule, declarations: [ButtonComponent, ButtonGroupComponent], imports: [CommonModule, LoadingModule, DCommonModule], exports: [ButtonComponent, ButtonGroupComponent] });
ButtonModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: ButtonModule, providers: [], imports: [[CommonModule, LoadingModule, DCommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: ButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, LoadingModule, DCommonModule],
                    exports: [ButtonComponent, ButtonGroupComponent],
                    declarations: [ButtonComponent, ButtonGroupComponent],
                    providers: [],
                }]
        }] });
