// ng-devui/esm2020/quadrant-diagram/quadrant-diagram.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { QuadrantDiagramAxisComponent } from './quadrant-axis/quadrant-axis.component';
import { QuadrantDiagramComponent } from './quadrant-diagram.component';
import { DragDropModule } from 'ng-devui/dragdrop';
import { FullscreenModule } from 'ng-devui/fullscreen';
import { PopoverModule } from 'ng-devui/popover';
import { ProgressModule } from 'ng-devui/progress';
import { SafePipeModule } from 'ng-devui/utils';
import { QuadrantDiagramService } from './quadrant-diagram.service';
import { QuadrantLabelComponent } from './quadrant-label/quadrant-label.component';
import { QuadrantRegionComponent } from './quadrant-region/quadrant-region.component';
import * as i0 from "@angular/core";
export class QuadrantDiagramModule {
}
QuadrantDiagramModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: QuadrantDiagramModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
QuadrantDiagramModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: QuadrantDiagramModule, declarations: [QuadrantDiagramComponent,
        QuadrantDiagramAxisComponent,
        QuadrantRegionComponent,
        QuadrantLabelComponent], imports: [CommonModule,
        FormsModule,
        DragDropModule,
        ProgressModule,
        PopoverModule,
        ProgressModule,
        FullscreenModule,
        SafePipeModule], exports: [QuadrantDiagramComponent] });
QuadrantDiagramModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: QuadrantDiagramModule, providers: [QuadrantDiagramService], imports: [[
            CommonModule,
            FormsModule,
            DragDropModule,
            ProgressModule,
            PopoverModule,
            ProgressModule,
            FullscreenModule,
            SafePipeModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: QuadrantDiagramModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        DragDropModule,
                        ProgressModule,
                        PopoverModule,
                        ProgressModule,
                        FullscreenModule,
                        SafePipeModule
                    ],
                    exports: [
                        QuadrantDiagramComponent
                    ],
                    declarations: [
                        QuadrantDiagramComponent,
                        QuadrantDiagramAxisComponent,
                        QuadrantRegionComponent,
                        QuadrantLabelComponent
                    ],
                    providers: [QuadrantDiagramService]
                }]
        }] });
