// ng-devui/esm2020/quadrant-diagram/quadrant-diagram.service.mjs
import { DOCUMENT } from '@angular/common';
import { Inject, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class QuadrantDiagramService {
    constructor(doc) {
        this.doc = doc;
        this.document = this.doc;
    }
    showAxisLine(x, y, diagramId, view, axisConfigs) {
        const horizontalLine = this.document.querySelector('d-quadrant-diagram#' + diagramId + ' .devui-horizontal-line');
        const verticalLine = this.document.querySelector('d-quadrant-diagram#' + diagramId + ' .devui-vertical-line');
        const labelXAxisValue = this.document.querySelector('d-quadrant-diagram#' + diagramId + ' #devui-label-x-axis-value');
        const labelYAxisValue = this.document.querySelector('d-quadrant-diagram#' + diagramId + ' #devui-label-y-axis-value');
        labelXAxisValue.textContent = this.getXAxisValue(view, axisConfigs, x);
        labelYAxisValue.textContent = this.getYAxisValue(view, axisConfigs, y);
        horizontalLine.style.top = y + 'px';
        horizontalLine.style.display = '';
        verticalLine.style.left = x + 'px';
        verticalLine.style.display = '';
    }
    hideAxisLine(diagramId) {
        const horizontalLine = this.document.querySelector('d-quadrant-diagram#' + diagramId + ' .devui-horizontal-line');
        const verticalLine = this.document.querySelector('d-quadrant-diagram#' + diagramId + ' .devui-vertical-line');
        verticalLine.style.display = 'none';
        horizontalLine.style.display = 'none';
    }
    setListPointerEvents(diagramId, value) {
        const ele = this.document.querySelectorAll('d-quadrant-diagram#' + diagramId + ' .devui-list-style');
        ele.forEach(element => {
            element.style.pointerEvents = value;
        });
    }
    getXAxisValue(view, axisConfigs, left) {
        return ((left - axisConfigs.originPosition.left) / axisConfigs.xTickSpacing).toFixed(1);
    }
    getYAxisValue(view, axisConfigs, top) {
        return ((view.height - top - axisConfigs.originPosition.bottom)
            / axisConfigs.yTickSpacing).toFixed(1);
    }
}
QuadrantDiagramService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: QuadrantDiagramService, deps: [{ token: DOCUMENT }], target: i0.ɵɵFactoryTarget.Injectable });
QuadrantDiagramService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: QuadrantDiagramService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: QuadrantDiagramService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }]; } });
