// ng-devui/esm2020/devui.module.mjs
import { NgModule } from '@angular/core';
import { AccordionModule } from 'ng-devui/accordion';
import { AlertModule } from 'ng-devui/alert';
import { AnchorModule } from 'ng-devui/anchor';
import { AutoCompleteModule } from 'ng-devui/auto-complete';
import { AvatarModule } from 'ng-devui/avatar';
import { BackTopModule } from 'ng-devui/back-top';
import { BadgeModule } from 'ng-devui/badge';
import { BreadcrumbModule } from 'ng-devui/breadcrumb';
import { ButtonModule } from 'ng-devui/button';
import { CardModule } from 'ng-devui/card';
import { CarouselModule } from 'ng-devui/carousel';
import { CascaderModule } from 'ng-devui/cascader';
import { CategorySearchModule } from 'ng-devui/category-search';
import { CheckBoxModule } from 'ng-devui/checkbox';
import { DCommonModule } from 'ng-devui/common';
import { DataTableModule } from 'ng-devui/data-table';
import { DatepickerModule } from 'ng-devui/datepicker';
import { DatepickerProModule } from 'ng-devui/datepicker-pro';
import { DragDropModule } from 'ng-devui/dragdrop';
import { DrawerModule } from 'ng-devui/drawer';
import { DropDownModule } from 'ng-devui/dropdown';
import { EditableSelectModule } from 'ng-devui/editable-select';
import { FormModule } from 'ng-devui/form';
import { FullscreenModule } from 'ng-devui/fullscreen';
import { GanttModule } from 'ng-devui/gantt';
import { IconModule } from 'ng-devui/icon';
import { ImagePreviewModule } from 'ng-devui/image-preview';
import { InputNumberModule } from 'ng-devui/input-number';
import { LayoutModule } from 'ng-devui/layout';
import { LoadingModule } from 'ng-devui/loading';
import { MentionModule } from 'ng-devui/mention';
import { ModalModule } from 'ng-devui/modal';
import { MultiAutoCompleteModule } from 'ng-devui/multi-auto-complete';
import { NavSpriteModule } from 'ng-devui/nav-sprite';
import { PaginationModule } from 'ng-devui/pagination';
import { PanelModule } from 'ng-devui/panel';
import { PopoverModule } from 'ng-devui/popover';
import { ProgressModule } from 'ng-devui/progress';
import { QuadrantDiagramModule } from 'ng-devui/quadrant-diagram';
import { RadioModule } from 'ng-devui/radio';
import { RateModule } from 'ng-devui/rate';
import { ReadTipModule } from 'ng-devui/read-tip';
import { SearchModule } from 'ng-devui/search';
import { SelectModule } from 'ng-devui/select';
import { SliderModule } from 'ng-devui/slider';
import { SplitterModule } from 'ng-devui/splitter';
import { StatusModule } from 'ng-devui/status';
import { StepsGuideModule } from 'ng-devui/steps-guide';
import { StickyModule } from 'ng-devui/sticky';
import { TabsModule } from 'ng-devui/tabs';
import { TagsModule } from 'ng-devui/tags';
import { TagsInputModule } from 'ng-devui/tags-input';
import { TextInputModule } from 'ng-devui/text-input';
import { TextareaModule } from 'ng-devui/textarea';
import { TimeAxisModule } from 'ng-devui/time-axis';
import { TimePickerModule } from 'ng-devui/time-picker';
import { ToastModule } from 'ng-devui/toast';
import { ToggleModule } from 'ng-devui/toggle';
import { TooltipModule } from 'ng-devui/tooltip';
import { TransferModule } from 'ng-devui/transfer';
import { TreeModule } from 'ng-devui/tree';
import { TreeSelectModule } from 'ng-devui/tree-select';
import { UploadModule } from 'ng-devui/upload';
import { WindowRefModule } from 'ng-devui/window-ref';
import * as i0 from "@angular/core";
export * from 'ng-devui/accordion';
export * from 'ng-devui/alert';
export * from 'ng-devui/anchor';
export * from 'ng-devui/auto-complete';
export * from 'ng-devui/avatar';
export * from 'ng-devui/back-top';
export * from 'ng-devui/badge';
export * from 'ng-devui/breadcrumb';
export * from 'ng-devui/button';
export * from 'ng-devui/card';
export * from 'ng-devui/carousel';
export * from 'ng-devui/cascader';
export * from 'ng-devui/category-search';
export * from 'ng-devui/checkbox';
export * from 'ng-devui/common';
export * from 'ng-devui/data-table';
export * from 'ng-devui/datepicker';
export * from 'ng-devui/datepicker-pro';
export * from 'ng-devui/dragdrop';
export * from 'ng-devui/drawer';
export * from 'ng-devui/dropdown';
export * from 'ng-devui/editable-select';
export * from 'ng-devui/form';
export * from 'ng-devui/fullscreen';
export * from 'ng-devui/gantt';
export * from 'ng-devui/icon';
export * from 'ng-devui/image-preview';
export * from 'ng-devui/input-number';
export * from 'ng-devui/layout';
export * from 'ng-devui/loading';
export * from 'ng-devui/mention';
export * from 'ng-devui/modal';
export * from 'ng-devui/multi-auto-complete';
export * from 'ng-devui/nav-sprite';
export * from 'ng-devui/pagination';
export * from 'ng-devui/panel';
export * from 'ng-devui/popover';
export * from 'ng-devui/progress';
export * from 'ng-devui/quadrant-diagram';
export * from 'ng-devui/radio';
export * from 'ng-devui/rate';
export * from 'ng-devui/read-tip';
export * from 'ng-devui/search';
export * from 'ng-devui/select';
export * from 'ng-devui/slider';
export * from 'ng-devui/splitter';
export * from 'ng-devui/status';
export * from 'ng-devui/steps-guide';
export * from 'ng-devui/sticky';
export * from 'ng-devui/tabs';
export * from 'ng-devui/tags';
export * from 'ng-devui/tags-input';
export * from 'ng-devui/text-input';
export * from 'ng-devui/textarea';
export * from 'ng-devui/time-axis';
export * from 'ng-devui/time-picker';
export * from 'ng-devui/toast';
export * from 'ng-devui/toggle';
export * from 'ng-devui/tooltip';
export * from 'ng-devui/transfer';
export * from 'ng-devui/tree';
export * from 'ng-devui/tree-select';
export * from 'ng-devui/upload';
export * from 'ng-devui/window-ref';
export * from './version';
export class DevUIModule {
    static forRoot() {
        return {
            ngModule: DevUIModule,
        };
    }
}
DevUIModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DevUIModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DevUIModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DevUIModule, exports: [AccordionModule,
        AlertModule,
        AnchorModule,
        AutoCompleteModule,
        AvatarModule,
        DCommonModule,
        BadgeModule,
        BackTopModule,
        ButtonModule,
        BreadcrumbModule,
        CardModule,
        CarouselModule,
        CheckBoxModule,
        DataTableModule,
        DatepickerModule,
        DragDropModule,
        DrawerModule,
        DropDownModule,
        EditableSelectModule,
        FormModule,
        FullscreenModule,
        ImagePreviewModule,
        LoadingModule,
        LayoutModule,
        IconModule,
        ModalModule,
        MultiAutoCompleteModule,
        PaginationModule,
        PanelModule,
        PopoverModule,
        ProgressModule,
        RadioModule,
        RateModule,
        SearchModule,
        SelectModule,
        StatusModule,
        StickyModule,
        TabsModule,
        TagsModule,
        TagsInputModule,
        ToastModule,
        ToggleModule,
        TooltipModule,
        TimeAxisModule,
        TreeModule,
        UploadModule,
        InputNumberModule,
        SliderModule,
        SplitterModule,
        TreeSelectModule,
        TransferModule,
        TextInputModule,
        TextareaModule,
        WindowRefModule,
        StepsGuideModule,
        GanttModule,
        QuadrantDiagramModule,
        TimePickerModule,
        CascaderModule,
        CategorySearchModule,
        DatepickerProModule,
        MentionModule,
        NavSpriteModule,
        ReadTipModule] });
DevUIModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DevUIModule, imports: [[], AccordionModule,
        AlertModule,
        AnchorModule,
        AutoCompleteModule,
        AvatarModule,
        DCommonModule,
        BadgeModule,
        BackTopModule,
        ButtonModule,
        BreadcrumbModule,
        CardModule,
        CarouselModule,
        CheckBoxModule,
        DataTableModule,
        DatepickerModule,
        DragDropModule,
        DrawerModule,
        DropDownModule,
        EditableSelectModule,
        FormModule,
        FullscreenModule,
        ImagePreviewModule,
        LoadingModule,
        LayoutModule,
        IconModule,
        ModalModule,
        MultiAutoCompleteModule,
        PaginationModule,
        PanelModule,
        PopoverModule,
        ProgressModule,
        RadioModule,
        RateModule,
        SearchModule,
        SelectModule,
        StatusModule,
        StickyModule,
        TabsModule,
        TagsModule,
        TagsInputModule,
        ToastModule,
        ToggleModule,
        TooltipModule,
        TimeAxisModule,
        TreeModule,
        UploadModule,
        InputNumberModule,
        SliderModule,
        SplitterModule,
        TreeSelectModule,
        TransferModule,
        TextInputModule,
        TextareaModule,
        WindowRefModule,
        StepsGuideModule,
        GanttModule,
        QuadrantDiagramModule,
        TimePickerModule,
        CascaderModule,
        CategorySearchModule,
        DatepickerProModule,
        MentionModule,
        NavSpriteModule,
        ReadTipModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DevUIModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [],
                    exports: [
                        AccordionModule,
                        AlertModule,
                        AnchorModule,
                        AutoCompleteModule,
                        AvatarModule,
                        DCommonModule,
                        BadgeModule,
                        BackTopModule,
                        ButtonModule,
                        BreadcrumbModule,
                        CardModule,
                        CarouselModule,
                        CheckBoxModule,
                        DataTableModule,
                        DatepickerModule,
                        DragDropModule,
                        DrawerModule,
                        DropDownModule,
                        EditableSelectModule,
                        FormModule,
                        FullscreenModule,
                        ImagePreviewModule,
                        LoadingModule,
                        LayoutModule,
                        IconModule,
                        ModalModule,
                        MultiAutoCompleteModule,
                        PaginationModule,
                        PanelModule,
                        PopoverModule,
                        ProgressModule,
                        RadioModule,
                        RateModule,
                        SearchModule,
                        SelectModule,
                        StatusModule,
                        StickyModule,
                        TabsModule,
                        TagsModule,
                        TagsInputModule,
                        ToastModule,
                        ToggleModule,
                        TooltipModule,
                        TimeAxisModule,
                        TreeModule,
                        UploadModule,
                        InputNumberModule,
                        SliderModule,
                        SplitterModule,
                        TreeSelectModule,
                        TransferModule,
                        TextInputModule,
                        TextareaModule,
                        WindowRefModule,
                        StepsGuideModule,
                        GanttModule,
                        QuadrantDiagramModule,
                        TimePickerModule,
                        CascaderModule,
                        CategorySearchModule,
                        DatepickerProModule,
                        MentionModule,
                        NavSpriteModule,
                        ReadTipModule
                    ],
                    declarations: [],
                }]
        }] });
