// ng-devui/esm2020/sticky/sticky.module.mjs
import { NgModule } from '@angular/core';
import { WindowRefModule } from 'ng-devui/window-ref';
import { StickyComponent } from './sticky.component';
import * as i0 from "@angular/core";
export class StickyModule {
}
StickyModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: StickyModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
StickyModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: StickyModule, declarations: [StickyComponent], imports: [WindowRefModule], exports: [StickyComponent] });
StickyModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: StickyModule, imports: [[
            WindowRefModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: StickyModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        WindowRefModule
                    ],
                    declarations: [
                        StickyComponent
                    ],
                    exports: [
                        StickyComponent,
                    ],
                }]
        }] });
