// ng-devui/esm2020/tags/tags.component.mjs
import { ChangeDetectorRef, Component, ContentChild, ElementRef, EventEmitter, Input, Output, QueryList, TemplateRef, ViewChild, ViewChildren } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./tag.component";
import * as i2 from "@angular/common";
import * as i3 from "ng-devui/popover";
export class TagsComponent {
    constructor(cdr) {
        this.cdr = cdr;
        /**
         * 【必选】记录输入的标签
         */
        this.tags = [];
        /**
         * 【可选】使用的属性名
         */
        this.displayProperty = '';
        /**
         * @deprecated
         */
        this.deletable = false;
        this.mode = 'default';
        this.titleProperty = '';
        this.hideBeyondTags = false;
        /**
         * tag被删除后触发
         */
        this.tagDelete = new EventEmitter();
        /**
         * 多标签超出显示相关数据
         */
        this.showTags = [];
        this.beyondTags = [];
        this.showMore = false;
        this.MORE_TAG_WIDTH = 30;
        this.TAG_MARGIN = 4;
    }
    ngOnInit() {
        this.maxShowNumber = this.tags.length;
        // TODO: 使用cloneDeep导致category-search卡顿，lodash-es报Illegal invocation非法调用错误，需调查原因，暂用循环方式替代深拷贝
        this.tags.forEach((item) => this.showTags.push(typeof item === 'object' ? { ...item } : item));
    }
    ngAfterViewInit() {
        if (this.hideBeyondTags) {
            this.calculateHideTagsStatus();
        }
    }
    ngOnChanges(changes) {
        if (this.hideBeyondTags && changes['tags'] && !changes['tags'].firstChange) {
            this.maxShowNumber = this.tags.length;
            this.showTags = [...this.tags];
            this.calculateHideTagsStatus();
        }
    }
    calculateHideTagsStatus() {
        this.cdr.detectChanges();
        this.tagsFatherWidth = this.tagsElement.nativeElement.getBoundingClientRect().width;
        let curWidth = 0;
        this.viewChildren.forEach((element, index) => {
            if (curWidth + element.nativeElement.offsetWidth + this.TAG_MARGIN < this.tagsFatherWidth - this.MORE_TAG_WIDTH) {
                curWidth += element.nativeElement.offsetWidth;
                curWidth += this.TAG_MARGIN;
                this.maxShowNumber = index + 1;
            }
        });
        if (this.maxShowNumber !== this.tags.length) {
            this.showMore = true;
            this.beyondTags = this.showTags.slice(this.maxShowNumber);
            this.showTags.splice(this.maxShowNumber);
            this.cdr.detectChanges();
        }
        else {
            this.showMore = false;
        }
    }
    removeTag($event, tag, index) {
        this.tagDelete.emit({ tag: tag, index: index, event: $event });
    }
}
TagsComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TagsComponent, deps: [{ token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Component });
TagsComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: TagsComponent, selector: "d-tags", inputs: { tags: "tags", displayProperty: "displayProperty", deletable: "deletable", mode: "mode", titleProperty: "titleProperty", hideBeyondTags: "hideBeyondTags" }, outputs: { tagDelete: "tagDelete" }, queries: [{ propertyName: "customViewTemplate", first: true, predicate: TemplateRef, descendants: true }], viewQueries: [{ propertyName: "tagsElement", first: true, predicate: ["tagsElement"], descendants: true }, { propertyName: "viewChildren", predicate: ["tagElement"], descendants: true, read: ElementRef }], exportAs: ["Tags"], usesOnChanges: true, ngImport: i0, template: "<div class=\"devui-tags\" #tagsElement>\n  <d-tag\n    #tagElement\n    *ngFor=\"let tag of hideBeyondTags ? showTags : tags; let index = index\"\n    [tag]=\"displayProperty ? tag[displayProperty] : tag\"\n    [mode]=\"mode\"\n    [deletable]=\"deletable\"\n    [checked]=\"tag['checked']\"\n    [labelStyle]=\"tag['labelStyle']\"\n    [customColor]=\"tag['customColor']\"\n    [titleContent]=\"titleProperty ? tag[titleProperty] : displayProperty ? tag[displayProperty] : tag\"\n    [customViewTemplate]=\"customViewTemplate\"\n    (tagDelete)=\"removeTag($event, tag, index)\"\n  >\n  </d-tag>\n  <d-tag\n    *ngIf=\"hideBeyondTags && showMore\"\n    dPopover\n    [content]=\"beyondTagsTpl\"\n    [controlled]=\"true\"\n    [tag]=\"'+ ' + beyondTags.length\"\n    [mode]=\"'checkable'\"\n    [popoverStyle]=\"{ backgroundColor: 'white', color: 'black' }\"\n  ></d-tag>\n</div>\n\n<ng-template #beyondTagsTpl>\n  <div class=\"beyond-tags-wrapper\">\n    <d-tag\n      class=\"beyond-tag-style\"\n      *ngFor=\"let tag of beyondTags; let index = index\"\n      [tag]=\"displayProperty ? tag[displayProperty] : tag\"\n      [mode]=\"mode\"\n      [deletable]=\"deletable\"\n      [checked]=\"tag['checked']\"\n      [labelStyle]=\"tag['labelStyle']\"\n      [customColor]=\"tag['customColor']\"\n      [titleContent]=\"titleProperty ? tag[titleProperty] : displayProperty ? tag[displayProperty] : tag\"\n      [customViewTemplate]=\"customViewTemplate\"\n      (tagDelete)=\"removeTag($event, tag, index)\"\n    >\n    </d-tag>\n  </div>\n</ng-template>\n", styles: [".devui-tags{overflow:hidden;word-wrap:break-word;height:100%;line-height:24px}.beyond-tags-wrapper{max-width:300px;margin-top:8px}.beyond-tags-wrapper .beyond-tag-style{margin-bottom:8px}.beyond-tags-wrapper d-tag:not(:last-of-type){margin-right:8px}d-tag:not(:last-of-type){margin-right:4px}\n"], components: [{ type: i1.TagComponent, selector: "d-tag", inputs: ["tag", "labelStyle", "customColor", "deletable", "titleContent", "mode", "checked", "maxWidth", "customViewTemplate"], outputs: ["tagDelete", "checkedChange"], exportAs: ["Tag"] }], directives: [{ type: i2.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i3.PopoverDirective, selector: "[dPopover]", inputs: ["content", "controlled", "position", "showAnimation", "showAnimate", "scrollElement", "appendToBody", "zIndex", "popType", "popMaxWidth", "trigger", "hoverToContent", "hoverDelayTime", "popoverStyle", "mouseEnterDelay", "mouseLeaveDelay", "visible"], exportAs: ["dPopover"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TagsComponent, decorators: [{
            type: Component,
            args: [{ selector: 'd-tags', exportAs: 'Tags', preserveWhitespaces: false, template: "<div class=\"devui-tags\" #tagsElement>\n  <d-tag\n    #tagElement\n    *ngFor=\"let tag of hideBeyondTags ? showTags : tags; let index = index\"\n    [tag]=\"displayProperty ? tag[displayProperty] : tag\"\n    [mode]=\"mode\"\n    [deletable]=\"deletable\"\n    [checked]=\"tag['checked']\"\n    [labelStyle]=\"tag['labelStyle']\"\n    [customColor]=\"tag['customColor']\"\n    [titleContent]=\"titleProperty ? tag[titleProperty] : displayProperty ? tag[displayProperty] : tag\"\n    [customViewTemplate]=\"customViewTemplate\"\n    (tagDelete)=\"removeTag($event, tag, index)\"\n  >\n  </d-tag>\n  <d-tag\n    *ngIf=\"hideBeyondTags && showMore\"\n    dPopover\n    [content]=\"beyondTagsTpl\"\n    [controlled]=\"true\"\n    [tag]=\"'+ ' + beyondTags.length\"\n    [mode]=\"'checkable'\"\n    [popoverStyle]=\"{ backgroundColor: 'white', color: 'black' }\"\n  ></d-tag>\n</div>\n\n<ng-template #beyondTagsTpl>\n  <div class=\"beyond-tags-wrapper\">\n    <d-tag\n      class=\"beyond-tag-style\"\n      *ngFor=\"let tag of beyondTags; let index = index\"\n      [tag]=\"displayProperty ? tag[displayProperty] : tag\"\n      [mode]=\"mode\"\n      [deletable]=\"deletable\"\n      [checked]=\"tag['checked']\"\n      [labelStyle]=\"tag['labelStyle']\"\n      [customColor]=\"tag['customColor']\"\n      [titleContent]=\"titleProperty ? tag[titleProperty] : displayProperty ? tag[displayProperty] : tag\"\n      [customViewTemplate]=\"customViewTemplate\"\n      (tagDelete)=\"removeTag($event, tag, index)\"\n    >\n    </d-tag>\n  </div>\n</ng-template>\n", styles: [".devui-tags{overflow:hidden;word-wrap:break-word;height:100%;line-height:24px}.beyond-tags-wrapper{max-width:300px;margin-top:8px}.beyond-tags-wrapper .beyond-tag-style{margin-bottom:8px}.beyond-tags-wrapper d-tag:not(:last-of-type){margin-right:8px}d-tag:not(:last-of-type){margin-right:4px}\n"] }]
        }], ctorParameters: function () { return [{ type: i0.ChangeDetectorRef }]; }, propDecorators: { tags: [{
                type: Input
            }], displayProperty: [{
                type: Input
            }], deletable: [{
                type: Input
            }], mode: [{
                type: Input
            }], titleProperty: [{
                type: Input
            }], hideBeyondTags: [{
                type: Input
            }], customViewTemplate: [{
                type: ContentChild,
                args: [TemplateRef]
            }], tagsElement: [{
                type: ViewChild,
                args: ['tagsElement']
            }], viewChildren: [{
                type: ViewChildren,
                args: ['tagElement', { read: ElementRef }]
            }], tagDelete: [{
                type: Output
            }] } });
