// ng-devui/esm2020/tags/tags.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { PopoverModule } from 'ng-devui/popover';
import { TagComponent } from './tag.component';
import { TagsComponent } from './tags.component';
import * as i0 from "@angular/core";
export class TagsModule {
}
TagsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TagsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TagsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TagsModule, declarations: [TagsComponent,
        TagComponent], imports: [CommonModule,
        FormsModule,
        PopoverModule], exports: [TagsComponent,
        TagComponent] });
TagsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TagsModule, imports: [[
            CommonModule,
            FormsModule,
            PopoverModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TagsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        PopoverModule
                    ],
                    exports: [
                        TagsComponent,
                        TagComponent
                    ],
                    declarations: [
                        TagsComponent,
                        TagComponent
                    ]
                }]
        }] });
