// ng-devui/esm2020/category-search/category-search.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { ButtonModule } from 'ng-devui/button';
import { CheckBoxModule } from 'ng-devui/checkbox';
import { DatepickerProModule } from 'ng-devui/datepicker-pro';
import { DropDownModule } from 'ng-devui/dropdown';
import { FormModule } from 'ng-devui/form';
import { InputNumberModule } from 'ng-devui/input-number';
import { LoadingModule } from 'ng-devui/loading';
import { SearchModule } from 'ng-devui/search';
import { TagsModule } from 'ng-devui/tags';
import { TextInputModule } from 'ng-devui/text-input';
import { TreeModule } from 'ng-devui/tree';
import { PopperModule } from 'ng-devui/utils';
import { CategorySearchComponent } from './category-search.component';
import { ContentTemplateDirective } from './content-template.directive';
import { DefaultTemplateDirective } from './default-template.directive';
import * as i0 from "@angular/core";
export class CategorySearchModule {
}
CategorySearchModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CategorySearchModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
CategorySearchModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CategorySearchModule, declarations: [CategorySearchComponent, ContentTemplateDirective, DefaultTemplateDirective], imports: [CommonModule,
        FormModule,
        FormsModule,
        DropDownModule,
        TextInputModule,
        SearchModule,
        TagsModule,
        ButtonModule,
        CheckBoxModule,
        InputNumberModule,
        LoadingModule,
        TreeModule,
        PopperModule,
        DatepickerProModule], exports: [CategorySearchComponent, ContentTemplateDirective, DefaultTemplateDirective] });
CategorySearchModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CategorySearchModule, imports: [[
            CommonModule,
            FormModule,
            FormsModule,
            DropDownModule,
            TextInputModule,
            SearchModule,
            TagsModule,
            ButtonModule,
            CheckBoxModule,
            InputNumberModule,
            LoadingModule,
            TreeModule,
            PopperModule,
            DatepickerProModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CategorySearchModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormModule,
                        FormsModule,
                        DropDownModule,
                        TextInputModule,
                        SearchModule,
                        TagsModule,
                        ButtonModule,
                        CheckBoxModule,
                        InputNumberModule,
                        LoadingModule,
                        TreeModule,
                        PopperModule,
                        DatepickerProModule,
                    ],
                    declarations: [CategorySearchComponent, ContentTemplateDirective, DefaultTemplateDirective],
                    exports: [CategorySearchComponent, ContentTemplateDirective, DefaultTemplateDirective],
                }]
        }] });
