// ng-devui/esm2020/theme-collection/extend-theme.mjs
import { devuiDarkTheme, Theme } from 'ng-devui/theme';
import { devuiCustomComponentsDarkThemeData, devuiCustomComponentsThemeData } from './theme-custom-components-data';
import { devuiInitialThemeData } from './theme-initial-data';
export const infinityTheme = new Theme({
    id: 'infinity-theme',
    name: '无限主题',
    data: { ...devuiInitialThemeData, ...devuiCustomComponentsThemeData },
    extends: 'devui-light-theme',
    isDark: false,
});
export const provenceTheme = new Theme({
    id: 'provence-theme',
    name: '紫罗兰主题',
    data: {
        ...infinityTheme.data,
        'devui-brand': '#7B69EE',
        'devui-brand-foil': '#F5F5F9',
        'devui-brand-active-focus': '#7B69EE',
        'devui-primary-active': '#7B69EE',
        'devui-brand-hover': '#7B69EE',
        'devui-global-bg': '#f9fafb',
        'devui-base-bg': '#ffffff',
        'devui-text': '#070036',
        'devui-aide-text': '#717087',
        'devui-placeholder': '#babbc0',
        'devui-disabled-text': '#cfd0d3',
        'devui-disabled-bg': '#f5f5f6',
        'devui-line': '#E2E2E5',
        'devui-dividing-line': '#F2F2F3',
        'devui-list-item-hover-bg': '#F5F5F9',
        'devui-list-item-active-bg': '#7B69EE',
        'devui-list-item-active-hover-bg': '#7B69EE',
        'devui-list-item-selected-bg': '#F4F2FF',
        'devui-list-item-hover-text': '#252b3a',
        'devui-list-item-active-text': '#ffffff',
        'devui-form-control-line': '#D7D8DA',
        'devui-icon-text': '#babbc0',
        'devui-brand-active': '#7B69EE',
        'devui-primary': '#7B69EE',
        'devui-primary-hover': '#7B69EE',
        'devui-form-control-line-active': '#7B69EE',
        'devui-form-control-line-active-hover': '#7B69EE',
        'devui-icon-fill-active': '#7B69EE',
        'devui-icon-fill-active-hover': '#7B69EE',
        'devui-label-bg': '#F4F2FF',
        'devui-embed-search-bg': '#F4F2FF',
        'devui-connected-overlay-line': '#7B69EE',
        'devui-primary-disabled': '#d8d2fa',
        'devui-icon-fill-active-disabled': '#d8d2fa',
        'devui-link-active': '#3f1a9c'
    },
    extends: 'infinity-theme',
    isDark: false
});
export const sweetTheme = new Theme({
    id: 'sweet-theme',
    name: '蜜糖主题',
    data: {
        ...infinityTheme.data,
        'devui-brand': '#ec66ab',
        'devui-brand-foil': '#f8f1f5',
        'devui-brand-active-focus': '#ec66ab',
        'devui-primary-active': '#ec66ab',
        'devui-brand-hover': '#ec66ab',
        'devui-global-bg': '#f9fafb',
        'devui-base-bg': '#ffffff',
        'devui-text': '#2f272f',
        'devui-aide-text': '#827d82',
        'devui-placeholder': '#bdb8bd',
        'devui-disabled-text': '#cbcacb',
        'devui-disabled-bg': '#f6f6f6',
        'devui-line': '#aea6ad',
        'devui-dividing-line': '#eae7e9',
        'devui-list-item-hover-bg': '#f8f1f5',
        'devui-list-item-active-bg': '#ffdcee',
        'devui-list-item-active-hover-bg': '#ffdcee',
        'devui-list-item-selected-bg': '#ffdcee',
        'devui-list-item-hover-text': '#252b3a',
        'devui-list-item-active-text': '#252b3a',
        'devui-form-control-line': '#D7D8DA',
        'devui-icon-text': '#babbc0',
        'devui-brand-active': '#ec66ab',
        'devui-primary': '#ec66ab',
        'devui-primary-hover': '#ec66ab',
        'devui-form-control-line-active': '#ec66ab',
        'devui-form-control-line-active-hover': '#ec66ab',
        'devui-icon-fill-active': '#ec66ab',
        'devui-icon-fill-active-hover': '#ec66ab',
        'devui-label-bg': '#ffdcee',
        'devui-embed-search-bg': '#ffdcee',
        'devui-connected-overlay-line': '#ec66ab',
        'devui-primary-disabled': '#fad1e6',
        'devui-icon-fill-active-disabled': '#fad1e6',
        'devui-link-active': '#96114d'
    },
    extends: 'infinity-theme',
    isDark: false
});
export const deepTheme = new Theme({
    id: 'deep-theme',
    name: '深邃夜空主题',
    data: {
        ...infinityTheme.data,
        'devui-brand': '#252b3a',
        'devui-brand-foil': '#f3f4f7',
        'devui-brand-active-focus': '#252b3a',
        'devui-primary-active': '#252b3a',
        'devui-brand-active': '#252b3a',
        'devui-brand-hover': '#252b3a',
        'devui-global-bg': '#f7f8fa',
        'devui-base-bg': '#ffffff',
        'devui-text': '#252b3a',
        'devui-aide-text': '#505c7c',
        'devui-placeholder': '#9ba6bf',
        'devui-disabled-text': '#a8b1c7',
        'devui-disabled-bg': '#f7f8fa',
        'devui-line': '#cdd2df',
        'devui-dividing-line': '#e6e9ef',
        'devui-list-item-hover-bg': '#f3f4f7',
        'devui-list-item-active-bg': '#252b3a',
        'devui-list-item-active-hover-bg': '#252b3a',
        'devui-list-item-selected-bg': '#252b3a',
        'devui-list-item-hover-text': '#252b3a',
        'devui-list-item-active-text': '#ffffff',
        'devui-form-control-line': '#D7D8DA',
        'devui-icon-text': '#babbc0',
        'devui-primary': '#252b3a',
        'devui-primary-hover': '#252b3a',
        'devui-form-control-line-active': '#252b3a',
        'devui-form-control-line-active-hover': '#252b3a',
        'devui-icon-fill-active': '#252b3a',
        'devui-icon-fill-active-hover': '#252b3a',
        'devui-connected-overlay-line': '#252b3a',
        'devui-primary-disabled': '#bebfc4',
        'devui-icon-fill-active-disabled': '#bebfc4',
    },
    extends: 'infinity-theme',
    isDark: false
});
export const galaxyTheme = new Theme({
    id: 'galaxy-theme',
    name: '追光主题',
    data: {
        ...devuiDarkTheme.data,
        ...devuiCustomComponentsDarkThemeData,
        // use '526ECC' for active status except 'devui-list-item-active-bg'
        'devui-brand': '#526ECC',
        'devui-info': '#526ECC',
        'devui-icon-fill-active': '#526ECC',
        'devui-primary': '#526ECC',
        'devui-default-line': '#526ECC',
        'devui-brand-hover': '#5E7CE0',
        'devui-primary-hover': '#5E7CE0',
        'devui-brand-foil': '#F2F2F3',
        'devui-global-bg': '#1A1A1C',
        'devui-base-bg': '#292A2E',
        'devui-text': '#CED1DB',
        'devui-aide-text': '#868A99',
        'devui-placeholder': '#4E5057',
        'devui-disabled-text': '#48494D',
        'devui-disabled-bg': '#2D2E32',
        'devui-line': '#3C3D42',
        'devui-dividing-line': '#323438',
        'devui-list-item-hover-bg': '#393A3E',
        'devui-list-item-active-bg': '#303548',
        'devui-list-item-active-hover-bg': '#303548',
        'devui-list-item-selected-bg': '#303548',
        'devui-list-item-hover-text': '#F5F5F5',
        'devui-list-item-active-text': '#4660B8',
        'devui-primary-disabled': '#3f3f3f',
        'devui-form-control-line': '#3c3d42',
        'devui-form-control-bg': '#292A2E',
        'devui-form-control-line-hover': '#696c75',
        'devui-form-control-line-active': '#526ECC',
        'devui-icon-text': '#A3A3A3',
        'devui-connected-overlay-bg': '#2E2F33',
        'devui-fullscreen-overlay-bg': '#2E2F33',
        'devui-warning-line': '#EB8F3D',
        'devui-warning-bg': '#4b2e14',
        'devui-success-line': '#2AA37E ',
        'devui-success-bg': '#123d32',
        'devui-danger-line': '#D15854',
        'devui-danger-bg': '#4a2120',
        'devui-info-line': '#4482BD',
        'devui-info-bg': '#1c2543',
        'devui-default-bg': '#313131',
        'devui-label-bg': '#353537',
        'devui-border-radius': '4px',
        'devui-font-size': '14px',
        'devui-font-size-md': '14px',
        'devui-font-size-card-title': '16px',
        'devui-shadow-length-fullscreen-overlay': '0 0 6px 0',
        'devui-border-radius-card': '4px'
    },
    extends: 'devui-dark-theme',
    isDark: true
});
