// ng-devui/esm2020/checkbox/checkbox.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CheckBoxGroupComponent } from './checkbox-group.component';
import { CheckBoxComponent } from './checkbox.component';
import * as i0 from "@angular/core";
export class CheckBoxModule {
}
CheckBoxModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CheckBoxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
CheckBoxModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CheckBoxModule, declarations: [CheckBoxComponent, CheckBoxGroupComponent], imports: [CommonModule, FormsModule], exports: [CheckBoxComponent, CheckBoxGroupComponent] });
CheckBoxModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CheckBoxModule, providers: [], imports: [[CommonModule, FormsModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CheckBoxModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, FormsModule],
                    exports: [CheckBoxComponent, CheckBoxGroupComponent],
                    declarations: [CheckBoxComponent, CheckBoxGroupComponent],
                    providers: [],
                }]
        }] });
