// ng-devui/esm2020/pagination/pagination.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { DropDownModule } from 'ng-devui/dropdown';
import { IconModule } from 'ng-devui/icon';
import { SelectModule } from 'ng-devui/select';
import { TextInputModule } from 'ng-devui/text-input';
import { SafePipeModule } from 'ng-devui/utils';
import { PaginationComponent } from './pagination.component';
import * as i0 from "@angular/core";
export class PaginationModule {
}
PaginationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PaginationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PaginationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PaginationModule, declarations: [PaginationComponent], imports: [CommonModule,
        SelectModule,
        FormsModule,
        SafePipeModule,
        TextInputModule,
        IconModule,
        DropDownModule], exports: [PaginationComponent] });
PaginationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PaginationModule, imports: [[
            CommonModule,
            SelectModule,
            FormsModule,
            SafePipeModule,
            TextInputModule,
            IconModule,
            DropDownModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: PaginationModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        SelectModule,
                        FormsModule,
                        SafePipeModule,
                        TextInputModule,
                        IconModule,
                        DropDownModule
                    ],
                    declarations: [PaginationComponent],
                    exports: [PaginationComponent]
                }]
        }] });
