// ng-devui/esm2020/select/select.module.mjs
import { OverlayModule } from '@angular/cdk/overlay';
import { ScrollingModule } from '@angular/cdk/scrolling';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CheckBoxModule } from 'ng-devui/checkbox';
import { LoadingModule } from 'ng-devui/loading';
import { TagsModule } from 'ng-devui/tags';
import { LazyLoadModule } from 'ng-devui/utils';
import { WindowRefModule } from 'ng-devui/window-ref';
import { SelectComponent } from './select.component';
import * as i0 from "@angular/core";
export class SelectModule {
}
SelectModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: SelectModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SelectModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: SelectModule, declarations: [SelectComponent], imports: [CommonModule,
        FormsModule,
        OverlayModule,
        ScrollingModule,
        CheckBoxModule,
        LoadingModule,
        LazyLoadModule,
        WindowRefModule,
        TagsModule], exports: [SelectComponent] });
SelectModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: SelectModule, imports: [[
            CommonModule,
            FormsModule,
            OverlayModule,
            ScrollingModule,
            CheckBoxModule,
            LoadingModule,
            LazyLoadModule,
            WindowRefModule,
            TagsModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: SelectModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        OverlayModule,
                        ScrollingModule,
                        CheckBoxModule,
                        LoadingModule,
                        LazyLoadModule,
                        WindowRefModule,
                        TagsModule,
                    ],
                    exports: [SelectComponent],
                    declarations: [SelectComponent],
                }]
        }] });
