// ng-devui/esm2020/editable-select/editable-select.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { AutoCompleteModule } from 'ng-devui/auto-complete';
import { PositioningModule } from 'ng-devui/position';
import { EditableSelectComponent } from './editable-select.component';
import * as i0 from "@angular/core";
export class EditableSelectModule {
}
EditableSelectModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: EditableSelectModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
EditableSelectModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: EditableSelectModule, declarations: [EditableSelectComponent], imports: [CommonModule, FormsModule, AutoCompleteModule, PositioningModule], exports: [EditableSelectComponent] });
EditableSelectModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: EditableSelectModule, imports: [[CommonModule, FormsModule, AutoCompleteModule, PositioningModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: EditableSelectModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, FormsModule, AutoCompleteModule, PositioningModule],
                    exports: [EditableSelectComponent],
                    declarations: [EditableSelectComponent],
                }]
        }] });
