// ng-devui/esm2020/dragdrop/services/drag-drop-desc-reg.service.mjs
import { Directive, Injectable, QueryList } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { debounceTime } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class DescendantRegisterService {
    constructor() {
        this._result = [];
        this.changeSubject = new BehaviorSubject([]);
        this.changes = this.changeSubject.asObservable().pipe(debounceTime(200));
    }
    register(t) {
        if (!t) {
            return;
        }
        const index = this._result.indexOf(t);
        if (index === -1) {
            this._result.push(t);
            this.changeSubject.next(this._result);
        }
    }
    unregister(t) {
        if (!t) {
            return;
        }
        const index = this._result.indexOf(t);
        if (index > -1) {
            this._result.splice(index, 1);
            this.changeSubject.next(this._result);
        }
    }
    queryResult() {
        return this._result.concat([]);
    }
}
DescendantRegisterService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DescendantRegisterService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
DescendantRegisterService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DescendantRegisterService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DescendantRegisterService, decorators: [{
            type: Injectable
        }] });
/* eslint-disable-next-line @angular-eslint/directive-class-suffix */
export class DescendantChildren {
    constructor(drs) {
        this.drs = drs;
    }
    ngOnInit() {
        this.drs.register(this.descendantItem);
    }
    ngOnDestroy() {
        this.drs.unregister(this.descendantItem);
    }
}
DescendantChildren.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DescendantChildren, deps: [{ token: DescendantRegisterService }], target: i0.ɵɵFactoryTarget.Directive });
DescendantChildren.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: DescendantChildren, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DescendantChildren, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: DescendantRegisterService }]; } });
export class DescendantRoot extends QueryList {
    constructor(drs) {
        super();
        this.drs = drs;
    }
    on() {
        if (this.sub) {
            return;
        }
        this.reset(this.drs.queryResult());
        this.sub = this.drs.changes.subscribe(result => { this.reset(result); });
    }
    off() {
        if (this.sub) {
            this.sub.unsubscribe();
        }
    }
}
