// ng-devui/esm2020/dragdrop/services/drag-drop-sync.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class DragDropSyncService {
    updateDragSyncList(list) {
        this.dragSyncList = list;
    }
    getDragSyncByGroup(groupName) {
        if (groupName === '') {
            return [];
        }
        return this.dragSyncList ? this.dragSyncList.filter(dragSync => dragSync.dragSyncGroup === groupName) : [];
    }
    updateDropSyncList(list) {
        this.dropSortSyncList = list;
    }
    getDropSyncByGroup(groupName) {
        if (groupName === '') {
            return [];
        }
        return this.dropSortSyncList ? this.dropSortSyncList.filter(dropSync => dropSync.dropSyncGroup === groupName) : [];
    }
}
DragDropSyncService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DragDropSyncService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
DragDropSyncService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DragDropSyncService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DragDropSyncService, decorators: [{
            type: Injectable
        }] });
