// ng-devui/esm2020/read-tip/read-tip.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { OverlayContainerModule } from 'ng-devui/overlay-container';
import { PositioningModule } from 'ng-devui/position';
import { ReadTipComponent } from './read-tip.component';
import { ReadTipDirective } from './read-tip.directive';
import * as i0 from "@angular/core";
export class ReadTipModule {
}
ReadTipModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: ReadTipModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ReadTipModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: ReadTipModule, declarations: [ReadTipComponent, ReadTipDirective], imports: [CommonModule, PositioningModule, OverlayContainerModule], exports: [ReadTipComponent, ReadTipDirective] });
ReadTipModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: ReadTipModule, imports: [[CommonModule, PositioningModule, OverlayContainerModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: ReadTipModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, PositioningModule, OverlayContainerModule],
                    exports: [ReadTipComponent, ReadTipDirective],
                    declarations: [ReadTipComponent, ReadTipDirective],
                }]
        }] });
