// ng-devui/fesm2015/ng-devui-card.mjs
import { CommonModule } from '@angular/common';
import * as i0 from '@angular/core';
import { Directive, HostBinding, Input, Component, ViewEncapsulation, ChangeDetectionStrategy, NgModule } from '@angular/core';

class CardContentDirective {
    constructor() {
        this.default = true;
    }
}
CardContentDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardContentDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
CardContentDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: CardContentDirective, selector: "d-card-content, [dCardContent]", host: { properties: { "class.devui-card-content": "this.default" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardContentDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'd-card-content, [dCardContent]'
                }]
        }], propDecorators: { default: [{
                type: HostBinding,
                args: ['class.devui-card-content']
            }] } });
class CardTitleDirective {
    constructor() {
        this.default = true;
    }
}
CardTitleDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardTitleDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
CardTitleDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: CardTitleDirective, selector: "d-card-title, [dCardTitle]", host: { properties: { "class.devui-card-title": "this.default" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardTitleDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: `d-card-title, [dCardTitle]`
                }]
        }], propDecorators: { default: [{
                type: HostBinding,
                args: ['class.devui-card-title']
            }] } });
class CardSubtitleDirective {
    constructor() {
        this.default = true;
    }
}
CardSubtitleDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardSubtitleDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
CardSubtitleDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: CardSubtitleDirective, selector: "d-card-subtitle, [dCardSubtitle]", host: { properties: { "class.devui-card-subtitle": "this.default" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardSubtitleDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: `d-card-subtitle, [dCardSubtitle]`,
                }]
        }], propDecorators: { default: [{
                type: HostBinding,
                args: ['class.devui-card-subtitle']
            }] } });
class CardActionsDirective {
    constructor() {
        this.align = 'start';
        this.default = true;
    }
    get alignEnd() {
        return this.align === 'end';
    }
    get alignSpaceBetween() {
        return this.align === 'spaceBetween';
    }
}
CardActionsDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardActionsDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
CardActionsDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: CardActionsDirective, selector: "d-card-actions,[dCardActions]", inputs: { align: "align" }, host: { properties: { "class.devui-card-actions": "this.default", "class.devui-card-actions-align-end": "this.alignEnd", "class.devui-card-actions-align-space-between": "this.alignSpaceBetween" } }, exportAs: ["dCardActions"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardActionsDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'd-card-actions,[dCardActions]',
                    exportAs: 'dCardActions',
                }]
        }], propDecorators: { align: [{
                type: Input
            }], default: [{
                type: HostBinding,
                args: ['class.devui-card-actions']
            }], alignEnd: [{
                type: HostBinding,
                args: ['class.devui-card-actions-align-end']
            }], alignSpaceBetween: [{
                type: HostBinding,
                args: ['class.devui-card-actions-align-space-between']
            }] } });
class CardMetaDirective {
    constructor() {
        this.default = true;
    }
}
CardMetaDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardMetaDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
CardMetaDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: CardMetaDirective, selector: "[dCardMeta]", host: { properties: { "class.devui-card-meta": "this.default" } }, exportAs: ["dCardMeta"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardMetaDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[dCardMeta]',
                    exportAs: 'dCardMeta'
                }]
        }], propDecorators: { default: [{
                type: HostBinding,
                args: ['class.devui-card-meta']
            }] } });
class CardAvatarDirective {
    constructor() {
        this.default = true;
    }
}
CardAvatarDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardAvatarDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
CardAvatarDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: CardAvatarDirective, selector: "[dCardAvatar]", host: { properties: { "class.devui-card-avatar": "this.default" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardAvatarDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[dCardAvatar]'
                }]
        }], propDecorators: { default: [{
                type: HostBinding,
                args: ['class.devui-card-avatar']
            }] } });
class CardComponent {
    constructor() {
        this.default = true;
    }
}
CardComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
CardComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: CardComponent, selector: "d-card", host: { properties: { "class.devui-card": "this.default" } }, exportAs: ["dCard"], ngImport: i0, template: "<ng-content></ng-content>\n", styles: [".devui-font-size-base{font-size:12px;font-size:var(--devui-font-size, 12px)}.devui-font-base{font-size:12px;font-size:var(--devui-font-size, 12px);font-weight:400;line-height:1.5}.devui-font-size-modal-title{font-size:18px;font-size:var(--devui-font-size-modal-title, 18px)}.devui-font-modal-title{font-size:18px;font-size:var(--devui-font-size-modal-title, 18px);font-weight:700;line-height:1.5}.devui-font-size-page-title{font-size:16px;font-size:var(--devui-font-size-page-title, 16px)}.devui-font-page-title{font-size:16px;font-size:var(--devui-font-size-page-title, 16px);font-weight:700;line-height:1.5}.devui-font-size-secondary-title{font-size:14px;font-size:var(--devui-font-size-card-title, 14px)}.devui-font-secondary-title{font-size:14px;font-size:var(--devui-font-size-card-title, 14px);font-weight:700;line-height:1.5}.devui-card{position:relative;display:block;padding:16px 20px;border-radius:6px;border-radius:var(--devui-border-radius-card, 6px);box-shadow:0 1px 4px rgba(37,43,58,.1);box-shadow:var(--devui-shadow-length-base, 0 1px 4px 0) var(--devui-light-shadow, rgba(37, 43, 58, .1))}.devui-card-title{display:block;font-size:14px;font-size:var(--devui-font-size-card-title, 14px);font-weight:700;color:#252b3a;color:var(--devui-text, #252b3a)}.devui-card-subtitle{display:block;font-size:12px;font-size:var(--devui-font-size, 12px);color:#575d6c;color:var(--devui-aide-text-stress, #575d6c)}.devui-card-header{display:flex;flex-direction:row;align-items:center;margin-bottom:16px;margin-bottom:var(--card-block-space, 16px)}.devui-card-avatar{margin-right:16px;margin-right:var(--card-block-space, 16px)}.devui-card-content{font-size:12px;font-size:var(--devui-font-size, 12px);color:#71757f;color:var(--devui-aide-text, #71757f)}.devui-card-actions{margin-top:16px;margin-top:var(--card-block-space, 16px);display:flex;justify-content:flex-start;align-items:center}.devui-card-title+.devui-card-subtitle{margin-top:8px;margin-top:var(--card-ele-space, 8px)}.devui-card-actions-align-end{justify-content:flex-end}.devui-card-actions-align-space-between{justify-content:space-between}.devui-card-meta{width:calc(100% + 40px);margin:0 -20px 16px;margin:0 -20px var(--card-block-space, 16px) -20px;height:auto}.devui-card-meta:first-child{margin-top:-var(--card-block-space,16px)}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardComponent, decorators: [{
            type: Component,
            args: [{ selector: 'd-card', exportAs: 'dCard', encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, template: "<ng-content></ng-content>\n", styles: [".devui-font-size-base{font-size:12px;font-size:var(--devui-font-size, 12px)}.devui-font-base{font-size:12px;font-size:var(--devui-font-size, 12px);font-weight:400;line-height:1.5}.devui-font-size-modal-title{font-size:18px;font-size:var(--devui-font-size-modal-title, 18px)}.devui-font-modal-title{font-size:18px;font-size:var(--devui-font-size-modal-title, 18px);font-weight:700;line-height:1.5}.devui-font-size-page-title{font-size:16px;font-size:var(--devui-font-size-page-title, 16px)}.devui-font-page-title{font-size:16px;font-size:var(--devui-font-size-page-title, 16px);font-weight:700;line-height:1.5}.devui-font-size-secondary-title{font-size:14px;font-size:var(--devui-font-size-card-title, 14px)}.devui-font-secondary-title{font-size:14px;font-size:var(--devui-font-size-card-title, 14px);font-weight:700;line-height:1.5}.devui-card{position:relative;display:block;padding:16px 20px;border-radius:6px;border-radius:var(--devui-border-radius-card, 6px);box-shadow:0 1px 4px rgba(37,43,58,.1);box-shadow:var(--devui-shadow-length-base, 0 1px 4px 0) var(--devui-light-shadow, rgba(37, 43, 58, .1))}.devui-card-title{display:block;font-size:14px;font-size:var(--devui-font-size-card-title, 14px);font-weight:700;color:#252b3a;color:var(--devui-text, #252b3a)}.devui-card-subtitle{display:block;font-size:12px;font-size:var(--devui-font-size, 12px);color:#575d6c;color:var(--devui-aide-text-stress, #575d6c)}.devui-card-header{display:flex;flex-direction:row;align-items:center;margin-bottom:16px;margin-bottom:var(--card-block-space, 16px)}.devui-card-avatar{margin-right:16px;margin-right:var(--card-block-space, 16px)}.devui-card-content{font-size:12px;font-size:var(--devui-font-size, 12px);color:#71757f;color:var(--devui-aide-text, #71757f)}.devui-card-actions{margin-top:16px;margin-top:var(--card-block-space, 16px);display:flex;justify-content:flex-start;align-items:center}.devui-card-title+.devui-card-subtitle{margin-top:8px;margin-top:var(--card-ele-space, 8px)}.devui-card-actions-align-end{justify-content:flex-end}.devui-card-actions-align-space-between{justify-content:space-between}.devui-card-meta{width:calc(100% + 40px);margin:0 -20px 16px;margin:0 -20px var(--card-block-space, 16px) -20px;height:auto}.devui-card-meta:first-child{margin-top:-var(--card-block-space,16px)}\n"] }]
        }], propDecorators: { default: [{
                type: HostBinding,
                args: ['class.devui-card']
            }] } });
class CardHeaderComponent {
    constructor() {
        this.default = true;
    }
}
CardHeaderComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardHeaderComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
CardHeaderComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: CardHeaderComponent, selector: "d-card-header", host: { properties: { "class.devui-card-header": "this.default" } }, ngImport: i0, template: "<ng-content select=\"[d-card-avatar], [dCardAvatar]\"></ng-content>\n<div class=\"devui-card-header-title-area\">\n  <ng-content\n    select=\"d-card-title, d-card-subtitle,\n      [d-card-title], [d-card-subtitle],\n      [dCardTitle], [dCardSubtitle]\"\n  ></ng-content>\n</div>\n", changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardHeaderComponent, decorators: [{
            type: Component,
            args: [{ selector: 'd-card-header', encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, template: "<ng-content select=\"[d-card-avatar], [dCardAvatar]\"></ng-content>\n<div class=\"devui-card-header-title-area\">\n  <ng-content\n    select=\"d-card-title, d-card-subtitle,\n      [d-card-title], [d-card-subtitle],\n      [dCardTitle], [dCardSubtitle]\"\n  ></ng-content>\n</div>\n" }]
        }], propDecorators: { default: [{
                type: HostBinding,
                args: ['class.devui-card-header']
            }] } });
// waiting for design
class CardExtendComponent {
    constructor() {
        this.default = true;
    }
}
CardExtendComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardExtendComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
CardExtendComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: CardExtendComponent, selector: "d-card-extend", host: { properties: { "class.devui-card-extend": "this.default" } }, ngImport: i0, template: "<!-- waiting for design -->\n", changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardExtendComponent, decorators: [{
            type: Component,
            args: [{ selector: 'd-card-extend', encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, template: "<!-- waiting for design -->\n" }]
        }], propDecorators: { default: [{
                type: HostBinding,
                args: ['class.devui-card-extend']
            }] } });

class CardModule {
}
CardModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
CardModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardModule, declarations: [CardComponent, CardHeaderComponent, CardAvatarDirective, CardMetaDirective,
        CardActionsDirective, CardSubtitleDirective, CardTitleDirective, CardContentDirective, CardExtendComponent], imports: [CommonModule], exports: [CardComponent, CardHeaderComponent, CardAvatarDirective, CardMetaDirective,
        CardActionsDirective, CardSubtitleDirective, CardTitleDirective, CardContentDirective, CardExtendComponent] });
CardModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardModule, imports: [[
            CommonModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: CardModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [CardComponent, CardHeaderComponent, CardAvatarDirective, CardMetaDirective,
                        CardActionsDirective, CardSubtitleDirective, CardTitleDirective, CardContentDirective, CardExtendComponent],
                    imports: [
                        CommonModule
                    ],
                    exports: [CardComponent, CardHeaderComponent, CardAvatarDirective, CardMetaDirective,
                        CardActionsDirective, CardSubtitleDirective, CardTitleDirective, CardContentDirective, CardExtendComponent]
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { CardActionsDirective, CardAvatarDirective, CardComponent, CardContentDirective, CardExtendComponent, CardHeaderComponent, CardMetaDirective, CardModule, CardSubtitleDirective, CardTitleDirective };
