// ng-devui/fesm2015/ng-devui-theme-collection.mjs
import { Theme, devuiDarkTheme } from 'ng-devui/theme';

const devuiCustomComponentsThemeData = {
    'devui-btn-sm-padding': '0 12px',
    'devui-btn-padding': '0 16px',
    'devui-btn-lg-padding': '0 20px',
    'devui-btn-common-bg': '#EBEBEB',
    'devui-btn-common-border-color': 'transparent',
    'devui-btn-common-bg-hover': '#D1D1D1',
    'devui-btn-common-color-hover': '#252B3A',
    'devui-btn-common-border-color-hover': 'transparent',
    'devui-btn-common-bg-active': '#BDBDBD',
    'devui-btn-common-color-active': '#252B3A',
    'devui-btn-common-border-color-active': 'transparent',
    'devui-btn-common-border-disabled': 'transparent',
    'devui-icon-hover-bg': '#EBEBEB',
    'devui-icon-active-bg': '#D1D1D1'
};
const devuiCustomComponentsDarkThemeData = {
    'devui-btn-sm-padding': '0 12px',
    'devui-btn-padding': '0 16px',
    'devui-btn-lg-padding': '0 20px',
    'devui-btn-common-bg': '#393a3e',
    'devui-btn-common-border-color': 'transparent',
    'devui-btn-common-bg-hover': '#4e5057',
    'devui-btn-common-color-hover': '#CED1DB',
    'devui-btn-common-border-color-hover': 'transparent',
    'devui-btn-common-bg-active': '#4e5057',
    'devui-btn-common-color-active': '#CED1DB',
    'devui-btn-common-border-color-active': 'transparent',
    'devui-btn-common-border-disabled': 'transparent'
};

const devuiInitialThemeData = {
    'devui-global-bg': '#f8f8fa',
    'devui-global-bg-normal': '#ffffff',
    'devui-base-bg': '#ffffff',
    'devui-base-bg-dark': '#252b3a',
    'devui-brand': '#5e7ce0',
    'devui-brand-foil': '#f2f2f3',
    'devui-brand-hover': '#7693f5',
    'devui-brand-active': '#526ecc',
    'devui-brand-active-focus': '#344899',
    'devui-contrast': '#c7000b',
    'devui-text': '#252b3a',
    'devui-text-weak': '#575d6c',
    'devui-aide-text': '#71757f',
    'devui-aide-text-hover': '#252b3a',
    'devui-aide-text-stress': '#575d6c',
    'devui-placeholder': '#babbc0',
    'devui-light-text': '#ffffff',
    'devui-dark-text': '#252b3a',
    'devui-link': '#526ecc',
    'devui-link-active': '#526ecc',
    'devui-link-light': '#96adfa',
    'devui-link-light-active': '#beccfa',
    'devui-line': '#d7d8da',
    'devui-dividing-line': '#f2f2f3',
    'devui-block': '#ffffff',
    'devui-area': '#f5f5f5',
    'devui-danger': '#f66f6a',
    'devui-warning': '#fac20a',
    'devui-waiting': '#beccfa',
    'devui-success': '#50d4ab',
    'devui-info': '#5e7ce0',
    'devui-initial': '#e9edfa',
    'devui-unavailable': '#f5f5f5',
    'devui-shadow': 'rgba(37, 43, 58, 0.2)',
    'devui-light-shadow': 'rgba(37, 43, 58, 0.1)',
    'devui-connected-overlay-shadow': 'rgba(37, 43, 58, 0.12)',
    'devui-feedback-overlay-shadow': 'rgba(37, 43, 58, 0.16)',
    'devui-icon-text': '#71757f',
    'devui-icon-bg': '#ffffff',
    'devui-icon-fill': '#71757f',
    'devui-icon-fill-weak': '#babbc0',
    'devui-icon-fill-hover': '#252b3a',
    'devui-icon-fill-active': '#252b3a',
    'devui-icon-fill-active-hover': '#252b3a',
    'devui-shape-icon-fill': '#d7d8da',
    'devui-shape-icon-fill-hover': '#babbc0',
    'devui-shape-icon-fill-active': '#babbc0',
    'devui-shape-icon-fill-disabled': '#f5f5f5',
    'devui-form-control-line': '#d7d8da',
    'devui-form-control-bg': '#ffffff',
    'devui-form-control-line-hover': '#9b9fa8',
    'devui-form-control-line-active': '#5e7ce0',
    'devui-form-control-line-active-hover': '#344899',
    'devui-list-item-active-bg': '#f2f5fc',
    'devui-list-item-active-text': '#252b3a',
    'devui-list-item-active-hover-bg': '#f2f5fc',
    'devui-list-item-hover-bg': '#f2f2f3',
    'devui-list-item-hover-text': '#252b3a',
    'devui-list-item-selected-bg': '#f2f5fc',
    'devui-list-item-strip-bg': '#f2f5fc',
    'devui-disabled-bg': '#f5f5f5',
    'devui-disabled-line': '#dfe1e6',
    'devui-disabled-text': '#cfd0d3',
    'devui-primary-disabled': '#beccfa',
    'devui-icon-fill-active-disabled': '#beccfa',
    'devui-label-bg': '#e9edfa',
    'devui-connected-overlay-bg': '#ffffff',
    'devui-connected-overlay-line': '#526ecc',
    'devui-fullscreen-overlay-bg': '#ffffff',
    'devui-feedback-overlay-bg': '#464d6e',
    'devui-feedback-overlay-text': '#dfe1e6',
    'devui-embed-search-bg': '#f2f2f3',
    'devui-embed-search-bg-hover': '#eef0f5',
    'devui-float-block-shadow': 'rgba(94, 124, 224, 0.3)',
    'devui-highlight-overlay': 'rgba(255, 255, 255, 0.8)',
    'devui-range-item-hover-bg': '#e9edfa',
    'devui-primary': '#5e7ce0',
    'devui-primary-hover': '#7693f5',
    'devui-primary-active': '#344899',
    'devui-contrast-hover': '#d64a52',
    'devui-contrast-active': '#b12220',
    'devui-danger-line': '#f66f6a',
    'devui-danger-bg': '#ffd5d4',
    'devui-warning-line': '#fa9841',
    'devui-warning-bg': '#ffe1c7',
    'devui-info-line': '#5e7ce0',
    'devui-info-bg': '#e9edfa',
    'devui-success-line': '#50d4ab',
    'devui-success-bg': '#cffcee',
    'devui-primary-line': '#5e7ce0',
    'devui-primary-bg': '#f2f5fc',
    'devui-default-line': '#5e7ce0',
    'devui-default-bg': '#f7f8fa',
    'devui-shadow-length-base': '0 1px 4px 0',
    'devui-shadow-length-slide-left': '-2px 0 8px 0',
    'devui-shadow-length-slide-right': '2px 0 8px 0',
    'devui-shadow-length-connected-overlay': '0 4px 8px 0',
    'devui-shadow-length-hover': '0 8px 16px 0',
    'devui-shadow-length-feedback-overlay': '0 8px 16px 0',
    'devui-shadow-length-fullscreen-overlay': '0 12px 24px 0',
    'devui-border-radius': '4px',
    'devui-border-radius-feedback': '4px',
    'devui-border-radius-card': '8px',
    'devui-border-radius-full': '100px',
    'devui-font-size': '14px',
    'devui-font-size-card-title': '16px',
    'devui-font-size-page-title': '16px',
    'devui-font-size-modal-title': '18px',
    'devui-font-size-price': '20px',
    'devui-font-size-data-overview': '24px',
    'devui-font-size-icon': '16px',
    'devui-font-size-sm': '12px',
    'devui-font-size-md': '14px',
    'devui-font-size-lg': '14px',
    'devui-font-title-weight': 'bold',
    'devui-font-content-weight': 'normal',
    'devui-line-height-base': '1.5',
    'devui-animation-ease-in-smooth': 'cubic-bezier(0.645, 0.045, 0.355, 1)',
    'devui-animation-duration-slow': '300ms',
    'devui-animation-duration-base': '200ms',
    'devui-animation-duration-fast': '100ms',
    'devui-animation-ease-in': 'cubic-bezier(0.5, 0, 0.84, 0.25)',
    'devui-animation-ease-out': 'cubic-bezier(0.16, 0.75, 0.5, 1)',
    'devui-animation-ease-in-out': 'cubic-bezier(0.5, 0.05, 0.5, 0.95)',
    'devui-animation-ease-in-out-smooth': 'cubic-bezier(0.645, 0.045, 0.355, 1)',
    'devui-animation-linear': 'cubic-bezier(0, 0, 1, 1)',
    'devui-z-index-full-page-overlay': '1080',
    'devui-z-index-pop-up': '1060',
    'devui-z-index-dropdown': '1052',
    'devui-z-index-modal': '1050',
    'devui-z-index-drawer': '1040',
    'devui-z-index-framework': '1000',
};

const infinityTheme = new Theme({
    id: 'infinity-theme',
    name: '无限主题',
    data: Object.assign(Object.assign({}, devuiInitialThemeData), devuiCustomComponentsThemeData),
    extends: 'devui-light-theme',
    isDark: false,
});
const provenceTheme = new Theme({
    id: 'provence-theme',
    name: '紫罗兰主题',
    data: Object.assign(Object.assign({}, infinityTheme.data), { 'devui-brand': '#7B69EE', 'devui-brand-foil': '#F5F5F9', 'devui-brand-active-focus': '#7B69EE', 'devui-primary-active': '#7B69EE', 'devui-brand-hover': '#7B69EE', 'devui-global-bg': '#f9fafb', 'devui-base-bg': '#ffffff', 'devui-text': '#070036', 'devui-aide-text': '#717087', 'devui-placeholder': '#babbc0', 'devui-disabled-text': '#cfd0d3', 'devui-disabled-bg': '#f5f5f6', 'devui-line': '#E2E2E5', 'devui-dividing-line': '#F2F2F3', 'devui-list-item-hover-bg': '#F5F5F9', 'devui-list-item-active-bg': '#7B69EE', 'devui-list-item-active-hover-bg': '#7B69EE', 'devui-list-item-selected-bg': '#F4F2FF', 'devui-list-item-hover-text': '#252b3a', 'devui-list-item-active-text': '#ffffff', 'devui-form-control-line': '#D7D8DA', 'devui-icon-text': '#babbc0', 'devui-brand-active': '#7B69EE', 'devui-primary': '#7B69EE', 'devui-primary-hover': '#7B69EE', 'devui-form-control-line-active': '#7B69EE', 'devui-form-control-line-active-hover': '#7B69EE', 'devui-icon-fill-active': '#7B69EE', 'devui-icon-fill-active-hover': '#7B69EE', 'devui-label-bg': '#F4F2FF', 'devui-embed-search-bg': '#F4F2FF', 'devui-connected-overlay-line': '#7B69EE', 'devui-primary-disabled': '#d8d2fa', 'devui-icon-fill-active-disabled': '#d8d2fa', 'devui-link-active': '#3f1a9c' }),
    extends: 'infinity-theme',
    isDark: false
});
const sweetTheme = new Theme({
    id: 'sweet-theme',
    name: '蜜糖主题',
    data: Object.assign(Object.assign({}, infinityTheme.data), { 'devui-brand': '#ec66ab', 'devui-brand-foil': '#f8f1f5', 'devui-brand-active-focus': '#ec66ab', 'devui-primary-active': '#ec66ab', 'devui-brand-hover': '#ec66ab', 'devui-global-bg': '#f9fafb', 'devui-base-bg': '#ffffff', 'devui-text': '#2f272f', 'devui-aide-text': '#827d82', 'devui-placeholder': '#bdb8bd', 'devui-disabled-text': '#cbcacb', 'devui-disabled-bg': '#f6f6f6', 'devui-line': '#aea6ad', 'devui-dividing-line': '#eae7e9', 'devui-list-item-hover-bg': '#f8f1f5', 'devui-list-item-active-bg': '#ffdcee', 'devui-list-item-active-hover-bg': '#ffdcee', 'devui-list-item-selected-bg': '#ffdcee', 'devui-list-item-hover-text': '#252b3a', 'devui-list-item-active-text': '#252b3a', 'devui-form-control-line': '#D7D8DA', 'devui-icon-text': '#babbc0', 'devui-brand-active': '#ec66ab', 'devui-primary': '#ec66ab', 'devui-primary-hover': '#ec66ab', 'devui-form-control-line-active': '#ec66ab', 'devui-form-control-line-active-hover': '#ec66ab', 'devui-icon-fill-active': '#ec66ab', 'devui-icon-fill-active-hover': '#ec66ab', 'devui-label-bg': '#ffdcee', 'devui-embed-search-bg': '#ffdcee', 'devui-connected-overlay-line': '#ec66ab', 'devui-primary-disabled': '#fad1e6', 'devui-icon-fill-active-disabled': '#fad1e6', 'devui-link-active': '#96114d' }),
    extends: 'infinity-theme',
    isDark: false
});
const deepTheme = new Theme({
    id: 'deep-theme',
    name: '深邃夜空主题',
    data: Object.assign(Object.assign({}, infinityTheme.data), { 'devui-brand': '#252b3a', 'devui-brand-foil': '#f3f4f7', 'devui-brand-active-focus': '#252b3a', 'devui-primary-active': '#252b3a', 'devui-brand-active': '#252b3a', 'devui-brand-hover': '#252b3a', 'devui-global-bg': '#f7f8fa', 'devui-base-bg': '#ffffff', 'devui-text': '#252b3a', 'devui-aide-text': '#505c7c', 'devui-placeholder': '#9ba6bf', 'devui-disabled-text': '#a8b1c7', 'devui-disabled-bg': '#f7f8fa', 'devui-line': '#cdd2df', 'devui-dividing-line': '#e6e9ef', 'devui-list-item-hover-bg': '#f3f4f7', 'devui-list-item-active-bg': '#252b3a', 'devui-list-item-active-hover-bg': '#252b3a', 'devui-list-item-selected-bg': '#252b3a', 'devui-list-item-hover-text': '#252b3a', 'devui-list-item-active-text': '#ffffff', 'devui-form-control-line': '#D7D8DA', 'devui-icon-text': '#babbc0', 'devui-primary': '#252b3a', 'devui-primary-hover': '#252b3a', 'devui-form-control-line-active': '#252b3a', 'devui-form-control-line-active-hover': '#252b3a', 'devui-icon-fill-active': '#252b3a', 'devui-icon-fill-active-hover': '#252b3a', 'devui-connected-overlay-line': '#252b3a', 'devui-primary-disabled': '#bebfc4', 'devui-icon-fill-active-disabled': '#bebfc4' }),
    extends: 'infinity-theme',
    isDark: false
});
const galaxyTheme = new Theme({
    id: 'galaxy-theme',
    name: '追光主题',
    data: Object.assign(Object.assign(Object.assign({}, devuiDarkTheme.data), devuiCustomComponentsDarkThemeData), { 
        // use '526ECC' for active status except 'devui-list-item-active-bg'
        'devui-brand': '#526ECC', 'devui-info': '#526ECC', 'devui-icon-fill-active': '#526ECC', 'devui-primary': '#526ECC', 'devui-default-line': '#526ECC', 'devui-brand-hover': '#5E7CE0', 'devui-primary-hover': '#5E7CE0', 'devui-brand-foil': '#F2F2F3', 'devui-global-bg': '#1A1A1C', 'devui-base-bg': '#292A2E', 'devui-text': '#CED1DB', 'devui-aide-text': '#868A99', 'devui-placeholder': '#4E5057', 'devui-disabled-text': '#48494D', 'devui-disabled-bg': '#2D2E32', 'devui-line': '#3C3D42', 'devui-dividing-line': '#323438', 'devui-list-item-hover-bg': '#393A3E', 'devui-list-item-active-bg': '#303548', 'devui-list-item-active-hover-bg': '#303548', 'devui-list-item-selected-bg': '#303548', 'devui-list-item-hover-text': '#F5F5F5', 'devui-list-item-active-text': '#4660B8', 'devui-primary-disabled': '#3f3f3f', 'devui-form-control-line': '#3c3d42', 'devui-form-control-bg': '#292A2E', 'devui-form-control-line-hover': '#696c75', 'devui-form-control-line-active': '#526ECC', 'devui-icon-text': '#A3A3A3', 'devui-connected-overlay-bg': '#2E2F33', 'devui-fullscreen-overlay-bg': '#2E2F33', 'devui-warning-line': '#EB8F3D', 'devui-warning-bg': '#4b2e14', 'devui-success-line': '#2AA37E ', 'devui-success-bg': '#123d32', 'devui-danger-line': '#D15854', 'devui-danger-bg': '#4a2120', 'devui-info-line': '#4482BD', 'devui-info-bg': '#1c2543', 'devui-default-bg': '#313131', 'devui-label-bg': '#353537', 'devui-border-radius': '4px', 'devui-font-size': '14px', 'devui-font-size-md': '14px', 'devui-font-size-card-title': '16px', 'devui-shadow-length-fullscreen-overlay': '0 0 6px 0', 'devui-border-radius-card': '4px' }),
    extends: 'devui-dark-theme',
    isDark: true
});

/**
 * Generated bundle index. Do not edit.
 */

export { deepTheme, galaxyTheme, infinityTheme, provenceTheme, sweetTheme };
