// ng-devui/esm2020/search/search.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { DCommonModule } from 'ng-devui/common';
import { SearchComponent } from './search.component';
import * as i0 from "@angular/core";
export class SearchModule {
}
SearchModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: SearchModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SearchModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: SearchModule, declarations: [SearchComponent], imports: [CommonModule,
        FormsModule,
        DCommonModule], exports: [SearchComponent] });
SearchModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: SearchModule, imports: [[
            CommonModule,
            FormsModule,
            DCommonModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: SearchModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        DCommonModule
                    ],
                    exports: [
                        SearchComponent,
                    ],
                    declarations: [
                        SearchComponent,
                    ]
                }]
        }] });
