// ng-devui/esm2020/anchor/anchor.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { AnchorBoxHashSupportDirective } from './anchor-box-hash.directive';
import { AnchorBoxDirective } from './anchor-box.directive';
import { AnchorLinkDirective } from './anchor-link.directive';
import { AnchorDirective } from './anchor.directive';
import * as i0 from "@angular/core";
export class AnchorModule {
}
AnchorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AnchorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AnchorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AnchorModule, declarations: [AnchorLinkDirective,
        AnchorBoxDirective,
        AnchorDirective,
        AnchorBoxHashSupportDirective], imports: [CommonModule,
        RouterModule], exports: [AnchorLinkDirective,
        AnchorBoxDirective,
        AnchorDirective,
        AnchorBoxHashSupportDirective] });
AnchorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AnchorModule, imports: [[
            CommonModule,
            RouterModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AnchorModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        RouterModule
                    ],
                    declarations: [
                        AnchorLinkDirective,
                        AnchorBoxDirective,
                        AnchorDirective,
                        AnchorBoxHashSupportDirective
                    ],
                    exports: [
                        AnchorLinkDirective,
                        AnchorBoxDirective,
                        AnchorDirective,
                        AnchorBoxHashSupportDirective
                    ],
                }]
        }] });
