// ng-devui/esm2020/anchor/anchor-box.directive.mjs
import { ContentChildren, Directive, Input, QueryList } from '@angular/core';
import { Subject, Subscription } from 'rxjs';
import { filter } from 'rxjs/operators';
import { AnchorDirective } from './anchor.directive';
import * as i0 from "@angular/core";
export class AnchorBoxDirective {
    constructor() {
        this.isScrollingToTarget = false;
        this.activeChangeSubject = new Subject();
        this.activeChange = this.activeChangeSubject.asObservable();
        this.refreshAnchorMap = new Subject();
    }
    set anchorList(list) {
        if (this.sub) {
            this.sub.unsubscribe();
        }
        this.sub = new Subscription();
        this.anchorMap = {};
        this._anchorList = list;
        this._anchorList.toArray().forEach(targetAnchor => {
            this.anchorMap[targetAnchor.anchor] = targetAnchor;
            targetAnchor.boxElement = this;
            this.sub.add(targetAnchor.activeChangeSubject.pipe(filter(bool => !!bool)).subscribe(() => {
                this.activeChangeSubject.next(targetAnchor);
            }));
        });
        this.refreshAnchorMap.next();
    }
    get anchorList() {
        return this._anchorList;
    }
    forceActiveAnchor(anchorName, forceActiveSource = 'scroll', deactivateOtherAnchor = true) {
        this.anchorMap[anchorName].activeChangeBy = forceActiveSource;
        this.anchorMap[anchorName].isActive = true;
        if (deactivateOtherAnchor) {
            Object.keys(this.anchorMap)
                .filter(name => name !== anchorName)
                .map(name => this.anchorMap[name])
                .forEach(anchor => {
                anchor.activeChangeBy = forceActiveSource;
                anchor.isActive = false;
            });
        }
    }
    ngOnDestroy() {
        this.activeChangeSubject.complete();
        if (this.sub) {
            this.sub.unsubscribe();
        }
    }
}
AnchorBoxDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AnchorBoxDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
AnchorBoxDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: AnchorBoxDirective, selector: "[dAnchorBox]", inputs: { view: "view", defaultAnchor: "defaultAnchor", scrollTarget: "scrollTarget" }, queries: [{ propertyName: "anchorList", predicate: AnchorDirective, descendants: true }], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AnchorBoxDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[dAnchorBox]'
                }]
        }], propDecorators: { view: [{
                type: Input
            }], defaultAnchor: [{
                type: Input
            }], scrollTarget: [{
                type: Input
            }], anchorList: [{
                type: ContentChildren,
                args: [AnchorDirective, { descendants: true }]
            }] } });
