// ng-devui/esm2020/anchor/anchor-box-hash.directive.mjs
import { Directive, Input } from '@angular/core';
import { ActivatedRoute, NavigationEnd, Router } from '@angular/router';
import { Subscription } from 'rxjs';
import { debounceTime, filter } from 'rxjs/operators';
import { AnchorBoxDirective } from './anchor-box.directive';
import { AnchorLinkDirective } from './anchor-link.directive';
import * as i0 from "@angular/core";
import * as i1 from "./anchor-box.directive";
import * as i2 from "@angular/router";
export class AnchorBoxHashSupportDirective {
    constructor(box, router, route) {
        this.box = box;
        this.router = router;
        this.route = route;
        this.updateUrlWhenAnchorActive = true;
        this.scrollToAnchorByHashOnlyInit = false;
        this.sub = new Subscription();
        this.manual = false;
        this.navigateToHash = (targetAnchor) => {
            if (targetAnchor.activeChangeBy === 'initial') {
                return;
            }
            this.router.navigate([], { fragment: targetAnchor.anchor, replaceUrl: true });
            this.manual = true;
        };
        this.navigateToAnchor = (event) => {
            if (this.manual) {
                this.manual = false;
                return;
            }
            if (this.scrollToAnchorByHashOnlyInit) {
                return;
            }
            const frag = this.router.parseUrl(event.url).fragment;
            this.scrollToFragment(frag);
        };
        this.scrollToFragment = (frag) => {
            if (!frag) {
                return;
            }
            if (this.box.anchorMap[frag]) {
                const tempAnchor = new AnchorLinkDirective(this.box);
                tempAnchor.anchorName = frag;
                tempAnchor.anchorBlock = this.box.anchorMap[frag];
                tempAnchor.scrollToAnchor('fragment');
            }
        };
    }
    ngOnInit() {
        this.sub.add(this.box.activeChange.pipe(debounceTime(300), filter(anchor => this.updateUrlWhenAnchorActive)).subscribe(this.navigateToHash));
        this.sub.add(this.router.events.pipe(filter(event => event instanceof NavigationEnd)).subscribe(this.navigateToAnchor));
    }
    ngAfterViewInit() {
        const frag = this.route.snapshot.fragment;
        setTimeout(() => {
            this.scrollToFragment(frag);
        }, 120);
    }
    ngOnDestroy() {
        if (this.sub) {
            this.sub.unsubscribe();
        }
    }
}
AnchorBoxHashSupportDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AnchorBoxHashSupportDirective, deps: [{ token: i1.AnchorBoxDirective }, { token: i2.Router }, { token: i2.ActivatedRoute }], target: i0.ɵɵFactoryTarget.Directive });
AnchorBoxHashSupportDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: AnchorBoxHashSupportDirective, selector: "[dAnchorBox][dAnchorHashSupport]", inputs: { updateUrlWhenAnchorActive: "updateUrlWhenAnchorActive", scrollToAnchorByHashOnlyInit: "scrollToAnchorByHashOnlyInit" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AnchorBoxHashSupportDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[dAnchorBox][dAnchorHashSupport]'
                }]
        }], ctorParameters: function () { return [{ type: i1.AnchorBoxDirective }, { type: i2.Router }, { type: i2.ActivatedRoute }]; }, propDecorators: { updateUrlWhenAnchorActive: [{
                type: Input
            }], scrollToAnchorByHashOnlyInit: [{
                type: Input
            }] } });
