// ng-devui/esm2020/utils/highlight/highlight.component.mjs
import { DOCUMENT } from '@angular/common';
import { ChangeDetectionStrategy, Component, ElementRef, HostBinding, Inject, Input } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import * as i0 from "@angular/core";
import * as i1 from "@angular/platform-browser";
export class HighlightComponent {
    constructor(translateHtml, eleRef, doc) {
        this.translateHtml = translateHtml;
        this.eleRef = eleRef;
        this.doc = doc;
        this.display = 'inline';
        /**
         * @deprecated
         */
        this.highlightClass = 'devui-match-highlight';
        this.document = this.doc;
    }
    ngOnChanges(changes) {
        this.addDom(this.value, this.term);
    }
    addDom(value, term) {
        if (value && term) {
            this.highlight(value, term);
        }
        else {
            const container = this.eleRef.nativeElement;
            this.emptyChildren(container);
            container.textContent = value;
        }
    }
    emptyChildren(container) {
        while (container.hasChildNodes()) {
            container.removeChild(container.firstChild);
        }
    }
    highlight(value, term) {
        const container = this.eleRef.nativeElement;
        this.emptyChildren(container);
        const reg = (str) => str.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&');
        const regExp = new RegExp('(' + reg(term) + ')', 'gi');
        const temp = value.split(regExp);
        const createHighLight = (text) => {
            const spanDOM = this.document.createElement('span');
            spanDOM.classList.add('devui-match-highlight');
            spanDOM.textContent = text;
            return spanDOM;
        };
        temp.forEach((element, index) => {
            if (index % 2 === 0) {
                container.appendChild(this.document.createTextNode(element));
            }
            else {
                container.appendChild(createHighLight(element));
            }
        });
    }
}
HighlightComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: HighlightComponent, deps: [{ token: i1.DomSanitizer }, { token: i0.ElementRef }, { token: DOCUMENT }], target: i0.ɵɵFactoryTarget.Component });
HighlightComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: HighlightComponent, selector: "d-highlight", inputs: { value: "value", term: "term", highlightClass: "highlightClass" }, host: { properties: { "style.display": "this.display" } }, usesOnChanges: true, ngImport: i0, template: ``, isInline: true, styles: [":host ::ng-deep .devui-match-highlight{font-weight:700}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: HighlightComponent, decorators: [{
            type: Component,
            args: [{ selector: 'd-highlight', template: ``, changeDetection: ChangeDetectionStrategy.OnPush, styles: [":host ::ng-deep .devui-match-highlight{font-weight:700}\n"] }]
        }], ctorParameters: function () { return [{ type: i1.DomSanitizer }, { type: i0.ElementRef }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }]; }, propDecorators: { display: [{
                type: HostBinding,
                args: ['style.display']
            }], value: [{
                type: Input
            }], term: [{
                type: Input
            }], highlightClass: [{
                type: Input
            }] } });
