// ng-devui/esm2020/dashboard/dashboard.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { DashboardComponent } from './dashboard.component';
import { DashboardLibraryPanelDirective } from './widget-library/library-panel.directive';
import { DashboardLibraryTrashDirective } from './widget-library/library-trash.directive';
import { DashboardLibraryWidgetDirective } from './widget-library/library-widget.directive';
import { DashboardWidgetComponent } from './widget/widget.component';
import * as i0 from "@angular/core";
export class DashboardModule {
}
DashboardModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DashboardModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DashboardModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DashboardModule, declarations: [DashboardComponent,
        DashboardWidgetComponent,
        DashboardLibraryWidgetDirective,
        DashboardLibraryTrashDirective,
        DashboardLibraryPanelDirective], imports: [CommonModule], exports: [DashboardComponent,
        DashboardWidgetComponent,
        DashboardLibraryWidgetDirective,
        DashboardLibraryTrashDirective,
        DashboardLibraryPanelDirective] });
DashboardModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DashboardModule, imports: [[
            CommonModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DashboardModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                    ],
                    exports: [
                        DashboardComponent,
                        DashboardWidgetComponent,
                        DashboardLibraryWidgetDirective,
                        DashboardLibraryTrashDirective,
                        DashboardLibraryPanelDirective,
                    ],
                    declarations: [
                        DashboardComponent,
                        DashboardWidgetComponent,
                        DashboardLibraryWidgetDirective,
                        DashboardLibraryTrashDirective,
                        DashboardLibraryPanelDirective,
                    ]
                }]
        }] });
