// ng-devui/esm2020/gantt/gantt.module.mjs
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ButtonModule } from 'ng-devui/button';
import { DropDownModule } from 'ng-devui/dropdown';
import { PopoverModule } from 'ng-devui/popover';
import { TooltipModule } from 'ng-devui/tooltip';
import { GanttBarParentComponent } from './gantt-bar-parent/gantt-bar-parent.component';
import { GanttBarComponent } from './gantt-bar/gantt-bar.component';
import { GanttMarkerDirective } from './gantt-marker.directive';
import { GanttMilestoneComponent } from './gantt-milestone/gantt-milestone.component';
import { GanttScaleComponent } from './gantt-scale/gantt-scale.component';
import { GanttToolsComponent } from './gantt-tools/gantt-tools.component';
import { GanttService } from './gantt.service';
import { ResizeHandleDirective } from './resize-handle.directive';
import * as i0 from "@angular/core";
export class GanttModule {
}
GanttModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: GanttModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
GanttModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: GanttModule, declarations: [GanttScaleComponent,
        GanttBarComponent,
        GanttMarkerDirective,
        GanttMilestoneComponent,
        GanttBarParentComponent,
        ResizeHandleDirective,
        GanttToolsComponent], imports: [CommonModule,
        OverlayModule,
        PopoverModule,
        TooltipModule,
        ButtonModule,
        DropDownModule], exports: [GanttScaleComponent,
        GanttBarComponent,
        GanttMilestoneComponent,
        GanttBarParentComponent,
        ResizeHandleDirective,
        GanttToolsComponent] });
GanttModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: GanttModule, providers: [GanttService], imports: [[
            CommonModule,
            OverlayModule,
            PopoverModule,
            TooltipModule,
            ButtonModule,
            DropDownModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: GanttModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        GanttScaleComponent,
                        GanttBarComponent,
                        GanttMarkerDirective,
                        GanttMilestoneComponent,
                        GanttBarParentComponent,
                        ResizeHandleDirective,
                        GanttToolsComponent
                    ],
                    imports: [
                        CommonModule,
                        OverlayModule,
                        PopoverModule,
                        TooltipModule,
                        ButtonModule,
                        DropDownModule
                    ],
                    providers: [GanttService],
                    exports: [
                        GanttScaleComponent,
                        GanttBarComponent,
                        GanttMilestoneComponent,
                        GanttBarParentComponent,
                        ResizeHandleDirective,
                        GanttToolsComponent
                    ]
                }]
        }] });
