// ng-devui/esm2020/tooltip/tooltip.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { OverlayContainerRef } from 'ng-devui/overlay-container';
import { PortalModule } from 'ng-devui/portal';
import { PositioningModule } from 'ng-devui/position';
import { TooltipComponent } from './tooltip.component';
import { TooltipDirective } from './tooltip.directive';
import * as i0 from "@angular/core";
export class TooltipModule {
}
TooltipModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TooltipModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TooltipModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TooltipModule, declarations: [TooltipComponent, TooltipDirective], imports: [CommonModule,
        PortalModule,
        PositioningModule], exports: [TooltipComponent, TooltipDirective] });
TooltipModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TooltipModule, providers: [
        OverlayContainerRef,
    ], imports: [[
            CommonModule,
            PortalModule,
            PositioningModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: TooltipModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PortalModule,
                        PositioningModule
                    ],
                    exports: [TooltipComponent, TooltipDirective],
                    declarations: [TooltipComponent, TooltipDirective],
                    providers: [
                        OverlayContainerRef,
                    ],
                }]
        }] });
