// ng-devui/esm2020/dashboard/widget-library/library-widget.directive.mjs
import { DomPortalOutlet, TemplatePortal } from '@angular/cdk/portal';
import { ApplicationRef, ComponentFactoryResolver, Directive, ElementRef, HostBinding, Injector, Input, Optional, TemplateRef, ViewContainerRef } from '@angular/core';
import { DashboardComponent } from '../dashboard.component';
import { GridStackService } from '../grid-stack.service';
import { DashboardLibraryPanelDirective } from '../widget-library/library-panel.directive';
import * as i0 from "@angular/core";
import * as i1 from "../widget-library/library-panel.directive";
export class DashboardLibraryWidgetDirective {
    constructor(el, cfr, appRef, injector, vcf, libraryPanel) {
        this.el = el;
        this.cfr = cfr;
        this.appRef = appRef;
        this.injector = injector;
        this.vcf = vcf;
        this.libraryPanel = libraryPanel;
        this.dragMode = 'copy';
        this.dragDisabled = false;
        this.dragCopyStyle = false;
        this.hostBinding = true;
        this.isSetup = false;
        this.notify = (event) => {
            if (this.libraryPanel) {
                if (event === 'dragStart') {
                    return (...args) => {
                        this.libraryPanel.dragStartHandler();
                    };
                }
                if (event === 'dragStop') {
                    return (...args) => {
                        this.libraryPanel.dragStopHandler();
                    };
                }
            }
            else {
                return () => { };
            }
        };
        this.helper = (event) => {
            if (this.dragMode === 'move') {
                // grid stack 有高度变化的bug， 暂时就不解决了
                return this.el.nativeElement;
            }
            if (!this.dragTemplate) {
                const el = this.el.nativeElement.cloneNode(true);
                el.style.zIndex = '1060';
                this.copyCanvas(this.el.nativeElement, el);
                if (this.dragCopyStyle) {
                    this.copyStyle(this.el.nativeElement, el);
                }
                return el;
            }
            else {
                const el = document.createElement('div');
                el.classList.add('grid-stack-item-adding-item-template');
                el.style.zIndex = '1060';
                if (this.width !== undefined) {
                    el.setAttribute('gs-w', `${this.width}`);
                }
                if (this.height !== undefined) {
                    el.setAttribute('gs-h', `${this.height}`);
                }
                this.setDragWidthHeight(el);
                const domPortalOutlet = new DomPortalOutlet(el, this.cfr, this.appRef, this.injector);
                domPortalOutlet.attachTemplatePortal(new TemplatePortal(this.dragTemplate, this.vcf, {
                    $implicit: this.widgetData,
                    width: this.width,
                    height: this.height,
                }));
                const result = el.cloneNode(true);
                this.copyCanvas(el, result);
                if (this.dragCopyStyle) {
                    this.copyStyle(el, result);
                }
                domPortalOutlet.detach();
                domPortalOutlet.dispose();
                return result;
            }
        };
    }
    get gridStackService() {
        return this.targetDashboard?.gridStackService;
    }
    ngOnChanges(changes) {
        if (changes.targetDashboard) {
            this.setup();
        }
        if (changes.dragDisabled) {
            if (this.dragDisabled) {
                GridStackService.disableDrag(this.el.nativeElement);
            }
            else {
                GridStackService.enableDrag(this.el.nativeElement);
            }
        }
    }
    ngAfterViewInit() {
        if (!this.isSetup) {
            GridStackService.cleanDragIn(this.el.nativeElement);
            this.setup();
        }
    }
    ngOnDestroy() {
        if (this.gridStackService) {
            this.gridStackService.destroyDragIn(this.el.nativeElement);
        }
    }
    setup() {
        if (this.targetDashboard && !this.isSetup) {
            setTimeout(() => {
                if (this.gridStackService && this.gridStackService.gridStack) {
                    GridStackService.cleanDragIn(this.el.nativeElement);
                    this.gridStackService.setupDragIn(this.el.nativeElement, this, this.helper, this.notify);
                    if (this.dragDisabled) {
                        GridStackService.disableDrag(this.el.nativeElement);
                    }
                }
            });
            this.isSetup = true;
        }
    }
    copyCanvas(origin, target) {
        // 拷贝canvas的内容
        const originCanvasArr = origin.querySelectorAll('canvas');
        const targetCanvasArr = target.querySelectorAll('canvas');
        [].forEach.call(targetCanvasArr, (canvas, index) => {
            canvas.getContext('2d').drawImage(originCanvasArr[index], 0, 0);
        });
    }
    setDragWidthHeight(el) {
        const columnWidth = this.targetDashboard.getCurrentColumnWidth();
        const cellHeight = this.targetDashboard.getCurrentCellHeight();
        const margin = this.targetDashboard.getCurrentMargin();
        if (this.width) {
            el.style.width = columnWidth * (this.width || 1) - margin * ((this.width || 1) - 1) + 'px';
        }
        if (this.height) {
            el.style.height = cellHeight * (this.height || 1) - margin * 2 + 'px';
        }
    }
    copyStyle(source, target) {
        ['id', 'style', 'draggable'].forEach((att) => {
            target.removeAttribute(att);
        });
        // copy style (without transitions)
        const computedStyle = getComputedStyle(source);
        for (let i = 0; i < computedStyle.length; i++) {
            const key = computedStyle[i];
            if (key.indexOf('transition') < 0) {
                target.style[key] = computedStyle[key];
            }
        }
        target.style.pointerEvents = 'none';
        // and repeat for all children
        for (let i = 0; i < source.children.length; i++) {
            this.copyStyle(source.children[i], target.children[i]);
        }
    }
}
DashboardLibraryWidgetDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DashboardLibraryWidgetDirective, deps: [{ token: i0.ElementRef }, { token: i0.ComponentFactoryResolver }, { token: i0.ApplicationRef }, { token: i0.Injector }, { token: i0.ViewContainerRef }, { token: i1.DashboardLibraryPanelDirective, optional: true }], target: i0.ɵɵFactoryTarget.Directive });
DashboardLibraryWidgetDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: DashboardLibraryWidgetDirective, selector: "[dDashboardLibraryWidget]", inputs: { width: "width", height: "height", widgetData: "widgetData", dragMode: "dragMode", dragTemplate: "dragTemplate", dragDisabled: "dragDisabled", dragCopyStyle: "dragCopyStyle", targetDashboard: "targetDashboard" }, host: { properties: { "attr.gs-w": "this.width", "attr.gs-h": "this.height", "class.grid-stack-new-item": "this.hostBinding", "attr.gs-instance": "this.gridStackId" } }, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DashboardLibraryWidgetDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[dDashboardLibraryWidget]',
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.ComponentFactoryResolver }, { type: i0.ApplicationRef }, { type: i0.Injector }, { type: i0.ViewContainerRef }, { type: i1.DashboardLibraryPanelDirective, decorators: [{
                    type: Optional
                }] }]; }, propDecorators: { width: [{
                type: HostBinding,
                args: ['attr.gs-w']
            }, {
                type: Input
            }], height: [{
                type: HostBinding,
                args: ['attr.gs-h']
            }, {
                type: Input
            }], widgetData: [{
                type: Input
            }], dragMode: [{
                type: Input
            }], dragTemplate: [{
                type: Input
            }], dragDisabled: [{
                type: Input
            }], dragCopyStyle: [{
                type: Input
            }], hostBinding: [{
                type: HostBinding,
                args: ['class.grid-stack-new-item']
            }], gridStackId: [{
                type: HostBinding,
                args: ['attr.gs-instance']
            }], targetDashboard: [{
                type: Input
            }] } });
