// ng-devui/esm2020/dashboard/widget-library/library-panel.directive.mjs
import { Directive, EventEmitter, Output } from '@angular/core';
import * as i0 from "@angular/core";
export class DashboardLibraryPanelDirective {
    constructor() {
        this.widgetDragStart = new EventEmitter();
        this.widgetDragStop = new EventEmitter();
    }
    dragStartHandler() {
        this.widgetDragStart.emit();
    }
    dragStopHandler() {
        this.widgetDragStop.emit();
    }
}
DashboardLibraryPanelDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DashboardLibraryPanelDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
DashboardLibraryPanelDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: DashboardLibraryPanelDirective, selector: "[dDashboardLibraryPanel]", outputs: { widgetDragStart: "widgetDragStart", widgetDragStop: "widgetDragStop" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DashboardLibraryPanelDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[dDashboardLibraryPanel]',
                }]
        }], propDecorators: { widgetDragStart: [{
                type: Output
            }], widgetDragStop: [{
                type: Output
            }] } });
