// ng-devui/esm2020/dashboard/widget-library/library-trash.directive.mjs
import { Directive, ElementRef, HostBinding, Input } from '@angular/core';
import { DashboardComponent } from '../dashboard.component';
import { GridStackService } from '../grid-stack.service';
import * as i0 from "@angular/core";
export class DashboardLibraryTrashDirective {
    constructor(el) {
        this.el = el;
        this.dropDisabled = false;
        this.hostBinding = true;
        this.isSetup = false;
    }
    get gridStackService() {
        return this.targetDashboard?.gridStackService;
    }
    ngAfterViewInit() {
        this.setup();
    }
    ngOnChanges(changes) {
        if (changes.targetDashBoard) {
            this.setup();
        }
        if (changes.dropDisabled) {
            if (this.dropDisabled) {
                GridStackService.disableDrop(this.el.nativeElement);
            }
            else {
                GridStackService.enableDrop(this.el.nativeElement);
            }
        }
    }
    ngOnDestroy() {
        if (this.gridStackService && this.isSetup) {
            this.gridStackService.destroyRemoveDropAreas(this.el.nativeElement);
        }
    }
    setup() {
        if (this.targetDashboard && !this.isSetup) {
            setTimeout(() => {
                if (this.gridStackService) {
                    this.gridStackService.setupRemoveDropArea(this.el.nativeElement, this.targetDashboard, this);
                    if (this.dropDisabled) {
                        GridStackService.disableDrop(this.el.nativeElement);
                    }
                }
            });
            this.isSetup = true;
        }
    }
}
DashboardLibraryTrashDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DashboardLibraryTrashDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
DashboardLibraryTrashDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.3", type: DashboardLibraryTrashDirective, selector: "[dDashboardLibraryTrash]", inputs: { targetDashboard: "targetDashboard", trashData: "trashData", dropDisabled: "dropDisabled" }, host: { properties: { "class.grid-stack-library-trash": "this.hostBinding" } }, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DashboardLibraryTrashDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[dDashboardLibraryTrash]',
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; }, propDecorators: { targetDashboard: [{
                type: Input
            }], trashData: [{
                type: Input
            }], dropDisabled: [{
                type: Input
            }], hostBinding: [{
                type: HostBinding,
                args: ['class.grid-stack-library-trash']
            }] } });
