// ng-devui/esm2020/form/services/d-validate-sync.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
function resolvedPromiseFunc() { return Promise.resolve(null); }
const resolvedPromise = resolvedPromiseFunc();
export class DValidateSyncService {
    constructor() {
        this.controlsMap = new Map();
        this.controlsSubMap = new Map();
    }
    addControl(key, cd) {
        if (!this.controlsMap.get(key)) {
            this.controlsMap.set(key, []);
        }
        this.controlsMap.get(key).push(cd);
        this._addControlSubscription(key, cd);
    }
    removeControl(key, cd) {
        if (this.controlsMap.get(key)) {
            const controls = this.controlsMap.get(key);
            const index = controls.indexOf(cd);
            if (index !== -1) {
                controls.splice(index, 1);
            }
        }
        this._removeControlSubscription(cd);
    }
    validateControlsSync(key, cd) {
        const controls = this.controlsMap.get(key);
        if (controls) {
            controls.forEach(control => {
                this._removeControlSubscription(control);
            });
            controls.forEach(control => {
                if (control !== cd) {
                    resolvedPromise.then(() => {
                        control.updateValueAndValidity();
                    });
                }
            });
            controls.forEach(control => {
                resolvedPromise.then(() => {
                    this._addControlSubscription(key, control);
                });
            });
        }
    }
    _addControlSubscription(key, cd) {
        this._removeControlSubscription(cd);
        const sub = cd.valueChanges.subscribe(() => {
            this.validateControlsSync(key, cd);
        });
        this.controlsSubMap.set(cd, sub);
    }
    _removeControlSubscription(cd) {
        if (this.controlsSubMap.get(cd)) {
            this.controlsSubMap.get(cd).unsubscribe();
            this.controlsSubMap.delete(cd);
        }
    }
}
DValidateSyncService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DValidateSyncService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
DValidateSyncService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DValidateSyncService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DValidateSyncService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return []; } });
