// ng-devui/esm2020/data-table/data-table.component.mjs
import { CdkVirtualScrollViewport } from '@angular/cdk/scrolling';
import { DOCUMENT } from '@angular/common';
import { ChangeDetectorRef, Component, ContentChild, ContentChildren, ElementRef, EventEmitter, forwardRef, HostBinding, Inject, Input, NgZone, Output, QueryList, Renderer2, TemplateRef, ViewChild } from '@angular/core';
import { merge } from 'rxjs';
import { switchMap, takeUntil } from 'rxjs/operators';
import { DATA_TABLE } from './data-table.token';
import { TableTbodyComponent } from './table/body/tbody.component';
import { TableThComponent } from './table/head/th/th.component';
import { TableTheadComponent } from './table/head/thead.component';
import { DataTableColumnTmplComponent } from './tmpl/data-table-column-tmpl.component';
import * as i0 from "@angular/core";
import * as i1 from "./table/body/tbody.component";
import * as i2 from "./data-table-body.component";
import * as i3 from "@angular/cdk/scrolling";
import * as i4 from "./data-table-head.component";
import * as i5 from "@angular/common";
import * as i6 from "ng-devui/utils";
import * as i7 from "ng-devui/dragdrop";
export class DataTableComponent {
    constructor(elementRef, ngZone, renderer, cdr, doc) {
        this.elementRef = elementRef;
        this.ngZone = ngZone;
        this.renderer = renderer;
        this.cdr = cdr;
        this.doc = doc;
        /**
         * 【可选】表头checkbox是否可见
         */
        this.headerCheckVisible = true;
        /**
         * 【可选】默认表格使用的表格类型，可选值为'cell' @deprecated
         */
        this.editModel = 'cell';
        /**
         * 鼠标悬浮行时是否高亮
         */
        this.rowHoveredHighlight = true;
        /**
         * 表格宽度
         */
        this.tableWidth = '100%';
        /**
         * 固定表头指定高度是否包含表头的高度，`tableHeight`设置的高度默认是表格body的高度
         */
        this.containFixHeaderHeight = false;
        /**
         * 【可选】多列选择数组，用来指导那几列会被排序
         */
        this.multiSort = [];
        /**
         * 【可选】用来自定义不可拖拽的前几列
         */
        this.colDropFreezeTo = 0;
        /**
         * 【可选】同时绑定单击、双击事件时，用于区分点击的时间间隔,默认300ms，两个事件不同时使用可以指定为0
         */
        this.timeout = 300;
        /**
         * 【可选】配置表头操作未激活状态下是否显示操作区域，默认不显示
         */
        this.showOperationArea = false;
        /**
         * 【可选】是否显示排序未激活图标，默认不显示,
         */
        this.showSortIcon = false;
        /**
         * 【可选】是否显示筛选未激活图标，默认不显示,
         */
        this.showFilterIcon = false;
        /**
         * 多列选择Change事件，用来更新多列选择数组, column param
         * */
        this.multiSortChange = new EventEmitter();
        /**
         * 表格单元格点击事件
         */
        this.cellClick = new EventEmitter();
        /**
         * 表格单元格双击事件
         */
        this.cellDBClick = new EventEmitter();
        /**
         * 表格行点击事件
         */
        this.rowClick = new EventEmitter();
        /**
         * 表格行双击事件
         */
        this.rowDBClick = new EventEmitter();
        /**
        * 行detail toggle事件
        */
        this.detialToggle = new EventEmitter();
        /**
         * 表格单元格开始编辑事件
         */
        this.cellEditStart = new EventEmitter();
        /**
         * 表格单元格结束编辑事件
         */
        this.cellEditEnd = new EventEmitter();
        /**
         * 某行的勾选状态变化事件
         */
        this.rowCheckChange = new EventEmitter();
        /**
         * 当前页码全勾选状态变化事件
         */
        this.checkAllChange = new EventEmitter();
        /**
         * 延迟懒加载完成事件
         */
        this.loadMore = new EventEmitter();
        /**
         * 列宽变化事件
         */
        this.resize = new EventEmitter();
        /**
         * 当前表格层级，默认为0，在树形表格场景下自增长
         * 内部嵌套使用，不对外暴露
         */
        this.tableLevel = 0;
        /**
         * 配置树形表格的父子选中是否互相关联
         * upward：选中子关联父
         * downward： 选中父关联子
         */
        this.checkableRelation = { upward: true, downward: true };
        /**
         * 子列表关闭事件
         */
        this.childrenTableClose = new EventEmitter();
        /**
         * 全部子列表关闭事件
         */
        this.allChildrenTableClose = new EventEmitter();
        /**
         * 虚拟滚动配置
         */
        this.virtualItemSize = 40;
        this.virtualMinBufferPx = 80;
        this.virtualMaxBufferPx = 200;
        /**
         * 列宽配置
         */
        this.tableWidthConfig = [];
        /**
         * 表格布局
         */
        this.tableLayout = 'fixed';
        /**
         * 表格边框类型，默认有行边框，bordered：全边框，borderless：无边框
         */
        this.borderType = '';
        /**
         * 表格内部滚动事件
         */
        this.tableScrollEvent = new EventEmitter();
        this.columnDragEnd = new EventEmitter();
        /**
         * 表格尺寸，sm对应行高40px， md对应行高48px，lg对应行高56px
         */
        this.size = 'sm';
        this.shadowType = 'embed';
        this.tableOverflowType = 'auto';
        this._dataSource = [];
        this._pageAllChecked = false;
        this.selectable = true;
        this.allChecked = [];
        this.documentClickEvent = new EventEmitter();
        this.cellEditorClickEvent = new EventEmitter();
        this._hideColumn = [];
        this.searchQueryChange = new EventEmitter();
        this.halfChecked = false;
        this.scrollY = 0;
        this.BUILTIN_COL_WIDTH = '36px';
        this.BUILTIN_COL_WIDTH_EXTRA = '55px';
        this._tableTotalWidth = 0;
        this._lastColSize = 0;
        // 判断数据是否存在选中状态
        this.hasChecked = (data) => {
            if (data.$checked) {
                return true;
            }
            if (data.children) {
                return data.children.some(this.hasChecked);
            }
        };
        // 判断数据是否存在未选中状态
        this.hasUnChecked = (data) => {
            if (!data.$checked) {
                return true;
            }
            if (data.children) {
                return data.children.some(this.hasUnChecked);
            }
        };
        this.onDocumentClickListen = this.onDocumentClick.bind(this);
        this.document = this.doc;
    }
    get hostHeight() {
        return (this.tableHeight && this.dataSource.length) ? this.tableHeight : null;
    }
    get hasShadow() {
        return this.shadowType === 'normal';
    }
    set content(content) {
        setTimeout(() => {
            this.tableBodyEl = content;
            if (this.virtualScroll) {
                this.initVirtualBodyHeight();
            }
        });
    }
    set dataSource(dataSource) {
        if (dataSource === null || !dataSource) {
            /* eslint-disable-next-line no-param-reassign */
            dataSource = [];
        }
        this._dataSource = dataSource;
        const hasChecked = this.dataSource.some(this.hasChecked);
        const hasUnChecked = this.dataSource.some(this.hasUnChecked);
        this._pageAllChecked = dataSource && dataSource.length > 0 && !hasUnChecked;
        this.halfChecked = hasChecked && hasUnChecked;
        if (this.innerHeader) {
            this.innerHeader.setHeaderCheckStatus({ pageAllChecked: this._pageAllChecked, pageHalfChecked: this.halfChecked });
        }
        if (this.virtualScroll) {
            this.initVirtualBodyHeight();
        }
        this.initScrollStatus();
    }
    get dataSource() {
        return this._dataSource;
    }
    set hideColumn(hideColumn) {
        this._hideColumn = hideColumn;
        if (this._columns) {
            this.updateColumns();
        }
    }
    get hideColumn() {
        return this._hideColumn;
    }
    set pageAllChecked(pageAllChecked) {
        if (this.dataSource) {
            this._dataSource = this.setCheckedStatus(this.dataSource, pageAllChecked);
        }
        this._pageAllChecked = pageAllChecked;
        this.halfChecked = this.dataSource.some(this.hasChecked) && this.dataSource.some(this.hasUnChecked);
    }
    get pageAllChecked() {
        return this._pageAllChecked;
    }
    initVirtualBodyHeight() {
        setTimeout(() => {
            if (this.virtualScrollViewport) {
                this.virtualScrollViewport.checkViewportSize();
            }
        });
        if (this.tableHeight && this.tableHeight !== 'auto') {
            this.virtualBodyHeight = this.tableHeight;
            return;
        }
        if (!this.maxHeight) {
            this.virtualBodyHeight = null;
            return;
        }
        if (this.tableBodyEl) {
            const tableHeader = this.tableBodyEl.nativeElement.querySelector('thead');
            const tableHeaderHeight = (tableHeader?.offsetHeight + 8) || 0;
            const curTotalHeight = this.dataSource.length * this.virtualItemSize + tableHeaderHeight;
            this.virtualBodyHeight = curTotalHeight < parseInt(this.maxHeight, 10) ? curTotalHeight + 'px' : this.maxHeight;
            return;
        }
    }
    getColumns() {
        const cols = this.columns
            .filter(column => {
            return !this.hideColumn.some(field => column.field === field);
        });
        cols.sort((first, second) => first.order - second.order);
        return cols;
    }
    // life hook start
    ngOnInit() {
        this.ngZone.runOutsideAngular(() => {
            this.document.addEventListener('click', this.onDocumentClickListen);
        });
    }
    ngOnChanges(changes) {
        if (changes['checkable']) {
            const checkColIndex = this.tableWidthConfig.findIndex((config) => {
                return config.field === 'checkbox';
            });
            if (this.checkable) {
                if (checkColIndex < 0) {
                    if (this.showExpandToggle) {
                        this.tableWidthConfig.splice(1, 0, { field: 'checkbox', width: this.BUILTIN_COL_WIDTH });
                    }
                    else {
                        this.tableWidthConfig.unshift({ field: 'checkbox', width: this.BUILTIN_COL_WIDTH });
                    }
                }
            }
            else {
                if (checkColIndex > -1) {
                    this.tableWidthConfig.splice(checkColIndex, 1);
                }
            }
        }
        if (changes['showExpandToggle']) {
            const expandColIndex = this.tableWidthConfig.findIndex((config) => {
                return config.field === 'expand';
            });
            if (this.showExpandToggle) {
                if (expandColIndex < 0) {
                    this.tableWidthConfig.unshift({ field: 'expand', width: this.BUILTIN_COL_WIDTH });
                }
            }
            else {
                if (expandColIndex > -1) {
                    this.tableWidthConfig.splice(expandColIndex, 1);
                }
            }
        }
        if (this.virtualScroll &&
            ((changes['tableHeight'] && !changes.tableHeight.firstChange) ||
                (changes['maxHeight'] && !changes.maxHeight.firstChange) ||
                (changes['virtualScroll'] && !changes.virtualScroll.firstChange))) {
            this.initVirtualBodyHeight();
        }
    }
    onDocumentClick($event) {
        this.documentClickEvent.emit($event);
    }
    ngAfterContentInit() {
        if (this.columns.length > 0) {
            this.updateColumns();
            this.columns.forEach((col) => {
                col.orderChange.subscribe(() => {
                    this.updateColumns();
                });
                col.widthChange.subscribe(() => {
                    this.updateColumns();
                });
            });
        }
        if (this.innerHeader) {
            this.headerCheckStatusSubscription = this.innerHeader.headerCheckStatusEvent.subscribe((status) => {
                this.onCheckAllChange(status);
            });
            this.headertoggleTableSubscription = this.innerHeader.headerChildrenTableToggleEvent.subscribe((status) => {
                this.onToggleAllChildrenTable(status);
            });
        }
        else {
            this.columns.changes.subscribe(() => {
                this.updateColumns();
            });
        }
        setTimeout(() => {
            this.initScrollStatus();
        });
    }
    initScrollStatus() {
        if (this.tableOverflowType !== 'overlay') {
            return;
        }
        const ele = this.virtualScroll ? this.virtualScrollViewport?.elementRef?.nativeElement : this.normalScrollElement?.nativeElement;
        this.hasWidthScroll = ele && ele.scrollWidth > ele.clientWidth;
        this.hasHeightScroll = ele && ele.scrollHeight > ele.clientHeight;
    }
    ngAfterViewInit() {
        this.thList.forEach(th => {
            th.tableViewRefElement = this.tableViewRefElement;
        });
        this.thList.changes.subscribe((list) => {
            list.forEach(th => {
                th.tableViewRefElement = this.tableViewRefElement;
            });
        });
        if (this.onlyOneColumnSort) {
            this.resetThSortOrder();
        }
        this.initScrollClass();
    }
    // 初始化时判断是否存在横向滚动，并加上相应类名
    initScrollClass() {
        const ele = this.normalScrollElement?.nativeElement || this.vitualScrollElement?.elementRef?.nativeElement;
        if (ele.clientWidth !== ele.scrollWidth) {
            this.setScrollViewClass('left');
        }
    }
    resetThSortOrder() {
        merge(...this.thList?.map(th => th.sortChange)).pipe(takeUntil(this.thList.changes)).subscribe((sortEvent) => {
            this.thList.filter(th => th !== sortEvent.th).forEach(th => th.clearSortOrder());
        });
        this.thList.changes.pipe(switchMap(() => merge(...this.thList.map(th => th.sortChange)))).subscribe((sortEvent) => {
            this.thList.filter(th => th !== sortEvent.th).forEach(th => th.clearSortOrder());
        });
    }
    updateColumns() {
        this._columns = this.getColumns();
        this.tableWidthConfig = [];
        if (this.showExpandToggle) {
            this.tableWidthConfig.push({ field: 'expand', width: this.BUILTIN_COL_WIDTH });
        }
        if (this.checkable) {
            if (this.checkOptions && this.checkOptions.length > 0) {
                this.tableWidthConfig.push({ field: 'checkbox', width: this.BUILTIN_COL_WIDTH_EXTRA });
            }
            else {
                this.tableWidthConfig.push({ field: 'checkbox', width: this.BUILTIN_COL_WIDTH });
            }
        }
        this._columns.forEach((col) => {
            this.tableWidthConfig.push({ field: col.field, width: col.width });
        });
    }
    ngOnDestroy() {
        this.unSubscription();
        this.document.removeEventListener('click', this.onDocumentClickListen);
    }
    onHandleSort(column) {
        if (this.multiSort && this.multiSort.length > 0) {
            const multiSortIndex = this.multiSort.findIndex(item => item.field === column.field);
            if (multiSortIndex !== -1) {
                this.multiSort.splice(multiSortIndex, 1);
            }
            if (column.direction === '') {
                this.multiSortChange.emit(this.multiSort);
                return;
            }
            if (!this.onlyOneColumnSort) {
                this.multiSort.push(column);
            }
            else {
                this.multiSort = [column];
            }
        }
        else {
            this.multiSort = [column];
        }
        this.multiSortChange.emit(this.multiSort);
    }
    onCellClick($event) {
        this.selectedRowItem = $event.rowItem;
        this.selectedColumnItem = $event.column;
        this.ngZone.run(() => {
            this.cellClick.emit($event);
        });
    }
    onCellEditStart($event) {
        this.isCellEdit = true;
        this.cellEditStart.emit($event);
    }
    onCellEditEnd($event) {
        this.isCellEdit = false;
        this.cellEditEnd.emit($event);
    }
    onCellDBClick($event) {
        this.ngZone.run(() => {
            this.cellDBClick.emit($event);
        });
    }
    onRowClick($event) {
        this.selectedRowItem = $event.rowItem;
        this.ngZone.run(() => {
            this.rowClick.emit($event);
        });
    }
    onRowDBClick($event) {
        this.ngZone.run(() => {
            this.rowDBClick.emit($event);
        });
    }
    onDetailToggle($event) {
        this.detialToggle.emit($event);
    }
    setRowCheckStatus($event) {
        // 处理children的选中
        if ($event.rowItem.children && this.checkableRelation.downward) {
            this.setCheckedStatus($event.rowItem.children, $event.checked);
        }
        // 处理parents的选中
        if (this.checkableRelation.upward) {
            const nestedIndexArray = $event.nestedIndex.split(',');
            nestedIndexArray.shift();
            const nestedIndexArrayToInt = nestedIndexArray.map((value) => {
                return parseInt(value, 10);
            });
            // 通过选中行的父级索引设置父的选中状态
            this.setParentCheckStatus(nestedIndexArrayToInt);
        }
        // 处理整个table header的选中
        const hasChecked = this.dataSource.some(this.hasChecked);
        if ($event) {
            const hasUnChecked = this.dataSource.some(this.hasUnChecked);
            this._pageAllChecked = !hasUnChecked;
            this.halfChecked = hasChecked && hasUnChecked;
        }
        else {
            this._pageAllChecked = false;
            this.halfChecked = hasChecked;
        }
        if (this.innerHeader) {
            this.innerHeader.setHeaderCheckStatus({ pageAllChecked: this._pageAllChecked, pageHalfChecked: this.halfChecked });
        }
        this.rowCheckChange.emit($event);
    }
    setParentCheckStatus(nestedIndex) {
        if (nestedIndex.length > 0) {
            const topIndex = nestedIndex[0];
            const topParent = this.dataSource[topIndex];
            const argNestedIndex = [...nestedIndex];
            argNestedIndex.shift();
            const lastParent = this.findLastParent(topParent, argNestedIndex);
            this.setSelfCheckStatus(lastParent);
            nestedIndex.pop();
            if (nestedIndex.length > 0) {
                this.setParentCheckStatus(nestedIndex);
            }
        }
    }
    findLastParent(source, indexArray) {
        if (source && indexArray.length > 0) {
            const topIndex = indexArray[0];
            const topParent = source.children[topIndex];
            indexArray.shift();
            return this.findLastParent(topParent, indexArray);
        }
        else {
            return source;
        }
    }
    setSelfCheckStatus(data) {
        if (data && data.children) {
            const hasUnChecked = data.children.some((child) => {
                return !child.$checked;
            });
            const hasChecked = data.children.some((child) => {
                return child.$checked || child.$halfChecked;
            });
            data.$checked = !hasUnChecked;
            data.$halfChecked = hasUnChecked && hasChecked;
        }
    }
    setCheckedStatus(data, checked, toggle) {
        return data.map(item => {
            if (!(item.$checkDisabled || item.$disabled)) {
                if ((toggle && item.$checked === undefined) || !toggle) {
                    item.$checked = checked;
                    item.$halfChecked = false;
                }
            }
            if (item.children) {
                item.children = this.setCheckedStatus(item.children, checked, toggle);
            }
            return item;
        });
    }
    onCheckAllChange($event) {
        this.pageAllChecked = $event;
        this.checkAllChange.emit($event);
    }
    onSearchQueryChange($event) {
        this.searchQueryChange.emit($event);
    }
    getSelectedRowItem() {
        return this.selectedRowItem;
    }
    onLoadMore(event) {
        this.loadMore.emit(this);
    }
    updateColumnsWidth() {
        this.tableWidthConfig = [];
        if (this.showExpandToggle) {
            const expandWidth = this.elementRef.nativeElement.querySelector('.devui-detail-cell').clientWidth;
            this.tableWidthConfig.push({ field: 'expand', width: expandWidth + 'px' });
        }
        if (this.checkable) {
            const checkboxWidth = this.elementRef.nativeElement.querySelector('.devui-checkable-cell').clientWidth;
            this.tableWidthConfig.push({ field: 'checkbox', width: checkboxWidth + 'px' });
        }
        this._columns.forEach((col) => {
            this.tableWidthConfig.push({ field: col.field, width: col.width });
        });
    }
    beginResizeHandlerEvent($event) {
        const thRenderWidthList = $event.thRenderWidthList;
        if (thRenderWidthList.length > 0) {
            this._tableTotalWidth = this.elementRef.nativeElement.querySelector('.table-wrap').offsetWidth;
            // 兼容d-column表头分组场景
            const reverseThList = thRenderWidthList.reverse();
            this._columns.forEach(column => {
                const thItem = reverseThList.find(th => th.field === column.field);
                if (thItem) {
                    column.width = thItem.width + 'px';
                }
            });
            if (!this._lastColSize) {
                this._lastColSize = parseInt(this._columns.slice(-1)[0].width, 10);
            }
            this.updateColumnsWidth();
        }
        this.onDocumentClick($event.event);
    }
    onResizingFixedHandler({ field, width }) {
        if (this.resizeable) {
            const index = this.tableWidthConfig.findIndex((config) => {
                return config.field === field;
            });
            if (index > -1) {
                setTimeout(() => {
                    this.tableWidthConfig[index].width = width + 'px';
                });
            }
        }
    }
    onResizeHandler({ width, field }) {
        if (width < 0) {
            return;
        }
        const index = this.tableWidthConfig.findIndex((config) => {
            return config.field === field;
        });
        if (index > -1) {
            this.tableWidthConfig[index].width = width + 'px';
        }
        const curTotal = this.tableWidthConfig.reduce((pre, cur) => {
            const value = pre + parseInt(cur.width, 10);
            return value;
        }, 0);
        let columnResizeEventArg;
        this._columns = this._columns.map((column, colIndex) => {
            if (column.field === field) {
                column.width = parseInt(width, 10) + 'px';
                columnResizeEventArg = { currentColumn: column, nextColumn: this._columns[colIndex + 1] };
                this.resize.emit(columnResizeEventArg);
            }
            return column;
        });
        const changeSize = curTotal - this._tableTotalWidth;
        const lastCol = this._columns.slice(-1)[0];
        const lastColWidth = parseInt(lastCol.width, 10);
        if (changeSize < 0 && columnResizeEventArg.nextColumn) {
            const newSize = parseInt(lastCol.width) - changeSize + 'px';
            lastCol.width = newSize;
            this.tableWidthConfig[this.tableWidthConfig.length - 1].width = newSize;
        }
        else if (this._lastColSize < lastColWidth) {
            const lastChange = (lastColWidth - this._lastColSize) > changeSize ? changeSize : (lastColWidth - this._lastColSize);
            lastCol.width = lastColWidth - lastChange + 'px';
            this.tableWidthConfig[this.tableWidthConfig.length - 1].width = lastColWidth - lastChange + 'px';
        }
    }
    handleDragTable({ from, to }) {
        this.columnDragEnd.emit({ from, to });
        const sortArray = (array, fromE, toE) => {
            if (fromE < toE) {
                const fromEData = array[fromE];
                for (let i = 0; i < array.length; i++) {
                    if (i >= fromE && i < toE) {
                        this.ngZone.run(() => {
                            array[i] = array[i + 1];
                        });
                    }
                }
                this.ngZone.run(() => {
                    array[toE] = fromEData;
                });
            }
            if (fromE > toE) {
                const fromEData = array[fromE];
                for (let i = array.length; i > 0; i--) {
                    if (i <= fromE && i > toE) {
                        array[i] = array[i - 1];
                    }
                }
                array[toE] = fromEData;
            }
        };
        sortArray(this._columns, from, to);
        this._columns.forEach((item, index) => {
            item.order = index;
        });
    }
    onBodyScroll(event) {
        const target = event.target;
        if (this.isCellEdit) {
            // Y轴滚动距离超过tr高度时取消目前编辑状态
            if (this.scrollY === 0) {
                this.scrollY = target.scrollTop;
            }
            const offset = target.scrollTop - this.scrollY;
            if (offset > 40 || offset < -40) {
                this.cancelEditingStatus();
                this.scrollY = 0;
            }
        }
        const scrollLeft = target.scrollLeft;
        if (scrollLeft === 0) {
            if (target.clientWidth === target.scrollWidth) {
                this.setScrollViewClass('none');
            }
            else {
                this.setScrollViewClass('left');
            }
        }
        else if (scrollLeft + target.clientWidth === target.scrollWidth) {
            this.setScrollViewClass('right');
        }
        else {
            this.setScrollViewClass('middle');
        }
        if (this.fixHeader) {
            this.fixHeaderContainerRefElement.nativeElement.scrollLeft = scrollLeft;
        }
        this.tableScrollEvent.emit(event);
    }
    setScrollViewClass(position) {
        const element = this.tableViewRefElement.nativeElement;
        const className = 'devui-talbe-scorll-' + position;
        const elClassList = element.classList;
        if (!elClassList.contains(className)) {
            for (let index = 0; index < elClassList.length; index++) {
                const clName = elClassList[index];
                if (clName.startsWith('devui-talbe-scorll-')) {
                    this.renderer.removeClass(element, clName);
                }
            }
            this.renderer.addClass(element, className);
        }
    }
    unSubscription() {
        if (this.headerCheckStatusSubscription) {
            this.headerCheckStatusSubscription.unsubscribe();
            this.headerCheckStatusSubscription = null;
        }
        if (this.headertoggleTableSubscription) {
            this.headertoggleTableSubscription.unsubscribe();
            this.headertoggleTableSubscription = null;
        }
    }
    setRowChildToggleStatus(rowItem, open) {
        if (open) {
            let loadChildrenResult = Promise.resolve(true);
            if (this.loadChildrenTable) {
                loadChildrenResult = this.loadChildrenTable(rowItem);
            }
            loadChildrenResult.then(() => {
                // 异步加载子表格是检查选中状态
                if (rowItem.$checked && this.checkableRelation.downward) {
                    this.setCheckedStatus(rowItem.children, rowItem.$checked, true);
                }
            });
        }
        else {
            this.childrenTableClose.emit(rowItem);
        }
    }
    setTableChildrenToggleStatus(open) {
        this.onToggleAllChildrenTable(open);
        if (this.innerHeader) {
            this.innerHeader.setHeaderToggleStatus(open);
        }
        else {
            this.childrenTableOpen = open;
        }
    }
    travelChildrenToggleStatus(data, open) {
        return data.map(item => {
            if (item.children) {
                item.$isChildTableOpen = open;
                item.children = this.travelChildrenToggleStatus(item.children, open);
            }
            return item;
        });
    }
    // 切换表头的子表格展开收起
    onToggleAllChildrenTable(open) {
        this.childrenTableOpen = open;
        if (open) {
            let loadAllChildrenResult = Promise.resolve(true);
            if (this.loadAllChildrenTable) {
                loadAllChildrenResult = this.loadAllChildrenTable();
            }
            loadAllChildrenResult.then(() => {
                this.dataSource.forEach(item => {
                    if (item.$checked && item.children) {
                        this.setCheckedStatus(item.children, true, true);
                    }
                });
                this.travelChildrenToggleStatus(this.dataSource, open);
            });
        }
        else {
            this.travelChildrenToggleStatus(this.dataSource, open);
            this.allChildrenTableClose.emit();
        }
    }
    cancelEditingStatus() {
        this.documentClickEvent.emit('cancel');
    }
    collectCheckedRows(dist, source) {
        source.forEach(row => {
            if (row.$checked) {
                dist.push(row);
            }
            if (row.children) {
                this.collectCheckedRows(dist, row.children);
            }
        });
    }
    getCheckedRows() {
        if (this.checkableRelation.upward) {
            // 如果children的选中状态关联parent的选中状态,只需返回最外层的数据
            return this.dataSource ? this.dataSource.filter(item => item.$checked || item.$halfChecked) : [];
        }
        else {
            // 如果children的选中状态不关联parent的选中状态,遍历dataSource,将所有的选中行平级返回
            const checkedRows = [];
            this.collectCheckedRows(checkedRows, this.dataSource);
            return checkedRows;
        }
    }
    setTableCheckStatus(status) {
        if (status.pageAllChecked !== undefined) { // 设置全选
            if (this.dataSource) {
                this._dataSource = this.setCheckedStatus(this.dataSource, status.pageAllChecked);
            }
            this._pageAllChecked = status.pageAllChecked;
            if (status.pageAllChecked) { // 全选为true
                this.halfChecked = false;
            }
            else {
                this.halfChecked = this.dataSource.some(this.hasChecked) && this.dataSource.some(this.hasUnChecked);
            }
        }
        if (status.pageHalfChecked !== undefined) { // 设置半选
            this.halfChecked = status.pageHalfChecked;
        }
        if (this.innerHeader) {
            this.innerHeader.setHeaderCheckStatus({ pageAllChecked: this._pageAllChecked, pageHalfChecked: this.halfChecked });
        }
    }
    // 更新cdk虚拟滚动viewport size并重新渲染，解决父层高度变化渲染数据size没有更新问题
    updateVirtualScrollSize() {
        this.virtualScrollViewport.checkViewportSize();
    }
}
DataTableComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DataTableComponent, deps: [{ token: i0.ElementRef }, { token: i0.NgZone }, { token: i0.Renderer2 }, { token: i0.ChangeDetectorRef }, { token: DOCUMENT }], target: i0.ɵɵFactoryTarget.Component });
DataTableComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: DataTableComponent, selector: "d-data-table", inputs: { checkable: "checkable", headerCheckDisabled: "headerCheckDisabled", headerCheckVisible: "headerCheckVisible", checkOptions: "checkOptions", showExpandToggle: "showExpandToggle", fixHeader: "fixHeader", scrollable: "scrollable", editModel: "editModel", maxWidth: "maxWidth", maxHeight: "maxHeight", type: "type", rowHoveredHighlight: "rowHoveredHighlight", generalRowHoveredData: "generalRowHoveredData", cssClass: "cssClass", tableWidth: "tableWidth", tableHeight: "tableHeight", containFixHeaderHeight: "containFixHeaderHeight", onlyOneColumnSort: "onlyOneColumnSort", multiSort: "multiSort", resizeable: "resizeable", colDraggable: "colDraggable", colDropFreezeTo: "colDropFreezeTo", detailTemplateRef: "detailTemplateRef", timeout: "timeout", showOperationArea: "showOperationArea", showSortIcon: "showSortIcon", showFilterIcon: "showFilterIcon", beforeCellEdit: "beforeCellEdit", tableLevel: "tableLevel", checkableRelation: "checkableRelation", loadChildrenTable: "loadChildrenTable", loadAllChildrenTable: "loadAllChildrenTable", virtualScroll: "virtualScroll", headerExpandConfig: "headerExpandConfig", virtualItemSize: "virtualItemSize", virtualMinBufferPx: "virtualMinBufferPx", virtualMaxBufferPx: "virtualMaxBufferPx", lazy: "lazy", tableWidthConfig: "tableWidthConfig", headerBg: "headerBg", tableLayout: "tableLayout", borderType: "borderType", striped: "striped", minHeight: "minHeight", size: "size", shadowType: "shadowType", tableOverflowType: "tableOverflowType", dataSource: "dataSource", hideColumn: "hideColumn", pageAllChecked: "pageAllChecked" }, outputs: { multiSortChange: "multiSortChange", cellClick: "cellClick", cellDBClick: "cellDBClick", rowClick: "rowClick", rowDBClick: "rowDBClick", detialToggle: "detialToggle", cellEditStart: "cellEditStart", cellEditEnd: "cellEditEnd", rowCheckChange: "rowCheckChange", checkAllChange: "checkAllChange", loadMore: "loadMore", resize: "resize", childrenTableClose: "childrenTableClose", allChildrenTableClose: "allChildrenTableClose", tableScrollEvent: "tableScrollEvent", columnDragEnd: "columnDragEnd" }, host: { properties: { "style.height": "this.hostHeight", "class.devui-table-shadow": "this.hasShadow" } }, providers: [{
            provide: DATA_TABLE,
            useExisting: forwardRef(() => DataTableComponent)
        }], queries: [{ propertyName: "innerHeader", first: true, predicate: TableTheadComponent, descendants: true }, { propertyName: "innerBody", first: true, predicate: TableTbodyComponent, descendants: true }, { propertyName: "noResultTemplate", first: true, predicate: ["noResultTemplateRef"], descendants: true }, { propertyName: "columns", predicate: DataTableColumnTmplComponent }, { propertyName: "thList", predicate: TableThComponent, descendants: true }], viewQueries: [{ propertyName: "fixHeaderContainerRefElement", first: true, predicate: ["fixHeaderContainerRef"], descendants: true }, { propertyName: "tableViewRefElement", first: true, predicate: ["tableView"], descendants: true, static: true }, { propertyName: "virtualScrollViewport", first: true, predicate: ["cdkVirtualScrollViewport"], descendants: true }, { propertyName: "normalScrollElement", first: true, predicate: ["normalScroll"], descendants: true }, { propertyName: "vitualScrollElement", first: true, predicate: ["scrollViewTpl"], descendants: true }, { propertyName: "devuiNormalScrollBody", first: true, predicate: ["devuiNormalScrollBody"], descendants: true, read: ElementRef }, { propertyName: "content", first: true, predicate: ["tableBody"], descendants: true }], exportAs: ["dataTable"], usesOnChanges: true, ngImport: i0, template: "<div class=\"devui-data-table {{ cssClass }}\">\n  <div class=\"devui-table-view {{ containFixHeaderHeight ? 'contain-fix-height' : '' }}\" #tableView>\n    <div\n      *ngIf=\"fixHeader\"\n      #fixHeaderContainerRef\n      class=\"table-wrap\"\n      [style.overflow-x]=\"'hidden'\"\n      [style.scrollbar-width]=\"'none'\"\n      [style.overflow-y]=\"tableOverflowType === 'overlay' ? 'overlay' : 'scroll'\"\n      [style.max-height]=\"maxHeight ? maxHeight : null\"\n      [style.max-width]=\"maxWidth ? maxWidth : null\"\n      [style.width]=\"!maxWidth ? tableWidth : null\"\n    >\n      <table\n        class=\"devui-table devui-table-{{ size }} {{ borderType ? 'table-' + borderType : '' }} {{ fixHeader ? 'table-fix-header' : '' }} {{\n          headerBg ? 'header-bg' : ''\n        }} {{ striped ? 'table-striped' : '' }}\"\n        [ngClass]=\"{ 'area-active': showOperationArea }\"\n        [style.table-layout]=\"tableLayout\"\n      >\n        <colgroup>\n          <col *ngFor=\"let config of tableWidthConfig\" [style.width]=\"config.width\" />\n        </colgroup>\n        <ng-container *ngTemplateOutlet=\"HeaderContent\"></ng-container>\n        <ng-template *ngIf=\"!innerHeader\" [ngTemplateOutlet]=\"headerTpl\"></ng-template>\n      </table>\n    </div>\n\n    <div\n      #normalScroll\n      cdkScrollable\n      dLazyLoad\n      [enableLazyLoad]=\"lazy\"\n      (loadMore)=\"onLoadMore($event)\"\n      *ngIf=\"!virtualScroll; else scrollViewTpl\"\n      class=\"devui-scrollbar\"\n      [ngClass]=\"{\n        'scroll-view': scrollable,\n        'overflow-overlay': tableOverflowType === 'overlay',\n        'table-wrap': !fixHeader,\n        'has-width-scroll': hasWidthScroll,\n        'has-height-scroll': hasHeightScroll\n      }\"\n      [style.height]=\"tableHeight ? tableHeight : null\"\n      [style.max-height]=\"maxHeight ? maxHeight : null\"\n      [style.max-width]=\"maxWidth ? maxWidth : null\"\n      [style.width]=\"!maxWidth ? tableWidth : null\"\n      (scroll)=\"onBodyScroll($event)\"\n    >\n      <table\n        #tableBody\n        class=\"devui-table devui-table-{{ size }} {{ borderType ? 'table-' + borderType : '' }} {{\n          rowHoveredHighlight ? 'table-hover' : ''\n        }} {{ headerBg ? 'header-bg' : '' }} {{ striped ? 'table-striped' : '' }}\"\n        [ngClass]=\"{ 'area-active': showOperationArea }\"\n        [style.table-layout]=\"tableLayout\"\n        [style.height]=\"minHeight ? minHeight : null\"\n        [style.position]=\"tableOverflowType === 'overlay' ? null : 'relative'\"\n      >\n        <colgroup>\n          <col *ngFor=\"let config of tableWidthConfig\" [style.width]=\"config.width\" />\n        </colgroup>\n        <ng-container *ngIf=\"!fixHeader\">\n          <ng-container *ngTemplateOutlet=\"HeaderContent\"></ng-container>\n          <ng-template *ngIf=\"!innerHeader\" [ngTemplateOutlet]=\"headerTpl\"></ng-template>\n        </ng-container>\n        <tbody\n          dTableBody\n          dSortable\n          *ngIf=\"innerBody\"\n          [dataSource]=\"dataSource\"\n          [tableWidthConfig]=\"tableWidthConfig\"\n          [rowTemplete]=\"innerBody.rowTempleteForSelect\"\n          [minHeight]=\"minHeight\"\n          [minHeightStretchRow]=\"innerBody.minHeightStretchRow\"\n        ></tbody>\n        <tbody\n          *ngIf=\"!innerBody\"\n          dDataTableBody\n          #devuiNormalScrollBody\n          [dataSource]=\"dataSource\"\n          [resizeable]=\"resizeable\"\n          [columns]=\"_columns\"\n          [checkable]=\"checkable\"\n          [allChecked]=\"_pageAllChecked\"\n          [selectable]=\"selectable\"\n          [editModel]=\"editModel\"\n          [editRowItem]=\"editRowItem\"\n          [showExpandToggle]=\"showExpandToggle\"\n          [detailTemplateRef]=\"detailTemplateRef\"\n          [timeout]=\"timeout\"\n          [type]=\"type\"\n          [rowHoveredHighlight]=\"rowHoveredHighlight\"\n          [tableLevel]=\"tableLevel\"\n          [tableWidthConfig]=\"tableWidthConfig\"\n          [headerExpandConfig]=\"headerExpandConfig\"\n          [generalRowHoveredData]=\"generalRowHoveredData\"\n        ></tbody>\n      </table>\n    </div>\n  </div>\n</div>\n\n<div\n  class=\"devui-no-result-wrapper\"\n  [ngClass]=\"{ 'overlay-table-no-result': tableOverflowType === 'overlay' }\"\n  *ngIf=\"!dataSource?.length && noResultTemplate != null\"\n>\n  <ng-template [ngTemplateOutlet]=\"noResultTemplate\" [ngTemplateOutletContext]=\"{ $implicit: this }\"></ng-template>\n</div>\n\n<ng-template #scrollViewTpl>\n  <cdk-virtual-scroll-viewport\n    #cdkVirtualScrollViewport\n    dLazyLoad\n    [enableLazyLoad]=\"lazy\"\n    (loadMore)=\"onLoadMore($event)\"\n    [itemSize]=\"virtualItemSize\"\n    [minBufferPx]=\"virtualMinBufferPx\"\n    [maxBufferPx]=\"virtualMaxBufferPx\"\n    class=\"devui-scrollbar viewport-wrapper\"\n    [ngClass]=\"{\n      'scroll-view': scrollable,\n      'overflow-overlay': tableOverflowType === 'overlay',\n      'virtual-scroll-wrap': !fixHeader,\n      'has-width-scroll': hasWidthScroll,\n      'has-height-scroll': hasHeightScroll\n    }\"\n    [style.height]=\"virtualBodyHeight\"\n    [style.max-height]=\"maxHeight ? maxHeight : null\"\n    [style.max-width]=\"maxWidth ? maxWidth : null\"\n    [style.width]=\"!maxWidth ? tableWidth : null\"\n    (scroll)=\"onBodyScroll($event)\"\n  >\n    <table\n      #tableBody\n      class=\"devui-table devui-table-{{ size }} {{ borderType ? 'table-' + borderType : '' }} {{\n        rowHoveredHighlight ? 'table-hover' : ''\n      }} {{ headerBg ? 'header-bg' : '' }} {{ striped ? 'table-striped' : '' }}\"\n      [ngClass]=\"{ 'area-active': showOperationArea }\"\n      [style.table-layout]=\"tableLayout\"\n      [style.height]=\"minHeight ? minHeight : null\"\n    >\n      <colgroup>\n        <col *ngFor=\"let config of tableWidthConfig\" [style.width]=\"config.width\" />\n      </colgroup>\n      <ng-container *ngIf=\"!fixHeader\">\n        <ng-container *ngTemplateOutlet=\"HeaderContent\"></ng-container>\n        <ng-template *ngIf=\"!innerHeader\" [ngTemplateOutlet]=\"headerTpl\"></ng-template>\n      </ng-container>\n      <tbody\n        dTableBody\n        dSortable\n        *ngIf=\"innerBody\"\n        [dataSource]=\"dataSource\"\n        [tableWidthConfig]=\"tableWidthConfig\"\n        [virtualScroll]=\"virtualScroll\"\n        [rowTemplete]=\"innerBody.rowTempleteForSelect\"\n        [minHeight]=\"minHeight\"\n        [minHeightStretchRow]=\"innerBody.minHeightStretchRow\"\n      ></tbody>\n      <tbody\n        *ngIf=\"!innerBody\"\n        dDataTableBody\n        [dataSource]=\"dataSource\"\n        [resizeable]=\"resizeable\"\n        [columns]=\"_columns\"\n        [checkable]=\"checkable\"\n        [allChecked]=\"_pageAllChecked\"\n        [selectable]=\"selectable\"\n        [editModel]=\"editModel\"\n        [editRowItem]=\"editRowItem\"\n        [showExpandToggle]=\"showExpandToggle\"\n        [detailTemplateRef]=\"detailTemplateRef\"\n        [timeout]=\"timeout\"\n        [type]=\"type\"\n        [rowHoveredHighlight]=\"rowHoveredHighlight\"\n        [tableLevel]=\"tableLevel\"\n        [virtualScroll]=\"virtualScroll\"\n        [tableWidthConfig]=\"tableWidthConfig\"\n        [headerExpandConfig]=\"headerExpandConfig\"\n        [generalRowHoveredData]=\"generalRowHoveredData\"\n      ></tbody>\n    </table>\n  </cdk-virtual-scroll-viewport>\n</ng-template>\n\n<ng-template #headerTpl>\n  <thead\n    dDataTableHead\n    [resizeable]=\"resizeable\"\n    [tableViewRefElement]=\"tableViewRefElement\"\n    [tableBodyEl]=\"tableBodyEl\"\n    [multiSort]=\"multiSort\"\n    [columns]=\"_columns\"\n    [pageAllChecked]=\"_pageAllChecked\"\n    [checkable]=\"checkable\"\n    [checkOptions]=\"checkOptions\"\n    [headerCheckDisabled]=\"headerCheckDisabled\"\n    [headerCheckVisible]=\"headerCheckVisible\"\n    [showExpandToggle]=\"showExpandToggle\"\n    [showSortIcon]=\"showSortIcon\"\n    [showFilterIcon]=\"showFilterIcon\"\n    (resizeHandlerEvent)=\"onResizeHandler($event)\"\n    (headClickSortEvent)=\"onHandleSort($event)\"\n    [halfChecked]=\"halfChecked\"\n    [maxHeight]=\"maxHeight\"\n    (beginResizeHandlerEvent)=\"beginResizeHandlerEvent($event)\"\n    (resizingHandlerEvent)=\"onResizingFixedHandler($event)\"\n    [fixHeader]=\"fixHeader\"\n    (dragTableEndEvent)=\"handleDragTable($event)\"\n    [dataSource]=\"dataSource\"\n    [colDropFreezeTo]=\"colDropFreezeTo\"\n    [colDraggable]=\"colDraggable\"\n    [childrenTableOpen]=\"childrenTableOpen\"\n  ></thead>\n</ng-template>\n<ng-template #HeaderContent><ng-content select=\"thead\"></ng-content></ng-template>\n", styles: [".devui-font-size-base{font-size:12px;font-size:var(--devui-font-size, 12px)}.devui-font-base{font-size:12px;font-size:var(--devui-font-size, 12px);font-weight:400;line-height:1.5}.devui-font-size-modal-title{font-size:18px;font-size:var(--devui-font-size-modal-title, 18px)}.devui-font-modal-title{font-size:18px;font-size:var(--devui-font-size-modal-title, 18px);font-weight:700;line-height:1.5}.devui-font-size-page-title{font-size:16px;font-size:var(--devui-font-size-page-title, 16px)}.devui-font-page-title{font-size:16px;font-size:var(--devui-font-size-page-title, 16px);font-weight:700;line-height:1.5}.devui-font-size-secondary-title{font-size:14px;font-size:var(--devui-font-size-card-title, 14px)}.devui-font-secondary-title{font-size:14px;font-size:var(--devui-font-size-card-title, 14px);font-weight:700;line-height:1.5}:host{display:block}:host.devui-table-shadow{border-radius:2px;border-radius:var(--devui-border-radius, 2px);overflow:hidden;box-shadow:0 1px 4px rgba(37,43,58,.1);box-shadow:var(--devui-shadow-length-base, 0 1px 4px 0) var(--devui-light-shadow, rgba(37, 43, 58, .1))}:host.devui-table-shadow ::ng-deep table tbody tr:last-child td{border-bottom-color:transparent}.devui-data-table ::ng-deep{border:none;height:100%}.devui-data-table ::ng-deep .devui-table-view{padding:0;border:none;position:relative;height:100%}.devui-data-table ::ng-deep .devui-table-view.contain-fix-height{display:flex;flex-direction:column}.devui-data-table ::ng-deep .devui-table-view.contain-fix-height .scroll-view{flex:1}.devui-data-table ::ng-deep .devui-table-view .scroll-view{overflow-x:auto;overflow-y:scroll}.devui-data-table ::ng-deep .devui-table-view .scroll-view.overflow-overlay{overflow:hidden}.devui-data-table ::ng-deep .devui-table-view .scroll-view.overflow-overlay:hover{overflow:auto;overflow:overlay}.devui-data-table ::ng-deep .devui-table-view .scroll-view.has-width-scroll:not(:hover):after{content:\"\";position:absolute;right:0;bottom:0;top:0;width:50px;background:linear-gradient(90deg,transparent,#ffffff);background:linear-gradient(90deg,transparent,var(--devui-base-bg, #ffffff))}.devui-data-table ::ng-deep .devui-table-view .scroll-view.has-height-scroll:not(:hover):before{content:\"\";position:absolute;left:0;bottom:0;right:0;height:20px;background:linear-gradient(180deg,transparent,#ffffff);background:linear-gradient(180deg,transparent,var(--devui-base-bg, #ffffff))}.devui-data-table ::ng-deep .devui-table-view .devui-table{border-collapse:separate;border-spacing:0;table-layout:fixed;width:100%;max-width:100%;margin:0}.devui-data-table ::ng-deep .devui-table-view .devui-table>tbody+tbody{border-top:none}.devui-data-table ::ng-deep .devui-table-view .viewport-wrapper{overflow-x:auto;overflow-y:scroll;position:relative}.devui-data-table ::ng-deep .devui-table-view .viewport-wrapper.overflow-overlay{overflow:hidden}.devui-data-table ::ng-deep .devui-table-view .viewport-wrapper.overflow-overlay:hover{overflow:auto;overflow:overlay}.devui-data-table ::ng-deep .devui-table-view .viewport-wrapper.has-width-scroll:not(:hover):after{content:\"\";position:absolute;right:0;bottom:0;top:0;width:50px;background:linear-gradient(90deg,transparent,#ffffff)}.devui-data-table ::ng-deep .devui-table-view .viewport-wrapper.has-height-scroll:not(:hover):before{content:\"\";position:absolute;left:0;bottom:0;right:0;height:20px;background:linear-gradient(180deg,transparent,#ffffff)}.devui-data-table ::ng-deep .devui-table-view .resize-bar{top:0;bottom:0;position:absolute;cursor:col-resize;background:#5e7ce0;background:var(--devui-form-control-line-active, #5e7ce0);width:2px;z-index:9999;display:none}.devui-data-table ::ng-deep .devui-table-view .resize-overlay{position:absolute;display:block;top:0;left:0;bottom:0;right:0;z-index:1000}.devui-data-table ::ng-deep .table-wrap{overflow-y:auto}.devui-data-table ::ng-deep .virtual-scroll-wrap .cdk-virtual-scroll-content-wrapper{padding-top:8px}.devui-data-table ::ng-deep .devui-scrollbar{scrollbar-width:thin}.devui-table ::ng-deep thead span{font-weight:700;color:#252b3a;color:var(--devui-text, #252b3a);font-size:12px;font-size:var(--devui-font-size-sm, 12px)}.devui-table ::ng-deep thead tr{background-color:#fff;background-color:var(--devui-base-bg, #ffffff)}.devui-table ::ng-deep thead tr th{text-align:left;background-clip:padding-box!important;border-bottom:1px solid #f2f2f3;border-bottom:1px solid var(--devui-dividing-line, #f2f2f3);display:table-cell;padding:0;font-size:0;background-color:inherit}.devui-table ::ng-deep thead tr th .drag-icon{display:inline-block;margin-right:2px;visibility:hidden}.devui-table ::ng-deep thead tr th .drag-icon+.childtable-toggler>.customized-icon{font-size:16px;font-size:var(--devui-font-size-icon, 16px)}.devui-table ::ng-deep thead tr th>*{display:inline-block}.devui-table ::ng-deep thead tr th .title,.devui-table ::ng-deep thead tr th .parent-title{display:inline-block;max-width:calc(100% - 18px);line-height:42px;vertical-align:middle;white-space:nowrap;text-overflow:ellipsis;overflow:hidden;font-size:12px;font-size:var(--devui-font-size-sm, 12px);font-weight:700;color:#252b3a;color:var(--devui-text, #252b3a)}.devui-table ::ng-deep thead tr th .title.can-filter{width:calc(100% - 34px)}.devui-table ::ng-deep thead tr th .childtable-toggler+.title{width:calc(100% - 38px)}.devui-table ::ng-deep thead tr th .childtable-toggler+.title.can-filter{width:calc(100% - 54px)}.devui-table ::ng-deep thead tr th:hover,.devui-table ::ng-deep thead tr th.hover{border-radius:2px 0 0 2px;border-radius:var(--devui-border-radius, 2px) 0 0 var(--devui-border-radius, 2px)}.devui-table ::ng-deep thead tr th:hover d-table-sort,.devui-table ::ng-deep thead tr th.hover d-table-sort{display:block}.devui-table ::ng-deep thead tr th:hover d-table-filter,.devui-table ::ng-deep thead tr th.hover d-table-filter{display:block}.devui-table ::ng-deep thead tr th:hover .filter-icon,.devui-table ::ng-deep thead tr th.hover .filter-icon{visibility:visible}.devui-table ::ng-deep thead tr th:hover .sort-icon-default,.devui-table ::ng-deep thead tr th.hover .sort-icon-default{visibility:visible}.devui-table ::ng-deep thead tr th.operable:hover{background-color:#f2f2f3;background-color:var(--devui-list-item-hover-bg, #f2f2f3)}.devui-table ::ng-deep thead tr th.sindu_handle:hover,.devui-table ::ng-deep thead tr th.sindu_handle.hover{border-radius:2px 0 0 2px;border-radius:var(--devui-border-radius, 2px) 0 0 var(--devui-border-radius, 2px)}.devui-table ::ng-deep thead tr th.sindu_handle:hover .drag-icon,.devui-table ::ng-deep thead tr th.sindu_handle.hover .drag-icon{visibility:visible}.devui-table ::ng-deep thead tr th.sindu_handle:hover .hidden-icon.drag-icon,.devui-table ::ng-deep thead tr th.sindu_handle.hover .hidden-icon.drag-icon{visibility:hidden}.devui-table ::ng-deep thead tr th.devui-checkable-cell{padding:10px 0 10px 20px}.devui-table ::ng-deep thead tr th.devui-detail-cell{width:36px;text-align:center;padding:0;cursor:pointer}.devui-table ::ng-deep thead tr th.devui-sticky-left-cell,.devui-table ::ng-deep thead tr th.devui-sticky-right-cell{position:-webkit-sticky;position:sticky;z-index:5}.devui-table ::ng-deep thead tr th .header-container{position:relative;display:flex;align-items:center;padding-left:2px;padding-right:8px}.devui-table ::ng-deep thead tr .sort-active{background:#f2f2f3;background:var(--devui-list-item-hover-bg, #f2f2f3);border-radius:2px 0 0 2px;border-radius:var(--devui-border-radius, 2px) 0 0 var(--devui-border-radius, 2px)}.devui-table ::ng-deep thead tr .filter-active{background:#f2f2f3;background:var(--devui-list-item-hover-bg, #f2f2f3);border-radius:2px 0 0 2px;border-radius:var(--devui-border-radius, 2px) 0 0 var(--devui-border-radius, 2px)}.devui-table ::ng-deep thead tr .check-box-all{vertical-align:middle;margin-right:5px}.devui-table ::ng-deep tbody>tr{background:#ffffff;background:var(--devui-base-bg, #ffffff);-webkit-user-select:auto;-moz-user-select:auto;-ms-user-select:auto;user-select:auto}.devui-table ::ng-deep tbody>tr.table-row-selected{background:#f2f5fc;background:var(--devui-list-item-selected-bg, #f2f5fc)}.devui-table ::ng-deep tbody>tr>td{background-clip:padding-box;vertical-align:middle;padding:10px 20px 9px;border-top:none;word-wrap:break-word;word-break:normal;line-height:22px;font-size:12px;font-size:var(--devui-font-size, 12px);color:#252b3a;color:var(--devui-text, #252b3a);background-color:inherit;border-bottom:1px solid #f2f2f3;border-bottom:1px solid var(--devui-dividing-line, #f2f2f3)}.devui-table ::ng-deep tbody>tr>td.editable-cell{position:relative}.devui-table ::ng-deep tbody>tr>td.editable-cell:hover:after{content:\"\";display:block;width:calc(100% - 32px);position:absolute;height:calc(100% - 14px);border:1px solid #d7d8da;border:1px solid var(--devui-line, #d7d8da);left:12px;top:8px;border-radius:2px;border-radius:var(--devui-border-radius, 2px);pointer-events:none}.devui-table ::ng-deep tbody>tr>td.devui-checkable-cell{padding:10px 0 10px 20px}.devui-table ::ng-deep tbody>tr>td.devui-detail-cell{width:36px;text-align:center;padding:0;cursor:pointer}.devui-table ::ng-deep tbody>tr>td.devui-sticky-left-cell,.devui-table ::ng-deep tbody>tr>td.devui-sticky-right-cell{position:-webkit-sticky;position:sticky;z-index:5}.devui-table ::ng-deep .clickable{cursor:pointer;vertical-align:middle}.devui-table ::ng-deep .d-checkbox>label{margin:auto}.devui-table ::ng-deep svg.svg-icon rect{stroke:#cfd0d3;stroke:var(--devui-disabled-text, #cfd0d3)}.devui-table ::ng-deep svg.svg-icon.svg-icon-close rect:last-child{stroke:none;fill:#cfd0d3;fill:var(--devui-disabled-text, #cfd0d3)}.devui-table ::ng-deep path{fill:#cfd0d3;fill:var(--devui-disabled-text, #cfd0d3)}.devui-table ::ng-deep svg.svg-icon:hover g path{fill:#252b3a;fill:var(--devui-icon-fill-active, #252b3a)}.devui-table ::ng-deep svg.svg-icon:hover g rect{stroke:#252b3a;stroke:var(--devui-icon-fill-active, #252b3a)}.devui-table ::ng-deep svg.svg-icon.svg-icon-close:hover rect:last-child{stroke:none;fill:#252b3a;fill:var(--devui-icon-fill-active, #252b3a)}.devui-table ::ng-deep .resize-handle{display:inline-block;position:absolute;right:0;top:0;bottom:0;width:5px;cursor:col-resize}.devui-table ::ng-deep .resizeable:hover .resize-handle{border-right:1px solid #d7d8da;border-right:1px solid var(--devui-line, #d7d8da)}.devui-table ::ng-deep .resizeable .resize-handle:hover{border-right:2px solid #5e7ce0;border-right:2px solid var(--devui-form-control-line-active, #5e7ce0)}.devui-no-result-wrapper{padding:40px 0;background:#ffffff;background:var(--devui-base-bg, #ffffff);width:calc(100% - 8px)}.devui-no-result-wrapper.overlay-table-no-result{width:100%}.devui-table.area-active ::ng-deep thead>tr>th.operable{background-color:#f2f2f3;background-color:var(--devui-list-item-hover-bg, #f2f2f3)}.devui-table.area-active ::ng-deep thead>tr>th.operable .filter-icon{visibility:visible}.devui-table.area-active ::ng-deep thead>tr>th.operable .sort-icon-default{visibility:visible}.devui-table ::ng-deep thead>tr>th.can-sort d-table-filter{flex:1;text-align:right}.devui-table.devui-table-md ::ng-deep tbody>tr>td{padding:11px 20px 12px}.devui-table.devui-table-md ::ng-deep thead>tr>th.devui-checkable-cell{padding:10px 20px}.devui-table.devui-table-lg ::ng-deep tbody>tr>td{padding:15px 20px 16px}.devui-table.devui-table-lg ::ng-deep thead>tr>th.devui-checkable-cell{padding:10px 20px}.devui-table.devui-table-xs ::ng-deep tbody>tr>td{padding:4px 20px}.devui-table.devui-table-xs ::ng-deep thead>tr>th.devui-checkable-cell{padding:10px 20px}.devui-table.devui-table-xs ::ng-deep thead tr th .title{line-height:32px}.devui-table.table-striped ::ng-deep tbody>tr:nth-of-type(even){background:#f2f5fc;background:var(--devui-list-item-strip-bg, #f2f5fc)}.devui-table.table-borderless ::ng-deep tbody>tr>td{border-bottom:none}.devui-table.header-bg ::ng-deep thead tr{background-color:#f2f5fc;background-color:var(--devui-list-item-strip-bg, #f2f5fc)}.devui-table.header-bg ::ng-deep thead tr .devui-first-sticky-right-cell{background:linear-gradient(to right,transparent,#f2f5fc 10px);background:linear-gradient(to right,transparent,var(--devui-list-item-strip-bg, #f2f5fc) 10px)}.devui-table.header-bg ::ng-deep thead tr .devui-last-sticky-left-cell{background:linear-gradient(to left,transparent,#f2f5fc 10px);background:linear-gradient(to left,transparent,var(--devui-list-item-strip-bg, #f2f5fc) 10px)}.devui-table.table-bordered ::ng-deep tr td:first-child{border-left:1px solid #f2f2f3;border-left:1px solid var(--devui-dividing-line, #f2f2f3)}.devui-table.table-bordered ::ng-deep tr td{border-right:1px solid #f2f2f3;border-right:1px solid var(--devui-dividing-line, #f2f2f3)}.devui-table.table-bordered ::ng-deep thead tr th:first-child{border-left:1px solid #f2f2f3;border-left:1px solid var(--devui-dividing-line, #f2f2f3)}.devui-table.table-bordered ::ng-deep thead tr th{border-top:1px solid #f2f2f3;border-top:1px solid var(--devui-dividing-line, #f2f2f3);border-right:1px solid #f2f2f3;border-right:1px solid var(--devui-dividing-line, #f2f2f3)}.devui-table.table-hover ::ng-deep tbody>tr:not(.table-row-selected):not(.expand-row):not(.child-table):hover{background:#f2f2f3;background:var(--devui-list-item-hover-bg, #f2f2f3)}.devui-table.table-hover ::ng-deep tbody>tr:not(.table-row-selected):not(.expand-row):not(.child-table):hover .devui-first-sticky-right-cell{background:linear-gradient(to right,transparent,#f2f2f3 10px);background:linear-gradient(to right,transparent,var(--devui-list-item-hover-bg, #f2f2f3) 10px)}.devui-table.table-hover ::ng-deep tbody>tr:not(.table-row-selected):not(.expand-row):not(.child-table):hover .devui-last-sticky-left-cell{background:linear-gradient(to left,transparent,#f2f2f3 10px);background:linear-gradient(to left,transparent,var(--devui-list-item-hover-bg, #f2f2f3) 10px)}:host ::ng-deep .devui-toggle-childtable{cursor:pointer;margin-right:5px}:host ::ng-deep .devui-toggle-childtable>svg{vertical-align:middle}:host ::ng-deep .devui-toggle-childtable>svg>g>path{fill:#babbc0;fill:var(--devui-placeholder, #babbc0)}:host ::ng-deep .full-width{width:100%}:host ::ng-deep .hover-bg{background:#f2f2f3;background:var(--devui-list-item-hover-bg, #f2f2f3);pointer-events:none}:host ::ng-deep .hover-bg+th{pointer-events:none}:host ::ng-deep .table-view-selector{-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;cursor:col-resize}:host ::ng-deep .edit-padding-fix{margin-top:-6px;margin-bottom:-6px}:host ::ng-deep .childtable-toggler{cursor:pointer;padding-right:8px;font-size:unset!important}:host ::ng-deep .childtable-toggler>svg{vertical-align:middle}:host ::ng-deep .childtable-toggler>svg>g>path{fill:#babbc0;fill:var(--devui-placeholder, #babbc0)}:host ::ng-deep .table.table-bordered.table-fix-header{border-bottom:0}:host ::ng-deep .table.table-bordered.table-fix-header th{border-bottom:0}:host ::ng-deep table.table-fix-header thead tr th:first-child{border-top-left-radius:2px;border-top-left-radius:var(--devui-border-radius, 2px)}:host ::ng-deep table.table-fix-header thead tr th:last-child{border-top-right-radius:2px;border-top-right-radius:var(--devui-border-radius, 2px)}:host ::ng-deep table thead tr th:first-child{border-top-left-radius:2px;border-top-left-radius:var(--devui-border-radius, 2px)}:host ::ng-deep table thead tr th:last-child{border-top-right-radius:2px;border-top-right-radius:var(--devui-border-radius, 2px)}:host ::ng-deep table tbody tr:last-child td:first-child{border-bottom-left-radius:2px;border-bottom-left-radius:var(--devui-border-radius, 2px)}:host ::ng-deep table tbody tr:last-child td:last-child{border-bottom-right-radius:2px;border-bottom-right-radius:var(--devui-border-radius, 2px)}:host ::ng-deep .devui-talbe-scorll-middle tr .devui-last-sticky-left-cell,:host ::ng-deep .devui-talbe-scorll-middle th .devui-last-sticky-left-cell,:host ::ng-deep .devui-talbe-scorll-right tr .devui-last-sticky-left-cell,:host ::ng-deep .devui-talbe-scorll-right th .devui-last-sticky-left-cell{background:linear-gradient(to left,transparent,#ffffff 10px);background:linear-gradient(to left,transparent,var(--devui-base-bg, #ffffff) 10px)}:host ::ng-deep .devui-talbe-scorll-middle tr .devui-last-sticky-left-cell:after,:host ::ng-deep .devui-talbe-scorll-middle th .devui-last-sticky-left-cell:after,:host ::ng-deep .devui-talbe-scorll-right tr .devui-last-sticky-left-cell:after,:host ::ng-deep .devui-talbe-scorll-right th .devui-last-sticky-left-cell:after{content:\"\";position:absolute;top:0;right:0;bottom:0;width:10px;pointer-events:none;box-shadow:inset 8px 0 8px -4px rgba(37,43,58,.1);box-shadow:inset var(--devui-table-inset-shadow-left, 8px 0 8px -4px) var(--devui-light-shadow, rgba(37, 43, 58, .1))}:host ::ng-deep .devui-talbe-scorll-middle tr .devui-first-sticky-right-cell,:host ::ng-deep .devui-talbe-scorll-middle th .devui-first-sticky-right-cell,:host ::ng-deep .devui-talbe-scorll-left tr .devui-first-sticky-right-cell,:host ::ng-deep .devui-talbe-scorll-left th .devui-first-sticky-right-cell{background:linear-gradient(to right,transparent,#ffffff 10px);background:linear-gradient(to right,transparent,var(--devui-base-bg, #ffffff) 10px)}:host ::ng-deep .devui-talbe-scorll-middle tr .devui-first-sticky-right-cell:before,:host ::ng-deep .devui-talbe-scorll-middle th .devui-first-sticky-right-cell:before,:host ::ng-deep .devui-talbe-scorll-left tr .devui-first-sticky-right-cell:before,:host ::ng-deep .devui-talbe-scorll-left th .devui-first-sticky-right-cell:before{content:\"\";position:absolute;top:0;left:0;bottom:0;width:10px;pointer-events:none;box-shadow:inset -8px 0 8px -4px rgba(37,43,58,.1);box-shadow:inset var(--devui-table-inset-shadow-right, -8px 0 8px -4px) var(--devui-light-shadow, rgba(37, 43, 58, .1))}:host ::ng-deep .gu-unselectable{-webkit-user-select:none!important;-moz-user-select:none!important;-ms-user-select:none!important;user-select:none!important}:host ::ng-deep .gu-mirror{position:fixed!important;margin:0!important;z-index:9999!important;background-color:#fff;box-shadow:0 2px 8px 0 var(--dark-1-boxshadow, rgba(41, 48, 64, .2));box-shadow:0 2px 8px rgba(41,48,64,.2);box-shadow:var(--devui-shadow-connected-overlay, 0 2px 8px 0) var(--dark-1-boxshadow, rgba(41, 48, 64, .2));list-style:none}:host ::ng-deep .gu-hide{display:none!important}:host ::ng-deep .sindu_dragger{list-style:none;margin:0;padding:0;box-sizing:border-box;display:flex}:host ::ng-deep .sindu_handle>.header-container>.drag-icon{cursor:move}:host ::ng-deep .sindu_dragger li{margin:0;padding:0;list-style:none;text-align:inherit}:host ::ng-deep .sindu_dragger li table,:host ::ng-deep .sindu_dragger tr,:host ::ng-deep .sindu_dragger th,:host ::ng-deep .sindu_dragger td{box-sizing:border-box}:host ::ng-deep .sindu_dragging .sindu_origin_table{visibility:hidden}:host ::ng-deep .gu-mirror li{margin:0;padding:0;list-style:none;text-align:inherit}:host ::ng-deep .gu-mirror li table,:host ::ng-deep .gu-mirror tr,:host ::ng-deep .gu-mirror th,:host ::ng-deep .gu-mirror td{box-sizing:border-box}:host ::ng-deep .gu-transit{opacity:.5}:host ::ng-deep .devui-table.table-bordered thead tr th.devui-checkable-cell{padding:10px}:host ::ng-deep .devui-table.table-bordered tbody>tr>td.devui-checkable-cell{padding:10px}\n", ":host ::ng-deep th .datatable-svg g#order-default path{fill:#71757f;fill:var(--devui-icon-fill, #71757f)}:host ::ng-deep th .datatable-svg g#order-desc polygon#Path-4{fill:#71757f;fill:var(--devui-icon-fill, #71757f)}:host ::ng-deep th .datatable-svg g#order-desc polygon#Path-4-Copy{fill:#252b3a;fill:var(--devui-icon-fill-active, #252b3a)}:host ::ng-deep th .datatable-svg g#order-asc polygon#Path-4{fill:#252b3a;fill:var(--devui-icon-fill-active, #252b3a)}:host ::ng-deep th .datatable-svg g#order-asc polygon#Path-4-Copy{fill:#71757f;fill:var(--devui-icon-fill, #71757f)}\n"], components: [{ type: i1.TableTbodyComponent, selector: "[dTableBody]", inputs: ["dataSource", "rowTemplete", "nestedLayer", "nestedIndex", "virtualScroll", "tableWidthConfig", "minHeight", "minHeightStretchRow"] }, { type: i2.DataTableBodyComponent, selector: "d-data-table-body, [dDataTableBody]", inputs: ["checkable", "showExpandToggle", "allChecked", "selectable", "editModel", "editRowItem", "dataSource", "resizeable", "columns", "detailTemplateRef", "timeout", "type", "rowHoveredHighlight", "tableLevel", "nestedIndex", "virtualScroll", "tableWidthConfig", "headerExpandConfig", "generalRowHoveredData"] }, { type: i3.CdkVirtualScrollViewport, selector: "cdk-virtual-scroll-viewport", inputs: ["orientation", "appendOnly"], outputs: ["scrolledIndexChange"] }, { type: i4.DataTableHeadComponent, selector: "d-data-table-head,[dDataTableHead]", inputs: ["checkable", "headerCheckDisabled", "headerCheckVisible", "checkOptions", "showExpandToggle", "pageAllChecked", "columns", "multiSort", "resizeable", "maxHeight", "showSortIcon", "showFilterIcon", "colDropFreezeTo", "colDraggable", "fixHeader", "dataSource", "tableViewRefElement", "tableBodyEl", "checkableColumn", "showExpandToggleColumn", "halfChecked", "childrenTableOpen"], outputs: ["headClickSortEvent", "resizeHandlerEvent", "beginResizeHandlerEvent", "resizingHandlerEvent", "dragTableEndEvent"] }], directives: [{ type: i5.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i5.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { type: i5.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i5.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet"] }, { type: i3.CdkScrollable, selector: "[cdk-scrollable], [cdkScrollable]" }, { type: i6.LazyLoadDirective, selector: "[dLazyLoad]", inputs: ["enableLazyLoad", "contentMode", "target", "imgLoadSrc"], outputs: ["loadMore"] }, { type: i7.SortableDirective, selector: "[dSortable]", inputs: ["dSortable", "dSortableZMode"] }, { type: i3.CdkFixedSizeVirtualScroll, selector: "cdk-virtual-scroll-viewport[itemSize]", inputs: ["itemSize", "minBufferPx", "maxBufferPx"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DataTableComponent, decorators: [{
            type: Component,
            args: [{ selector: 'd-data-table', exportAs: 'dataTable', preserveWhitespaces: false, providers: [{
                            provide: DATA_TABLE,
                            useExisting: forwardRef(() => DataTableComponent)
                        }], template: "<div class=\"devui-data-table {{ cssClass }}\">\n  <div class=\"devui-table-view {{ containFixHeaderHeight ? 'contain-fix-height' : '' }}\" #tableView>\n    <div\n      *ngIf=\"fixHeader\"\n      #fixHeaderContainerRef\n      class=\"table-wrap\"\n      [style.overflow-x]=\"'hidden'\"\n      [style.scrollbar-width]=\"'none'\"\n      [style.overflow-y]=\"tableOverflowType === 'overlay' ? 'overlay' : 'scroll'\"\n      [style.max-height]=\"maxHeight ? maxHeight : null\"\n      [style.max-width]=\"maxWidth ? maxWidth : null\"\n      [style.width]=\"!maxWidth ? tableWidth : null\"\n    >\n      <table\n        class=\"devui-table devui-table-{{ size }} {{ borderType ? 'table-' + borderType : '' }} {{ fixHeader ? 'table-fix-header' : '' }} {{\n          headerBg ? 'header-bg' : ''\n        }} {{ striped ? 'table-striped' : '' }}\"\n        [ngClass]=\"{ 'area-active': showOperationArea }\"\n        [style.table-layout]=\"tableLayout\"\n      >\n        <colgroup>\n          <col *ngFor=\"let config of tableWidthConfig\" [style.width]=\"config.width\" />\n        </colgroup>\n        <ng-container *ngTemplateOutlet=\"HeaderContent\"></ng-container>\n        <ng-template *ngIf=\"!innerHeader\" [ngTemplateOutlet]=\"headerTpl\"></ng-template>\n      </table>\n    </div>\n\n    <div\n      #normalScroll\n      cdkScrollable\n      dLazyLoad\n      [enableLazyLoad]=\"lazy\"\n      (loadMore)=\"onLoadMore($event)\"\n      *ngIf=\"!virtualScroll; else scrollViewTpl\"\n      class=\"devui-scrollbar\"\n      [ngClass]=\"{\n        'scroll-view': scrollable,\n        'overflow-overlay': tableOverflowType === 'overlay',\n        'table-wrap': !fixHeader,\n        'has-width-scroll': hasWidthScroll,\n        'has-height-scroll': hasHeightScroll\n      }\"\n      [style.height]=\"tableHeight ? tableHeight : null\"\n      [style.max-height]=\"maxHeight ? maxHeight : null\"\n      [style.max-width]=\"maxWidth ? maxWidth : null\"\n      [style.width]=\"!maxWidth ? tableWidth : null\"\n      (scroll)=\"onBodyScroll($event)\"\n    >\n      <table\n        #tableBody\n        class=\"devui-table devui-table-{{ size }} {{ borderType ? 'table-' + borderType : '' }} {{\n          rowHoveredHighlight ? 'table-hover' : ''\n        }} {{ headerBg ? 'header-bg' : '' }} {{ striped ? 'table-striped' : '' }}\"\n        [ngClass]=\"{ 'area-active': showOperationArea }\"\n        [style.table-layout]=\"tableLayout\"\n        [style.height]=\"minHeight ? minHeight : null\"\n        [style.position]=\"tableOverflowType === 'overlay' ? null : 'relative'\"\n      >\n        <colgroup>\n          <col *ngFor=\"let config of tableWidthConfig\" [style.width]=\"config.width\" />\n        </colgroup>\n        <ng-container *ngIf=\"!fixHeader\">\n          <ng-container *ngTemplateOutlet=\"HeaderContent\"></ng-container>\n          <ng-template *ngIf=\"!innerHeader\" [ngTemplateOutlet]=\"headerTpl\"></ng-template>\n        </ng-container>\n        <tbody\n          dTableBody\n          dSortable\n          *ngIf=\"innerBody\"\n          [dataSource]=\"dataSource\"\n          [tableWidthConfig]=\"tableWidthConfig\"\n          [rowTemplete]=\"innerBody.rowTempleteForSelect\"\n          [minHeight]=\"minHeight\"\n          [minHeightStretchRow]=\"innerBody.minHeightStretchRow\"\n        ></tbody>\n        <tbody\n          *ngIf=\"!innerBody\"\n          dDataTableBody\n          #devuiNormalScrollBody\n          [dataSource]=\"dataSource\"\n          [resizeable]=\"resizeable\"\n          [columns]=\"_columns\"\n          [checkable]=\"checkable\"\n          [allChecked]=\"_pageAllChecked\"\n          [selectable]=\"selectable\"\n          [editModel]=\"editModel\"\n          [editRowItem]=\"editRowItem\"\n          [showExpandToggle]=\"showExpandToggle\"\n          [detailTemplateRef]=\"detailTemplateRef\"\n          [timeout]=\"timeout\"\n          [type]=\"type\"\n          [rowHoveredHighlight]=\"rowHoveredHighlight\"\n          [tableLevel]=\"tableLevel\"\n          [tableWidthConfig]=\"tableWidthConfig\"\n          [headerExpandConfig]=\"headerExpandConfig\"\n          [generalRowHoveredData]=\"generalRowHoveredData\"\n        ></tbody>\n      </table>\n    </div>\n  </div>\n</div>\n\n<div\n  class=\"devui-no-result-wrapper\"\n  [ngClass]=\"{ 'overlay-table-no-result': tableOverflowType === 'overlay' }\"\n  *ngIf=\"!dataSource?.length && noResultTemplate != null\"\n>\n  <ng-template [ngTemplateOutlet]=\"noResultTemplate\" [ngTemplateOutletContext]=\"{ $implicit: this }\"></ng-template>\n</div>\n\n<ng-template #scrollViewTpl>\n  <cdk-virtual-scroll-viewport\n    #cdkVirtualScrollViewport\n    dLazyLoad\n    [enableLazyLoad]=\"lazy\"\n    (loadMore)=\"onLoadMore($event)\"\n    [itemSize]=\"virtualItemSize\"\n    [minBufferPx]=\"virtualMinBufferPx\"\n    [maxBufferPx]=\"virtualMaxBufferPx\"\n    class=\"devui-scrollbar viewport-wrapper\"\n    [ngClass]=\"{\n      'scroll-view': scrollable,\n      'overflow-overlay': tableOverflowType === 'overlay',\n      'virtual-scroll-wrap': !fixHeader,\n      'has-width-scroll': hasWidthScroll,\n      'has-height-scroll': hasHeightScroll\n    }\"\n    [style.height]=\"virtualBodyHeight\"\n    [style.max-height]=\"maxHeight ? maxHeight : null\"\n    [style.max-width]=\"maxWidth ? maxWidth : null\"\n    [style.width]=\"!maxWidth ? tableWidth : null\"\n    (scroll)=\"onBodyScroll($event)\"\n  >\n    <table\n      #tableBody\n      class=\"devui-table devui-table-{{ size }} {{ borderType ? 'table-' + borderType : '' }} {{\n        rowHoveredHighlight ? 'table-hover' : ''\n      }} {{ headerBg ? 'header-bg' : '' }} {{ striped ? 'table-striped' : '' }}\"\n      [ngClass]=\"{ 'area-active': showOperationArea }\"\n      [style.table-layout]=\"tableLayout\"\n      [style.height]=\"minHeight ? minHeight : null\"\n    >\n      <colgroup>\n        <col *ngFor=\"let config of tableWidthConfig\" [style.width]=\"config.width\" />\n      </colgroup>\n      <ng-container *ngIf=\"!fixHeader\">\n        <ng-container *ngTemplateOutlet=\"HeaderContent\"></ng-container>\n        <ng-template *ngIf=\"!innerHeader\" [ngTemplateOutlet]=\"headerTpl\"></ng-template>\n      </ng-container>\n      <tbody\n        dTableBody\n        dSortable\n        *ngIf=\"innerBody\"\n        [dataSource]=\"dataSource\"\n        [tableWidthConfig]=\"tableWidthConfig\"\n        [virtualScroll]=\"virtualScroll\"\n        [rowTemplete]=\"innerBody.rowTempleteForSelect\"\n        [minHeight]=\"minHeight\"\n        [minHeightStretchRow]=\"innerBody.minHeightStretchRow\"\n      ></tbody>\n      <tbody\n        *ngIf=\"!innerBody\"\n        dDataTableBody\n        [dataSource]=\"dataSource\"\n        [resizeable]=\"resizeable\"\n        [columns]=\"_columns\"\n        [checkable]=\"checkable\"\n        [allChecked]=\"_pageAllChecked\"\n        [selectable]=\"selectable\"\n        [editModel]=\"editModel\"\n        [editRowItem]=\"editRowItem\"\n        [showExpandToggle]=\"showExpandToggle\"\n        [detailTemplateRef]=\"detailTemplateRef\"\n        [timeout]=\"timeout\"\n        [type]=\"type\"\n        [rowHoveredHighlight]=\"rowHoveredHighlight\"\n        [tableLevel]=\"tableLevel\"\n        [virtualScroll]=\"virtualScroll\"\n        [tableWidthConfig]=\"tableWidthConfig\"\n        [headerExpandConfig]=\"headerExpandConfig\"\n        [generalRowHoveredData]=\"generalRowHoveredData\"\n      ></tbody>\n    </table>\n  </cdk-virtual-scroll-viewport>\n</ng-template>\n\n<ng-template #headerTpl>\n  <thead\n    dDataTableHead\n    [resizeable]=\"resizeable\"\n    [tableViewRefElement]=\"tableViewRefElement\"\n    [tableBodyEl]=\"tableBodyEl\"\n    [multiSort]=\"multiSort\"\n    [columns]=\"_columns\"\n    [pageAllChecked]=\"_pageAllChecked\"\n    [checkable]=\"checkable\"\n    [checkOptions]=\"checkOptions\"\n    [headerCheckDisabled]=\"headerCheckDisabled\"\n    [headerCheckVisible]=\"headerCheckVisible\"\n    [showExpandToggle]=\"showExpandToggle\"\n    [showSortIcon]=\"showSortIcon\"\n    [showFilterIcon]=\"showFilterIcon\"\n    (resizeHandlerEvent)=\"onResizeHandler($event)\"\n    (headClickSortEvent)=\"onHandleSort($event)\"\n    [halfChecked]=\"halfChecked\"\n    [maxHeight]=\"maxHeight\"\n    (beginResizeHandlerEvent)=\"beginResizeHandlerEvent($event)\"\n    (resizingHandlerEvent)=\"onResizingFixedHandler($event)\"\n    [fixHeader]=\"fixHeader\"\n    (dragTableEndEvent)=\"handleDragTable($event)\"\n    [dataSource]=\"dataSource\"\n    [colDropFreezeTo]=\"colDropFreezeTo\"\n    [colDraggable]=\"colDraggable\"\n    [childrenTableOpen]=\"childrenTableOpen\"\n  ></thead>\n</ng-template>\n<ng-template #HeaderContent><ng-content select=\"thead\"></ng-content></ng-template>\n", styles: [".devui-font-size-base{font-size:12px;font-size:var(--devui-font-size, 12px)}.devui-font-base{font-size:12px;font-size:var(--devui-font-size, 12px);font-weight:400;line-height:1.5}.devui-font-size-modal-title{font-size:18px;font-size:var(--devui-font-size-modal-title, 18px)}.devui-font-modal-title{font-size:18px;font-size:var(--devui-font-size-modal-title, 18px);font-weight:700;line-height:1.5}.devui-font-size-page-title{font-size:16px;font-size:var(--devui-font-size-page-title, 16px)}.devui-font-page-title{font-size:16px;font-size:var(--devui-font-size-page-title, 16px);font-weight:700;line-height:1.5}.devui-font-size-secondary-title{font-size:14px;font-size:var(--devui-font-size-card-title, 14px)}.devui-font-secondary-title{font-size:14px;font-size:var(--devui-font-size-card-title, 14px);font-weight:700;line-height:1.5}:host{display:block}:host.devui-table-shadow{border-radius:2px;border-radius:var(--devui-border-radius, 2px);overflow:hidden;box-shadow:0 1px 4px rgba(37,43,58,.1);box-shadow:var(--devui-shadow-length-base, 0 1px 4px 0) var(--devui-light-shadow, rgba(37, 43, 58, .1))}:host.devui-table-shadow ::ng-deep table tbody tr:last-child td{border-bottom-color:transparent}.devui-data-table ::ng-deep{border:none;height:100%}.devui-data-table ::ng-deep .devui-table-view{padding:0;border:none;position:relative;height:100%}.devui-data-table ::ng-deep .devui-table-view.contain-fix-height{display:flex;flex-direction:column}.devui-data-table ::ng-deep .devui-table-view.contain-fix-height .scroll-view{flex:1}.devui-data-table ::ng-deep .devui-table-view .scroll-view{overflow-x:auto;overflow-y:scroll}.devui-data-table ::ng-deep .devui-table-view .scroll-view.overflow-overlay{overflow:hidden}.devui-data-table ::ng-deep .devui-table-view .scroll-view.overflow-overlay:hover{overflow:auto;overflow:overlay}.devui-data-table ::ng-deep .devui-table-view .scroll-view.has-width-scroll:not(:hover):after{content:\"\";position:absolute;right:0;bottom:0;top:0;width:50px;background:linear-gradient(90deg,transparent,#ffffff);background:linear-gradient(90deg,transparent,var(--devui-base-bg, #ffffff))}.devui-data-table ::ng-deep .devui-table-view .scroll-view.has-height-scroll:not(:hover):before{content:\"\";position:absolute;left:0;bottom:0;right:0;height:20px;background:linear-gradient(180deg,transparent,#ffffff);background:linear-gradient(180deg,transparent,var(--devui-base-bg, #ffffff))}.devui-data-table ::ng-deep .devui-table-view .devui-table{border-collapse:separate;border-spacing:0;table-layout:fixed;width:100%;max-width:100%;margin:0}.devui-data-table ::ng-deep .devui-table-view .devui-table>tbody+tbody{border-top:none}.devui-data-table ::ng-deep .devui-table-view .viewport-wrapper{overflow-x:auto;overflow-y:scroll;position:relative}.devui-data-table ::ng-deep .devui-table-view .viewport-wrapper.overflow-overlay{overflow:hidden}.devui-data-table ::ng-deep .devui-table-view .viewport-wrapper.overflow-overlay:hover{overflow:auto;overflow:overlay}.devui-data-table ::ng-deep .devui-table-view .viewport-wrapper.has-width-scroll:not(:hover):after{content:\"\";position:absolute;right:0;bottom:0;top:0;width:50px;background:linear-gradient(90deg,transparent,#ffffff)}.devui-data-table ::ng-deep .devui-table-view .viewport-wrapper.has-height-scroll:not(:hover):before{content:\"\";position:absolute;left:0;bottom:0;right:0;height:20px;background:linear-gradient(180deg,transparent,#ffffff)}.devui-data-table ::ng-deep .devui-table-view .resize-bar{top:0;bottom:0;position:absolute;cursor:col-resize;background:#5e7ce0;background:var(--devui-form-control-line-active, #5e7ce0);width:2px;z-index:9999;display:none}.devui-data-table ::ng-deep .devui-table-view .resize-overlay{position:absolute;display:block;top:0;left:0;bottom:0;right:0;z-index:1000}.devui-data-table ::ng-deep .table-wrap{overflow-y:auto}.devui-data-table ::ng-deep .virtual-scroll-wrap .cdk-virtual-scroll-content-wrapper{padding-top:8px}.devui-data-table ::ng-deep .devui-scrollbar{scrollbar-width:thin}.devui-table ::ng-deep thead span{font-weight:700;color:#252b3a;color:var(--devui-text, #252b3a);font-size:12px;font-size:var(--devui-font-size-sm, 12px)}.devui-table ::ng-deep thead tr{background-color:#fff;background-color:var(--devui-base-bg, #ffffff)}.devui-table ::ng-deep thead tr th{text-align:left;background-clip:padding-box!important;border-bottom:1px solid #f2f2f3;border-bottom:1px solid var(--devui-dividing-line, #f2f2f3);display:table-cell;padding:0;font-size:0;background-color:inherit}.devui-table ::ng-deep thead tr th .drag-icon{display:inline-block;margin-right:2px;visibility:hidden}.devui-table ::ng-deep thead tr th .drag-icon+.childtable-toggler>.customized-icon{font-size:16px;font-size:var(--devui-font-size-icon, 16px)}.devui-table ::ng-deep thead tr th>*{display:inline-block}.devui-table ::ng-deep thead tr th .title,.devui-table ::ng-deep thead tr th .parent-title{display:inline-block;max-width:calc(100% - 18px);line-height:42px;vertical-align:middle;white-space:nowrap;text-overflow:ellipsis;overflow:hidden;font-size:12px;font-size:var(--devui-font-size-sm, 12px);font-weight:700;color:#252b3a;color:var(--devui-text, #252b3a)}.devui-table ::ng-deep thead tr th .title.can-filter{width:calc(100% - 34px)}.devui-table ::ng-deep thead tr th .childtable-toggler+.title{width:calc(100% - 38px)}.devui-table ::ng-deep thead tr th .childtable-toggler+.title.can-filter{width:calc(100% - 54px)}.devui-table ::ng-deep thead tr th:hover,.devui-table ::ng-deep thead tr th.hover{border-radius:2px 0 0 2px;border-radius:var(--devui-border-radius, 2px) 0 0 var(--devui-border-radius, 2px)}.devui-table ::ng-deep thead tr th:hover d-table-sort,.devui-table ::ng-deep thead tr th.hover d-table-sort{display:block}.devui-table ::ng-deep thead tr th:hover d-table-filter,.devui-table ::ng-deep thead tr th.hover d-table-filter{display:block}.devui-table ::ng-deep thead tr th:hover .filter-icon,.devui-table ::ng-deep thead tr th.hover .filter-icon{visibility:visible}.devui-table ::ng-deep thead tr th:hover .sort-icon-default,.devui-table ::ng-deep thead tr th.hover .sort-icon-default{visibility:visible}.devui-table ::ng-deep thead tr th.operable:hover{background-color:#f2f2f3;background-color:var(--devui-list-item-hover-bg, #f2f2f3)}.devui-table ::ng-deep thead tr th.sindu_handle:hover,.devui-table ::ng-deep thead tr th.sindu_handle.hover{border-radius:2px 0 0 2px;border-radius:var(--devui-border-radius, 2px) 0 0 var(--devui-border-radius, 2px)}.devui-table ::ng-deep thead tr th.sindu_handle:hover .drag-icon,.devui-table ::ng-deep thead tr th.sindu_handle.hover .drag-icon{visibility:visible}.devui-table ::ng-deep thead tr th.sindu_handle:hover .hidden-icon.drag-icon,.devui-table ::ng-deep thead tr th.sindu_handle.hover .hidden-icon.drag-icon{visibility:hidden}.devui-table ::ng-deep thead tr th.devui-checkable-cell{padding:10px 0 10px 20px}.devui-table ::ng-deep thead tr th.devui-detail-cell{width:36px;text-align:center;padding:0;cursor:pointer}.devui-table ::ng-deep thead tr th.devui-sticky-left-cell,.devui-table ::ng-deep thead tr th.devui-sticky-right-cell{position:-webkit-sticky;position:sticky;z-index:5}.devui-table ::ng-deep thead tr th .header-container{position:relative;display:flex;align-items:center;padding-left:2px;padding-right:8px}.devui-table ::ng-deep thead tr .sort-active{background:#f2f2f3;background:var(--devui-list-item-hover-bg, #f2f2f3);border-radius:2px 0 0 2px;border-radius:var(--devui-border-radius, 2px) 0 0 var(--devui-border-radius, 2px)}.devui-table ::ng-deep thead tr .filter-active{background:#f2f2f3;background:var(--devui-list-item-hover-bg, #f2f2f3);border-radius:2px 0 0 2px;border-radius:var(--devui-border-radius, 2px) 0 0 var(--devui-border-radius, 2px)}.devui-table ::ng-deep thead tr .check-box-all{vertical-align:middle;margin-right:5px}.devui-table ::ng-deep tbody>tr{background:#ffffff;background:var(--devui-base-bg, #ffffff);-webkit-user-select:auto;-moz-user-select:auto;-ms-user-select:auto;user-select:auto}.devui-table ::ng-deep tbody>tr.table-row-selected{background:#f2f5fc;background:var(--devui-list-item-selected-bg, #f2f5fc)}.devui-table ::ng-deep tbody>tr>td{background-clip:padding-box;vertical-align:middle;padding:10px 20px 9px;border-top:none;word-wrap:break-word;word-break:normal;line-height:22px;font-size:12px;font-size:var(--devui-font-size, 12px);color:#252b3a;color:var(--devui-text, #252b3a);background-color:inherit;border-bottom:1px solid #f2f2f3;border-bottom:1px solid var(--devui-dividing-line, #f2f2f3)}.devui-table ::ng-deep tbody>tr>td.editable-cell{position:relative}.devui-table ::ng-deep tbody>tr>td.editable-cell:hover:after{content:\"\";display:block;width:calc(100% - 32px);position:absolute;height:calc(100% - 14px);border:1px solid #d7d8da;border:1px solid var(--devui-line, #d7d8da);left:12px;top:8px;border-radius:2px;border-radius:var(--devui-border-radius, 2px);pointer-events:none}.devui-table ::ng-deep tbody>tr>td.devui-checkable-cell{padding:10px 0 10px 20px}.devui-table ::ng-deep tbody>tr>td.devui-detail-cell{width:36px;text-align:center;padding:0;cursor:pointer}.devui-table ::ng-deep tbody>tr>td.devui-sticky-left-cell,.devui-table ::ng-deep tbody>tr>td.devui-sticky-right-cell{position:-webkit-sticky;position:sticky;z-index:5}.devui-table ::ng-deep .clickable{cursor:pointer;vertical-align:middle}.devui-table ::ng-deep .d-checkbox>label{margin:auto}.devui-table ::ng-deep svg.svg-icon rect{stroke:#cfd0d3;stroke:var(--devui-disabled-text, #cfd0d3)}.devui-table ::ng-deep svg.svg-icon.svg-icon-close rect:last-child{stroke:none;fill:#cfd0d3;fill:var(--devui-disabled-text, #cfd0d3)}.devui-table ::ng-deep path{fill:#cfd0d3;fill:var(--devui-disabled-text, #cfd0d3)}.devui-table ::ng-deep svg.svg-icon:hover g path{fill:#252b3a;fill:var(--devui-icon-fill-active, #252b3a)}.devui-table ::ng-deep svg.svg-icon:hover g rect{stroke:#252b3a;stroke:var(--devui-icon-fill-active, #252b3a)}.devui-table ::ng-deep svg.svg-icon.svg-icon-close:hover rect:last-child{stroke:none;fill:#252b3a;fill:var(--devui-icon-fill-active, #252b3a)}.devui-table ::ng-deep .resize-handle{display:inline-block;position:absolute;right:0;top:0;bottom:0;width:5px;cursor:col-resize}.devui-table ::ng-deep .resizeable:hover .resize-handle{border-right:1px solid #d7d8da;border-right:1px solid var(--devui-line, #d7d8da)}.devui-table ::ng-deep .resizeable .resize-handle:hover{border-right:2px solid #5e7ce0;border-right:2px solid var(--devui-form-control-line-active, #5e7ce0)}.devui-no-result-wrapper{padding:40px 0;background:#ffffff;background:var(--devui-base-bg, #ffffff);width:calc(100% - 8px)}.devui-no-result-wrapper.overlay-table-no-result{width:100%}.devui-table.area-active ::ng-deep thead>tr>th.operable{background-color:#f2f2f3;background-color:var(--devui-list-item-hover-bg, #f2f2f3)}.devui-table.area-active ::ng-deep thead>tr>th.operable .filter-icon{visibility:visible}.devui-table.area-active ::ng-deep thead>tr>th.operable .sort-icon-default{visibility:visible}.devui-table ::ng-deep thead>tr>th.can-sort d-table-filter{flex:1;text-align:right}.devui-table.devui-table-md ::ng-deep tbody>tr>td{padding:11px 20px 12px}.devui-table.devui-table-md ::ng-deep thead>tr>th.devui-checkable-cell{padding:10px 20px}.devui-table.devui-table-lg ::ng-deep tbody>tr>td{padding:15px 20px 16px}.devui-table.devui-table-lg ::ng-deep thead>tr>th.devui-checkable-cell{padding:10px 20px}.devui-table.devui-table-xs ::ng-deep tbody>tr>td{padding:4px 20px}.devui-table.devui-table-xs ::ng-deep thead>tr>th.devui-checkable-cell{padding:10px 20px}.devui-table.devui-table-xs ::ng-deep thead tr th .title{line-height:32px}.devui-table.table-striped ::ng-deep tbody>tr:nth-of-type(even){background:#f2f5fc;background:var(--devui-list-item-strip-bg, #f2f5fc)}.devui-table.table-borderless ::ng-deep tbody>tr>td{border-bottom:none}.devui-table.header-bg ::ng-deep thead tr{background-color:#f2f5fc;background-color:var(--devui-list-item-strip-bg, #f2f5fc)}.devui-table.header-bg ::ng-deep thead tr .devui-first-sticky-right-cell{background:linear-gradient(to right,transparent,#f2f5fc 10px);background:linear-gradient(to right,transparent,var(--devui-list-item-strip-bg, #f2f5fc) 10px)}.devui-table.header-bg ::ng-deep thead tr .devui-last-sticky-left-cell{background:linear-gradient(to left,transparent,#f2f5fc 10px);background:linear-gradient(to left,transparent,var(--devui-list-item-strip-bg, #f2f5fc) 10px)}.devui-table.table-bordered ::ng-deep tr td:first-child{border-left:1px solid #f2f2f3;border-left:1px solid var(--devui-dividing-line, #f2f2f3)}.devui-table.table-bordered ::ng-deep tr td{border-right:1px solid #f2f2f3;border-right:1px solid var(--devui-dividing-line, #f2f2f3)}.devui-table.table-bordered ::ng-deep thead tr th:first-child{border-left:1px solid #f2f2f3;border-left:1px solid var(--devui-dividing-line, #f2f2f3)}.devui-table.table-bordered ::ng-deep thead tr th{border-top:1px solid #f2f2f3;border-top:1px solid var(--devui-dividing-line, #f2f2f3);border-right:1px solid #f2f2f3;border-right:1px solid var(--devui-dividing-line, #f2f2f3)}.devui-table.table-hover ::ng-deep tbody>tr:not(.table-row-selected):not(.expand-row):not(.child-table):hover{background:#f2f2f3;background:var(--devui-list-item-hover-bg, #f2f2f3)}.devui-table.table-hover ::ng-deep tbody>tr:not(.table-row-selected):not(.expand-row):not(.child-table):hover .devui-first-sticky-right-cell{background:linear-gradient(to right,transparent,#f2f2f3 10px);background:linear-gradient(to right,transparent,var(--devui-list-item-hover-bg, #f2f2f3) 10px)}.devui-table.table-hover ::ng-deep tbody>tr:not(.table-row-selected):not(.expand-row):not(.child-table):hover .devui-last-sticky-left-cell{background:linear-gradient(to left,transparent,#f2f2f3 10px);background:linear-gradient(to left,transparent,var(--devui-list-item-hover-bg, #f2f2f3) 10px)}:host ::ng-deep .devui-toggle-childtable{cursor:pointer;margin-right:5px}:host ::ng-deep .devui-toggle-childtable>svg{vertical-align:middle}:host ::ng-deep .devui-toggle-childtable>svg>g>path{fill:#babbc0;fill:var(--devui-placeholder, #babbc0)}:host ::ng-deep .full-width{width:100%}:host ::ng-deep .hover-bg{background:#f2f2f3;background:var(--devui-list-item-hover-bg, #f2f2f3);pointer-events:none}:host ::ng-deep .hover-bg+th{pointer-events:none}:host ::ng-deep .table-view-selector{-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;cursor:col-resize}:host ::ng-deep .edit-padding-fix{margin-top:-6px;margin-bottom:-6px}:host ::ng-deep .childtable-toggler{cursor:pointer;padding-right:8px;font-size:unset!important}:host ::ng-deep .childtable-toggler>svg{vertical-align:middle}:host ::ng-deep .childtable-toggler>svg>g>path{fill:#babbc0;fill:var(--devui-placeholder, #babbc0)}:host ::ng-deep .table.table-bordered.table-fix-header{border-bottom:0}:host ::ng-deep .table.table-bordered.table-fix-header th{border-bottom:0}:host ::ng-deep table.table-fix-header thead tr th:first-child{border-top-left-radius:2px;border-top-left-radius:var(--devui-border-radius, 2px)}:host ::ng-deep table.table-fix-header thead tr th:last-child{border-top-right-radius:2px;border-top-right-radius:var(--devui-border-radius, 2px)}:host ::ng-deep table thead tr th:first-child{border-top-left-radius:2px;border-top-left-radius:var(--devui-border-radius, 2px)}:host ::ng-deep table thead tr th:last-child{border-top-right-radius:2px;border-top-right-radius:var(--devui-border-radius, 2px)}:host ::ng-deep table tbody tr:last-child td:first-child{border-bottom-left-radius:2px;border-bottom-left-radius:var(--devui-border-radius, 2px)}:host ::ng-deep table tbody tr:last-child td:last-child{border-bottom-right-radius:2px;border-bottom-right-radius:var(--devui-border-radius, 2px)}:host ::ng-deep .devui-talbe-scorll-middle tr .devui-last-sticky-left-cell,:host ::ng-deep .devui-talbe-scorll-middle th .devui-last-sticky-left-cell,:host ::ng-deep .devui-talbe-scorll-right tr .devui-last-sticky-left-cell,:host ::ng-deep .devui-talbe-scorll-right th .devui-last-sticky-left-cell{background:linear-gradient(to left,transparent,#ffffff 10px);background:linear-gradient(to left,transparent,var(--devui-base-bg, #ffffff) 10px)}:host ::ng-deep .devui-talbe-scorll-middle tr .devui-last-sticky-left-cell:after,:host ::ng-deep .devui-talbe-scorll-middle th .devui-last-sticky-left-cell:after,:host ::ng-deep .devui-talbe-scorll-right tr .devui-last-sticky-left-cell:after,:host ::ng-deep .devui-talbe-scorll-right th .devui-last-sticky-left-cell:after{content:\"\";position:absolute;top:0;right:0;bottom:0;width:10px;pointer-events:none;box-shadow:inset 8px 0 8px -4px rgba(37,43,58,.1);box-shadow:inset var(--devui-table-inset-shadow-left, 8px 0 8px -4px) var(--devui-light-shadow, rgba(37, 43, 58, .1))}:host ::ng-deep .devui-talbe-scorll-middle tr .devui-first-sticky-right-cell,:host ::ng-deep .devui-talbe-scorll-middle th .devui-first-sticky-right-cell,:host ::ng-deep .devui-talbe-scorll-left tr .devui-first-sticky-right-cell,:host ::ng-deep .devui-talbe-scorll-left th .devui-first-sticky-right-cell{background:linear-gradient(to right,transparent,#ffffff 10px);background:linear-gradient(to right,transparent,var(--devui-base-bg, #ffffff) 10px)}:host ::ng-deep .devui-talbe-scorll-middle tr .devui-first-sticky-right-cell:before,:host ::ng-deep .devui-talbe-scorll-middle th .devui-first-sticky-right-cell:before,:host ::ng-deep .devui-talbe-scorll-left tr .devui-first-sticky-right-cell:before,:host ::ng-deep .devui-talbe-scorll-left th .devui-first-sticky-right-cell:before{content:\"\";position:absolute;top:0;left:0;bottom:0;width:10px;pointer-events:none;box-shadow:inset -8px 0 8px -4px rgba(37,43,58,.1);box-shadow:inset var(--devui-table-inset-shadow-right, -8px 0 8px -4px) var(--devui-light-shadow, rgba(37, 43, 58, .1))}:host ::ng-deep .gu-unselectable{-webkit-user-select:none!important;-moz-user-select:none!important;-ms-user-select:none!important;user-select:none!important}:host ::ng-deep .gu-mirror{position:fixed!important;margin:0!important;z-index:9999!important;background-color:#fff;box-shadow:0 2px 8px 0 var(--dark-1-boxshadow, rgba(41, 48, 64, .2));box-shadow:0 2px 8px rgba(41,48,64,.2);box-shadow:var(--devui-shadow-connected-overlay, 0 2px 8px 0) var(--dark-1-boxshadow, rgba(41, 48, 64, .2));list-style:none}:host ::ng-deep .gu-hide{display:none!important}:host ::ng-deep .sindu_dragger{list-style:none;margin:0;padding:0;box-sizing:border-box;display:flex}:host ::ng-deep .sindu_handle>.header-container>.drag-icon{cursor:move}:host ::ng-deep .sindu_dragger li{margin:0;padding:0;list-style:none;text-align:inherit}:host ::ng-deep .sindu_dragger li table,:host ::ng-deep .sindu_dragger tr,:host ::ng-deep .sindu_dragger th,:host ::ng-deep .sindu_dragger td{box-sizing:border-box}:host ::ng-deep .sindu_dragging .sindu_origin_table{visibility:hidden}:host ::ng-deep .gu-mirror li{margin:0;padding:0;list-style:none;text-align:inherit}:host ::ng-deep .gu-mirror li table,:host ::ng-deep .gu-mirror tr,:host ::ng-deep .gu-mirror th,:host ::ng-deep .gu-mirror td{box-sizing:border-box}:host ::ng-deep .gu-transit{opacity:.5}:host ::ng-deep .devui-table.table-bordered thead tr th.devui-checkable-cell{padding:10px}:host ::ng-deep .devui-table.table-bordered tbody>tr>td.devui-checkable-cell{padding:10px}\n", ":host ::ng-deep th .datatable-svg g#order-default path{fill:#71757f;fill:var(--devui-icon-fill, #71757f)}:host ::ng-deep th .datatable-svg g#order-desc polygon#Path-4{fill:#71757f;fill:var(--devui-icon-fill, #71757f)}:host ::ng-deep th .datatable-svg g#order-desc polygon#Path-4-Copy{fill:#252b3a;fill:var(--devui-icon-fill-active, #252b3a)}:host ::ng-deep th .datatable-svg g#order-asc polygon#Path-4{fill:#252b3a;fill:var(--devui-icon-fill-active, #252b3a)}:host ::ng-deep th .datatable-svg g#order-asc polygon#Path-4-Copy{fill:#71757f;fill:var(--devui-icon-fill, #71757f)}\n"] }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.NgZone }, { type: i0.Renderer2 }, { type: i0.ChangeDetectorRef }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }]; }, propDecorators: { checkable: [{
                type: Input
            }], headerCheckDisabled: [{
                type: Input
            }], headerCheckVisible: [{
                type: Input
            }], checkOptions: [{
                type: Input
            }], showExpandToggle: [{
                type: Input
            }], fixHeader: [{
                type: Input
            }], scrollable: [{
                type: Input
            }], editModel: [{
                type: Input
            }], maxWidth: [{
                type: Input
            }], maxHeight: [{
                type: Input
            }], type: [{
                type: Input
            }], rowHoveredHighlight: [{
                type: Input
            }], generalRowHoveredData: [{
                type: Input
            }], cssClass: [{
                type: Input
            }], tableWidth: [{
                type: Input
            }], tableHeight: [{
                type: Input
            }], containFixHeaderHeight: [{
                type: Input
            }], onlyOneColumnSort: [{
                type: Input
            }], multiSort: [{
                type: Input
            }], resizeable: [{
                type: Input
            }], colDraggable: [{
                type: Input
            }], colDropFreezeTo: [{
                type: Input
            }], detailTemplateRef: [{
                type: Input
            }], timeout: [{
                type: Input
            }], showOperationArea: [{
                type: Input
            }], showSortIcon: [{
                type: Input
            }], showFilterIcon: [{
                type: Input
            }], multiSortChange: [{
                type: Output
            }], cellClick: [{
                type: Output
            }], cellDBClick: [{
                type: Output
            }], rowClick: [{
                type: Output
            }], rowDBClick: [{
                type: Output
            }], detialToggle: [{
                type: Output
            }], beforeCellEdit: [{
                type: Input
            }], cellEditStart: [{
                type: Output
            }], cellEditEnd: [{
                type: Output
            }], rowCheckChange: [{
                type: Output
            }], checkAllChange: [{
                type: Output
            }], loadMore: [{
                type: Output
            }], resize: [{
                type: Output
            }], tableLevel: [{
                type: Input
            }], checkableRelation: [{
                type: Input
            }], loadChildrenTable: [{
                type: Input
            }], loadAllChildrenTable: [{
                type: Input
            }], virtualScroll: [{
                type: Input
            }], headerExpandConfig: [{
                type: Input
            }], childrenTableClose: [{
                type: Output
            }], allChildrenTableClose: [{
                type: Output
            }], virtualItemSize: [{
                type: Input
            }], virtualMinBufferPx: [{
                type: Input
            }], virtualMaxBufferPx: [{
                type: Input
            }], lazy: [{
                type: Input
            }], tableWidthConfig: [{
                type: Input
            }], headerBg: [{
                type: Input
            }], tableLayout: [{
                type: Input
            }], borderType: [{
                type: Input
            }], striped: [{
                type: Input
            }], tableScrollEvent: [{
                type: Output
            }], minHeight: [{
                type: Input
            }], columnDragEnd: [{
                type: Output
            }], size: [{
                type: Input
            }], shadowType: [{
                type: Input
            }], tableOverflowType: [{
                type: Input
            }], columns: [{
                type: ContentChildren,
                args: [DataTableColumnTmplComponent]
            }], innerHeader: [{
                type: ContentChild,
                args: [TableTheadComponent]
            }], innerBody: [{
                type: ContentChild,
                args: [TableTbodyComponent]
            }], thList: [{
                type: ContentChildren,
                args: [TableThComponent, { descendants: true }]
            }], noResultTemplate: [{
                type: ContentChild,
                args: ['noResultTemplateRef']
            }], fixHeaderContainerRefElement: [{
                type: ViewChild,
                args: ['fixHeaderContainerRef']
            }], tableViewRefElement: [{
                type: ViewChild,
                args: ['tableView', { static: true }]
            }], virtualScrollViewport: [{
                type: ViewChild,
                args: ['cdkVirtualScrollViewport']
            }], normalScrollElement: [{
                type: ViewChild,
                args: ['normalScroll']
            }], vitualScrollElement: [{
                type: ViewChild,
                args: ['scrollViewTpl']
            }], devuiNormalScrollBody: [{
                type: ViewChild,
                args: ['devuiNormalScrollBody', { read: ElementRef }]
            }], hostHeight: [{
                type: HostBinding,
                args: ['style.height']
            }], hasShadow: [{
                type: HostBinding,
                args: ['class.devui-table-shadow']
            }], content: [{
                type: ViewChild,
                args: ['tableBody']
            }], dataSource: [{
                type: Input
            }], hideColumn: [{
                type: Input
            }], pageAllChecked: [{
                type: Input
            }] } });
