// ng-devui/esm2020/data-table/display-cell-value.pipe.mjs
import { Pipe } from '@angular/core';
import { formatDate } from 'ng-devui/utils';
import * as i0 from "@angular/core";
export class DisPlayCellValuePipe {
    constructor() {
        this.datePicker = {
            format: {
                date: 'y/MM/dd',
                time: 'y/MM/dd HH:mm'
            }
        };
    }
    transform(rowItem, rowIndex, column) {
        if (!column || !column.field) {
            return null;
        }
        if (column.field === '$index') {
            return rowIndex + 1;
        }
        const cellItem = rowItem[column.field];
        switch (column.fieldType) {
            case 'date': {
                let pattern;
                if (column.extraOptions && column.extraOptions.dateFormat) {
                    pattern = column.extraOptions.dateFormat;
                }
                else {
                    pattern = column.extraOptions && column.extraOptions.showTime ?
                        this.datePicker.format.time : this.datePicker.format.date;
                }
                return cellItem ? formatDate(new Date(cellItem), pattern) : '';
            }
            case 'select':
            case 'treeSelect':
                if (column.extraOptions && column.extraOptions.multiple) {
                    if (cellItem && cellItem.length > 0) {
                        let displayValue;
                        if (column.extraOptions.filterKey) {
                            displayValue = cellItem.map(item => item[column.extraOptions.filterKey]).join(';');
                        }
                        else {
                            displayValue = cellItem.join(';');
                        }
                        return displayValue;
                    }
                    else {
                        return '--';
                    }
                }
                else {
                    let displayValue;
                    if (column.extraOptions && column.extraOptions.filterKey) {
                        displayValue = cellItem ? cellItem[column.extraOptions.filterKey] : cellItem;
                    }
                    else {
                        displayValue = cellItem;
                    }
                    return displayValue === 0 ? 0 : (displayValue || '--');
                }
            default:
                if (cellItem === null || cellItem === undefined || cellItem === '') {
                    return '--';
                }
                else {
                    return cellItem;
                }
        }
    }
}
DisPlayCellValuePipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DisPlayCellValuePipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
DisPlayCellValuePipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DisPlayCellValuePipe, name: "disPlayCellValuePipe" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DisPlayCellValuePipe, decorators: [{
            type: Pipe,
            args: [{ name: 'disPlayCellValuePipe' }]
        }], ctorParameters: function () { return []; } });
